---
title: "analyse_data"
author: "Bogna Smug"
date: '2022-09-06'
output: html_document
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE)
unaccepted.genes = data.table::fread(file = sprintf("%sunaccepted.genes.txt",OUTPUT.DATA.PATH))
families = data.table::fread(file = sprintf("%sfamilies.txt", OUTPUT.DATA.PATH)) %>%
  anti_join(unaccepted.genes %>% select(qname)) 
protein.similarity.data = data.table::fread(file = sprintf("%sprotein.similarity.data.txt",OUTPUT.DATA.PATH)) %>%
  anti_join(unaccepted.genes %>% select(qname)) %>%
  anti_join(unaccepted.genes %>% select(sname = qname))
repr.seq.lengths = data.table::fread(file = sprintf("%sprot-families/representative/repr-seqs-lengths.txt", DATA.PATH)) %>%
  anti_join(unaccepted.genes %>% select(name = qname)) 
ecod.domains.hits = data.table::fread(sprintf("%secod.domains.hits.txt",OUTPUT.DATA.PATH)) %>%
  anti_join(unaccepted.genes %>% select(qname)) 
ecod_id_to_names_map = data.table::fread(sprintf("%secod_id_to_names_map.txt",OUTPUT.DATA.PATH)) %>% as.data.frame()

surely.annotated.proteins.including.multi.annot = data.table::fread( sprintf("%ssurely.annotated.proteins.including.multi.annot",OUTPUT.DATA.PATH))  %>%
  left_join(families) %>%
  anti_join(unaccepted.genes %>% select(qname))
relaxely.annotated.proteins.including.multi.annot =data.table::fread(sprintf("%srelaxely.annotated.proteins.including.multi.annot",OUTPUT.DATA.PATH)) %>%
  left_join(families)  %>%
  anti_join(unaccepted.genes %>% select(qname))
mid.annotated.proteins.including.multi.annot =data.table::fread(sprintf("%smid.annotated.proteins.including.multi.annot",OUTPUT.DATA.PATH)) %>%
  left_join(families) %>%
  anti_join(unaccepted.genes %>% select(qname))


num.prot.in.reprseq = data.table::fread(file = sprintf("%snum.prot.in.reprseq.txt", OUTPUT.DATA.PATH)) %>%
  anti_join(unaccepted.genes %>% select(qname))


classes = rep("character", 16)
ecod.domain.descriptions = read.delim(ECOD.DOMAIN.DESCRIPTION.FILEPATH, 
                                      skip = 4, header = TRUE, 
                                      colClasses = classes, sep = "\t") 
dir.create(sprintf("%sFigure1", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure2", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure3", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure4", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure5", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure6", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sAdditional_Figures", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure_Supplementary", OUTPUT.FIGURES.PATH))


dir.create(sprintf("%sFigure_Supplementary/FigS1_tradeoff/", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure_Supplementary/FigS2_annot_network/", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure_Supplementary/FigS3_num_F_per_H/", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure_Supplementary/FigS4_all_T_domains/", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure_Supplementary/FigS5_p_vs_cov/", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure_Supplementary/FigS6/", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure_Supplementary/FigS7_folds_shared_by_mosaic/", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure_Supplementary/FigS8_annot_low_cov/", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure_Supplementary/FigS9_ecod_hist/", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure_Supplementary/FigS10_poor_ecod/", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure_Supplementary/FigS11_recent_HGT/", OUTPUT.FIGURES.PATH))
dir.create(sprintf("%sFigure_Supplementary/FigS12/", OUTPUT.FIGURES.PATH))

color.map.df = PHROG.COLOR.MAP %>% stack()
names(color.map.df) = c("color", "category")



# disregard the ones with multiple annotations
# now surely.annotated.prteins includes all small categories
index.cols = which(names(ecod.domains.hits) %in% c("x_id", "h_id", "t_id", "f_id"))
ecod.domains.hits[ , (index.cols) := lapply(.SD, as.character), .SDcols = index.cols]
relaxely.annotated.proteins = GetUniquelyAnnotatedProteins(relaxely.annotated.proteins.including.multi.annot)
surely.annotated.proteins = GetUniquelyAnnotatedProteins(surely.annotated.proteins.including.multi.annot) 
mid.annotated.proteins = GetUniquelyAnnotatedProteins(mid.annotated.proteins.including.multi.annot)
data.table::fwrite(surely.annotated.proteins, sprintf("%ssurely.annotated.proteins.txt", OUTPUT.DATA.PATH))


# We only include categories that have MIN.NUM.PROT (including the proteins with multiple annotation) as defined at high coverage threshold
reprseq.function.status = repr.seq.lengths %>%
  select(qname = name) %>%
  left_join(families) %>%
  mutate(has.unknown.function = !(qname %in% surely.annotated.proteins.including.multi.annot$qname),
         not.unique.function = (qname %in% surely.annotated.proteins.including.multi.annot$qname & !(qname %in% surely.annotated.proteins$qname)))


category.sizes = surely.annotated.proteins.including.multi.annot %>%
  left_join(reprseq.function.status) %>%
  group_by(annotation.index, annotation, category, include) %>%
  summarise(
    num.proteins.including.multi.annot = n_distinct(qname),
    num.families.including.multi.annot = n_distinct(family)) %>%
  ungroup() %>%
  as.data.frame() %>%
  rbind(
    data.frame(
      annotation.index = as.integer(UNKNOWN.ANNOTATION.INDEX),
      annotation = "unknown" , 
      category = "unknown", 
      include = FALSE,
      num.proteins.including.multi.annot = reprseq.function.status %>% filter(has.unknown.function) %>% nrow(),
      num.families.including.multi.annot = reprseq.function.status %>% filter(has.unknown.function) %>% distinct(family) %>% nrow()))  %>%
  mutate(unspecific = annotation %in% UNSPECIFIC.ANNOTATIONS) 


included.category.sizes = category.sizes %>%
  filter(num.proteins.including.multi.annot >= MIN.NUM.PROT & include) %>%
  distinct(annotation.index, annotation, category, num.proteins.including.multi.annot, unspecific) %>%
  inner_join(surely.annotated.proteins %>%
              group_by(annotation.index, annotation, category) %>%
              summarise(
                  num.proteins = n_distinct(qname),
                  num.families = n_distinct(family)) %>%
               ungroup())
data.table::fwrite(included.category.sizes, file = sprintf("%sincluded.category.sizes",OUTPUT.DATA.PATH))

```




# Fig 1: ANNOTATION UNCERTAINTY
```{r, eval = FALSE}
hhr.phrogs.annotated = data.table::fread(sprintf("%shhr.phrogs.annotated.txt",OUTPUT.DATA.PATH))  %>%
  anti_join(unaccepted.genes %>% select(qname)) 
num.proteins = n_distinct(repr.seq.lengths$name)


# NOTE THAT IN FIGURE 1 WE DO NOT RESTRICT OURSELVES TO CATEGORIES THAT ARE BIG ENOUGH
annotation.uncertainty.data = data.frame(annotation.coverage = character(0), minimum.probability = character(0), maximum.eval = character(0),
                                         num.prot = integer(0), num.annotated = integer(0), perc.annotated = numeric(0), perc.unique.annotation = numeric(0))


annotation.uncertainty.data.prob = Calculate.Annotation.Tradeoff(hhr.phrogs.annotated,
                                                                 max.eval.range = DEFAULT.NAXIMUM.EVAL.FOR.ANNOTATION, 
                                                                        min.cov.range = MAIN.COVS.FOR.ANNOTATION, 
                                                                        min.prob.range = MINIMUM.PROBABILITIES.FOR.ANNOTATION.EXPLORATION)
data.table::fwrite(annotation.uncertainty.data.prob, sprintf("%sFigure_Supplementary/FigS1_tradeoff/Tradeoff_prob.txt", OUTPUT.FIGURES.PATH)) 

annotation.uncertainty.data.cov= Calculate.Annotation.Tradeoff(hhr.phrogs.annotated,
                                                               max.eval.range = DEFAULT.NAXIMUM.EVAL.FOR.ANNOTATION, 
                                                                        min.cov.range = MINIMUM.COVERAGES.FOR.ANNOTATION.EXPLORATION, 
                                                                        min.prob.range = MAIN.PROBS.FOR.ANNOTATION)
data.table::fwrite(annotation.uncertainty.data.cov, sprintf("%sFigure_Supplementary/FigS1_tradeoff/Tradeoff_cov.txt", OUTPUT.FIGURES.PATH)) 
```



```{r}

included.categories = included.category.sizes %>% 
  distinct(annotation.index)


node.color.map = included.category.sizes %>% 
                                      select(annotation.index, category) %>%
                                      left_join(color.map.df, by = "category") %>%
                                      select(node = annotation.index, color)
node.size.map = included.category.sizes %>% 
                                      mutate(node = annotation.index, size = log(num.families) + 1) %>%#num.proteins/100) %>%
                                      select(node, size) 
node.labels = included.category.sizes %>% 
                                      rowwise() %>%
                                      mutate(annot = gsub(" ", "\n", annotation)) %>%
                                      select(node = annotation.index, label = annot)
annotation.network.metadata = node.color.map %>%
  inner_join(node.size.map, by = "node") %>%
  inner_join(node.labels, by = "node") %>%
 rbind(
    data.frame(
      node = as.integer(UNKNOWN.ANNOTATION.INDEX),
      color = "#808080",
      size = (reprseq.function.status %>% filter(has.unknown.function | not.unique.function) %>% distinct(family) %>% nrow() %>% log()) + 1,
      label = "unknown"))


# metadata fro all annotations, not just the included ones, here we will look at the size inluding multiannots
# otherwise it will be hard to interpret
node.color.map.all.cat = category.sizes %>% 
                                      select(annotation.index, category) %>%
                                      left_join(color.map.df, by = "category") %>%
                                      select(node = annotation.index, color)
node.size.map.all.cat = category.sizes %>% 
                                      mutate(node = annotation.index, size = log(num.families.including.multi.annot) + 1) %>% #size = num.proteins.including.multi.annot/100) %>%
                                      select(node, size) 
node.labels.all.cat = category.sizes %>% 
                                      rowwise() %>%
                                      mutate(annot = gsub(" ", "\n", annotation)) %>%
                                      select(node = annotation.index, label = annot)
annotation.network.metadata.all.cat = node.color.map.all.cat %>%
  inner_join(node.size.map.all.cat, by = "node") %>%
  inner_join(node.labels.all.cat, by = "node") %>%
  mutate(color = if_else(label == "unknown", "#808080", color))

data.table::fwrite(annotation.network.metadata, 
                   sprintf("%sFigure_Supplementary/FigS2_annot_network/annotation.network.metadata.txt", OUTPUT.FIGURES.PATH))
data.table::fwrite(annotation.network.metadata.all.cat, 
                   sprintf("%sFigure_Supplementary/FigS2_annot_network/annotation.network.metadata.all.cat.txt", OUTPUT.FIGURES.PATH))
```


```{r}
# we onlcy consider included categories here to make it easier (i,e the ones we consider in fig 2, 3, 4, we only look at all categories when looking into recent mosaicism)
included.proteins.relaxely.annotated =   mid.annotated.proteins.including.multi.annot %>%
  #relaxely.annotated.proteins.including.multi.annot %>%
  inner_join(included.categories) %>% 
  distinct(annotation.index, qname)

annotation.netwok.relaxed.threshold = included.proteins.relaxely.annotated %>%
  inner_join(included.proteins.relaxely.annotated, by = "qname") %>%
  filter(annotation.index.x != annotation.index.y) %>%
  group_by(annotation.index.x, annotation.index.y) %>%
  summarise(n = n_distinct(qname)) %>%
  ungroup() %>%
  filter(n >= MIN.NUM.SHARED.SEQ.FOR.SHARED.ANNOT.GRAPH) %>%
  select(from = annotation.index.x, to = annotation.index.y)


included.proteins.surely.annotated =   surely.annotated.proteins.including.multi.annot %>%
  inner_join(included.categories) %>% 
  distinct(annotation.index, qname)

annotation.netwok.default.threshold = included.proteins.surely.annotated %>%
  inner_join(included.proteins.surely.annotated, by = "qname") %>%
  filter(annotation.index.x != annotation.index.y) %>%
  group_by(annotation.index.x, annotation.index.y) %>%
  summarise(n = n_distinct(qname)) %>%
  ungroup() %>%
  filter(n >= MIN.NUM.SHARED.SEQ.FOR.SHARED.ANNOT.GRAPH) %>%
  select(from = annotation.index.x, to = annotation.index.y)



data.table::fwrite(annotation.netwok.relaxed.threshold, 
                   sprintf("%sFigure_Supplementary/FigS2_annot_network/annotation.netwok.relaxed.threshold.txt", OUTPUT.FIGURES.PATH))
data.table::fwrite(annotation.netwok.default.threshold, 
                   sprintf("%sFigure_Supplementary/FigS2_annot_network/annotation.netwok.default.threshold.txt", OUTPUT.FIGURES.PATH))

```

```{r}
surely.annotated.ecod.domains.high.confidence = ecod.domains.hits %>% 
  select(qname, qstart, qend, qlength, sstart, send, slength, x_id, h_id, t_id, f_id) %>%
  # note that some families may be unclassified i.e. d_id = NA
  inner_join(surely.annotated.proteins)  %>%  # take the conservatively defined domains only
  inner_join(included.category.sizes, by = c("annotation.index", "annotation", "category")) 

surely.annotated.proteins.from.included.categories = surely.annotated.proteins %>%
  inner_join(included.category.sizes) %>%
  select(qname, family, annotation.index , annotation, category)
```

## Heatmap of H ECOD domains found in different functions
```{r, fig.height = 13}

# here we only look at included categories and only the specific ones

h.domain.occurence.data = surely.annotated.ecod.domains.high.confidence %>% 
  filter(!(annotation %in% UNSPECIFIC.ANNOTATIONS)) %>%
  distinct(qname, family, h_id, annotation, category) %>%
  group_by(annotation, category, h_id) %>%
  summarise(num.prot.with.this.h.domain = n_distinct(qname),
            num.families.with.this.h.domain = n_distinct(family)) %>%
  inner_join(included.category.sizes %>% 
               select(annotation, category, num.proteins)) %>%
  mutate(occurence = num.prot.with.this.h.domain >= MIN.NUM.REPRSEQS.FOR.ECOD.PRESENCE) %>%
  filter(occurence) %>%
  mutate(h_id = as.character(h_id)) %>%
  left_join(ecod_id_to_names_map %>% filter(level == "H") %>% select(h_id = id, h_name = name)) %>%
         # delete singletons
         group_by(h_id) %>%
         mutate(num_annot = n_distinct(annotation)) %>%
         filter(num_annot >= 3) %>%
  group_by(h_name) %>%
  mutate(num.categories = n_distinct(category)) %>%
  mutate(all.categories = if_else(num.categories == 1, 
                                  category, 
                                  "multiple categories")) %>%
  ungroup()
data.table::fwrite(h.domain.occurence.data, sprintf("%sFigure1/H_domains.txt", OUTPUT.FIGURES.PATH))




t.domain.occurence.data.full = surely.annotated.ecod.domains.high.confidence %>% 
    filter(!(annotation %in% UNSPECIFIC.ANNOTATIONS)) %>%
  distinct(qname, family, t_id, annotation, category) %>%
  group_by(annotation, category, t_id) %>%
  summarise(num.prot.with.this.t.domain = n_distinct(qname),
            num.families.with.this.h.domain = n_distinct(family)) %>%
  inner_join(included.category.sizes %>% 
               select(annotation, category, num.proteins)) %>%
  mutate(occurence = num.prot.with.this.t.domain >= MIN.NUM.REPRSEQS.FOR.ECOD.PRESENCE) %>%
  filter(occurence) %>%
  mutate(t_id = as.character(t_id)) %>%
  left_join(ecod_id_to_names_map %>% filter(level == "T") %>% select(t_id = id, t_name = name)) %>%
  group_by(t_name) %>%
  mutate(num.categories = n_distinct(category)) %>%
  mutate(all.categories = if_else(num.categories == 1, 
                                  category, 
                                  "multiple categories")) %>%
  ungroup()
data.table::fwrite(t.domain.occurence.data.full, sprintf("%sFigure_Supplementary/FigS4_all_T_domains/T_domains.txt", OUTPUT.FIGURES.PATH))



# here we also include annotations from included categories only
num.f.within.h = surely.annotated.ecod.domains.high.confidence  %>% 
    filter(!(annotation %in% UNSPECIFIC.ANNOTATIONS)) %>%
  group_by(h_id) %>% 
  summarise(num_f = n_distinct(f_id), 
            num_t = n_distinct(t_id), 
            num_prot = n_distinct(qname)) %>%
  left_join(h.domain.occurence.data %>% distinct(h_id, h_name, all.categories) %>% mutate(in.ecod.heatmap = TRUE)) %>%
  distinct(h_name, num_f, num_t, num_prot, all.categories, in.ecod.heatmap)
data.table::fwrite(num.f.within.h, sprintf("%sFigure_Supplementary/FigS3_num_F_per_H/num.f.within.h.txt", OUTPUT.FIGURES.PATH))
```









# Figure 3a: density plot
```{r}
# only filter to those with minimum length and high probability
 filtered.protein.similarity.data = protein.similarity.data %>% filter(share.any.fragment)
data.table::fwrite(filtered.protein.similarity.data, sprintf("%sFigure_Supplementary/FigS5_p_vs_cov/cov.vs.pident.txt", OUTPUT.FIGURES.PATH))
```

# network of DNA, RNA, and nucleotide metabolism domains
```{r}
selected_categories = c("DNA, RNA and nucleotide metabolism","tail", "lysis")
domain.network.selected.categories = data.frame(from = character(0), to = character(0), category = character(0), num.prot = numeric(0))


for (selected_category in selected_categories) {
  domains.in.selected.category = surely.annotated.ecod.domains.high.confidence %>%
    filter(category == selected_category) %>%
    distinct(t_id, category)
  # here we look only at sequences that include a domain from selected category
  this.category.domain.hits = surely.annotated.ecod.domains.high.confidence %>%
    inner_join(domains.in.selected.category, by = c("t_id", "category")) %>%
    distinct(qname, t_id, category)
  # and we see what other doains coexist with these domains
  domain.network.this.category = this.category.domain.hits %>%
    inner_join(this.category.domain.hits, by = c("qname", "category")) %>%
    filter(t_id.x < t_id.y) %>%
    rowwise() %>%
    mutate(pair = paste(sort(c(t_id.x, t_id.y)), collapse = " AND ", sep = ""),
           category = selected_category) %>%
    group_by(t_id.x, t_id.y, pair, category) %>%
    summarise(num.prot = n_distinct(qname)) %>%
    ungroup() %>%
    select(from = t_id.x, to = t_id.y, category, num.prot) #%>%
    #filter(num.prot >= 5)
    domain.network.selected.categories = rbind(domain.network.selected.categories, domain.network.this.category)
}



  domain.annotations = surely.annotated.ecod.domains.high.confidence %>%
    filter(category %in% selected_categories)  %>%
    group_by(t_id, category) %>%
    summarise(n = n_distinct(annotation),
              n.prot = n_distinct(qname),
              annotation = if_else(n == 1, annotation[1], "multiple")) %>%
    left_join(stack(PHROG.COLOR.MAP) %>% select(color = values, category = ind), by = "category") %>%
    mutate(multi = annotation == "multiple")  %>%
    mutate(color = if_else(multi, "black", color)) %>%
    select(node = t_id, category, color, n.prot)
  
  domain.network.metadata = data.frame(node = unique(c(domain.network.selected.categories$from, domain.network.selected.categories$to))) %>%
    left_join(ecod_id_to_names_map %>% filter(level == "T") %>% select(node = id, label = name)) %>%
    left_join(domain.annotations, by = "node")
  
data.table::fwrite(domain.network.metadata, sprintf("%sAdditional_Figures/domain.network.metadata.txt", OUTPUT.FIGURES.PATH))
data.table::fwrite(domain.network.selected.categories, sprintf("%sAdditional_Figures/domain.network.selected.categories.txt", OUTPUT.FIGURES.PATH))

```




#Fig 4a and 4b Diversity of functional categories
```{r, fig.width = 10}
any.hit.stats = surely.annotated.ecod.domains.high.confidence %>%
  distinct(qname) %>%
  mutate(has.ecod.hit = 1) %>%
  right_join(surely.annotated.proteins.from.included.categories %>% distinct(qname, annotation, category, annotation.index), by = "qname") %>%
  mutate(has.ecod.hit = if_else(is.na(has.ecod.hit), 0, has.ecod.hit)) %>%
  group_by(annotation, category, annotation.index) %>%
  summarise(num.proteins.with.ecod.hit = sum(has.ecod.hit)) %>%
  ungroup() %>%
  inner_join(included.category.sizes, by = c("annotation", "category", "annotation.index")) %>%
  mutate(prop.proteins.with.ecod.hit = num.proteins.with.ecod.hit/num.proteins)

# only look at included categories (big enough)
num.domains.per.annnot = surely.annotated.ecod.domains.high.confidence %>%
  group_by(annotation, category, annotation.index) %>%
  summarise(num.t = n_distinct(t_id),
            num.h = n_distinct(h_id),
            num.x = n_distinct(x_id))

num.domain.combinations.per.qname = surely.annotated.ecod.domains.high.confidence %>%
  group_by(qname, annotation, category, annotation.index) %>%
  summarise(all.t = paste(sort(unique(t_id)), collapse = " & "),
            all.h = paste(sort(unique(h_id)), collapse = " & "),
            all.x = paste(sort(unique(x_id)), collapse = " & "),
            num.x = n_distinct(x_id),
            num.h = n_distinct(h_id)) %>%
  ungroup()

num.domain.combinations.per.annnot = num.domain.combinations.per.qname %>%
  group_by(annotation, category, annotation.index) %>%
  summarise(num.t.combinations = n_distinct(all.t),
            num.h.combinations = n_distinct(all.h),
            num.x.combinations = n_distinct(all.x))

diversity.data0 = included.category.sizes %>%
  left_join(num.domains.per.annnot, by = c('annotation', 'category', 'annotation.index'))  %>%
  left_join(num.domain.combinations.per.annnot, by = c('annotation', 'category', 'annotation.index')) %>%
  left_join(any.hit.stats, by = c('annotation', 'category', 'annotation.index', 'num.proteins', 'num.families')) %>%
  mutate(mean.no.of.folds.per.fold.type = num.t.combinations/num.proteins) %>%
  select(annotation, category, annotation.index, num.families, num.proteins, num.topologies = num.t, num.topology.combinations = num.t.combinations, prop.proteins.with.ecod.hit, mean.no.of.folds.per.fold.type) %>%
  mutate(poorly.ecod.annotated = prop.proteins.with.ecod.hit < MIN.PROP.PROTS.WITH.ECOD.HIT) %>%
  filter(!(annotation %in% UNSPECIFIC.ANNOTATIONS ))




no.fam.25.quant <- floor(as.numeric(quantile(diversity.data0 %>% 
                                                   filter(!poorly.ecod.annotated) %>% 
                                                   pull(num.families))[2]))

no.fam.plotting.thr <- floor(as.numeric(quantile(diversity.data0 %>% 
                                                   filter(!poorly.ecod.annotated) %>% 
                                                   pull(num.families))[3]))
no.prot.plotting.thr <- floor(as.numeric(quantile(diversity.data0 %>% 
                                                   filter(!poorly.ecod.annotated) %>% 
                                                   pull(num.proteins))[3]))
no.ft.plotting.thr <- floor(as.numeric(quantile(diversity.data0 %>% 
                                                   filter(!poorly.ecod.annotated) %>% 
                                                   pull(num.topology.combinations))[3]))


no.prot.75.quant <- floor(as.numeric(quantile(diversity.data0 %>% 
                                                   filter(!poorly.ecod.annotated) %>% 
                                                   pull(num.proteins))[4]))
no.fam.75.quant <- floor(as.numeric(quantile(diversity.data0 %>% 
                                                   filter(!poorly.ecod.annotated) %>% 
                                                   pull(num.families))[4]))
no.ft.75.quant <- floor(as.numeric(quantile(diversity.data0 %>% 
                                                   filter(!poorly.ecod.annotated) %>% 
                                                   pull(num.topology.combinations))[4]))



diversity.data1 = diversity.data0 %>% 
  mutate(median.num.prot = no.prot.plotting.thr,
         median.num.fam = no.fam.plotting.thr,
        median.num.topology.combinations = no.ft.plotting.thr,
        first.quantile.num.fam  = no.fam.25.quant ) %>%
  mutate(#top.quantile = num.families >= median.num.fam & num.topology.combinations >= median.num.topology.combinations & !poorly.ecod.annotated,
         #bottom.quantile =  num.families <= median.num.fam & num.topology.combinations <= median.num.topology.combinations & !poorly.ecod.annotated,
         top.quantile = num.families >= first.quantile.num.fam & num.topology.combinations >= median.num.topology.combinations & !poorly.ecod.annotated,
         bottom.quantile =  num.families <= first.quantile.num.fam & num.topology.combinations <= median.num.topology.combinations & !poorly.ecod.annotated,
         diverse.for.label = (num.families > no.fam.75.quant & num.topology.combinations > no.ft.75.quant) & !poorly.ecod.annotated) %>%
  mutate(data.type = if_else(top.quantile, "top.diversity", if_else(bottom.quantile, "low.diversity", if_else(poorly.ecod.annotated, "poorly annotated", "unclear.diversity"))))

```




# sequence mosaicism 
```{r}
# Only looking at pairs of proteins where both are from the same functional category
seq.mosaicism.from.various.classes = protein.similarity.data %>%
  inner_join( surely.annotated.proteins.from.included.categories %>% distinct(qname, annotation, category) %>% select(qname = qname, qannot = annotation)) %>%
  inner_join( surely.annotated.proteins.from.included.categories %>% distinct(qname, annotation, category) %>% select(sname = qname, sannot = annotation)) %>%
  filter(qannot == sannot) %>%
  group_by(qannot) %>%
  # proportion of pairs that show any similarity not all possible pairs
  summarise(prop.mosaic.pairs.sequence.thr0p1 = sum(mosaic.pident10)/n(),
            prop.mosaic.pairs.sequence.thr0p3 = sum(mosaic.pident30)/n(),
            prop.mosaic.pairs.sequence.thr0p5 = sum(mosaic.pident50)/n()) %>%
  select(annotation = qannot, prop.mosaic.pairs.sequence.thr0p1, prop.mosaic.pairs.sequence.thr0p3, prop.mosaic.pairs.sequence.thr0p5)


```


# Mosaicism and diversity per fold architecture (i.e. combination of domains)
```{r}

fold.types.per.qname.and.domain = ecod.domains.hits %>%
  distinct(qname, x_id, t_id) %>%
  group_by(qname) %>%
  mutate(fold.type = paste(sort(unique(t_id)), collapse = " & ")) %>%
  ungroup()

fold.types.per.qname = fold.types.per.qname.and.domain %>%
  distinct(qname, fold.type) 

ecods.across.fold.types = fold.types.per.qname.and.domain %>%
  left_join(surely.annotated.proteins.from.included.categories) %>%
  distinct(t_id, x_id, fold.type, annotation.index, annotation, category) %>%
  group_by(fold.type, annotation,annotation.index, category) %>%
  mutate(all.x = paste(sort(unique(c(x_id))), collapse = " & "),
            num.x = n_distinct(x_id),
            num.t = n_distinct(t_id)) %>%
  ungroup()

#ecod.domains.in.qnames = ecods.across.fold.types.all  %>%
#  distinct(qname, family, annotation, annotation.index, category, x_id, t_id)

#ecod.domains.in.qnames = surely.annotated.ecod.domains.high.confidence  %>%
#  distinct(qname,family, annotation, annotation.index, category, x_id, t_id)


# first find all X level domains found


num.folds.per.fold.type.data = all.x.and.t.within.fold.type %>%
  group_by(annotation, category) %>%
  summarise(mean.num.t = mean(num.t))

fold.type.pairs.that.share.t.and.disshare.x = Get.Pairs.That.Share.T.And.Disshare.X(ecods.across.fold.types)
mosaicism.per.fold.type = Calculate.Mosaicism.Per.Fold.Type(ecods.across.fold.types)
fold.type.mosaicism = Calculate.Num.Mosaic.Fold.Types(ecods.across.fold.types)
  

diversity.data = diversity.data1 %>%
  left_join(mosaicism.per.fold.type) %>%
  left_join(num.folds.per.fold.type.data) %>%
  left_join(seq.mosaicism.from.various.classes) %>%
  filter(!(annotation %in% UNSPECIFIC.ANNOTATIONS ))  

fold.types.involved.in.mosaic.pairs = unique(c(fold.type.pairs.that.share.t.and.disshare.x %>%
  distinct(fold.type.x) %>% pull(fold.type.x),
  fold.type.pairs.that.share.t.and.disshare.x %>%
  distinct(fold.type.y) %>% pull(fold.type.y)))


fold.types.involved.in.mosaic.pairs.pident50 = unique(c(qname.pairs.sharing.t.and.disshare.x2 %>%
  filter(mosaic.pident50) %>% distinct(fold.type.x) %>% pull(fold.type.x),
  qname.pairs.sharing.t.and.disshare.x2 %>%
  filter(mosaic.pident50) %>% distinct(fold.type.y) %>% pull(fold.type.y)))


all.fold.types = ecods.across.fold.types %>%
  distinct(fold.type, t_id, x_id) %>%
  mutate(fold.type.mosaic = fold.type %in% fold.types.involved.in.mosaic.pairs,
         fold.type.mosaic.pident50 = fold.type %in% fold.types.involved.in.mosaic.pairs.pident50) 


all.t.in.fold.types = Get_Domain_Odss_In_Fold_Types(all.fold.types, ecod_id_to_names_map)    
data.table::fwrite(all.t.in.fold.types, file = sprintf("%sAdditional_figures/all.t.in.fold.types.txt", OUTPUT.FIGURES.PATH))

all.t.in.fold.types.pident50 = Get_Domain_Odss_In_Fold_Types(all.fold.types %>% 
                                                               mutate(fold.type.mosaic = fold.type.mosaic.pident50), 
                                                             ecod_id_to_names_map)    
data.table::fwrite(all.t.in.fold.types.pident50, file = sprintf("%sAdditional_figures/all.t.in.fold.types_pident50.txt", OUTPUT.FIGURES.PATH))
```





#FIGURE 3B MOSAICISM
#ECOD MOSAICISM
```{r}
seq.mosaic.pairs.all.pident.symmetric = rbind(
  seq.mosaic.pairs.all.pident %>%
  select(qname.x = qname, qname.y = sname, mosaic.pident50, mosaic.pident30, mosaic.pident70),
  seq.mosaic.pairs.all.pident %>%
  select(qname.y = qname, qname.x = sname, mosaic.pident50, mosaic.pident30, mosaic.pident70)) %>%
  group_by(qname.x, qname.y) %>%
  summarise(mosaic.pident30 = any(mosaic.pident30, na.rm = TRUE),
            mosaic.pident50 = any(mosaic.pident50, na.rm = TRUE),
            mosaic.pident70 = any(mosaic.pident70, na.rm = TRUE))

qname.pairs.sharing.t.and.disshare.x2 = 
  #qname.pairs.sharing.t.and.disshare.x %>%
  fold.type.pairs.that.share.t.and.disshare.x %>% 
  distinct(fold.type.x, fold.type.y) %>% 
  left_join(fold.types.per.qname %>% select(qname.x = qname, fold.type.x = fold.type), by = "fold.type.x", relationship = "many-to-many") %>%
  left_join(fold.types.per.qname %>% select(qname.y = qname,  fold.type.y = fold.type), by = "fold.type.y", relationship = "many-to-many") %>%
  left_join(seq.mosaic.pairs.all.pident.symmetric %>% select(qname.x, qname.y, mosaic.pident50, mosaic.pident30, mosaic.pident70)) %>%
  group_by(fold.type.x, fold.type.y) %>%
  summarise(mosaic.pident30 = any(mosaic.pident30, na.rm = TRUE),
            mosaic.pident50 = any(mosaic.pident50, na.rm = TRUE),
            mosaic.pident70 = any(mosaic.pident70, na.rm = TRUE))
  
  
``` 

```{r}
x.domains.per.qname = ecod.domains.hits %>%
  left_join(families) %>%
  group_by(qname, family) %>%
  summarise(num.x = n_distinct(x_id),
            all.x.names = paste(sort(unique(x_id)), collapse = " & ", sep = "")) 

ecod.domains.in.qnames.wih.multi.x = ecod.domains.hits %>%
  distinct(qname, t_id,x_id) %>%
  inner_join(x.domains.per.qname %>% filter(num.x > 1)) 

qname.pairs.that.share.t = ecod.domains.in.qnames.wih.multi.x %>%
  inner_join(ecod.domains.in.qnames.wih.multi.x, by = c("t_id")) 

qname.pairs.sharing.t.and.disshare.x = qname.pairs.that.share.t %>%
  filter(all.x.names.y!= all.x.names.x) %>%
  left_join(ecod.domains.in.qnames.wih.multi.x %>% select(qname.x = qname, x.id.x = x_id)) %>%
  left_join(ecod.domains.in.qnames.wih.multi.x %>% select(qname.y = qname, x.id.y = x_id)) %>%
  group_by(qname.x, qname.y, t_id) %>%
  mutate(all.x.within.both.families = paste(sort(unique(c(x.id.x, x.id.y))), collapse = " & ", sep = "")) %>%
  ungroup() %>%
  filter(all.x.within.both.families != all.x.names.x & all.x.within.both.families != all.x.names.y) 

# here we keep only qname.x < qname.y so one row per pair
qname.pairs.sharing.t.and.disshare.x.distinct = qname.pairs.sharing.t.and.disshare.x %>%
  left_join(surely.annotated.proteins.from.included.categories %>% distinct(qname,annotation.index) %>% select(qname.x = qname, annotation.index.from = annotation.index)) %>%
  left_join(surely.annotated.proteins.from.included.categories %>% distinct(qname,annotation.index) %>% select(qname.y = qname, annotation.index.to = annotation.index)) %>%
  distinct(qname.x, family.x, qname.y, family.y, t_id, all.x.names.x, all.x.names.y, all.x.within.both.families, annotation.index.from, annotation.index.to) %>%
  group_by(family.x, family.y) %>%
  mutate(family.pair = paste(sort(c(unique(family.x), unique(family.y))), collapse = "&", sep = "")) %>%
  ungroup() %>%
  group_by(qname.x, qname.y) %>%
  mutate(reprseq.pair = paste(sort(c(qname.x, qname.y)), collapse = "&", sep = "")) %>%
  ungroup() 
 # filtering is deleted so that we can group bt from annotation and to annotation and get the same results 
 #%>% 
  #filter(qname.x < qname.y) 
data.table::fwrite(qname.pairs.sharing.t.and.disshare.x.distinct, file = sprintf("%sFigure_Supplementary/FigS7_folds_shared_by_mosaic/qname.pairs.sharing.t.and.disshare.x.distinct.txt", OUTPUT.FIGURES.PATH))


#
```

#FIG 3C
# what domains are shared in mosaic pairs
```{r}
domains.shared.in.mosaic.pairs = qname.pairs.sharing.t.and.disshare.x.distinct %>%
  group_by(t_id) %>%
  summarise(num.family.pairs = n_distinct(family.pair),
            num.reprseq.pairs = n_distinct(reprseq.pair),
            num.unknown.family.pairs = n_distinct(family.pair[is.na(annotation.index.from) | is.na(annotation.index.to)]),
            num.known.family.pairs = n_distinct(family.pair[!is.na(annotation.index.from) & !is.na(annotation.index.to)])) %>%
  ungroup() %>%
  mutate(prop.family.pairs = num.family.pairs/n_distinct(qname.pairs.sharing.t.and.disshare.x.distinct$family.pair),
         prop.protein.pairs = num.reprseq.pairs/n_distinct(qname.pairs.sharing.t.and.disshare.x.distinct$reprseq.pair))


domains.shared.in.mosaic.pairs$h.index <- sapply(1:nrow(domains.shared.in.mosaic.pairs), function(index){
  this.topology <- domains.shared.in.mosaic.pairs$t_id[index]
  this.topology.h.index <- strsplit(this.topology, ".", fixed = T)[[1]]
  this.topology.h.index <- this.topology.h.index[-length(this.topology.h.index)]
  this.topology.h.index <- as.character(paste(this.topology.h.index, collapse = "."))
})

domains.shared.in.mosaic.pairs = domains.shared.in.mosaic.pairs %>%
    left_join(ecod_id_to_names_map %>% filter(level == "H") %>% select(h.index = id, h_name = name)) %>%
    left_join(ecod_id_to_names_map %>% filter(level == "T") %>% select(t_id = id, t_name = name)) 

data.table::fwrite(domains.shared.in.mosaic.pairs, file = sprintf("%sFigure_Supplementary/FigS7_folds_shared_by_mosaic/domains-shared-by-mosaic-pairs.txt", OUTPUT.FIGURES.PATH))
```



# Fig 3B
```{r}

seq.mosaic.pairs.all.pident = protein.similarity.data %>%
  filter(qname < sname & mosaic.pident10) %>%
  left_join(surely.annotated.proteins.from.included.categories %>% distinct(qname,annotation.index) %>% select(qname, annotation.index.from = annotation.index)) %>%
  left_join(surely.annotated.proteins.from.included.categories %>% distinct(qname,annotation.index) %>% select(sname = qname, annotation.index.to = annotation.index)) %>%
  left_join(families %>% select(qname,qfamily = family)) %>%
  left_join(families %>% select(sname=qname,sfamily = family)) %>%
  group_by(qfamily, sfamily) %>%
  mutate(family.pair = paste(sort(c(unique(qfamily), unique(sfamily))), collapse = "&", sep = "")) %>%
  ungroup() %>%
  group_by(qname, sname) %>%
  mutate(reprseq.pair = paste(sort(c(qname, sname)), collapse = "&", sep = "")) %>%
  ungroup() 


seq.mosaic.pairs.from.included.funct = seq.mosaic.pairs.all.pident %>%
  # narrow down to included annotations only
  inner_join(included.category.sizes %>% 
               select(annotation.index.from = annotation.index) %>% 
               distinct(annotation.index.from)) %>%
  inner_join(included.category.sizes %>% 
               select(annotation.index.to = annotation.index) %>% 
               distinct(annotation.index.to)) %>%
  distinct(qname, sname, annotation.index.from, annotation.index.to, family.pair, reprseq.pair, qfamily, sfamily, mosaic,  mosaic.pident10, mosaic.pident30, mosaic.pident50, mosaic.pident70, mosaic.pident90) %>%
  mutate(mosaicism.type = "seq")


ecod.mosaic.pairs.from.included.funct = 
  qname.pairs.sharing.t.and.disshare.x.distinct %>%
    # narrow down to included annotations only
  inner_join(included.category.sizes %>% 
               select(annotation.index.from = annotation.index) %>% 
               distinct(annotation.index.from)) %>%
  inner_join(included.category.sizes %>% 
               select(annotation.index.to = annotation.index) %>% 
               distinct(annotation.index.to)) %>%
  select(qname = qname.x, sname=qname.y, annotation.index.from, annotation.index.to, family.pair, reprseq.pair, qfamily = family.x, sfamily = family.y) %>%
  distinct() %>%
  mutate(mosaicism.type = "ecod")


unspecific.annotation.indices = included.category.sizes %>% 
  filter(unspecific) %>% 
  pull(annotation.index)

mosaic.pairs.from.included.funct = # we may have the same pair twice if it is both mosaic by seq and by ecod
  rbind(seq.mosaic.pairs.from.included.funct %>% 
          filter(mosaic) %>%
          select(qname, sname, annotation.index.from, annotation.index.to, family.pair, reprseq.pair,
                 qfamily, sfamily, mosaicism.type), 
         ecod.mosaic.pairs.from.included.funct %>%
          select(qname, sname, annotation.index.from, annotation.index.to, family.pair, reprseq.pair,
                 qfamily, sfamily, mosaicism.type),
         seq.mosaic.pairs.from.included.funct %>% 
          filter(mosaic.pident50) %>%
          mutate(mosaicism.type = "recent seq. mosaicism 50% ident") %>%
          select(qname, sname, annotation.index.from, annotation.index.to, family.pair, reprseq.pair,
                 qfamily, sfamily, mosaicism.type),
        seq.mosaic.pairs.from.included.funct %>% 
          filter(mosaic.pident30) %>%
          mutate(mosaicism.type = "recent seq. mosaicism 30% ident") %>%
          select(qname, sname, annotation.index.from, annotation.index.to, family.pair, reprseq.pair,
                 qfamily, sfamily, mosaicism.type),
        seq.mosaic.pairs.from.included.funct %>% 
          filter(mosaic.pident70) %>%
          mutate(mosaicism.type = "recent seq. mosaicism 70% ident") %>%
          select(qname, sname, annotation.index.from, annotation.index.to, family.pair, reprseq.pair,
                 qfamily, sfamily, mosaicism.type)) %>%
  group_by(annotation.index.from, annotation.index.to) %>%
  summarise(
    num.families.any.mosaic = n_distinct(family.pair),
    num.family.pairs.domain = n_distinct(family.pair[mosaicism.type == "ecod"]),
    num.family.pairs.seq = n_distinct(family.pair[mosaicism.type == "seq"]),
    num.family.pairs.recent.mosaic.seq.pident30 = n_distinct(family.pair[mosaicism.type == "recent seq. mosaicism 30% ident"]),
    num.reprseq.pairs.recent.mosaic.seq.pident30 = n_distinct(reprseq.pair[mosaicism.type == "recent seq. mosaicism 30% ident"]),
    num.family.pairs.recent.mosaic.seq.pident50 = n_distinct(family.pair[mosaicism.type == "recent seq. mosaicism 50% ident"]),
    num.reprseq.pairs.recent.mosaic.seq.pident50 = n_distinct(reprseq.pair[mosaicism.type == "recent seq. mosaicism 50% ident"]),
    num.family.pairs.recent.mosaic.seq.pident70 = n_distinct(family.pair[mosaicism.type == "recent seq. mosaicism 70% ident"]),
    num.reprseq.pairs.recent.mosaic.seq.pident70 = n_distinct(reprseq.pair[mosaicism.type == "recent seq. mosaicism 70% ident"]),
    num.reprseq.any.mossaic = n_distinct(reprseq.pair),
    num.reprseq.pairs.domain = n_distinct(reprseq.pair[mosaicism.type == "ecod"]),
    num.reprseq.pairs.seq = n_distinct(reprseq.pair[mosaicism.type == "seq"]),
    ) %>%
  ungroup() %>%
  #filter(num.reprseq.any.mossaic >= MIN.NUM.REPRSEQ.PAIRS.FOR.BETWEEN.FUNCTION.MOSAICISM) %>%
  #mutate(mosaicism.type = if_else(num.reprseq.pairs.domain > 0 & num.reprseq.pairs.seq > 0, "both",
  #                                  if_else(num.reprseq.pairs.domain > 0, "ecod","seq"))) %>%
  full_join(fold.type.mosaicism %>% 
              select(annotation.index.from = annotation.index.x, annotation.index.to = annotation.index.y, num.mosaic.fold.combination.pairs = num.mosaic.pairs)) %>%
  filter(annotation.index.from <= annotation.index.to) %>% 
  filter(!(annotation.index.from %in% unspecific.annotation.indices) & !(annotation.index.to %in% unspecific.annotation.indices)) %>%
  filter(!is.na(num.mosaic.fold.combination.pairs) & num.mosaic.fold.combination.pairs > 3)   


mosaic.network = mosaic.pairs.from.included.funct %>%
  select(from.index = annotation.index.from, 
         to.index = annotation.index.to, 
         #dataset = mosaicism.type, 
         num.reprseq.any.mossaic, 
         num.families.any.mosaic, 
         num.reprseq.pairs.domain, 
         num.reprseq.pairs.seq, 
         num.family.pairs.domain, 
         num.family.pairs.seq,
         num.family.pairs.recent.mosaic.seq.pident30,
         num.reprseq.pairs.recent.mosaic.seq.pident30,
         num.family.pairs.recent.mosaic.seq.pident50,
         num.reprseq.pairs.recent.mosaic.seq.pident50,
         num.family.pairs.recent.mosaic.seq.pident70,
         num.reprseq.pairs.recent.mosaic.seq.pident70,
         num.mosaic.fold.combination.pairs) 

  
#diversity.data1 %>%
data.table::fwrite(mosaic.network, sprintf("%sFigure3/3a_mosaic.network.txt", OUTPUT.FIGURES.PATH))


mosaic.metadata = included.category.sizes %>%
  select(node = annotation.index, func = annotation, class = category, num.families, num.proteins) %>%
  left_join(diversity.data %>% select(node = annotation.index, num.topology.combinations))
data.table::fwrite(mosaic.metadata, sprintf("%sFigure3/3a_mosaic.metadata.txt", OUTPUT.FIGURES.PATH), sep = "\t")
```

```{r}

#dd=Annotate_Graph(mosaic.network, 
#                               edge.colors = NULL, #mosaic.network %>% distinct(from.index, to.index,dataset),
#                               node.color.map = mosaic.metadata %>% select(node, category = class) %>% distinct() %>% left_join(color.map.df) %>% distinct(node, color),
#                               node.size.map = NULL,
#                               line.color.variable = 'num.reprseq.any.mossaic',
#                               vertex.size = 1)
#VisualiseGGnetwork(dd, vertex.size = 0.5, cex = 0.1, label_size = 3, layout = igraph::nicely(), main = "")



mosaic.pairs.by.ecod.and.seq.from.top.diverse = ecod.mosaic.pairs.from.included.funct %>%
  filter(annotation.index.from == annotation.index.from) %>%
  rename(annotation.index = annotation.index.from) %>%
  left_join(seq.mosaic.pairs.from.included.funct %>% 
              select(reprseq.pair, mosaic.pident10, mosaic.pident30, mosaic.pident50, mosaic.pident70, mosaic.pident90), 
            by = c("reprseq.pair")) %>%
  group_by(reprseq.pair, annotation.index) %>%
  summarise(mosaic.pident10 = any(mosaic.pident10),
            mosaic.pident30 = any(mosaic.pident30),
            mosaic.pident50 = any(mosaic.pident50),
            mosaic.pident70 = any(mosaic.pident70),
            mosaic.pident90 = any(mosaic.pident90)) %>%
  right_join(diversity.data %>% filter(top.quantile)) %>%
  group_by(annotation.index, annotation, category) %>%
  # here we have one row per reprseq pair
  summarise(num.reprseq.pairs.mosaic.by.domain  = n(),
            num.reprseq.pairs.mosaic.by.domain.and.pident10 = sum(mosaic.pident10, na.rm = TRUE),
            num.reprseq.pairs.mosaic.by.domain.and.pident30 = sum(mosaic.pident30, na.rm = TRUE),
            num.reprseq.pairs.mosaic.by.domain.and.pident50 = sum(mosaic.pident50, na.rm = TRUE),
            num.reprseq.pairs.mosaic.by.domain.and.pident70 = sum(mosaic.pident70, na.rm = TRUE),
            num.reprseq.pairs.mosaic.by.domain.and.pident90 = sum(mosaic.pident90, na.rm = TRUE)) %>%
  ungroup() %>%
  mutate(prop.reprseq.pairs.mosaic.by.domain.that.have.pident10 = num.reprseq.pairs.mosaic.by.domain.and.pident10/num.reprseq.pairs.mosaic.by.domain,
         prop.reprseq.pairs.mosaic.by.domain.that.have.pident30 = num.reprseq.pairs.mosaic.by.domain.and.pident30/num.reprseq.pairs.mosaic.by.domain,
         prop.reprseq.pairs.mosaic.by.domain.that.have.pident50 = num.reprseq.pairs.mosaic.by.domain.and.pident50/num.reprseq.pairs.mosaic.by.domain,
         prop.reprseq.pairs.mosaic.by.domain.that.have.pident70 = num.reprseq.pairs.mosaic.by.domain.and.pident70/num.reprseq.pairs.mosaic.by.domain,
         prop.reprseq.pairs.mosaic.by.domain.that.have.pident90 = num.reprseq.pairs.mosaic.by.domain.and.pident90/num.reprseq.pairs.mosaic.by.domain,)

data.table::fwrite(mosaic.pairs.by.ecod.and.seq.from.top.diverse, sprintf("%sFigure5/mosaic.pairs.by.ecod.and.seq.from.top.diverse.txt", OUTPUT.FIGURES.PATH))
  
```


# new Fig 3c
```{r}

protein.similarity.data.no.self.hits = protein.similarity.data %>%
  filter(qname != sname & prob > 0 & scov > 0 & qcov > 0)
all.proteins.with.hit.to.anything = unique(c(protein.similarity.data.no.self.hits$qname, protein.similarity.data.no.self.hits$sname))
all.proteins.with.domain.mosaic.signal = unique(c(qname.pairs.sharing.t.and.disshare.x.distinct$qname.x, qname.pairs.sharing.t.and.disshare.x.distinct$qname.y))
all.proteins.with.mosaic.signal.pident30 = unique(c(seq.mosaic.pairs.all.pident %>% filter(mosaic.pident30) %>% pull(qname),
                                                seq.mosaic.pairs.all.pident %>% filter(mosaic.pident30) %>% pull(sname)))
all.proteins.with.mosaic.signal.pident50 = unique(c(seq.mosaic.pairs.all.pident %>% filter(mosaic.pident50) %>% pull(qname),
                                                seq.mosaic.pairs.all.pident %>% filter(mosaic.pident50) %>% pull(sname)))
all.proteins.with.mosaic.signal.pident70 = unique(c(seq.mosaic.pairs.all.pident %>% filter(mosaic.pident70) %>% pull(qname),
                                                seq.mosaic.pairs.all.pident %>% filter(mosaic.pident70) %>% pull(sname)))
all.proteins.with.mosaic.signal.pident90 = unique(c(seq.mosaic.pairs.all.pident %>% filter(mosaic.pident90) %>% pull(qname),
                                                seq.mosaic.pairs.all.pident %>% filter(mosaic.pident90) %>% pull(sname)))
all.proteins.with.seq.mosaic.signal = unique(c(seq.mosaic.pairs.all.pident %>% filter(mosaic) %>% pull(qname),
                                                seq.mosaic.pairs.all.pident %>% filter(mosaic) %>% pull(sname)))
all.proteins.with.any.mosaic.signal = unique(c(all.proteins.with.domain.mosaic.signal, all.proteins.with.seq.mosaic.signal))

# qname.pairs.sharing.t.and.disshare.x.distinct only have annotationd from included categories
all.proteins.with.domain.mosaic.signal.within.specific.class = unique(c(
  qname.pairs.sharing.t.and.disshare.x.distinct %>% 
   filter(!is.na(annotation.index.from) & !is.na(annotation.index.to)) %>%
  filter(!(annotation.index.from %in% unspecific.annotation.indices | annotation.index.to %in% unspecific.annotation.indices)) %>%
  filter(annotation.index.from == annotation.index.to) %>%
  pull(qname.x),
  qname.pairs.sharing.t.and.disshare.x.distinct %>% 
  filter(!is.na(annotation.index.from) & !is.na(annotation.index.to)) %>%
  filter(!(annotation.index.from %in% unspecific.annotation.indices | annotation.index.to %in% unspecific.annotation.indices)) %>%
  filter(annotation.index.from == annotation.index.to) %>%
  pull(qname.y)))

all.proteins.with.domain.mosaic.signal.to.other.specific.class = unique(c(
  qname.pairs.sharing.t.and.disshare.x.distinct %>% 
  filter(!is.na(annotation.index.from) & !is.na(annotation.index.to)) %>%
  filter(!(annotation.index.from %in% unspecific.annotation.indices | annotation.index.to %in% unspecific.annotation.indices)) %>%
  filter(annotation.index.from != annotation.index.to) %>%
  pull(qname.x),
  qname.pairs.sharing.t.and.disshare.x.distinct %>% 
  filter(!is.na(annotation.index.from) & !is.na(annotation.index.to)) %>%
  filter(!(annotation.index.from %in% unspecific.annotation.indices | annotation.index.to %in% unspecific.annotation.indices)) %>%
  filter(annotation.index.from != annotation.index.to) %>%
  pull(qname.y)))

all.proteins.with.domain.mosaic.signal.to.unknown.excluded.or.unspecific.class = unique(c(
  qname.pairs.sharing.t.and.disshare.x.distinct %>% 
  filter(is.na(annotation.index.to) | annotation.index.to %in% unspecific.annotation.indices) %>%
  pull(qname.x),
  qname.pairs.sharing.t.and.disshare.x.distinct %>% 
  filter(is.na(annotation.index.from) | annotation.index.from %in% unspecific.annotation.indices) %>%
  pull(qname.y)))

# look at the annotations at relaxed threhsold
all.proteins = reprseq.function.status %>%
  mutate(mosaic = qname %in% all.proteins.with.any.mosaic.signal,
         mosaic.domain = qname %in% all.proteins.with.domain.mosaic.signal,
         mosaic.seq = qname %in% all.proteins.with.seq.mosaic.signal,
         mosaic.seq.pident30 = qname %in% all.proteins.with.mosaic.signal.pident30,
         mosaic.seq.pident50 = qname %in% all.proteins.with.mosaic.signal.pident50,
         mosaic.seq.pident70 = qname %in% all.proteins.with.mosaic.signal.pident70,
         mosaic.seq.pident90 = qname %in% all.proteins.with.mosaic.signal.pident90,
         has.hit.to.anything.in.the.data = qname %in% all.proteins.with.hit.to.anything,
         mosaic.domain.within.specific.class = qname %in% all.proteins.with.domain.mosaic.signal.within.specific.class,
         mosaic.domain.to.other.specific.class = qname %in% all.proteins.with.domain.mosaic.signal.to.other.specific.class,
         mosaic.domain.to.unknown.excluded.or.unspecific.class = qname %in% all.proteins.with.domain.mosaic.signal.to.unknown.excluded.or.unspecific.class)


prop.mosaic.and.multidomain.data = all.proteins %>%
  left_join(surely.annotated.proteins.from.included.categories) %>%
  left_join(num.domain.combinations.per.qname %>% select(qname, num.x, num.h), by = "qname") %>%
  left_join(included.category.sizes) %>%
  mutate(num.x = if_else(is.na(num.x), 0, as.numeric(num.x)),
         num.h = if_else(is.na(num.h), 0, as.numeric(num.h))) %>%
  distinct(qname,annotation.index, annotation, category, num.x,num.h, mosaic.domain, num.proteins) %>%
  group_by(annotation.index, annotation, category, num.proteins) %>%
  summarise(num.with.multi.x.domains = sum(num.x > 1),
            num.with.multi.h.domains = sum(num.h > 1),
            num.with.domain.mosaic.signal = sum(mosaic.domain)) %>%
  ungroup() %>%
  mutate(prop.with.multi.x.domains = num.with.multi.x.domains/num.proteins,
         prop.with.multi.h.domains = num.with.multi.h.domains/num.proteins,
         prop.with.domain.mosaic.signal = num.with.domain.mosaic.signal/num.proteins)
  

proteins.function.data = all.proteins %>%
  # filter(has.hit.to.anything.in.the.data & has.unknown.function) %>%
  left_join(families) %>%
  left_join(relaxely.annotated.proteins.including.multi.annot %>% select(qname, annotation, category, include)) %>%
  filter(include) %>%
  mutate(annotation = if_else(is.na(annotation), "unknown", annotation)) %>%
  mutate(category = if_else(is.na(category), "unknown", category))  %>%
  group_by(qname, family, mosaic) %>%
  mutate(num.hits = n_distinct(annotation[annotation != "unknown"]),
         annotation.weight = 1/num.hits,
         is.unannotated = if_else(num.hits == 0,1,0)) %>%
  ungroup() %>%
  mutate(this.annot.weight = if_else(annotation != "unknown", annotation.weight, is.unannotated))


table.seq.functions.new = proteins.function.data %>%
  #group_by(family, annotation, category, mosaic) %>%
  #summarise(annotation.weight = mean(this.annot.weight)) %>%
  group_by(annotation, category, mosaic) %>%
  summarise(weight = sum(annotation.weight)) %>%
  ungroup() %>%
  arrange(desc(weight)) %>%
  mutate(mosaic = if_else(mosaic, "Mosaic", "Not mosaic")) %>%
  filter(annotation != "unknown") %>%
  group_by(mosaic) %>%
  mutate(total.weight = sum(weight)) %>%
  ungroup() %>%
  mutate(freq = weight/total.weight) 

data.table::fwrite(table.seq.functions.new, sprintf("%sFigure3/table.seq.functions.low.cov.txt", OUTPUT.FIGURES.PATH))

mosaic.domain.classes = all.proteins %>%
  left_join(surely.annotated.proteins.from.included.categories) %>%
  inner_join(included.category.sizes) %>%
  group_by(annotation.index, annotation, category, num.proteins, num.families) %>%
  summarise(num.prot.mosaic.domain.within.specific.class = sum(mosaic.domain.within.specific.class),
            num.prot.mosaic.domain.to.other.specific.class = sum(mosaic.domain.to.other.specific.class),
            num.prot.mosaic.domain.to.unknown.excluded.or.unspecific.class = sum(mosaic.domain.to.unknown.excluded.or.unspecific.class),
            num.prot.mosaic.domain = sum(mosaic.domain),
            num.families.mosaic.domain = n_distinct(family[mosaic.domain]),
            num.prot.mosaic.domain.only.within.class.or.unspecific = 
              sum(mosaic.domain.within.specific.class & !mosaic.domain.to.other.specific.class),
            num.prot.mosaic.domain.any.within.class = 
              sum(mosaic.domain.within.specific.class),
            num.prot.mosaic.domain.only.to.other.specific.class = 
              sum(!mosaic.domain.within.specific.class & mosaic.domain.to.other.specific.class),
            num.prot.mosaic.domain.within.and.to.other.specific.class = 
              sum(mosaic.domain.within.specific.class & mosaic.domain.to.other.specific.class),
            num.prot.mosaic.domain.only.to.any.speciific.class =
              sum(!mosaic.domain.within.specific.class & !mosaic.domain.to.other.specific.class & mosaic.domain),          
            num.prot.mosaic.domain.only.to.unpecific.or.excluded.class =
              sum(!mosaic.domain.within.specific.class & !mosaic.domain.to.other.specific.class & mosaic.domain.to.unknown.excluded.or.unspecific.class),
            num.prot.mosaic.to.anything.with.specific.annotation = sum(mosaic.domain.within.specific.class | mosaic.domain.to.other.specific.class),
            num.families.mosaic.to.anything.with.specific.annotation = n_distinct(family[mosaic.domain.within.specific.class | mosaic.domain.to.other.specific.class]),
            num.prot.mosaic.seq.or.domain = sum(mosaic),
             num.families.mosaic.seq.or.domain = n_distinct(family[mosaic]),
            num.mprot.mosaic.seq = sum(mosaic.seq)) %>%
  ungroup() %>%
  mutate(prop.prot.mosaic.domain.within.any.pair.in.specific.class = num.prot.mosaic.domain.within.specific.class/num.prot.mosaic.domain,
         prop.prot.mosaic.domain.within.any.pair.to.other.specific.class = num.prot.mosaic.domain.to.other.specific.class/num.prot.mosaic.domain,
         prop.prot.mosaic.domain.within.any.pair.to.unknown.excluded.or.unspecific.class = num.prot.mosaic.domain.to.unknown.excluded.or.unspecific.class/num.prot.mosaic.domain) %>%
  mutate(num.families.nonmosaic.domain = num.families - num.families.mosaic.domain,
         num.families.nonmosaic.seq.or.domain = num.families - num.families.mosaic.seq.or.domain) %>%
  mutate(odds.seq.to.ecod = NA,#odds.seq.to.ecod = (num.families.mosaic.seq.or.domain/num.families.nonmosaic.seq.or.domain)/(num.families.mosaic.domain/num.families.nonmosaic.domain),
         pval = NA) 


  for (this.annotation.index in unique(mosaic.domain.classes$annotation.index)) {
    this.annot.stats = mosaic.domain.classes %>% filter(annotation.index == this.annotation.index)
    M1 = matrix(0, nrow=2, ncol = 2)
    M1[1,1] <- this.annot.stats$num.families.mosaic.seq.or.domain
    M1[1,2] = this.annot.stats$num.families.nonmosaic.seq.or.domain
    M1[2,1] = this.annot.stats$num.families.mosaic.domain 
    M1[2,2] = this.annot.stats$num.families.nonmosaic.domain
    fisher_test_res = fisher.test(M1, alternative = "greater")
    mosaic.domain.classes$pval[which(mosaic.domain.classes$annotation.index == this.annotation.index)] = fisher_test_res$p.value
    mosaic.domain.classes$odds.seq.to.ecod[which(mosaic.domain.classes$annotation.index == this.annotation.index)] =fisher_test_res$estimate
    
  }
 mosaic.domain.classes$pval.corr = mosaic.domain.classes$pval*length(unique(mosaic.domain.classes$annotation.index))

data.table::fwrite(mosaic.domain.classes, sprintf("%sFigure5/mosaic.domain.classes.txt", OUTPUT.FIGURES.PATH))


 
```









```{r}
annotated.qnames = surely.annotated.ecod.domains.high.confidence %>% distinct(qname) %>% pull(qname)
ecod.hits.data.table = surely.annotated.ecod.domains.high.confidence %>% 
        distinct(qname, annotation, qlength, qstart, qend, domain = t_id, family) %>% as.data.table()
ecod.hits.data.table = setDT(ecod.hits.data.table, key = 'qname')
seq2ecod.cov.list <- lapply(1:length(annotated.qnames), function(reprseq.index){
  this.reprseq <- annotated.qnames[reprseq.index]
  this.reprseq.hits <- ecod.hits.data.table[.(this.reprseq)]
  this.reprseq.domain.pos.list <- lapply(1:nrow(this.reprseq.hits), function(k) this.reprseq.hits$qstart[k]:this.reprseq.hits$qend[k])
  this.reprseq.domain.pos.unique <- unique(unlist(this.reprseq.domain.pos.list))
  return(ecod.pos = length(this.reprseq.domain.pos.unique))
})
seq2ecod.cov.df = data.frame(qname = annotated.qnames)
seq2ecod.cov.df$ecod.pos = seq2ecod.cov.list %>% unlist()
seq2ecod.cov.df = seq2ecod.cov.df %>%
  right_join(repr.seq.lengths %>% select(qname = name, length)) %>%
  inner_join(surely.annotated.proteins.from.included.categories) %>%
    rowwise() %>%
    # those that were not in ecod data are all undetected
  mutate(num.detected.positions = if_else(is.na(ecod.pos), as.numeric(0), as.numeric(ecod.pos)))

ecod.coverage.data = seq2ecod.cov.df %>%
  group_by(annotation.index, annotation, category) %>%
  summarise(num.detected.positions = sum(num.detected.positions),
            num.all.positions = sum(length)) %>%
  ungroup() %>%
  mutate(prop.ecod.annotated.positions = num.detected.positions/num.all.positions) %>%
  select(annotation.index, annotation, category, prop.ecod.annotated.positions)

classes.with.few.ecod.annotated.prots.new = ecod.coverage.data %>%
  left_join(diversity.data %>% select(annotation, category, prop.proteins.with.ecod.hit)) %>%
  rename(`prot. with any ECOD hit` = prop.proteins.with.ecod.hit,
         `proteome covered by ECOD hits` =prop.ecod.annotated.positions) %>%
  arrange(`prot. with any ECOD hit`)

data.table::fwrite(classes.with.few.ecod.annotated.prots.new, sprintf("%sFigure_Supplementary/FigS10_poor_ecod/classes.with.few.ecod.annotated.txt", OUTPUT.FIGURES.PATH))
```

```{r}

prop.mosaic.and.mean.length.per.category = all.proteins %>%  
  inner_join(surely.annotated.proteins.from.included.categories) %>%
  inner_join(repr.seq.lengths %>% select(qname = name, length)) %>%
  group_by(annotation.index, annotation, category) %>%
  summarise(prop.mosaic = n_distinct(qname[mosaic])/n_distinct(qname),
            mean.length = mean(length)) 


diversity.data2 = diversity.data %>%
  left_join(mosaic.pairs.from.included.funct %>% 
              filter(annotation.index.from == annotation.index.to) %>% 
              select(annotation.index = annotation.index.from, num.reprseq.pairs.domain)) %>%
  inner_join(prop.mosaic.and.mean.length.per.category) %>%
  left_join(prop.mosaic.and.multidomain.data) %>%
  left_join(classes.with.few.ecod.annotated.prots.new) %>%
  mutate(ecod.coverage.type = if_else(`prot. with any ECOD hit` < MIN.PROP.PROTS.WITH.ECOD.HIT, 
                                      "few prots. with >=1 detected domains",
                                      if_else(prop.with.multi.x.domains < MIN.PROP.PROTS.WITH.ECOD.HIT, 
                                              "few prots. with >=2 detected domains", "")))

   data.table::fwrite(diversity.data2, 
                      sprintf("%sFigure_Supplementary/FigS9_ecod_hist/Diversity_vs_mosaicism.txt", OUTPUT.FIGURES.PATH))        
```


```{r, fig.width = 8, fig.width = 12}
Get.Mosaicism.Odds.Ratio.Data = function(all.proteins, annotations, diversity.data) {
  all.proteins.with.any.relaxed.function = all.proteins %>%
    distinct(qname, mosaic) %>%
    left_join(annotations) %>%
    filter(!is.na(annotation.index)) %>%
    group_by(annotation.index) %>%
    mutate(num.proteins.in.this.annotation = n_distinct(qname)) %>%
    ungroup() 
  
  num.proteins.per.osaicism.data = all.proteins.with.any.relaxed.function %>%
    group_by(mosaic) %>%
    summarise(num.proteins = n_distinct(qname)) %>%
    ungroup()
    
  annot.stats.per.category.and.mosaicism = all.proteins.with.any.relaxed.function %>%
    inner_join(diversity.data) %>%
    group_by(annotation.index, mosaic, annotation, category) %>%
    summarise(num.proteins.with.this.annot = n_distinct(qname)) %>%
    ungroup() %>%
    left_join(num.proteins.per.osaicism.data) %>%
    mutate(num.proteins.with.different.annot = num.proteins - num.proteins.with.this.annot,
           prop.proteins.has.this.annot = num.proteins.with.this.annot/num.proteins) %>%
    tidyr::complete(annotation.index, mosaic, fill = list(num.proteins = 0, num.proteins.with.this.annot = 0, num.proteins.with.different.annot = 0, prop.proteins.has.this.annot = NA))
  
  
  
  annot.stats.per.category.and.mosaicism$pval = NA
  annot.stats.per.category.and.mosaicism$odds.ratio = NA
  for (this.annotation.index in unique(annot.stats.per.category.and.mosaicism$annotation.index)) {
    this.annot.stats = annot.stats.per.category.and.mosaicism %>% filter(annotation.index == this.annotation.index)
    M1 = this.annot.stats %>% select(mosaic, num.proteins.with.this.annot, num.proteins.with.different.annot) %>% select(num.proteins.with.this.annot, num.proteins.with.different.annot) %>% as.data.table()
  #  annot.stats.per.category.and.mosaicism$pval[which(annot.stats.per.category.and.mosaicism$annotation.index == this.annotation.index)] = chisq.test(M1)$p.value
      annot.stats.per.category.and.mosaicism$pval[which(annot.stats.per.category.and.mosaicism$annotation.index == this.annotation.index)] = fisher.test(M1)$p.value
    annot.stats.per.category.and.mosaicism$odds.ratio[which(annot.stats.per.category.and.mosaicism$annotation.index == this.annotation.index)] = (M1$num.proteins.with.this.annot[2]/M1$num.proteins.with.different.annot[2])/(M1$num.proteins.with.this.annot[1]/M1$num.proteins.with.different.annot[1])
  }
  num.comparisons = n_distinct(annot.stats.per.category.and.mosaicism$annotation.index)
  annot.stats.per.category.and.mosaicism = annot.stats.per.category.and.mosaicism %>%
    mutate(corrected.pval = pval*num.comparisons)
   return(annot.stats.per.category.and.mosaicism)
}

annot.stats.per.category.and.mosaicism = Get.Mosaicism.Odds.Ratio.Data(
  all.proteins = all.proteins, 
  annotations = relaxely.annotated.proteins.including.multi.annot, 
  diversity.data = diversity.data) %>%
  left_join(diversity.data2 %>% select(annotation.index, data.type, ecod.coverage.type))

annot.stats.per.category.and.mosaicism.by.domain = Get.Mosaicism.Odds.Ratio.Data(
  all.proteins = all.proteins %>% select(qname, mosaic = mosaic.domain), 
  annotations = relaxely.annotated.proteins.including.multi.annot, 
  diversity.data = diversity.data) %>%
  left_join(diversity.data2 %>% select(annotation.index, data.type, ecod.coverage.type))

annot.stats.per.category.and.mosaicism.by.seq = Get.Mosaicism.Odds.Ratio.Data(
  all.proteins = all.proteins %>% select(qname, mosaic = mosaic.seq), 
  annotations = relaxely.annotated.proteins.including.multi.annot, 
  diversity.data = diversity.data)%>%
  left_join(diversity.data2 %>% select(annotation.index, data.type, ecod.coverage.type))

data.table::fwrite(annot.stats.per.category.and.mosaicism, file = sprintf("%sFigure6/unknown_functions_mosaic_vs_nonmosaic.txt", OUTPUT.FIGURES.PATH))
data.table::fwrite(annot.stats.per.category.and.mosaicism.by.domain, file = sprintf("%sFigure6/unknown_functions_mosaic_vs_nonmosaic_by_domain.txt", OUTPUT.FIGURES.PATH))
data.table::fwrite(annot.stats.per.category.and.mosaicism.by.seq, file = sprintf("%sFigure6/unknown_functions_mosaic_vs_nonmosaic_by_seq.txt", OUTPUT.FIGURES.PATH))


annot.stats.per.category.and.mosaicism.sure.annot = Get.Mosaicism.Odds.Ratio.Data(
  all.proteins = all.proteins, 
  annotations = surely.annotated.proteins, 
  diversity.data = diversity.data) %>%
  left_join(diversity.data2 %>% select(annotation.index, data.type, ecod.coverage.type))

annot.stats.per.category.and.mosaicism.by.domain.sure.annot = Get.Mosaicism.Odds.Ratio.Data(
  all.proteins = all.proteins %>% select(qname, mosaic = mosaic.domain), 
  annotations = surely.annotated.proteins, 
  diversity.data = diversity.data) %>%
  left_join(diversity.data2 %>% select(annotation.index, data.type, ecod.coverage.type))

annot.stats.per.category.and.mosaicism.by.seq.sure.annot = Get.Mosaicism.Odds.Ratio.Data(
  all.proteins = all.proteins %>% select(qname, mosaic = mosaic.seq), 
  annotations = surely.annotated.proteins, 
  diversity.data = diversity.data)%>%
  left_join(diversity.data2 %>% select(annotation.index, data.type, ecod.coverage.type))

data.table::fwrite(annot.stats.per.category.and.mosaicism.sure.annot, file = sprintf("%sFigure6/known_functions_mosaic_vs_nonmosaic.txt", OUTPUT.FIGURES.PATH))
data.table::fwrite(annot.stats.per.category.and.mosaicism.by.domain.sure.annot, file = sprintf("%sFigure6/known_functions_mosaic_vs_nonmosaic_by_domain.txt", OUTPUT.FIGURES.PATH))
data.table::fwrite(annot.stats.per.category.and.mosaicism.by.seq.sure.annot, file = sprintf("%sFigure6/known_functions_mosaic_vs_nonmosaic_by_seq.txt", OUTPUT.FIGURES.PATH))



# test: domain Major cpasid protein GP7 is present in 121 proteins, 66 of them are known annotation (major capsid or structural protein)
# additionally 17 had a function that was not defined uniquely (capsid & structural protein)
# 116 of them were mosaic!
# 11 has sequence mosaicism and 111 had domain mosaicism!
# all 111 proteins with domain mosaicism had two domains: Major capsid protein gp5 & Major capsid protein GP7 (fold 3555.1.1, unknown X name)
# in the mosaic pairs they occured in the following combinations with 11 capsid proteins taht had a different fold architecture: 
# 1) Major capsid protein gp5 & Major capsid protein GP7  vs  Major capsid protein gp5 & Carbohydrate binding domain (probably also a capsid)
# 2) Major capsid protein gp5 & Major capsid protein GP7  vs  Major capsid protein gp5 & craddle loop barrel (probably also a capsid)
# 3) Major capsid protein gp5 & Major capsid protein GP7  vs  Major capsid protein gp5 & ClpP/crotonase (probably also a capsid)

```



# Domains within known and unknown proteins
```{r}
GetDomainsWithinMosaicProteins = function(all.proteins, families, ecod.domains.hits, ecod_id_to_names_map) {

    domains.within.proteins = all.proteins %>%
      # filter(has.hit.to.anything.in.the.data) %>%
      left_join(families) %>%
      inner_join(ecod.domains.hits, by = "qname") %>%
      group_by(mosaic) %>%
      mutate(total.num.families.with.any.ecod.hit = n_distinct(family),
             total.num.proteins.with.any.ecod.hit = n_distinct(qname)) %>%
      ungroup() %>%
      distinct(qname, family, t_id, t_name, mosaic, total.num.families.with.any.ecod.hit, total.num.proteins.with.any.ecod.hit) %>%
      group_by(t_id, t_name, mosaic, total.num.families.with.any.ecod.hit, total.num.proteins.with.any.ecod.hit) %>%
      summarise(num.families = n_distinct(family),
                num.proteins.with.this.ecod = n_distinct(qname)) %>%
      ungroup() %>%
      mutate(freq = num.families/total.num.families.with.any.ecod.hit,
             freq.q = num.proteins.with.this.ecod/total.num.proteins.with.any.ecod.hit,
             num.proteins.other.ecod.hits = total.num.proteins.with.any.ecod.hit - num.proteins.with.this.ecod) %>%
      tidyr::complete(t_id, mosaic, fill = list(num.proteins.with.this.ecod = 0, num.proteins.other.ecod.hits = 0))
    
    
    domains.within.proteins$h.index <- sapply(1:nrow(domains.within.proteins), function(index){
      this.topology <- domains.within.proteins$t_id[index]
      this.topology.h.index <- strsplit(this.topology, ".", fixed = T)[[1]]
      this.topology.h.index <- this.topology.h.index[-length(this.topology.h.index)]
      this.topology.h.index <- as.character(paste(this.topology.h.index, collapse = "."))
    })
    
    
    
    domains.within.proteins$pval = NA
    domains.within.proteins$odds.ratio = NA
    for (this.t.id in unique(domains.within.proteins$t_id)) {
      this.domain.stats = domains.within.proteins %>% filter(t_id == this.t.id)
      M1 = this.domain.stats %>% select(mosaic, num.proteins.with.this.ecod, num.proteins.other.ecod.hits) %>% select(num.proteins.with.this.ecod, num.proteins.other.ecod.hits) %>% as.data.table()
      #domains.within.proteins$pval[which(domains.within.proteins$t_id == this.t.id)] = chisq.test(M1)$p.value
      domains.within.proteins$pval[which(domains.within.proteins$t_id == this.t.id)] = fisher.test(M1)$p.value
      domains.within.proteins$odds.ratio[which(domains.within.proteins$t_id == this.t.id)] = (M1$num.proteins.with.this.ecod[2]/M1$num.proteins.other.ecod.hits[2])/(M1$num.proteins.with.this.ecod[1]/M1$num.proteins.other.ecod.hits[1])
    }
    
    num.comparisons = n_distinct(domains.within.proteins$t_id)
    domains.within.proteins = domains.within.proteins %>%
      mutate(corrected.pval = pval*num.comparisons)
    
    domains.within.proteins = domains.within.proteins %>%
        left_join(ecod_id_to_names_map %>% filter(level == "H") %>% select(h.index = id, h_name = name)) %>%
      mutate(mosaic = if_else(mosaic, "Mosaic", "Not mosaic"))
    return(domains.within.proteins)
}

domains.within.proteins = GetDomainsWithinMosaicProteins(all.proteins %>% select(qname, mosaic), families, ecod.domains.hits, ecod_id_to_names_map)
data.table::fwrite(domains.within.proteins, file = sprintf("%sFigure3/domains-in-known-and-unknown.txt", OUTPUT.FIGURES.PATH))

domains.within.proteins.with.domain.mosaicism = GetDomainsWithinMosaicProteins(all.proteins %>% select(qname, mosaic = mosaic.domain), families, ecod.domains.hits, ecod_id_to_names_map)
data.table::fwrite(domains.within.proteins.with.domain.mosaicism, file = sprintf("%sAdditional_Figures/domains-in-known-and-unknown-domain-mosaic.txt", OUTPUT.FIGURES.PATH))

domains.within.proteins.with.domain.mosaicism.and.pident30 = GetDomainsWithinMosaicProteins(
  all.proteins %>% 
     mutate(mosaic = mosaic.domain & mosaic.seq.pident30) %>%
     select(qname, mosaic), 
 families, ecod.domains.hits, ecod_id_to_names_map)
data.table::fwrite(domains.within.proteins.with.domain.mosaicism.and.pident30, file = sprintf("%sAdditional_Figures/domains-in-known-and-unknown-domain-mosaic-and-pident30.txt", OUTPUT.FIGURES.PATH))

domains.within.proteins.with.domain.mosaicism.and.pident50 = GetDomainsWithinMosaicProteins(
  all.proteins %>% 
     mutate(mosaic = mosaic.domain & mosaic.seq.pident50) %>%
     select(qname, mosaic), 
 families, ecod.domains.hits, ecod_id_to_names_map)
data.table::fwrite(domains.within.proteins.with.domain.mosaicism.and.pident50, file = sprintf("%sAdditional_Figures/domains-in-known-and-unknown-domain-mosaic-and-pident50.txt", OUTPUT.FIGURES.PATH))

domains.within.proteins.with.domain.mosaicism.and.pident70 = GetDomainsWithinMosaicProteins(
  all.proteins %>% 
     mutate(mosaic = mosaic.domain & mosaic.seq.pident70) %>%
     select(qname, mosaic), 
 families, ecod.domains.hits, ecod_id_to_names_map)
data.table::fwrite(domains.within.proteins.with.domain.mosaicism.and.pident70, file = sprintf("%sAdditional_Figures/domains-in-known-and-unknown-domain-mosaic-and-pident70.txt", OUTPUT.FIGURES.PATH))


domains.within.proteins.with.seq.mosaicism = GetDomainsWithinMosaicProteins(all.proteins %>% select(qname, mosaic = mosaic.seq), families, ecod.domains.hits, ecod_id_to_names_map)
data.table::fwrite(domains.within.proteins.with.domain.mosaicism, file = sprintf("%sAdditional_Figures/domains-in-known-and-unknown-seq-mosaic.txt", OUTPUT.FIGURES.PATH))

```

```{r}
# add some stats to diversity data
# mean seq length length and prop mosaic by sequence or domain

excel.filename = "Statistics.xlsx"
data.table::fwrite(diversity.data2 %>%
  filter(data.type == "top.diversity") %>%
    select(mean.num.t, prop.mosaic.pairs.sequence.thr0p1, prop.mosaic.pairs.sequence.thr0p3, prop.mosaic.pairs.sequence.thr0p5, 
           prob.mosaicism, num.mosaic.pairs, annotation, category, 
           prop.mosaic, num.reprseq.pairs.domain, num.theoretical.fold.type.pairs.with.multi.x,
           prop.with.multi.x.domains,
           mean.length,
           `proteome covered by ECOD hits`,
           `prot. with any ECOD hit`) %>%
  tidyr::gather(key = "proportion", value = "value", -c("annotation", "category")), 
  sprintf("%sFigure3/mosaicism.top.diverse.classes.txt", OUTPUT.FIGURES.PATH))


# mean length of sequences per data type
mean.seq.length.table = diversity.data2 %>% 
  filter(data.type %in% c("top.diversity", "low.diversity")) %>%
  group_by(data.type) %>% 
  summarise(mean.sequence.length = round(median(mean.length, na.rm = TRUE),1))
write.xlsx(x=mean.seq.length.table, 
           file = sprintf("%stables/%s", OUTPUT.FIGURES.PATH, excel.filename), 
           sheetName = "Mean seq length per data type", 
  col.names = TRUE, row.names = TRUE, append = TRUE)
test.longer.top.diverse = wilcox.test(x= diversity.data2 %>% filter(data.type == "top.diversity") %>% pull(mean.length), 
            y= diversity.data2 %>% filter(data.type == "low.diversity") %>% pull(mean.length), 
            paired = FALSE, alternative = "greater")
wilcox.test.data3 = data.frame(pval = test.longer.top.diverse$p.value,
                              statistic = test.longer.top.diverse$statistic)
write.xlsx(x=wilcox.test.data3, 
           file = sprintf("%stables/%s", OUTPUT.FIGURES.PATH, excel.filename), 
           sheetName = "Wilcox test seq length data type", 
  col.names = TRUE, row.names = TRUE, append = TRUE)





# test if top diversity functions are more mosaic than the low diversity oens
prop.mosaic.in.top.diverse.functions = diversity.data2 %>% filter(data.type == "top.diversity") %>% pull(prop.mosaic)
prop.mosaic.in.low.diverse.functions = diversity.data2 %>% filter(data.type == "low.diversity") %>% pull(prop.mosaic)
mean(prop.mosaic.in.top.diverse.functions)
mean(prop.mosaic.in.low.diverse.functions)
test.more.mosaic.top.diverse = wilcox.test(x= prop.mosaic.in.top.diverse.functions, 
            y= prop.mosaic.in.low.diverse.functions, 
            paired = FALSE, alternative = "greater")
wilcox.test.data = data.frame(pval = test.more.mosaic.top.diverse$p.value,
                              statistic = test.more.mosaic.top.diverse$statistic)
write.xlsx(x=wilcox.test.data, 
           file = sprintf("%stables/%s", OUTPUT.FIGURES.PATH, excel.filename), 
           sheetName = "Wilcox test prop. mosaic per data type", 
  col.names = TRUE, row.names = TRUE, append = TRUE)


# number of functional categories poorly and well annotated by ECOD
num.functional.categories.data = diversity.data %>% 
  distinct(annotation, category, poorly.ecod.annotated) %>%
  group_by(poorly.ecod.annotated) %>%
  summarise(n.classes = n())
write.xlsx(x=num.functional.categories.data, 
           file = sprintf("%stables/%s", OUTPUT.FIGURES.PATH, excel.filename), 
           sheetName = "Number of functional categories per data type", 
  col.names = TRUE, row.names = TRUE, append = TRUE)


# Correlation between num families/ num proteins and numbr of topology combinations
diversity.ok.ecod = diversity.data %>% filter(!poorly.ecod.annotated)
#ctest = cor.test(diversity.ok.ecod$num.families, diversity.ok.ecod$num.topology.combinations, method = "spearman")
ctest = cor.test(diversity.ok.ecod$num.proteins, diversity.ok.ecod$num.topology.combinations, method = "spearman")
ctesttable = data.frame(rho = ctest$estimate, p = ctest$p.value)
write.xlsx(x=ctesttable, 
           file = sprintf("%stables/%s", OUTPUT.FIGURES.PATH, excel.filename), 
           sheetName = "Correlation between two diversity measures", 
  col.names = TRUE, row.names = TRUE, append = TRUE)


# mean number of folds in low and top diversity functios
fold.number.per.data.type = diversity.data %>% 
  filter(data.type %in% c("top.diversity", "low.diversity")) %>%
  group_by(data.type) %>% 
  summarise(mean.num.folds = mean(mean.num.t, na.rm = TRUE))
write.xlsx(x=fold.number.per.data.type, 
           file = sprintf("%stables/%s", OUTPUT.FIGURES.PATH,excel.filename), 
           sheetName = "Av. number of folds per data type", 
  col.names = TRUE, row.names = TRUE, append = TRUE)

test.more.folds.top.diverse = wilcox.test(x= diversity.data %>% filter(data.type == "top.diversity") %>% pull(mean.num.t), 
            y= diversity.data %>% filter(data.type == "low.diversity") %>% pull(mean.num.t), 
            paired = FALSE, alternative = "greater")
wilcox.test.data2 = data.frame(pval = test.more.folds.top.diverse$p.value,
                              statistic = test.more.folds.top.diverse$statistic)
write.xlsx(x=wilcox.test.data2, 
           file = sprintf("%stables/%s", OUTPUT.FIGURES.PATH, excel.filename), 
           sheetName = "Wilcox test num folds per data type", 
  col.names = TRUE, row.names = TRUE, append = TRUE)




# some of the annotated may have hits to phrog classes with < 500 seq or may have multiple hits
num.proteins.with.unknown.function = all.proteins %>%
  group_by(has.unknown.function) %>%
  summarise(n = n())
write.xlsx(x=num.proteins.with.unknown.function, 
           file = sprintf("%stables/%s", OUTPUT.FIGURES.PATH, excel.filename), 
           sheetName = "num proteins.with unknown function", 
  col.names = TRUE, row.names = TRUE, append = TRUE)


aa = diversity.data %>% select(annotation, num.families, data.type) 
poorly_annotated_classes =  aa %>% filter(data.type== "poorly annotated") %>% arrange(desc(num.families))
write.xlsx(x=poorly_annotated_classes, 
           file = sprintf("%stables/%s", OUTPUT.FIGURES.PATH, excel.filename), 
           sheetName = " Poorly annotated classes", 
  col.names = TRUE, row.names = TRUE, append = TRUE)

n_distinct(surely.annotated.proteins.including.multi.annot %>% filter(num.annots == 1) %>% pull(qname))
n_distinct(repr.seq.lengths)

#mosaic.network %>% filter(num.reprseq.pairs.seq > 0 & num.reprseq.pairs.domain > 0 & from.index == to.index) %>% nrow()
#mosaic.network %>% filter(num.reprseq.pairs.seq > 0 | num.reprseq.pairs.domain > 0 & from.index == to.index) %>% nrow()
aa=all.proteins %>%
  inner_join(surely.annotated.proteins.from.included.categories) %>%
  group_by(annotation.index) %>%
  summarise(any.signal = any(mosaic.seq) | any(mosaic.domain),
            both.signal = any(mosaic.seq) & any(mosaic.domain))

bb = data.frame(
 any.signal = sum(aa$any.signal),
 both.signal = sum(aa$both.signal),
 all.classes = nrow(aa))
write.xlsx(x=bb, 
           file = sprintf("%stables/%s", OUTPUT.FIGURES.PATH, excel.filename), 
           sheetName = "num annots with mosaic signal", 
  col.names = TRUE, row.names = TRUE, append = TRUE)

```


#SHINY PLOTS: DOMAINS WITHIN FUNCTIONS
```{r, eval = TRUE}

#most.mosaic.annotations = diversity.data %>% filter(data.type == "top.diversity") %>% arrange(desc(prob.mosaicism)) %>% head(12) %>% pull(annotation.index)
most.mosaic.annotations = diversity.data  %>% filter(annotation %in% c("tail fiber","tail spike", "endolysin", "DNA polymerase")) %>% pull(annotation.index)
# too many domains if we include "tails"
tile.data.list = list()
i=0
domain_levels = c("T")
for (this.domain.level in domain_levels) {
this.domain.level.id = paste0(tolower(this.domain.level), "_id")

  for (this.annotation.index in most.mosaic.annotations) {
    domains.within.this.annotation = surely.annotated.ecod.domains.high.confidence %>% 
      filter(annotation.index == this.annotation.index) %>% 
        select(qname, annotation.index, annotation, qlength, qstart, qend, domain = this.domain.level.id, family)  %>% 
        group_by(qname) %>%
        mutate(all.domains = paste0(sort(unique(domain)), collapse = "_and_"),
               # this is just used for ordering
               first.domain = domain[qstart == min(qstart)][1]) %>%
        ungroup()
    
    if (nrow(domains.within.this.annotation) > 0) {
      i=i+1
      tile.data = Get.Domain.Positions.Data(
        domains.within.this.annotation %>%
          # take only one representative per each domain combination and family combination
          group_by(all.domains, family) %>%
          mutate(first.domain = first.domain[1],
                  repr.qname = qname[1]) %>%
          ungroup() %>%
          filter(repr.qname == qname) %>%
        arrange(first.domain, family)) %>%
        mutate(annotation.index = this.annotation.index,
               annotation = unique(domains.within.this.annotation$annotation),
               domain.level = this.domain.level) %>%
        left_join(families, by = "qname")  %>%
      left_join(num.prot.in.reprseq, by = "qname")
      tile.data.list[[i]] = tile.data
          
    }
    else {print("no domains")}
  }
}


tile.data.raw = do.call('rbind', tile.data.list)
tile.data.raw.domains = tile.data.raw %>% distinct(domain)
tile.data.raw.domains$x.index <- sapply(1:nrow(tile.data.raw.domains), function(index){
  this.topology <- tile.data.raw.domains$domain[index]
  this.topology.x.index <- strsplit(this.topology, ".", fixed = T)[[1]][1] %>% as.character()
})



  x_index_ecod_map = ecod_id_to_names_map %>% 
                  filter(level == "X") %>%
                  mutate(id = as.character(id)) %>%
                  select(x.index = id, x_name = name) %>%
                  rbind(data.frame(x.index = c("undetected", "multiple domains")) %>%
                                   mutate(x_name = x.index))
  
    tile.data = tile.data.raw %>%
      left_join(ecod_id_to_names_map %>% 
                  select(domain = id, domain_name = name, domain.level = level) %>%
                  rbind(expand.grid(domain.level = unique(ecod_id_to_names_map$level), 
                                    domain = c("undetected", "multiple domains")) %>%
                                   mutate(domain_name = domain))) %>%
      left_join(tile.data.raw.domains) %>%
      left_join(x_index_ecod_map, by = "x.index") 
    
    data.table::fwrite(tile.data, sprintf("%sFigure4/domain.positions.txt", OUTPUT.FIGURES.PATH))
```

```{r}
num.folds.per.fold.type.data.within.category = all.x.and.t.within.fold.type %>%
  group_by(category) %>%
  summarise(mean.num.t = mean(num.t))

mosaicism.per.fold.type.within.category = Calculate.Mosaicism.Per.Fold.Type(ecods.across.fold.types.data = ecods.across.fold.types.data %>% mutate(annotation = category), 
                                                                            all.x.and.t.within.fold.type = all.x.and.t.within.fold.type %>% mutate(annotation = category))



seq.mosaicism.from.various.classes.within.category = protein.similarity.data %>%
  inner_join( surely.annotated.proteins.from.included.categories %>% distinct(qname, category) %>% select(qname = qname, qannot = category)) %>%
  inner_join( surely.annotated.proteins.from.included.categories %>% distinct(qname, category) %>% select(sname = qname, sannot = category)) %>%
  filter(qannot == sannot) %>%
  group_by(qannot) %>%
  summarise(prop.mosaic.pairs.sequence.thr0p1 = sum(mosaic.pident10)/n(),
            prop.mosaic.pairs.sequence.thr0p3 = sum(mosaic.pident30)/n(),
            prop.mosaic.pairs.sequence.thr0p5 = sum(mosaic.pident50)/n()) %>%
  select(annotation = qannot, prop.mosaic.pairs.sequence.thr0p1, prop.mosaic.pairs.sequence.thr0p3, prop.mosaic.pairs.sequence.thr0p5)

diversity.data.per.category = surely.annotated.proteins.from.included.categories %>%
  distinct(category) %>%
  left_join(num.folds.per.fold.type.data.within.category) %>%
  left_join(mosaicism.per.fold.type.within.category) %>%
  left_join(seq.mosaicism.from.various.classes.within.category)   %>%
  select(category, mean.num.t, prop.mosaic.pairs.sequence.thr0p1, prop.mosaic.pairs.sequence.thr0p3, prop.mosaic.pairs.sequence.thr0p5, 
           prob.mosaicism, num.mosaic.pairs,
         num.fold.types, num.fold.types.with.multi.x, num.theoretical.fold.type.pairs, num.theoretical.fold.type.pairs.with.multi.x, prop.pairs.mosaoc.by.domain, num.pairs.sharing.t) %>%
  tidyr::gather(key = "proportion", value = "value", -category)
data.table::fwrite(diversity.data.per.category, sprintf("%sFigure3/mosaicism.per.category.txt", OUTPUT.FIGURES.PATH))
```






# Network of recent HGT
```{r}
recent_HGT_pairs_raw.50 = protein.similarity.data %>%
  #filter(mosaic.pident70) %>%
  filter(mosaic.pident50) %>%
  select(qname, sname) %>%
  # we will get multiple rows if there are multiple observations
  left_join(surely.annotated.proteins.including.multi.annot %>% select(qname, qindex = annotation.index)) %>%
  left_join(surely.annotated.proteins.including.multi.annot %>% select(sname = qname, sindex = annotation.index)) %>%
  #filter(qindex != sindex & !is.na(qindex) & !is.na(sindex)) %>%
  rowwise() %>%
  mutate(pair = paste(sort(c(qname, sname)), collapse = " AND ", sep = "")) %>%
  ungroup()


data.table::fwrite(recent_HGT_pairs_raw.50, sprintf("%sFigure_Supplementary/FigS11_recent_HGT/recent_HGT_pars_50.txt", OUTPUT.FIGURES.PATH))

recent_HGT_pairs_raw.70 = protein.similarity.data %>%
  #filter(mosaic.pident70) %>%
  filter(mosaic.pident70) %>%
  select(qname, sname) %>%
  left_join(surely.annotated.proteins.including.multi.annot %>% select(qname, qindex = annotation.index)) %>%
  left_join(surely.annotated.proteins.including.multi.annot %>% select(sname = qname, sindex = annotation.index)) %>%
  #filter(qindex != sindex & !is.na(qindex) & !is.na(sindex)) %>%
  rowwise() %>%
  mutate(pair = paste(sort(c(qname, sname)), collapse = " AND ", sep = "")) %>%
  ungroup()

recent_HGT_pairs.70 = rbind( recent_HGT_pairs_raw.70 %>%
  select(from = qindex, to = sindex, pair),
  recent_HGT_pairs_raw.70 %>%
  select(to = qindex, from = sindex, pair)) %>%
  mutate(from = if_else(is.na(from),as.integer(UNKNOWN.ANNOTATION.INDEX), from),
         to = if_else(is.na(to), as.integer(UNKNOWN.ANNOTATION.INDEX), to)) %>%
  group_by(from, to) %>%
  summarise(num.prot.pairs = n_distinct(pair)) %>%
  ungroup() %>%
  filter(num.prot.pairs >= 2) %>%
  filter(from <= to) %>%
  distinct() 

data.table::fwrite(recent_HGT_pairs.70, sprintf("%sFigure_Supplementary/FigS11_recent_HGT/recent_HGT_network_70.txt", OUTPUT.FIGURES.PATH))
data.table::fwrite(recent_HGT_pairs_raw.70, sprintf("%sFigure_Supplementary/FigS11_recent_HGT/recent_HGT_pars_70.txt", OUTPUT.FIGURES.PATH))


recent_HGT_pairs_raw.90 = protein.similarity.data %>%
  #filter(mosaic.pident70) %>%
  filter(mosaic.pident90) %>%
  select(qname, sname) %>%
  left_join(surely.annotated.proteins.including.multi.annot %>% select(qname, qindex = annotation.index)) %>%
  left_join(surely.annotated.proteins.including.multi.annot %>% select(sname = qname, sindex = annotation.index)) %>%
  #filter(qindex != sindex & !is.na(qindex) & !is.na(sindex)) %>%
  rowwise() %>%
  mutate(pair = paste(sort(c(qname, sname)), collapse = " AND ", sep = "")) %>%
  ungroup()


recent_HGT_pairs.90 = rbind( recent_HGT_pairs_raw.90 %>%
  select(from = qindex, to = sindex, pair),
  recent_HGT_pairs_raw.90 %>%
  select(to = qindex, from = sindex, pair)) %>%
  mutate(from = if_else(is.na(from),as.integer(UNKNOWN.ANNOTATION.INDEX), from),
         to = if_else(is.na(to), as.integer(UNKNOWN.ANNOTATION.INDEX), to)) %>%
  group_by(from, to) %>%
  summarise(num.prot.pairs = n_distinct(pair)) %>%
  ungroup() %>%
  filter(num.prot.pairs >= 2) %>%
  filter(from <= to) %>%
  distinct() 

data.table::fwrite(recent_HGT_pairs.90, sprintf("%sFigure_Supplementary/FigS11_recent_HGT/recent_HGT_network_90.txt", OUTPUT.FIGURES.PATH))
data.table::fwrite(recent_HGT_pairs_raw.90, sprintf("%sFigure_Supplementary/FigS11_recent_HGT/recent_HGT_pars_90.txt", OUTPUT.FIGURES.PATH))
```

