---
title: "generate_figures"
author: "Bogna Smug"
date: '2022-09-28'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(cowplot)
library(RColorBrewer)
library(gridExtra)
library(ggh4x)
ecod_id_to_names_map = data.table::fread(sprintf("%secod_id_to_names_map.txt",OUTPUT.DATA.PATH)) %>% as.data.frame()


PHROG.COLOR.MAP2 = PHROG.COLOR.MAP
idx = which(names(PHROG.COLOR.MAP2) == "DNA, RNA and nucleotide metabolism")
names(PHROG.COLOR.MAP2)[idx] =  "DNA, RNA\nand nucleotide metabolism"
idx = which(names(PHROG.COLOR.MAP2) == "moron, auxiliary metabolic gene and host takeover")
names(PHROG.COLOR.MAP2)[idx] =  "moron, auxiliary metabolic\ngene and host takeover"

```

# 
```{r}
# PHROG CLASS LEGEND
dat = stack(PHROG.COLOR.MAP)
names(dat) = c("values", "Category")
ggp <- ggplot(dat, aes(x=values, y=values, fill = Category)) +    # Create ggplot2 plot
  geom_col(size = 5) +
  scale_fill_manual(values = PHROG.COLOR.MAP, name = "Category")

jpeg(sprintf("%sCategory_Legend.jpg", OUTPUT.FIGURES.PATH),  width = 10, height = 10, units = "cm", res = 300)
ggp_legend <- cowplot::get_legend(ggp) 
grid::grid.newpage()                                     # Draw empty plot window
grid::grid.draw(ggp_legend)  
dev.off()
```


# Fig 1
```{r, eval = FALSE}
col1.uniqueness <- "darkgreen"
col2.coverage <- "darkred"
point.size = 3
line.size = 1
text.size = 12
annotation.uncertainty.data.vs.cov = data.table::fread(sprintf("%sFigure_Supplementary/FigS1_tradeoff/Tradeoff_cov.txt", OUTPUT.FIGURES.PATH)) %>% as.data.frame()
p.a=ggplot(data = annotation.uncertainty.data.vs.cov %>%
          mutate(`probability\nthreshold [%]` = as.factor(minimum.probability),
                  annotation.coverage = 100*annotation.coverage)) +
  geom_point(aes(x = annotation.coverage, y = perc.unique.annotation, linetype = `probability\nthreshold [%]`), col = col1.uniqueness, size = point.size) + 
  geom_line(aes(x = annotation.coverage, y = perc.unique.annotation, linetype = `probability\nthreshold [%]`), col = col1.uniqueness, size = line.size) + 
  geom_point(aes(x = annotation.coverage, y = perc.annotated, linetype = `probability\nthreshold [%]`), col = col2.coverage, size = point.size) + 
  geom_line(aes(x = annotation.coverage, y = perc.annotated, linetype = `probability\nthreshold [%]`), col = col2.coverage, size = line.size) + 
  Get.Theme(legend.position = "right") +
  xlab("sequence coverage threshold [%]") +
  theme( axis.title.y.right = element_text(colour = col2.coverage, size = text.size),
         axis.text.y.right = element_text(colour = col2.coverage, size = text.size),
          axis.title.y = element_text(colour = col1.uniqueness, size = text.size),
         axis.text.y = element_text(colour = col1.uniqueness, size = text.size),
           plot.title = element_text(size=14, face="bold", hjust = 0.5)) +
  scale_y_continuous(name = "functional uniqueness \n(% of annotated proteins with hits to a single function)",
                     limits = c(0,101),
                     breaks = seq(0,100,10),
                     sec.axis=sec_axis(~.*1,name="functional coverage \n(% proteins with a hit to any known function)"))  + 
  labs(title = "Impact of sequence coverage on functional annotation")


annotation.uncertainty.data.vs.prob = data.table::fread(sprintf("%sFigure_Supplementary/FigS1_tradeoff/Tradeoff_prob.txt", OUTPUT.FIGURES.PATH))
p.b=ggplot(data = annotation.uncertainty.data.vs.prob %>%
             filter(annotation.coverage %in% c(0.1, 0.8)) %>%
          mutate(`sequence coverage\nthreshold [%]`  = as.factor(annotation.coverage*100))) +
  geom_point(aes(x = minimum.probability, y = perc.unique.annotation, linetype = `sequence coverage\nthreshold [%]` ), col = col1.uniqueness, size = point.size) + 
  geom_line(aes(x = minimum.probability, y = perc.unique.annotation, linetype = `sequence coverage\nthreshold [%]` ), col = col1.uniqueness, size = line.size) + 
  geom_point(aes(x = minimum.probability, y = perc.annotated, linetype = `sequence coverage\nthreshold [%]` ), col = col2.coverage, size = point.size) + 
  geom_line(aes(x = minimum.probability, y = perc.annotated, linetype = `sequence coverage\nthreshold [%]` ), col = col2.coverage, size = line.size) + 
  Get.Theme(legend.position = "right") +
  xlab("probability [%]") +
  theme( axis.title.y.right = element_text(colour = col2.coverage, size = text.size),
         axis.text.y.right = element_text(colour = col2.coverage, size = text.size),
          axis.title.y = element_text(colour = col1.uniqueness, size = text.size),
         axis.text.y = element_text(colour = col1.uniqueness, size = text.size),
           plot.title = element_text(size=14, face="bold", hjust = 0.5)) +
  scale_y_continuous(name = "functional uniqueness \n(% of annotated proteins with hits to a single function)",
                     limits = c(0,101),
                     breaks = seq(0,100,10),
                     sec.axis=sec_axis(~.*1,name="functional coverage \n(% proteins with a hit to any known function)")) + 
  labs(title = "Impact of hit probability on functional annotation")
```


```{r, eval = FALSE}
p.figure <- plot_grid(p.a, p.b, labels = c("A", "B"), label_size = 25, ncol = 1)
ggsave(p.figure, filename = sprintf("%sFigure_Supplementary/FigS1_tradeoff/FigureS1.png", OUTPUT.FIGURES.PATH), bg = "white", width = 12, height = 10)
```


# Domain  network
```{r}
domain.network.metadata =   data.table::fread(sprintf("%sAdditional_Figures/domain.network.metadata.txt", OUTPUT.FIGURES.PATH)) 
domain.network.selected.categories = data.table::fread(sprintf("%sAdditional_Figures/domain.network.selected.categories.txt", OUTPUT.FIGURES.PATH))

for (selected_category in unique(domain.network.metadata$category)) {
  annotated.domain.network = Annotate_Graph(domain.network.selected.categories %>% filter(category == selected_category) %>% select(from, to), 
                    edge.colors = NULL,
                    node.color.map = domain.network.metadata %>%
                      filter(category == selected_category) %>%
                                        select(node, color),
                    # node these sizes are defined at high threshold
                    node.size.map = domain.network.metadata %>%
                      filter(category == selected_category) %>%
                      mutate(size = sqrt(n.prot)) %>%
                      select(node, size),
                    node.labels = domain.network.metadata %>% 
                      filter(category == selected_category) %>%
                      select(node, label) %>%
                      rowwise() %>%
                      mutate(label = gsub(" ", "\n", label))%>%
                      mutate(label = if_else(label == "Intramolecular\nchaperone\ndomain\nin\nvirus\ntail\nspike\nprotein", "Intramolecular\nchaperone\ndomain in\nvirus tail\nspike protein", label)) ,
                    line.color.variable =  "from")
  
  annotated.domain.network.figure = VisualiseGGnetwork(annotated.domain.network, 
                     cex = 0.1, label_size = 1.5, layout = igraph::nicely(), main = "", 
                     nudge_y = 0, alpha = 0.3)
  annotated.domain.network.figure
  ggsave(annotated.domain.network.figure, filename = sprintf("%sAdditional_Figures/annotated.domain.network.figure_%s.png", OUTPUT.FIGURES.PATH, selected_category), bg = "white", width = 8, height = 8)
}


```




```{r}

annotation.netwok.relaxed.threshold = data.table::fread( 
                   sprintf("%sFigure_Supplementary/FigS2_annot_network/annotation.netwok.relaxed.threshold.txt", OUTPUT.FIGURES.PATH))
annotation.netwok.default.threshold = data.table::fread( 
                   sprintf("%sFigure_Supplementary/FigS2_annot_network/annotation.netwok.default.threshold.txt", OUTPUT.FIGURES.PATH))
annotation.network.metadata = data.table::fread( 
                   sprintf("%sFigure_Supplementary/FigS2_annot_network/annotation.network.metadata.txt", OUTPUT.FIGURES.PATH)) %>%
  rowwise() %>%
  mutate(label = if_else(label == "DNA\npolymerase/primase", "DNA\npolymerase/\nprimase",  gsub("\n", " ", label)))

  
annotated.annotation.network.relaxed.threshold = Annotate_Graph(annotation.netwok.relaxed.threshold, 
                  edge.colors = NULL,
                  node.color.map = annotation.network.metadata%>%
                                      select(node, color),
                  # node these sizes are defined at high threshold
                  node.size.map = annotation.network.metadata %>%
                                      select(node, size),
                  node.labels = annotation.network.metadata %>% 
                    select(node, label),
                  line.color.variable =  "from")
annotated.annotation.network.default.threshold = Annotate_Graph(annotation.netwok.default.threshold, 
                  edge.colors = NULL,
                  node.color.map = annotation.network.metadata%>%
                                      select(node, color),
                  # node these sizes are defined at high threshold
                  node.size.map = annotation.network.metadata %>%
                                      select(node, size),
                  node.labels = annotation.network.metadata %>% 
                    select(node, label),
                  line.color.variable =  "from")



network.relaxed.figure = VisualiseGGnetwork(annotated.annotation.network.relaxed.threshold, 
                   cex = 0.1, label_size = 2.35, layout = igraph::nicely(), main = "", nudge_y = 0.005, alpha = 0.3) #nudge_y = 0.015,
network.relaxed.figure
#ggsave(network.relaxed.figure, filename = sprintf("%sFigure_Supplementary/FigS2_annot_network/Relaxed_network.png", OUTPUT.FIGURES.PATH), width = 9, height = 10)

network.default.figure = VisualiseGGnetwork(annotated.annotation.network.default.threshold, 
                   cex = 0.1, label_size = 2.35, layout = igraph::nicely(), main = "", nudge_y = 0.03, alpha = 0.5)
network.default.figure
#ggsave(network.default.figure, filename = sprintf("%sFigure_Supplementary/FigS2_annot_network/Default_network.png", OUTPUT.FIGURES.PATH), width = 5, height = 5)

p.figure <- plot_grid(network.default.figure, network.relaxed.figure, labels = c("A", "B"), label_size = 25, ncol = 1, rel_heights = c(0.27, 0.73))
ggsave(p.figure, filename = sprintf("%sFigure_Supplementary/FigS2_annot_network/Supplementary_Figure2.pdf", OUTPUT.FIGURES.PATH), width =8, height = 12)

```











# Fig 3 network
```{r}
annotation.network.metadata = data.table::fread( 
                   sprintf("%sFigure_Supplementary/FigS2_annot_network/annotation.network.metadata.txt", OUTPUT.FIGURES.PATH)) 


mosaic.network = data.table::fread(sprintf("%sFigure3/3a_mosaic.network.txt", OUTPUT.FIGURES.PATH))
#mosaic.metadata = data.table::fread(sprintf("%sFigure3/3a_mosaic.metadata.txt", OUTPUT.FIGURES.PATH), sep = "\t")



jpeg(filename = sprintf("%sFigure3/Figure3a_main_network_by_domain.jpg", OUTPUT.FIGURES.PATH), width = 10, height = 10, units = "cm", res = 300)
VisualiseIgraphnetwork(data.for.network = mosaic.network %>% 
                         filter(num.mosaic.fold.combination.pairs > 3) %>%
                         #filter(num.reprseq.pairs.domain > 5)  %>% 
                         #mutate(weight = log10(num.reprseq.pairs.domain)/20) %>%
                         mutate(weight = log10(1+num.mosaic.fold.combination.pairs/10),
                                dataset = if_else(num.family.pairs.recent.mosaic.seq.pident70 > 0, ">=70% pident", 
                                                         if_else(num.family.pairs.recent.mosaic.seq.pident30 > 0, ">=30% pident", 
                                                                 "ancient"))) %>%
                         select(from = from.index, to = to.index, weight, dataset),
                      edge.colors = data.frame(value = c(">=70% pident", ">=30% pident", "ancient"), color = c("darkred", "black", "gray50")),
                        #data.frame(value = c("recent", "ancient"), color = c("darkred", "gray")),
                      #edge.linetypes = data.frame(value = c("ecod", "seq", "both"), lty = c("solid", "dotted", "solid")),
                      node.color.map = annotation.network.metadata%>%
                                      select(node, color),
                      # node these sizes are defined at high threshold
                      node.size.map = annotation.network.metadata %>%
                                      select(node, size) %>% mutate(size = 2*size),
                      node.labels = annotation.network.metadata %>% 
                    select(node, label),
                      line.color.variable = "dataset",
                      line.color.legend.name = NULL,
                      font.size = 0.25,
                      main = "",
                      vertex.label.dist = 2,
                      include.labels = TRUE,
                      alpha = 1,
                    label.font = 2,
                    legend.title = "Mosaicism:",
                    node.label.color = "black",
                    add.legend = FALSE)
dev.off()

```





# Fig 2: heatmap of ECODS found in multiple categories
```{r}

text_size.x = 9
text_size.y = 9

h.domain.occurence.data = data.table::fread(sprintf("%sFigure1/H_domains.txt", OUTPUT.FIGURES.PATH)) %>% 
  as.data.frame() 
h.domain.occurence.data$category <- factor(h.domain.occurence.data$category, levels = phrogs.class.order)

phrog.class.labels2 = phrog.class.labels
phrog.class.labels2[which(names(phrog.class.labels2) == "head and packaging")] = "head\nand\npack."
phrog.class.labels2[which(names(phrog.class.labels2) == "DNA, RNA and nucleotide metabolism")] = "DNA,\nRNA\nand\nnucleot.\nmetab."
ggplot(h.domain.occurence.data) +
  geom_tile(aes(x=h_name, y = annotation, fill = num.prot.with.this.h.domain)) + #, hjust = 0, vjust = 0
  theme(axis.text.x = element_text(angle=45, hjust =1,vjust=1, size = text_size.x),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = text_size.y),
        strip.text.y = element_text(face = "bold", size = 6, angle = 360),
        strip.text.x = element_text(face = "bold", size = 6, angle = 360)) +
  facet_grid(category ~all.categories, scales = "free", space = "free", 
             labeller = labeller(category = phrog.class.labels,
                                 all.categories = c(phrog.class.labels2, "multiple categories" = "multiple\ncategories"))) + 
  xlab("") + ylab("") +
  scale_fill_gradient(name = "Num. rHHMs", low = "lightcyan", high = "darkblue", trans = "log10") #+
  #guides(fill = "none") #+
  #scale_fill_manual(values = PHROG.COLOR.MAP)

#ggsave(sprintf("%sFigure2/H_domains.jpg", OUTPUT.FIGURES.PATH), width = 7, height = 10)
ggsave(sprintf("%sFigure1/Figure1_H_domains.pdf", OUTPUT.FIGURES.PATH), width = 11, height = 10)
ggsave(sprintf("%sFigure1/Figure1_H_domains.tiff", OUTPUT.FIGURES.PATH), width = 11, height = 10)


num.f.within.h = data.table::fread(sprintf("%sFigure_Supplementary/FigS3_num_F_per_H/num.f.within.h.txt", OUTPUT.FIGURES.PATH)) 

num.f.within.h.bars = num.f.within.h %>%
  filter(in.ecod.heatmap) %>%
  rename(family = num_f, topology = num_t) %>%
  tidyr::gather(key = "ECOD level", value = "num.groups", family, topology)

text_size.x = 8
num.f.fig = ggplot(num.f.within.h.bars) +
  geom_col(aes(x = h_name, y = num.groups, fill = all.categories, alpha = `ECOD level`), position = position_dodge2(preserve = "single")) +
    theme(axis.text.x = element_text(angle=45, hjust =1,vjust=1, size = text_size.x),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = text_size.y),
        strip.text.y = element_text(face = "bold", size = 6, angle = 360),
        strip.text.x = element_text(face = "bold", size = 6, angle = 360),
        legend.position = "top") +
  facet_grid(. ~all.categories, scales = "free", space = "free", 
             labeller = labeller(all.categories = c(phrog.class.labels2, "multiple categories" = "multiple\ncategories"))) + 
  xlab("") + ylab("Num. of domain families in the group.") +
  guides(fill = "none") +
  scale_fill_manual(values = c(PHROG.COLOR.MAP, "multiple categories" = "#000000")) +
  scale_y_sqrt(breaks = c(1,10,25,50,100,150)) +
  scale_alpha_manual(values = c("family" = 0.5, "topology" =1))


ct=cor.test(num.f.within.h$num_f, num.f.within.h$num_prot, method = "spearman")
cor.fig = ggplot(num.f.within.h) +
  geom_point(aes(x = num_prot, y = num_f)) +
  scale_x_log10() +
  scale_y_log10() +
  geom_text(aes(x = 2, y = 100, label = paste0("Rho = ", as.numeric(round(ct$estimate,2)), ",\np.val = ",  as.numeric(round(ct$p.value,5))))) +
  theme_bw() + 
  theme(axis.text.x = element_text(size = text_size.x),
        axis.text.y = element_text(size = text_size.y)
        #plot.margin = unit(c(0, 0, 6, 0), "cm")
        ) +
  xlab("Number of representative proteins.") +
  ylab("Number of ECOD family groups.")

fig_diversity_ecod_f = plot_grid(num.f.fig, cor.fig, labels = c("A", "B"), label_size = 25, ncol = 1, rel_heights =  c(0.6, 0.4))#ncol = 2,rel_widths = c(0.6, 0.4))
ggsave(sprintf("%sFigure_Supplementary/FigS3_num_F_per_H/num.f.within.h.jpg", OUTPUT.FIGURES.PATH),width = 9, height = 12) #)width = 15, height = 7)


t.domain.occurence.data.full = data.table::fread(sprintf("%sFigure_Supplementary/FigS4_all_T_domains/T_domains.txt", OUTPUT.FIGURES.PATH)) %>%
  as.data.frame() 
t.domain.occurence.data.full$category <- factor(t.domain.occurence.data.full$category, levels = phrogs.class.order)
ggplot(t.domain.occurence.data.full) +
  geom_tile(aes(y=t_name, x = annotation, fill = category)) + #, hjust = 0, vjust = 0
  theme(axis.text.x = element_text(angle=45, hjust =1,vjust=1, size = text_size.x),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = text_size.y),
        strip.text.y = element_text(face = "bold", size = 6, angle = 360),
        strip.text.x = element_text(face = "bold", size = 6, angle = 360)) +
  facet_grid(all.categories ~category, scales = "free", space = "free", 
             labeller = labeller(category = phrog.class.labels,
                                 all.categories = c(phrog.class.labels2, "multiple categories" = "multiple\ncategories"))) + 
  xlab("") + ylab("") +
  guides(fill = "none") +
  scale_fill_manual(values = PHROG.COLOR.MAP)
ggsave(sprintf("%sFigure_Supplementary/FigS4_all_T_domains/T_domains.jpg", OUTPUT.FIGURES.PATH), width = 20, height = 20)


```





```{r, fig.width = 10}

most.mosaic.classes = c("tail fiber","tail spike", "endolysin", "DNA polymerase")
  tile.data = data.table::fread(sprintf("%sFigure4/domain.positions.txt", OUTPUT.FIGURES.PATH)) %>%
    filter(domain.level == "T" & annotation %in% most.mosaic.classes)

domain.position.plots =Plot.Domain.Combinations(tile.data, text.size.axis = 9, text.size.panel = 12)$domain.position.plots
pdf(sprintf("%sFigure4/Figure4_domain.positions_most_mosaic.pdf", OUTPUT.FIGURES.PATH), 
         width = 17,
         height = 20
   # units = "cm",
    #res = 300
   )
do.call("grid.arrange", c(domain.position.plots, ncol=1))
dev.off()
```

```{r, eval = FALSE}
tile.data.dnap.final = data.table::fread(sprintf("%sFigure5/domain.positions.dnap.uniclust.txt", OUTPUT.FIGURES.PATH))
domain.position.plots =Plot.Domain.Combinations(tile.data.dnap.final, text.size.axis = 9, text.size.panel = 12, interactive.plot = TRUE)$domain.position.plots
pdf(sprintf("%sFigure5/domain.positions_hypothetical_dnap.uniclust.pdf", OUTPUT.FIGURES.PATH), 
         width = 17,
         height = 12)
#gg_rect
#ggiraph:girafe(ggobj=gg_rect)
do.call("grid.arrange", c(domain.position.plots, ncol=1))
dev.off()

```




# Fig5: SHINY PLOTS
```{r, eval = FALSE}
text.size.axis = 6
text.size.panel = 12

  domain.position.plots = list()
  tile.data = data.table::fread(sprintf("%sFigure5/domain.pbottomositions.txt", OUTPUT.FIGURES.PATH)) %>% filter(domain.level == "T")

  colormap = tile.data %>% distinct(domain) %>% filter(!(domain %in% c("undetected", "multiple domains")))
  colormap$color = randomcoloR::distinctColorPalette(nrow(colormap))
  colormap = rbind(colormap, data.frame(domain = c("undetected", "multiple domains"), 
                                                  color = c( "#FFFFFF",  "#000000"), stringsAsFactors = FALSE))
  colm = colormap$color
  names(colm) = colormap$domain
        
      for (this.annotation in unique(tile.data$annotation)) {
        this.tile.data = tile.data %>% 
          filter(annotation == this.annotation)  %>% 
          mutate(annotation = this.annotation) %>%
          arrange(row)
        
        gg_rect = Show.Domain.Position.Within.Data(
          this.tile.data,
          colm, 
          theme.no.verical) +
          #scale_y_discrete(breaks =this.tile.data$qname, labels = this.tile.data$family) +
          
            theme(strip.text.x = element_text(size = text.size.panel), 
            legend.position = "none") +
              theme(
                axis.title.x = element_text(size = text.size.axis),
                axis.title.y = element_text(size = text.size.axis),
                #axis.text.x = element_blank(),
                axis.text.y = element_blank(),
                #axis.text.y = element_text(size = text.size.axis),
                #axis.ticks.x = element_blank(),
                axis.ticks.y = element_blank())
                
        domain.position.plots[[this.annotation]] =  gg_rect
  }
  
  
  #Create a subfigure with legend only, take only the most frequent
  colormap.with.names = colormap %>% inner_join(tile.data %>% 
                                                 group_by(domain_name) %>%
                                                 mutate(num_qnames = n_distinct(qname)) %>%
                                                 filter(num_qnames >= 3) %>%
                                                 distinct(domain, domain_name)) %>% distinct(color, domain_name)
  colm.with.names = colormap.with.names$color
  names(colm.with.names) = colormap.with.names$domain_name
  
  g.dummy = ggplot(tile.data, aes(x=qname, y = domain, fill = domain_name)) + geom_col() + 
    scale_fill_manual(values = colm.with.names, name = 'Domain topology') +
    theme(legend.margin =margin(r=0,l=0,t=0,b=0), text = element_text(size = 7)) +
    guides(fill=guide_legend(ncol=6))
  
  leg <- ggpubr::get_legend(g.dummy, position = "right")
  ggleg = ggpubr::as_ggplot(leg)
#grid::widthDetails(leg)
#grid::heightDetails(leg)


  
    ncol = ceiling(length(domain.position.plots)/4)
    jpeg(sprintf("%sFigure5/domain.positions.jpg", OUTPUT.FIGURES.PATH), 
         width = 50, 
         height = 30, units = "cm", res = 300)
    
    ggpubr::ggarrange(
      egg::ggarrange(plots = domain.position.plots,
                     ncol = ncol, nrow = 4),
      ggleg,
      nrow = 2,
      heights = c(0.77,0.25))
      #legend.prob = ggleg,
      #nrow = 1,
      #legend = "bottom")
  dev.off()

```

# Fig. 4: diversity
```{r}
diversity = data.table::fread(sprintf("%sFigure_Supplementary/FigS9_ecod_hist/Diversity_vs_mosaicism.txt", OUTPUT.FIGURES.PATH))

Fig4a = ggplot(diversity, aes(x=prop.proteins.with.ecod.hit)) + 
  geom_histogram(fill = "blue", position="dodge", bins = 20, color = "black") + 
  theme_minimal() + 
  labs(x = "Proportion of proteins with ECOD hit",
       y = "Number of functional classes")
ggsave(sprintf("%sFigure_Supplementary/FigS9_ecod_hist/ecod_histogram.pdf", OUTPUT.FIGURES.PATH), Fig4a, width = 12, height = 10)


# label the ones that have many families or fold types

diversity.ok.ecod = diversity %>% filter(!poorly.ecod.annotated)  %>%
  mutate(label = if_else(diverse.for.label,annotation , ""))


Fig4b = ggplot(diversity.ok.ecod) + 
  #geom_point(aes(x = num.proteins, y = num.topology.combinations, colour = category)) + 
  #geom_vline(aes(xintercept = median.num.prot), col = "red", linetype = "dashed") +
  geom_point(aes(x = num.families, y = num.topology.combinations, colour = category)) + 
  #geom_vline(aes(xintercept = median.num.fam), col = "red", linetype = "dashed") +
    geom_vline(aes(xintercept = first.quantile.num.fam), col = "red", linetype = "dashed") +
  geom_hline(aes(yintercept = median.num.topology.combinations), col = "red", linetype = "dashed") +
  scale_x_log10() + 
  #ggrepel::geom_label_repel(aes(x = num.proteins, y = num.topology.combinations, colour = category, label = label), size = 3) + 
  ggrepel::geom_label_repel(aes(x = num.families, y = num.topology.combinations, colour = category, label = label), size = 3) + 
  scale_colour_manual(values = PHROG.COLOR.MAP) + 
  labs(x = "Number of families", y = "Number of fold combinations") + 
  guides(colour = "none") + theme_minimal()

ggsave(sprintf("%sFigure3/Figure3b_diversity_scatterplot.pdf", OUTPUT.FIGURES.PATH), Fig4b, width = 10, height = 6)
ggsave(sprintf("%sFigure3/Figure3b_diversity_scatterplot.png", OUTPUT.FIGURES.PATH), Fig4b,bg = "white", width = 10, height = 6)
## supplementary version

# Fig 4c
# plot the relation between genetic and structural diversity


  levels = c("mean.length",
            "mean.num.t", 
             #"prop.mosaic.pairs.sequence.thr0p1",
             #"prop.mosaic.pairs.sequence.thr0p3",
             #"prop.mosaic.pairs.sequence.thr0p5", 
             #"num.theoretical.fold.type.pairs.with.multi.x",
             "num.mosaic.pairs",
             "prot. with any ECOD hit",
             "prop.with.multi.x.domains"
             #"proteome covered by ECOD hits"

            )
  
Get.Labeller = function(diversity.top, levels) {
   diversity.top = diversity.top %>%
    filter(proportion %in% levels)
  diversity.top = diversity.top %>% mutate(proportion = factor(proportion, levels = levels))
  diversity.levels <- levels(diversity.top$proportion)
  diversity.labeller <- character(length(diversity.levels))
  diversity.labeller[diversity.levels == "mean.num.t"] <- "Mean number of folds\nper fold architecture"
  #diversity.labeller[diversity.levels == "prop.mosaic.pairs.sequence.thr0p1"] <- "Proportion of mosaic pairs\nby sequence (10%)"
  #diversity.labeller[diversity.levels == "prop.mosaic.pairs.sequence.thr0p3"] <- "Proportion of mosaic pairs\nby sequence (30%)"
  #diversity.labeller[diversity.levels == "prop.mosaic.pairs.sequence.thr0p5"] <- "Proportion of mosaic pairs\nby sequence (50%)"
  #diversity.labeller[diversity.levels == "num.theoretical.fold.type.pairs.with.multi.x"] <- "Number of possible\npairs of fold types\nwith multiple X groups"
  diversity.labeller[diversity.levels == "num.mosaic.pairs"] <- "Number of pairs\nmosaic by domain"
  diversity.labeller[diversity.levels == "prop.with.multi.x.domains"] <- "Prop. of prot. with >= 2 X groups"
  #diversity.labeller[diversity.levels == "num.reprseq.pairs.domain"] <- "Number of protein pairs\nwith domain mosaicism"
  #diversity.labeller[diversity.levels == "prop.mosaic"] <- "Prop. of proteins\nhaving any mosaic signal."
  #diversity.labeller[diversity.levels == "proteome covered by ECOD hits"] <- "proteome covered by ECOD hits"
  diversity.labeller[diversity.levels == "prot. with any ECOD hit"] <-  "Prop. of prot. with >= 1 X group"
  diversity.labeller[diversity.levels == "mean.length"] = "Mean protein length (num AA)"
  names(diversity.labeller) <- diversity.levels
  return(diversity.labeller)
}

  
Plot.Mosaicism.Table = function(diversity.top, PHROG.COLOR.MAP) {
  p <- ggplot(diversity.top, aes(y = annotation, x = value, fill = category)) +
    geom_bar(stat = "identity", position=position_dodge()) +
    scale_fill_manual(values = PHROG.COLOR.MAP) + theme_minimal() + 
    theme(strip.text.y = element_text(face = "bold", size = 7, angle = 360)) + 
    labs(x = "", y = "") +
    scale_x_continuous(labels=function(x) sprintf("%.2f", x))
  
  p.most.diverse <- p + guides(fill = "none")
  return(p.most.diverse)
}

diversity.top = data.table::fread(sprintf("%sFigure3/mosaicism.top.diverse.classes.txt", OUTPUT.FIGURES.PATH)) %>%
  filter(proportion %in% levels)
diversity.labeller = Get.Labeller(diversity.top, levels)
p.most.diverse = Plot.Mosaicism.Table(diversity.top %>% 
                                        filter(proportion %in% names(diversity.labeller)), 
                                      PHROG.COLOR.MAP = PHROG.COLOR.MAP) +
    facet_grid(category ~ proportion, scales = "free", space = "free_y", 
                labeller = labeller(category = phrog.class.labels, proportion = diversity.labeller)
               ) #+ 
  #facetted_pos_scales(
  #  y = list(
  #    proportion == "mean.num.t" ~ scale_x_continuous(limits = c(1,3))
  #  )
  #)


#p.most.diverse = Plot.Mosaicism.Table(diversity.top %>% 
#                                        filter(proportion  == "mean.num.t"),#%in% names(diversity.labeller)), 
#                                      PHROG.COLOR.MAP = PHROG.COLOR.MAP) +
#    facet_wrap(proportion, scales = "free"
                #labeller = labeller(category = phrog.class.labels, proportion = diversity.labeller)
#               ) + 
#  facetted_pos_scales(
#    y = list(
#      proportion == "mean.num.t" ~ scale_x_continuous(limits = c(1,3))
#    )
#  )


ggsave(sprintf("%sFigure3/Figure3c.within_category_domain_mosaicism.pdf", OUTPUT.FIGURES.PATH), 
       plot_grid(p.most.diverse, ncol = 1, labels = c("B", "C"), label_size = 18, rel_heights = c(0.7, 0.3)), width = 15, height = 11)
ggsave(sprintf("%sFigure3/Figure3c.within_category_domain_mosaicism.png", OUTPUT.FIGURES.PATH), 
       plot_grid(p.most.diverse, ncol = 1, labels = c("B", "C"), label_size = 18, rel_heights = c(0.7, 0.3)), bg = "white", width = 15, height = 11)
## supplementary version
# Fig 4c
# plot the relation between genetic and structural diversity
```


```{r}
diversity.top.prop.mosaic.pairs.that.have.high.pident = data.table::fread(sprintf("%sFigure5/mosaic.pairs.by.ecod.and.seq.from.top.diverse.txt", OUTPUT.FIGURES.PATH)) %>%
  select(annotation, category, 
                prop.reprseq.pairs.mosaic.by.domain.that.have.pident10,
                prop.reprseq.pairs.mosaic.by.domain.that.have.pident30,
                prop.reprseq.pairs.mosaic.by.domain.that.have.pident50,
                prop.reprseq.pairs.mosaic.by.domain.that.have.pident70,
                prop.reprseq.pairs.mosaic.by.domain.that.have.pident90) %>%
  tidyr::gather(key = "proportion", value = "value",
                prop.reprseq.pairs.mosaic.by.domain.that.have.pident10,
                prop.reprseq.pairs.mosaic.by.domain.that.have.pident30,
                prop.reprseq.pairs.mosaic.by.domain.that.have.pident50,
                prop.reprseq.pairs.mosaic.by.domain.that.have.pident70,
                prop.reprseq.pairs.mosaic.by.domain.that.have.pident90)


p.most.diverse.perc.domain.pairs.with.pident = Plot.Mosaicism.Table(
  diversity.top.prop.mosaic.pairs.that.have.high.pident, 
  PHROG.COLOR.MAP = PHROG.COLOR.MAP) +
    facet_grid(category ~ proportion, scales = "free", space = "free_y", 
                labeller = labeller(category = phrog.class.labels, 
                                    proportion = c(prop.reprseq.pairs.mosaic.by.domain.that.have.pident10 = "Prop pairs with pident >=10%",
                                                   prop.reprseq.pairs.mosaic.by.domain.that.have.pident30 = "Prop pairs with pident >=30%",
                                                   prop.reprseq.pairs.mosaic.by.domain.that.have.pident50 = "Prop pairs with pident >=50%",
                                                   prop.reprseq.pairs.mosaic.by.domain.that.have.pident70 = "Prop pairs with pident >=70%",
                                                   prop.reprseq.pairs.mosaic.by.domain.that.have.pident90 = "Prop pairs with pident >=90%")))


ggsave(sprintf("%sFigure5/Figure5b_domain_mosaic_pairs_after_pident_check.pdf", OUTPUT.FIGURES.PATH), p.most.diverse.perc.domain.pairs.with.pident,
       width = 15, height = 11)
ggsave(sprintf("%sFigure5/Figure5b_domain_mosaic_pairs_after_pident_check.png", OUTPUT.FIGURES.PATH), p.most.diverse.perc.domain.pairs.with.pident,
       bg = "white",
       width = 15, height = 11)
```


```{r}
diversity.top.prop.mosaic.pairs.that.have.high.pident2 = data.table::fread(sprintf("%sFigure5/mosaic.pairs.by.ecod.and.seq.from.top.diverse.txt", OUTPUT.FIGURES.PATH)) %>%
  select(annotation, category, 
                num.reprseq.pairs.mosaic.by.domain,
                num.reprseq.pairs.mosaic.by.domain.and.pident10,
                num.reprseq.pairs.mosaic.by.domain.and.pident30,
                num.reprseq.pairs.mosaic.by.domain.and.pident50,
                num.reprseq.pairs.mosaic.by.domain.and.pident70,
                num.reprseq.pairs.mosaic.by.domain.and.pident90) %>%
  tidyr::gather(key = "num", value = "value",-annotation,-category)


p.most.diverse.perc.domain.pairs.with.pident2 = Plot.Mosaicism.Table(
  diversity.top.prop.mosaic.pairs.that.have.high.pident2, 
  PHROG.COLOR.MAP = PHROG.COLOR.MAP) +
    facet_grid(category ~ num, scales = "free", space = "free_y", 
                labeller = labeller(category = phrog.class.labels, 
                                    num = c(num.reprseq.pairs.mosaic.by.domain ="Num mosaic pairs",
                                              num.reprseq.pairs.mosaic.by.domain.and.pident10 = "Num. pairs with pident >=10%",
                                                   num.reprseq.pairs.mosaic.by.domain.and.pident30 = "Num. pairs with pident >=30%",
                                                   num.reprseq.pairs.mosaic.by.domain.and.pident50 = "Num. pairs with pident >=50%",
                                                   num.reprseq.pairs.mosaic.by.domain.and.pident70 = "Num. pairs with pident >=70%",
                                                   num.reprseq.pairs.mosaic.by.domain.and.pident90 = "Num. pairs with pident >=90%")))


ggsave(sprintf("%sFigure5/Figure5b_domain_mosaic_pairs_after_pident_check_v2.pdf", OUTPUT.FIGURES.PATH), p.most.diverse.perc.domain.pairs.with.pident2,
       width = 15, height = 11)
ggsave(sprintf("%sFigure5/Figure5b_domain_mosaic_pairs_after_pident_check_v2.png", OUTPUT.FIGURES.PATH), p.most.diverse.perc.domain.pairs.with.pident2,
        bg = "white",
       width = 15, height = 11)
```


```{r}
mosaic.domain.classes = data.table::fread(sprintf("%sFigure5/mosaic.domain.classes.txt", OUTPUT.FIGURES.PATH)) %>%
  select(annotation, category, annotation.index,
                prop.prot.mosaic.domain.within.any.pair.in.specific.class,
                prop.prot.mosaic.domain.within.any.pair.to.other.specific.class,
                prop.prot.mosaic.domain.within.any.pair.to.unknown.excluded.or.unspecific.class) %>%
  tidyr::gather(key = "proportion", value = "value",
                prop.prot.mosaic.domain.within.any.pair.in.specific.class,
                prop.prot.mosaic.domain.within.any.pair.to.other.specific.class,
                prop.prot.mosaic.domain.within.any.pair.to.unknown.excluded.or.unspecific.class) 


p.mosaic.domain.classes = Plot.Mosaicism.Table(
  mosaic.domain.classes, 
  PHROG.COLOR.MAP = PHROG.COLOR.MAP) +
    facet_grid(category ~ proportion, scales = "free", space = "free_y", 
                labeller = labeller(category = phrog.class.labels, 
                                    proportion = c(prop.prot.mosaic.domain.within.any.pair.in.specific.class = "% mosaic prot with\nany pair within the class",
                                                   prop.prot.mosaic.domain.within.any.pair.to.other.specific.class = "% mosaic prot with\nany pair to other class",
                                                   prop.prot.mosaic.domain.within.any.pair.to.unknown.excluded.or.unspecific.class = "% mosaic prot with\nany pair oustide of specific classes")))


ggsave(sprintf("%sFigure5/Figure5a_signal_of_mosaicism_within_and_outside_a_class_v3.pdf", OUTPUT.FIGURES.PATH), p.mosaic.domain.classes,
       width = 15, height = 11)
ggsave(sprintf("%sFigure5/Figure5a_signal_of_mosaicism_within_and_outside_a_class_v3.png", OUTPUT.FIGURES.PATH), p.mosaic.domain.classes, bg = "white",
       width = 15, height = 11)
```

```{r}
diversity.top = data.table::fread(sprintf("%sFigure3/mosaicism.top.diverse.classes.txt", OUTPUT.FIGURES.PATH))
mosaicism.underestimated.data = mosaic.domain.classes = data.table::fread(sprintf("%sFigure5/mosaic.domain.classes.txt", OUTPUT.FIGURES.PATH)) %>%
   select(annotation.index, annotation, category, num.prot.mosaic.domain, num.prot.mosaic.to.anything.with.specific.annotation, num.prot.mosaic.domain.only.to.unpecific.or.excluded.class, num.mprot.mosaic.seq, num.prot.mosaic.seq.or.domain) 
data.table::fwrite(mosaicism.underestimated.data, file = sprintf("%sFigure5/mosaicism.underestimated.data.txt", OUTPUT.FIGURES.PATH))

mosaicism.underestimated.data = mosaicism.underestimated.data %>%
  select(annotation.index, annotation, category, num.prot.mosaic.domain,num.prot.mosaic.seq.or.domain, num.prot.mosaic.to.anything.with.specific.annotation) %>%
  inner_join(included.category.sizes %>% select(annotation.index)) %>%
  inner_join(diversity.top %>% distinct(annotation, category), by = c("annotation", "category")) %>%
  filter(!(annotation.index %in% unspecific.annotation.indices)) %>%
  rename(mosaic.by.ECOD.to.annotated = num.prot.mosaic.to.anything.with.specific.annotation,#num.prot.mosaic.domain.only.within.class.or.unspecific, 
         mosaic.by.ECOD.to.anything =  num.prot.mosaic.domain,#num.prot.mosaic.domain.within.and.to.other.specific.class, 
         mosaic.by.ECOD.or.sequence =   num.prot.mosaic.seq.or.domain) %>%
  tidyr::gather(key = "type", value = "num.reprseq",-annotation.index, -annotation, -category)
mosaicism.underestimated.data$type = factor(mosaicism.underestimated.data$type,
                                            levels = c(
                                              'mosaic.by.ECOD.or.sequence',
                                              'mosaic.by.ECOD.to.anything',
                                              'mosaic.by.ECOD.to.annotated'))

g=ggplot(mosaicism.underestimated.data) +
  geom_bar(aes(y = annotation, x = num.reprseq, fill = category, col = category, alpha = type), stat = "identity", position  = "identity") +
  scale_alpha_manual(values = c(mosaic.by.ECOD.to.annotated = 1,
                                mosaic.by.ECOD.to.anything = 0.5, 
                                mosaic.by.ECOD.or.sequence = 0.05)) +
  #theme_bw() +
  #scale_fill_manual(values = c(mosaic.only.within.class = "gray50",
  #                              mosaic.to.other.class = "gray70", 
  #                              mosaic.to.unknowns.only = "gray90")) +
    scale_fill_manual(values = PHROG.COLOR.MAP) + 
    scale_color_manual(values = PHROG.COLOR.MAP) + 
   theme_minimal() + 
    theme(strip.text.y = element_text(face = "bold", size = 7, angle = 360)) + 
    labs(y = "", x = "Num. rHMMs with mosaic signal") +
    scale_x_continuous(labels=function(x) sprintf("%.2f", x)) +
    facet_grid(category ~ ., scales = "free", space = "free") +
  guides(color = "none", fill = "none")
  
g
ggsave(sprintf("%sFigure5/Figure5a_signal_of_mosaicism_within_and_outside_a_class_v5.pdf", OUTPUT.FIGURES.PATH), g,
       width = 13, height = 7)
ggsave(sprintf("%sFigure5/Figure5a_signal_of_mosaicism_within_and_outside_a_class_v5.png", OUTPUT.FIGURES.PATH), g, bg = "white",
       width = 13, height = 7)

```


```{r}
diversity.top = data.table::fread(sprintf("%sFigure3/mosaicism.top.diverse.classes.txt", OUTPUT.FIGURES.PATH))
mosaicism.underestimated.data  = data.table::fread(sprintf("%sFigure5/mosaic.domain.classes.txt", OUTPUT.FIGURES.PATH)) %>%
   select(annotation.index, annotation, category, num.families, num.families.mosaic.domain, num.families.mosaic.to.anything.with.specific.annotation, num.families.mosaic.seq.or.domain, pval.corr) 
data.table::fwrite(mosaicism.underestimated.data, file = sprintf("%sFigure5/mosaicism.underestimated.data.txt", OUTPUT.FIGURES.PATH))

mosaicism.underestimated.data = mosaicism.underestimated.data %>%
  select(annotation.index, annotation, category, 
         num.families,
         num.families.mosaic.domain,
         num.families.mosaic.seq.or.domain,
         pval.corr) %>%
  inner_join(included.category.sizes %>% select(annotation.index)) %>%
  inner_join(diversity.top %>% distinct(annotation, category), by = c("annotation", "category")) %>%
  filter(!(annotation.index %in% unspecific.annotation.indices)) %>%
  rename(`All` = num.families,
         `Mosaic by ECOD` =  num.families.mosaic.domain,#num.prot.mosaic.domain.within.and.to.other.specific.class, 
         `Mosaic by ECOD or seq` =   num.families.mosaic.seq.or.domain) %>%
  tidyr::gather(key = "type", value = "num.families", `All`, `Mosaic by ECOD`, `Mosaic by ECOD or seq`) %>% 
  mutate(significant = if_else(pval.corr < 0.05 & type == 'All', "*", "")) 
mosaicism.underestimated.data$type = factor(mosaicism.underestimated.data$type,
                                            levels = c(
                                              'All',
                                              'Mosaic by ECOD or seq',
                                              'Mosaic by ECOD'))

g=ggplot(mosaicism.underestimated.data, aes(y = annotation, x = num.families, fill = category, col = category)) +
  geom_bar(aes(alpha = type), stat = "identity", position  = "identity") +
  geom_text(aes(label = significant), nudge_x = 10) +
  scale_alpha_manual(values = c(All = 0.05,
                                `Mosaic by ECOD` = 1,
                                `Mosaic by ECOD or seq` = 0.3)) +
  #theme_bw() +
  #scale_fill_manual(values = c(mosaic.only.within.class = "gray50",
  #                              mosaic.to.other.class = "gray70", 
  #                              mosaic.to.unknowns.only = "gray90")) +
    scale_fill_manual(values = PHROG.COLOR.MAP) + 
    scale_color_manual(values = PHROG.COLOR.MAP) + 
   theme_minimal() + 
    theme(strip.text.y = element_text(face = "bold", size = 7, angle = 360)) + 
    labs(y = "", x = "Num. rHMM families with mosaic signal") +
    scale_x_continuous(labels=function(x) sprintf("%.2f", x)) +
    facet_grid(category ~ ., scales = "free", space = "free") +
  guides(color = "none", fill = "none")
  
g
ggsave(sprintf("%sFigure5/Figure5a_signal_of_mosaicism_within_and_outside_a_class_v5_families.pdf", OUTPUT.FIGURES.PATH), g,
       width = 13, height = 7)
ggsave(sprintf("%sFigure5/Figure5a_signal_of_mosaicism_within_and_outside_a_class_v5_families.png", OUTPUT.FIGURES.PATH), g, bg = "white",
       width = 13, height = 7)

```

```{r}
# proportions are not good because 35/1 loooks twice better than 35/2 which is not good


diversity.top = data.table::fread(sprintf("%sFigure3/mosaicism.top.diverse.classes.txt", OUTPUT.FIGURES.PATH))
mosaicism.underestimated.data = data.table::fread(sprintf("%sFigure5/mosaic.domain.classes.txt", OUTPUT.FIGURES.PATH)) %>%
   select(annotation.index, annotation, category, num.families.mosaic.domain, num.families.mosaic.seq.or.domain, num.prot.mosaic.domain, num.prot.mosaic.to.anything.with.specific.annotation, num.prot.mosaic.domain.only.to.unpecific.or.excluded.class, num.mprot.mosaic.seq, num.prot.mosaic.seq.or.domain) 
data.table::fwrite(mosaicism.underestimated.data, file = sprintf("%sFigure5/mosaicism.underestimated.data.txt", OUTPUT.FIGURES.PATH))

mosaicism.underestimated.data = mosaicism.underestimated.data %>%
  select(annotation.index, annotation, category, num.prot.mosaic.domain, num.prot.mosaic.seq.or.domain, num.prot.mosaic.to.anything.with.specific.annotation) %>%
  inner_join(included.category.sizes %>% select(annotation.index, num.proteins)) %>%
  inner_join(diversity.top %>% distinct(annotation, category), by = c("annotation", "category")) %>%
  filter(!(annotation.index %in% unspecific.annotation.indices)) %>%
  mutate(num.prot.nonmosaic.domain = num.proteins - num.prot.mosaic.domain,
         num.prot.nonmosaic.seq.or.domain = num.proteins - num.prot.mosaic.seq.or.domain) %>%
  mutate(odds.seq.to.ecod = (num.prot.mosaic.seq.or.domain/num.prot.nonmosaic.seq.or.domain)/(num.prot.mosaic.domain/num.prot.nonmosaic.domain),
         prop.mosaic.by.ecod.with.any.known.pairmate = num.prot.mosaic.to.anything.with.specific.annotation/num.prot.mosaic.domain) %>%
  rename(`odds ratio seq to ecod` = odds.seq.to.ecod,#num.prot.mosaic.domain.only.within.class.or.unspecific, 
         `Prop. mosaic by ECOD\nwith any specific pairmate` =  prop.mosaic.by.ecod.with.any.known.pairmate) %>%
  select(annotation, category, annotation.index, `odds ratio seq to ecod`, `Prop. mosaic by ECOD\nwith any specific pairmate`) %>%
  tidyr::gather(key = "type", value = "prop",-annotation.index, -annotation, -category)
mosaicism.underestimated.data$type = factor(mosaicism.underestimated.data$type,
                                            levels = c(
                                             "odds ratio seq to ecod",
                                             "Prop. mosaic by ECOD\nwith any specific pairmate"))

g=ggplot(mosaicism.underestimated.data, aes(y = annotation, x = prop, fill = category, col = category)) +
  geom_bar(stat = "identity", width =0.7) +
  facet_grid(category ~ type, scales = "free", space ="free_y") +
  
    scale_fill_manual(values = PHROG.COLOR.MAP) + 
    scale_color_manual(values = PHROG.COLOR.MAP) + 
   theme_minimal() + 
    theme(strip.text.y = element_text(face = "bold", size = 7, angle = 360)) + 
    labs(y = "", x = "Proportion") +
    scale_x_continuous(labels=function(x) sprintf("%.2f", x)) +
  guides(color = "none", fill = "none")
  
g
ggsave(sprintf("%sFigure5/Figure5a_signal_of_mosaicism_within_and_outside_a_class_v7.pdf", OUTPUT.FIGURES.PATH), g,
       width = 13, height = 7)
ggsave(sprintf("%sFigure5/Figure5a_signal_of_mosaicism_within_and_outside_a_class_v7.png", OUTPUT.FIGURES.PATH), g, bg = "white",
       width = 13, height = 7)







mosaicism.underestimated.data = mosaic.domain.classes = data.table::fread(sprintf("%sFigure5/mosaic.domain.classes.txt", OUTPUT.FIGURES.PATH)) %>%
  rename(`odds ratio seq to ecod` = odds.seq.to.ecod)  %>%
  filter(pval.corr < 0.05) %>%
  mutate(significant = if_else(pval.corr < 0.05, "*", "")) %>%
  arrange(`odds ratio seq to ecod`)
mosaicism.underestimated.data$annotation = factor(mosaicism.underestimated.data$annotation,
                                                  levels = unique(mosaicism.underestimated.data$annotation))

#%>%
 #inner_join(diversity.top %>% distinct(annotation, category), by = c("annotation", "category")) 

g=ggplot(mosaicism.underestimated.data, aes(y = annotation, x = `odds ratio seq to ecod`, fill = category, col = category)) +
  geom_bar(stat = "identity", width =0.7) +
  geom_text(aes(label = significant), nudge_x = 1) +
  #facet_grid(category ~ ., scales = "free", space ="free_y") +
    scale_fill_manual(values = PHROG.COLOR.MAP) + 
    scale_color_manual(values = PHROG.COLOR.MAP) + 
   theme_minimal() + 
    theme(strip.text.y = element_text(face = "bold", size = 7, angle = 360)) + 
    labs(y = "", x = "Odds of being mosaic by seq. to ECOD") +
    scale_x_continuous(labels=function(x) sprintf("%.2f", x)) +
  guides(color = "none", fill = "none")
  
g
ggsave(sprintf("%sFigure5/Figure5a_signal_of_mosaicism_within_and_outside_a_class_v9.pdf", OUTPUT.FIGURES.PATH), g,
       width = 13, height = 7)
ggsave(sprintf("%sFigure5/Figure5a_signal_of_mosaicism_within_and_outside_a_class_v9.png", OUTPUT.FIGURES.PATH), g, bg = "white",
       width = 13, height = 7)

```


```{r, eval = FALSE}

mosaic.domain.classes.long = mosaic.domain.classes = data.table::fread(sprintf("%sFigure5/mosaic.domain.classes.txt", OUTPUT.FIGURES.PATH)) %>%
  select(annotation.index, annotation, category, 
         num.prot.mosaic.domain.any.within.class,
         num.prot.mosaic.domain.only.to.other.specific.class,
         #num.prot.mosaic.domain.only.within.class.or.unspecific, 
         #num.prot.mosaic.domain.within.and.to.other.specific.class, 
         num.prot.mosaic.domain.only.to.unpecific.or.excluded.class) %>%
  inner_join(included.category.sizes %>% select(annotation.index)) %>%
  inner_join(diversity.top %>% distinct(annotation, category), by = c("annotation", "category")) %>%
  filter(!(annotation.index %in% unspecific.annotation.indices)) %>%
  rename(mosaic.within.class = num.prot.mosaic.domain.any.within.class,#num.prot.mosaic.domain.only.within.class.or.unspecific, 
         mosaic.to.other.class =  num.prot.mosaic.domain.only.to.other.specific.class,#num.prot.mosaic.domain.within.and.to.other.specific.class, 
         mosaic.to.unknowns.only =   num.prot.mosaic.domain.only.to.unpecific.or.excluded.class) %>%
  tidyr::gather(key = "type", value = "num.reprseq",-annotation.index, -annotation, -category)
mosaic.domain.classes.long$type = factor(mosaic.domain.classes.long$type,
                                            levels = c(
                                              'mosaic.to.unknowns.only',
                                              'mosaic.to.other.class',
                                              'mosaic.within.class'))

g=ggplot(mosaic.domain.classes.long) +
  geom_bar(aes(y = annotation, x = num.reprseq, fill = category, col = category, alpha = type), stat = "identity", position = "stack") +
  scale_alpha_manual(values = c(mosaic.within.class = 1,
                                mosaic.to.other.class = 0.5, 
                                mosaic.to.unknowns.only = 0.05)) +
  #theme_bw() +
  #scale_fill_manual(values = c(mosaic.only.within.class = "gray50",
  #                              mosaic.to.other.class = "gray70", 
  #                              mosaic.to.unknowns.only = "gray90")) +
    scale_fill_manual(values = PHROG.COLOR.MAP) + 
    scale_color_manual(values = PHROG.COLOR.MAP) + 
   theme_minimal() + 
    theme(strip.text.y = element_text(face = "bold", size = 7, angle = 360)) + 
    labs(y = "", x = "Num. rHMMs with mosaic signal") +
    scale_x_continuous(labels=function(x) sprintf("%.2f", x)) +
    facet_grid(category ~ ., scales = "free", space = "free") +
  guides(color = "none", fill = "none")
  
g
ggsave(sprintf("%sFigure5/Figure5a_signal_of_mosaicism_within_and_outside_a_class_v1.pdf", OUTPUT.FIGURES.PATH), g,
       width = 13, height = 7)
ggsave(sprintf("%sFigure5/Figure5a_signal_of_mosaicism_within_and_outside_a_class_v1.png", OUTPUT.FIGURES.PATH), g, bg = "white",
       width = 13, height = 7)
```


#Poorly annotated ECOD analysis
```{r}
classes.with.few.ecod.annotated.prots.new = data.table::fread(sprintf("%sFigure_Supplementary/FigS10_poor_ecod/classes.with.few.ecod.annotated.txt", OUTPUT.FIGURES.PATH))
classes.with.few.ecod.annotated.prots.new$annotation = factor(classes.with.few.ecod.annotated.prots.new$annotation , levels = unique(classes.with.few.ecod.annotated.prots.new$annotation))
classes.with.few.ecod.annotated.prots.long = classes.with.few.ecod.annotated.prots.new %>% 
  select(annotation, category, `prot. with any ECOD hit`,  `proteome covered by ECOD hits`) %>%
  tidyr::gather(key = "stat", value = "prop",`prot. with any ECOD hit`,  `proteome covered by ECOD hits`)

p=ggplot(classes.with.few.ecod.annotated.prots.long, aes(x = annotation, y = prop, fill = category, alpha = stat)) +
   geom_bar(stat = "identity", position=position_dodge()) +
  facet_grid(. ~ category, scales = "free_x", space = "free_x") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1 )) +
  scale_y_continuous(limits = c(0,1))+ 
  guides(fill = "none")+ 
  labs(x = " function", y = "Proportion") +
  scale_fill_manual(values = PHROG.COLOR.MAP) +
  scale_alpha_discrete(range = c(1, .4))
p


ggsave(sprintf("%sFigure_Supplementary/FigS10_poor_ecod/poor_ecod_annotation.png", OUTPUT.FIGURES.PATH), p, bg = "white",width = 20, height = 10)

```




```{r, fig.height =5, fig.width =10}
annot.stats.per.category.and.mosaicism = data.table::fread(file = sprintf("%sFigure6/unknown_functions_mosaic_vs_nonmosaic.txt", OUTPUT.FIGURES.PATH)) %>%
  arrange(category, annotation)
annot.stats.per.category.and.mosaicism.by.seq = data.table::fread(file = sprintf("%sFigure6/unknown_functions_mosaic_vs_nonmosaic_by_seq.txt", OUTPUT.FIGURES.PATH)) %>%
  arrange(category, annotation)
annot.stats.per.category.and.mosaicism.by.domain = data.table::fread(file = sprintf("%sFigure6/unknown_functions_mosaic_vs_nonmosaic_by_domain.txt", OUTPUT.FIGURES.PATH)) %>%
  arrange(category, annotation)


p.annot.low.cov = ggplot(annot.stats.per.category.and.mosaicism %>%
                           filter(!is.na(category) & category != "") %>%
  mutate(mosaic = if_else(mosaic, "Mosaic", "Not mosaic")) %>% as.data.frame()) +
  geom_col(aes(x=annotation, y = prop.proteins.has.this.annot, fill = category, alpha = mosaic), col = "black", position = position_dodge2(preserve = "single") ) +
  facet_grid(.~category, scales = "free", space = "free_x", labeller = labeller(category = phrog.class.labels)) +
  scale_alpha_manual(values = c("Mosaic" = 1, "Not mosaic" = 0.3)) +
  scale_fill_manual(values = PHROG.COLOR.MAP) +
  theme(axis.text.x = element_text(angle=45, size = 8, vjust = 1, hjust = 1),
        axis.text.y = element_text(size = 9)) +
  guides(fill = "none") + 
  theme(legend.text=element_text(size=4)) + 
  labs(x = "", y = "Proportion of proteins with this annotation") 
ggsave(filename = sprintf("%sFigure_Supplementary/FigS8_annot_low_cov/annotations.at.low.cov.jpg", 
                          OUTPUT.FIGURES.PATH) ,p.annot.low.cov , width = 15, height = 12)



annots.within.proteins.with.significant.higher.mosaic = annot.stats.per.category.and.mosaicism %>%
   filter(corrected.pval < 0.05 & odds.ratio > 1) %>%
  select(annotation, category, mosaic, prop.proteins.has.this.annot, odds.ratio, corrected.pval) %>%
   arrange(desc(odds.ratio))
write.xlsx(x = annots.within.proteins.with.significant.higher.mosaic, 
           file = sprintf("%stables/Supplementary_table_mosaic_vs_ninmosaic.xlsx", OUTPUT.FIGURES.PATH), 
           sheetName = "Annotations in mosaic vs non-mosaic proteins", 
  col.names = TRUE, row.names = TRUE, append = TRUE)


ss.by.domain.or.seq=annot.stats.per.category.and.mosaicism %>%
  arrange(desc(odds.ratio)) %>%
  filter(corrected.pval < 0.05) %>%
  #head(60) %>%
  mutate(mosaic = if_else(mosaic, "Mosaic", "Not mosaic")) 


ss.by.domain=annot.stats.per.category.and.mosaicism.by.domain %>%
  arrange(desc(odds.ratio)) %>%
  filter(corrected.pval < 0.05) %>%
  #head(60) %>%
  mutate(mosaic = if_else(mosaic, "Mosaic", "Not mosaic"))

ss.by.seq=annot.stats.per.category.and.mosaicism.by.seq %>%
  arrange(desc(odds.ratio)) %>%
  filter(corrected.pval < 0.05) %>%
  #head(60) %>%
  mutate(mosaic = if_else(mosaic, "Mosaic", "Not mosaic"))
  
ss = rbind(ss.by.domain %>% mutate(mosaicism.type = "Mosaic by domain"),
           ss.by.domain.or.seq %>% mutate(mosaicism.type = "Mosaic by domain or sequence")) %>%
      rbind(ss.by.seq %>% mutate(mosaicism.type = "Mosaic by sequence")) %>%
      mutate(ecod.coverage.type2 = if_else(ecod.coverage.type == "few prots. with >=1 detected domains" | ecod.coverage.type == "few prots. with >=2 detected domains", "< 20% prots. with >=2\ndetected domains", ">= 20% prots. with >=2\ndetected domains"))
  

ss$mosaic = factor(ss$mosaic, levels = c("Not mosaic", "Mosaic"))
ss$ecod.coverage.type2 = factor(ss$ecod.coverage.type2, levels = c(">= 20% prots. with >=2\ndetected domains", "< 20% prots. with >=2\ndetected domains"))
ss$category[ss$category == "DNA, RNA and nucleotide metabolism"] = "DNA, RNA\nand nucleotide metabolism"
ss$category[ss$category =="moron, auxiliary metabolic gene and host takeover" ] = "moron,\nauxiliary metabolic gene\nand host takeover" 
PHROG.COLOR.MAP2 = PHROG.COLOR.MAP
idx=which(names(PHROG.COLOR.MAP2) == "DNA, RNA and nucleotide metabolism")
names(PHROG.COLOR.MAP2)[idx] = "DNA, RNA\nand nucleotide metabolism"
idx=which(names(PHROG.COLOR.MAP2) ==  "moron, auxiliary metabolic gene and host takeover")
names(PHROG.COLOR.MAP2)[idx] = "moron,\nauxiliary metabolic gene\nand host takeover" 



ss.domain = ss  %>%  
  distinct(annotation, category, odds.ratio,ecod.coverage.type2, mosaicism.type) %>%
  filter(mosaicism.type == "Mosaic by domain") %>%
  arrange(desc(odds.ratio)) %>%
  head(30)
ss.domain$annotation = factor(ss.domain$annotation, levels = unique(ss.domain %>% arrange(ecod.coverage.type2, odds.ratio) %>% pull(annotation)))

ss.both = ss  %>%  
  distinct(annotation, category, odds.ratio,ecod.coverage.type2, mosaicism.type) %>%
  filter(mosaicism.type == "Mosaic by domain or sequence") %>%
  arrange(desc(odds.ratio)) %>%
  head(30)
ss.both$annotation = factor(ss.both$annotation, levels = unique(ss.both %>% arrange(ecod.coverage.type2, odds.ratio) %>% pull(annotation)))


p.odds.low.cov.by.domain = ggplot(ss.domain, aes(x=annotation, y = odds.ratio, fill = category)) +
  geom_bar(stat="identity", width = 0.8) + 
  scale_fill_manual(values = PHROG.COLOR.MAP2) +
  theme_minimal() + 
  #guides(fill = "none") + 
  theme(legend.text=element_text(size=8)) + 
  labs(x = "", y = "Odds ratio", fill = "category") + 
  coord_flip() +
  facet_grid(ecod.coverage.type2 ~ mosaicism.type, scales = "free", space = "free")
p.odds.low.cov.by.domain


p.odds.low.cov = ggplot(ss.both, aes(x=annotation, y = odds.ratio, fill = category)) +
  geom_bar(stat="identity", width = 0.8) + 
  scale_fill_manual(values = PHROG.COLOR.MAP2) +
  theme_minimal() + 
  #guides(fill = "none") + 
  theme(legend.text=element_text(size=8)) + 
  labs(x = "", y = "Odds ratio", fill = "category") + 
  coord_flip() +
  facet_grid(ecod.coverage.type2 ~ mosaicism.type, scales = "free", space = "free")
p.odds.low.cov


fig4ab = plot_grid(p.odds.low.cov.by.domain + theme(legend.position = "none"), p.odds.low.cov, labels = c("A", "B"), label_size = 25, ncol = 2,rel_widths = c(0.4, 0.6))
ggsave(sprintf("%s/Figure6/Figure6a_odds_ratio_domain_and_seq_v1_lowcov.pdf", OUTPUT.FIGURES.PATH), fig4ab,  width = 15, height = 6)
ggsave(sprintf("%s/Figure6/Figure6a_odds_ratio_domain_and_seq_v1_lowcov.png", OUTPUT.FIGURES.PATH), fig4ab, bg = "white", width = 15, height = 6)

```


```{r, fig.height =5, fig.width =10}
annot.stats.per.category.and.mosaicism.sure.annot = data.table::fread(file = sprintf("%sFigure6/known_functions_mosaic_vs_nonmosaic.txt", OUTPUT.FIGURES.PATH)) %>%
  arrange(category, annotation)
annot.stats.per.category.and.mosaicism.by.seq.sure.annot = data.table::fread(file = sprintf("%sFigure6/known_functions_mosaic_vs_nonmosaic_by_seq.txt", OUTPUT.FIGURES.PATH)) %>%
  arrange(category, annotation)
annot.stats.per.category.and.mosaicism.by.domain.sure.annot = data.table::fread(file = sprintf("%sFigure6/known_functions_mosaic_vs_nonmosaic_by_domain.txt", OUTPUT.FIGURES.PATH)) %>%
  arrange(category, annotation)



ss.by.domain.or.seq.sure.annot=annot.stats.per.category.and.mosaicism.sure.annot %>%
  arrange(desc(odds.ratio)) %>%
  filter(corrected.pval < 0.05) %>%
  #head(60) %>%
  mutate(mosaic = if_else(mosaic, "Mosaic", "Not mosaic")) 


ss.by.domain.sure.annot=annot.stats.per.category.and.mosaicism.by.domain.sure.annot %>%
  arrange(desc(odds.ratio)) %>%
  filter(corrected.pval < 0.05) %>%
  #head(60) %>%
  mutate(mosaic = if_else(mosaic, "Mosaic", "Not mosaic"))

ss.by.seq.sure.annot=annot.stats.per.category.and.mosaicism.by.seq.sure.annot %>%
  arrange(desc(odds.ratio)) %>%
  filter(corrected.pval < 0.05) %>%
  #head(60) %>%
  mutate(mosaic = if_else(mosaic, "Mosaic", "Not mosaic"))
  
ss.sure.annot = rbind(ss.by.domain.sure.annot %>% mutate(mosaicism.type = "Mosaic by domain"),
           ss.by.domain.or.seq.sure.annot %>% mutate(mosaicism.type = "Mosaic by domain or sequence")) %>%
      rbind(ss.by.seq.sure.annot %>% mutate(mosaicism.type = "Mosaic by sequence")) %>%
      mutate(ecod.coverage.type2 = if_else(ecod.coverage.type == "few prots. with >=1 detected domains" | ecod.coverage.type == "few prots. with >=2 detected domains", "< 20% prots. with >=2\ndetected domains", ">= 20% prots. with >=2\ndetected domains"))
  

ss.sure.annot$mosaic = factor(ss.sure.annot$mosaic, levels = c("Not mosaic", "Mosaic"))
ss.sure.annot$ecod.coverage.type2 = factor(ss.sure.annot$ecod.coverage.type2, levels = c(">= 20% prots. with >=2\ndetected domains", "< 20% prots. with >=2\ndetected domains"))
ss.sure.annot$category[ss.sure.annot$category == "DNA, RNA and nucleotide metabolism"] = "DNA, RNA\nand nucleotide metabolism"
ss.sure.annot$category[ss.sure.annot$category =="moron, auxiliary metabolic gene and host takeover" ] = "moron,\nauxiliary metabolic gene\nand host takeover" 


ss.domain.sure.annot = ss.sure.annot  %>%  
  distinct(annotation, category, odds.ratio,ecod.coverage.type2, mosaicism.type) %>%
  filter(mosaicism.type == "Mosaic by domain") %>%
  arrange(desc(odds.ratio)) %>%
  head(30)
ss.domain.sure.annot$annotation = factor(ss.domain.sure.annot$annotation, levels = unique(ss.domain.sure.annot %>% arrange(ecod.coverage.type2, odds.ratio) %>% pull(annotation)))

ss.domain.or.seq.sure.annot = ss.sure.annot  %>%  
  distinct(annotation, category, odds.ratio,ecod.coverage.type2, mosaicism.type) %>%
  filter(mosaicism.type ==  "Mosaic by domain or sequence") %>%
  arrange(desc(odds.ratio)) %>%
  head(30)
ss.domain.or.seq.sure.annot$annotation = factor(ss.domain.or.seq.sure.annot$annotation, levels = unique(ss.domain.or.seq.sure.annot %>% arrange(ecod.coverage.type2, odds.ratio) %>% pull(annotation)))


p.odds.high.cov.by.domain = ggplot(ss.domain.sure.annot, aes(x=annotation, y = odds.ratio, fill = category)) +
  geom_bar(stat="identity", width = 0.8) + 
  scale_fill_manual(values = PHROG.COLOR.MAP2) +
  theme_minimal() + 
  #guides(fill = "none") + 
  theme(legend.text=element_text(size=8)) + 
  labs(x = "", y = "Odds ratio", fill = "category") + 
  coord_flip() +
  facet_grid(ecod.coverage.type2 ~ mosaicism.type, scales = "free", space = "free")
p.odds.high.cov.by.domain
#ggsave(filename = sprintf("%sFigure3/low.cov.odds.ratio.by.domain.jpg", OUTPUT.FIGURES.PATH) ,p.odds.high.cov.by.domain , width = 10, height = 7)


p.odds.high.cov =ggplot(ss.domain.or.seq.sure.annot, 
                        aes(x=annotation, y = odds.ratio, fill = category))  +
  geom_bar(stat="identity", width = 0.8) + 
  #geom_text(aes(label = label), nudge_y = 1) +
  scale_fill_manual(values = PHROG.COLOR.MAP2) +
  theme_minimal() + 
  #guides(fill = "none") + 
  theme(legend.text=element_text(size=8)) + 
  labs(x = "", y = "Odds ratio", fill = "category") + 
  coord_flip() +
  facet_grid(ecod.coverage.type2 ~ mosaicism.type, scales = "free", space = "free") #+
  #heme(axis.text.y = element_text(colour = rev(ss.domain.or.seq.sure.annot$color)))
p.odds.high.cov
fig4ab = plot_grid(p.odds.high.cov.by.domain + theme(legend.position = "none"), p.odds.high.cov, labels = c("A", "B"), 
                   label_size = 25, ncol = 2,rel_widths = c(0.4, 0.6))
ggsave(sprintf("%s/Figure6/Figure6a_odds_ratio_domain_and_seq_v2_highcov.png", OUTPUT.FIGURES.PATH), fig4ab,  bg = "white", width = 15, height = 6)
ggsave(sprintf("%s/Figure6/Figure6a_odds_ratio_domain_and_seq_v2_highcov.pdf", OUTPUT.FIGURES.PATH), fig4ab,  width = 15, height = 6)


#ggsave(filename = sprintf("%sFigure3/low.cov.odds.ratio.by.domainor.seq.jpg", OUTPUT.FIGURES.PATH) ,p.odds.high.cov , width = 10, height = 7)
```





# new Fif 3c shared domains by mosaic pairs (known and unknown)
```{r}
domains.shared.in.mosaic.pairs = data.table::fread(file = sprintf("%sFigure_Supplementary/FigS7_folds_shared_by_mosaic/domains-shared-by-mosaic-pairs.txt", OUTPUT.FIGURES.PATH))

included.ecods = domains.shared.in.mosaic.pairs %>% 
  filter(prop.protein.pairs >= 0.005) %>%
  pull(t_id)

ecod.table <- domains.shared.in.mosaic.pairs %>% 
  filter(t_id %in% included.ecods)%>%
  arrange(prop.protein.pairs)

ecod.table$t_name = factor(ecod.table$t_name, 
                           levels = ecod.table %>% pull(t_name) %>% unique())

display.h.name <- ecod.table$h_name
display.h.name[ecod.table$t_name == ecod.table$h_name] <- "same"
display.t.name <- sprintf("%s\n(H: %s)", ecod.table$t_name, display.h.name)

getPalette <- colorRampPalette(RColorBrewer::brewer.pal(9, "Set1"))
no.hom <- length(unique(ecod.table$h_name))
p.shared.ecods <- ggplot(data=ecod.table, aes(x=t_name, y=prop.protein.pairs, fill=h_name)) + 
  geom_bar(stat="identity", position = position_dodge2(width = 0.9, preserve = "single"),
           alpha = 1, color="black") +#, fill = "gray90") +
  scale_x_discrete(breaks = ecod.table$t_name, labels = display.t.name) +
  theme_minimal() + theme(legend.text=element_text(size=6)) +
  labs(x = "", y = "Proportion of representative proteins pairs") + 
  guides(fill = "none", color = "none") + 
  coord_flip()

p.shared.ecods
#p.panelc <- plot_grid(p.seq, p.ecod, labels = c("C", "D"), label_size = 25, ncol = 2)
ggsave(filename = sprintf("%sFigure_Supplementary/FigS7_folds_shared_by_mosaic/folds_shared_by_mosaic_proteins.jpg", OUTPUT.FIGURES.PATH), p.shared.ecods, width = 12, height = 6)

```


# Fig 3c shared ecods in knwonn and unknown prots

```{r}

domains.within.proteins.domain.mosaic = data.table::fread(file = sprintf("%sAdditional_Figures/domains-in-known-and-unknown-domain-mosaic.txt", OUTPUT.FIGURES.PATH))
p.odds = Plot.Domains.Within.Mosaic.Proteins(domains.within.proteins.domain.mosaic) 
ggsave(filename = sprintf("%sAdditional_Figures/folds_within_domain_mosaic_proteins.odds.ratio.jpg", OUTPUT.FIGURES.PATH) ,p.odds , width = 15, height = 12)

domains.within.proteins.domain.mosaic.and.pident = data.table::fread(file =  sprintf("%sAdditional_Figures/domains-in-known-and-unknown-domain-mosaic-and-pident70.txt", OUTPUT.FIGURES.PATH))
p.odds = Plot.Domains.Within.Mosaic.Proteins(domains.within.proteins.domain.mosaic.and.pident) 
ggsave(filename = sprintf("%sAdditional_Figures/folds_within_domain_and_pident70_mosaic_proteins.odds.ratio.jpg", OUTPUT.FIGURES.PATH) ,p.odds , width = 15, height = 12)

domains.within.proteins = data.table::fread(file = sprintf("%sFigure3/domains-in-known-and-unknown.txt", OUTPUT.FIGURES.PATH))
p.odds = Plot.Domains.Within.Mosaic.Proteins(domains.within.proteins) 
ggsave(filename = sprintf("%sAdditional_Figures/folds_within_mosaic_proteins.odds.ratio.jpg", OUTPUT.FIGURES.PATH) ,p.odds , width = 15, height = 12)

domains.within.proteins.seq.mosaic = data.table::fread(file = sprintf("%sAdditional_Figures/domains-in-known-and-unknown-seq-mosaic.txt", OUTPUT.FIGURES.PATH))
p.odds = Plot.Domains.Within.Mosaic.Proteins(domains.within.proteins) 
ggsave(filename = sprintf("%sAdditional_Figures/folds_within_seq_mosaic_proteins.odds.ratio.jpg", OUTPUT.FIGURES.PATH) ,p.odds , width = 15, height = 12)



all.t.in.fold.types = data.table::fread(file = sprintf("%sAdditional_Figures/all.t.in.fold.types.txt", OUTPUT.FIGURES.PATH))
p.odds = Plot.Domains.Within.Mosaic.Proteins(all.t.in.fold.types) 
ggsave(filename = sprintf("%sAdditional_Figures/folds_within_mosaic_fold_types.jpg", OUTPUT.FIGURES.PATH) ,p.odds , width = 15, height = 12)


all.t.in.fold.types_pident50 = data.table::fread(file = sprintf("%sAdditional_Figures/all.t.in.fold.types_pident50.txt", OUTPUT.FIGURES.PATH))
p.odds = Plot.Domains.Within.Mosaic.Proteins(all.t.in.fold.types_pident50) 
ggsave(filename = sprintf("%sAdditional_Figures/folds_within_mosaic_fold_types_pident50.jpg", OUTPUT.FIGURES.PATH) ,p.odds , width = 15, height = 12)
```




```{r} 
annotated.hhr.table = data.table::fread(sprintf("%sFigure_Supplementary/FigS5_p_vs_cov/cov.vs.pident.txt", OUTPUT.FIGURES.PATH))
spectral.scale = rev(RColorBrewer::brewer.pal(11, "Spectral"))
fig3a = ggplot() + 
  stat_binhex(data = annotated.hhr.table, bins = 50, aes(x=pident, y=max.cov)) + 
  geom_rect(aes(xmin = MINIMUM.PIDENT.FOR.PAIRWISE.HIT/100, xmax = 1, 
            ymin = 0, ymax = MINIMUM.COV.FOR.PROTEIN.SIMILARITY, 
            color = "red"), fill = NA, size = 2) + guides(color = "none") +
  geom_text(aes(x = mean(c(MINIMUM.PIDENT.FOR.PAIRWISE.HIT/100, 1)), 
                y = mean(c(0, MINIMUM.COV.FOR.PROTEIN.SIMILARITY)), 
                label = "Mosaic\n pairs", col = "red")) +
  scale_fill_gradientn(colours = spectral.scale, trans = 'log10') + 
  scale_x_continuous(limits = c(0,1)) + 
  scale_y_continuous(limits = c(0,1)) + 
  theme_bw() + 
  labs(x = "Percentage identity", y = "Pairwise sequence coverage", fill = "no. pairs")
fig3a
ggsave(sprintf("%sFigure_Supplementary/FigS5_p_vs_cov/panelA_pairwise-seq.png", OUTPUT.FIGURES.PATH), fig3a, bg = "white",  width = 5, height = 4)
  
```

#join fig 3 panels
```{r}


recent_HGT_pairs_70 = data.table::fread(sprintf("%sFigure_Supplementary/FigS11_recent_HGT/recent_HGT_network_70.txt", OUTPUT.FIGURES.PATH)) %>%
  select(from, to, num.prot.pairs)
recent_HGT_pairs_90 = data.table::fread(sprintf("%sFigure_Supplementary/FigS11_recent_HGT/recent_HGT_network_90.txt", OUTPUT.FIGURES.PATH)) %>%
  select(from, to, num.prot.pairs)
annotation.network.metadata.all.cat = data.table::fread( 
                   sprintf("%sFigure_Supplementary/FigS2_annot_network/annotation.network.metadata.all.cat.txt", OUTPUT.FIGURES.PATH)) %>%
  rowwise() %>%
  mutate(size = min(size, 15))




pdf(sprintf("%s/Figure6/Figure6b_recentHGT.pdf", OUTPUT.FIGURES.PATH), width = 12, height = 24) #, units = "cm", res = 300)
VisualiseIgraphnetwork(data.for.network = recent_HGT_pairs_70 %>% mutate(weight = log10(num.prot.pairs)/10) %>% as.data.frame(),
                      edge.colors = NULL,
                      node.color.map = annotation.network.metadata.all.cat %>%
                                      select(node, color),
                      # node these sizes are defined at high threshold
                      node.size.map = annotation.network.metadata.all.cat %>%
                                      select(node, size), #%>% mutate(size=1),
                      node.labels = annotation.network.metadata.all.cat %>% 
                    select(node, label),
                      line.color.variable = "from",
                      line.color.legend.name = NULL,
                      font.size = 0.75,
                      vertex.size = 1,
                      main = "",
                      vertex.label.dist = 1,
                      include.labels = TRUE,
                      alpha = 0.3)
dev.off()



```


```{r}

pdf(sprintf("%sFigure_Supplementary/FigS11_recent_HGT//network_recent_HGT_igraph_70.pdf", OUTPUT.FIGURES.PATH), width = 6, height = 12) #, units = "cm", res = 300)
par(mfrow=c(2,1))
VisualiseIgraphnetwork(data.for.network = recent_HGT_pairs_70 %>% mutate(weight = log10(num.prot.pairs)/3),
                      edge.colors = NULL,
                      node.color.map = annotation.network.metadata.all.cat%>%
                                      select(node, color),
                      # node these sizes are defined at high threshold
                      node.size.map = annotation.network.metadata.all.cat %>%
                                      select(node, size),
                      node.labels = annotation.network.metadata.all.cat %>% 
                    select(node, label),
                      line.color.variable = "from",
                      line.color.legend.name = NULL,
                      font.size = 0.75,
                      vertex.size = 1,
                      main = "",
                      vertex.label.dist = 1,
                      include.labels = TRUE,
                      alpha = 0.3)
dev.off()


pdf(sprintf("%sFigure_Supplementary/FigS11_recent_HGT//network_recent_HGT_igraph_90.pdf", OUTPUT.FIGURES.PATH), width = 6, height = 6) # units = "cm", res = 300)
VisualiseIgraphnetwork(data.for.network = recent_HGT_pairs_90 %>% mutate(weight = log10(num.prot.pairs)/3),
                      edge.colors = NULL,
                      node.color.map = annotation.network.metadata.all.cat%>%
                                      select(node, color),
                      # node these sizes are defined at high threshold
                      node.size.map = annotation.network.metadata.all.cat %>%
                                      select(node, size),
                      node.labels = annotation.network.metadata.all.cat %>% 
                    select(node, label),
                      line.color.variable = "from",
                      line.color.legend.name = NULL,
                      font.size = 0.75,
                      vertex.size = 1,
                      main = "",
                      vertex.label.dist = 1,
                      include.labels = TRUE,
                      alpha = 0.3)
dev.off()
```

# recently transferred domains (recent HGT & domains)
```{r}
text_size.x = 6
text_size.y = 6
recent_HGT_pairs_raw.domain.data.all.pident70 = read.csv(sprintf("%sFigure_Supplementary/recent_HGT_domains.pident70.csv", OUTPUT.FIGURES.PATH))
recent_HGT_pairs_raw.domain.data.to.plot.all.pident70 = recent_HGT_pairs_raw.domain.data.all.pident70 %>%
  filter(n_pairs > 5)
recent_HGT_pairs_raw.domain.data.to.plot.all.pident70$t_name = factor(recent_HGT_pairs_raw.domain.data.to.plot.all.pident70$t_name, levels = unique(recent_HGT_pairs_raw.domain.data.to.plot.all.pident70$t_name))
p=ggplot(recent_HGT_pairs_raw.domain.data.to.plot.all.pident70) + geom_col(aes(x = t_name, y = n_pairs, fill = h_name), col = "black") +
  theme_bw() +
    theme(axis.text.x = element_text(angle=45, hjust =1,vjust=1, size = text_size.x),
        axis.text.y = element_text(hjust = 1, size = text_size.y),
        legend.position = "left") +
  scale_y_sqrt(breaks = c(2,10,25,50,100,150)) +
  xlab("") +
  ylab("Num. protein pairs")
ggsave(sprintf("%sFigure_Supplementary/recent_HGT_domains_pident70.pdf", OUTPUT.FIGURES.PATH), p, width = 15, height = 7)


recent_HGT_pairs_raw.domain.data.to.plot.all.pident90 = read.csv(sprintf("%sFigure_Supplementary/recent_HGT_domains.pident90.csv", OUTPUT.FIGURES.PATH))
recent_HGT_pairs_raw.domain.data.to.plot.all.pident90 = recent_HGT_pairs_raw.domain.data.to.plot.all.pident90 %>%
  filter(n_pairs > 2)
recent_HGT_pairs_raw.domain.data.to.plot.all.pident90$t_name = factor(recent_HGT_pairs_raw.domain.data.to.plot.all.pident90$t_name, levels = unique(recent_HGT_pairs_raw.domain.data.to.plot.all.pident90$t_name))
p=ggplot(recent_HGT_pairs_raw.domain.data.to.plot.all.pident90) + geom_col(aes(x = t_name, y = n_pairs, fill = h_name), col = "black") +
  theme_bw() +
  #facet_grid(stat~., scales = "free") +
    theme(axis.text.x = element_text(angle=45, hjust =1,vjust=1, size = text_size.x),
        axis.text.y = element_text(hjust = 1, size = text_size.y),
        legend.position = "left") +
  scale_y_sqrt(breaks = c(2,10,25,50,100,150)) +
  xlab("") +
  ylab("Num. protein pairs")
ggsave(sprintf("%sFigure_Supplementary/recent_HGT_domains_pident90.pdf", OUTPUT.FIGURES.PATH), p, width = 15, height = 7)
```


########### OLD CODE #############

```{r, eval = FALSE}
diversity.data.per.category = data.table::fread(sprintf("%sFigure3/mosaicism.per.category.txt", OUTPUT.FIGURES.PATH))
diversity.data.per.category$category[diversity.data.per.category$category == "DNA, RNA and nucleotide metabolism"] = "DNA, RNA\nand nucleotide metabolism"
diversity.data.per.category$category[diversity.data.per.category$category == "moron, auxiliary metabolic gene and host takeover"] =
"moron, auxiliary metabolic\ngene and host takeover"
fug4c = Plot.Mosaicism.Table(diversity.data.per.category %>% 
                               mutate(annotation = category) %>%
                               filter(proportion %in% names(diversity.labeller)), 
                             PHROG.COLOR.MAP = PHROG.COLOR.MAP2)  +
    facet_grid(. ~ proportion, scales = "free", space = "free_y", 
                labeller = labeller(proportion = diversity.labeller)) +
  theme(plot.margin = unit(c(0, 6, 0, 0), "cm"))


p.fig4 <- plot_grid(Fig4b, p.most.diverse,fug4c, ncol = 1, labels = c("A", "B", "C"), label_size = 18, rel_heights = c(0.3, 0.5, 0.2))
p.fig4.filename <- sprintf("%sFigure4/Figure4.png", OUTPUT.FIGURES.PATH)
ggsave(p.fig4.filename, p.fig4, bg = "white", width = 12, height = 15)
```

```{r, eval = FALSE}
#ggsave(filename = sprintf("%sFigure3/low.cov.odds.ratio.by.domain.jpg", OUTPUT.FIGURES.PATH) ,p.odds.low.cov.by.domain , width = 10, height = 7)


#ss.domain.or.seq= ss.both %>%
#  left_join(ss.domain %>% select(annotation, category) %>% mutate(top.odd.by.domain = TRUE)) %>%
#  mutate(color = if_else(is.na(top.odd.by.domain),"red",  "black"),
#         label = if_else(is.na(top.odd.by.domain),"*",  "")) %>% 
#  arrange(ecod.coverage.type2, desc(odds.ratio))
#ss.domain.or.seq$annotation = factor(ss.domain.or.seq$annotation, levels = unique(ss.domain.or.seq %>% arrange(ecod.coverage.type2, odds.ratio) %>% pull(annotation)))

p.odds.low.cov =ggplot(ss.domain.or.seq, 
                        aes(x=annotation, y = odds.ratio, fill = category))  +
  geom_bar(stat="identity", width = 0.8) + 
    geom_text(aes(label = label), nudge_y = 1) +
  scale_fill_manual(values = PHROG.COLOR.MAP2) +
  theme_minimal() + 
  #guides(fill = "none") + 
  theme(legend.text=element_text(size=8)) + 
  labs(x = "", y = "Odds ratio", fill = "category") + 
  coord_flip() +
  facet_grid(ecod.coverage.type2 ~ mosaicism.type, scales = "free", space = "free")
p.odds.low.cov
ggsave(filename = sprintf("%sFigure3/low.cov.odds.ratio.by.domainor.seq.jpg", OUTPUT.FIGURES.PATH) ,p.odds.low.cov , width = 10, height = 7)



p.odds.low.cov =ggplot(ss  %>% 
                         group_by(annotation) %>%
                         mutate(any.high.odds = any(odds.ratio > 2.5)) %>%
                          distinct(annotation, category, odds.ratio,ecod.coverage.type2, mosaicism.type, any.high.odds, mosaicism.type) %>%
                          filter(mosaicism.type == "Mosaic by domain or sequence" | mosaicism.type == "Mosaic by domain") %>%
                         filter(any.high.odds), 
                        aes(x=annotation, y = odds.ratio, fill = category)) +
  geom_bar(stat="identity", width = 0.8) + 
  scale_fill_manual(values = PHROG.COLOR.MAP2) +
  #theme_minimal() + 
  #guides(fill = "none") + 
  theme(legend.text=element_text(size=8)) + 
  labs(x = "", y = "Odds ratio", fill = "category") + 
  coord_flip() +
  facet_grid(ecod.coverage.type2 ~ ., scales = "free", space = "free_y")
p.odds.low.cov
ggsave(filename = sprintf("%sFigure3/low.cov.odds.ratio.by.both.jpg", OUTPUT.FIGURES.PATH) ,p.odds.low.cov , width = 10, height = 7)
```

```{r, eval = FALSE}
jpeg(filename = sprintf("%sFigure3/Figure3a_main_network.jpg", OUTPUT.FIGURES.PATH), width = 10, height = 10, units = "cm", res = 300)
VisualiseIgraphnetwork(data.for.network = mosaic.network %>% 
                         mutate(weight = log10(num.reprseq.any.mossaic)/20) %>% 
                         select(from = from.index, to = to.index, weight, dataset),
                      edge.colors = data.frame(value = c("ecod", "seq", "both"), color = c("gray", "gray", "darkred")),
                      edge.linetypes = data.frame(value = c("ecod", "seq", "both"), lty = c("solid", "dotted", "solid")),
                      node.color.map = annotation.network.metadata%>%
                                      select(node, color),
                      # node these sizes are defined at high threshold
                      node.size.map = annotation.network.metadata %>%
                                      select(node, size) %>% mutate(size = 2*size),
                      node.labels = annotation.network.metadata %>% 
                    select(node, label),
                      line.color.variable = "dataset",
                      line.color.legend.name = "Mosaicism via",
                      font.size = 0.25,
                      main = "",
                      vertex.label.dist = 1,
                      include.labels = TRUE,
                      alpha = 1,
                    label.font = 2,
                    legend.title = "Mosaicism via:",
                    node.label.color = "black")
dev.off()
```

