---
title: "Untitled"
author: "Bogna Smug"
date: '2023-08-08'
output: html_document
---

```{r setup, include=FALSE}
OUTPUT.FIGURES.METADATA.PATH = sprintf("%s/Metadata_analysis",OUTPUT.FIGURES.PATH)
dir.create(OUTPUT.FIGURES.METADATA.PATH)
knitr::opts_chunk$set(echo = TRUE)
unknown_words = c("unclassified", "unknown", "unspecified", "")
# use Genus and Family for old NCBI taxonomy

reprseq.metadata.table = data.table::fread(file = sprintf("%sreprseq.metadata.table_p09.txt", OUTPUT.DATA.PATH)) %>%
  select(-genus, - family) %>%
  rename(genus = genus_ICTV38, family = family_ICTV38) %>%
  mutate(genus = if_else(genus == "", "unknown", genus),
         family = if_else(family == "", "unknown", family))

#sreprseq.metadata.table.txt for THRESHOLDS 0.75 and 0.25
surely.annotated.proteins = data.table::fread(sprintf("%ssurely.annotated.proteins.txt", OUTPUT.DATA.PATH))
families = data.table::fread(file = sprintf("%sfamilies.txt", OUTPUT.DATA.PATH)) 
included.category.sizes = data.table::fread(file = sprintf("%sincluded.category.sizes",OUTPUT.DATA.PATH))
 
diversity = data.table::fread(sprintf("%sFigure4/diversity.top.diverse.classes.txt", OUTPUT.FIGURES.METADATA.PATH))
top.diversity.classes = diversity %>% filter(data.type == "top.diversity") %>% distinct(annotation.index)
all.proteins = data.table::fread(sprintf("%sall.proteins.txt", OUTPUT.DATA.PATH))
name.table = read.table(NAME_TABLE_PATH, sep = ",", header = TRUE)
metadata = data.table::fread(file = sprintf("%smetadata.txt", OUTPUT.DATA.PATH)) %>%
  select(-Genus, - Family) %>%
  rename(Genus = Genus_ICTV_38, Family = Family_ICTV38) %>%
  mutate(Genus = if_else(Genus == "", "unknown", Genus),
         Family = if_else(Family == "", "unknown", Family))


unaccepted.genes = data.table::fread(file = sprintf("%sunaccepted.genes.txt",OUTPUT.DATA.PATH))
ecod.domains.hits = data.table::fread(sprintf("%secod.domains.hits.txt",OUTPUT.DATA.PATH)) 
ecod_id_to_names_map = data.table::fread(sprintf("%secod_id_to_names_map.txt",OUTPUT.DATA.PATH)) %>% as.data.frame() %>% mutate(id = as.numeric(id))
# pairs
seq.mosaic.pairs.all.pident = data.table::fread(sprintf("%sseq.mosaic.pairs.all.pident.txt", OUTPUT.DATA.PATH))
qname.pairs.sharing.t.and.disshare.x = data.table::fread(sprintf("%sqname.pairs.sharing.t.and.disshare.x.txt", OUTPUT.DATA.PATH))
recent_HGT_pairs_raw.70 = data.table::fread(sprintf("%sFigure_Supplementary/network_recent_HGT/recent_HGT_pars_70.txt", OUTPUT.FIGURES.METADATA.PATH))
mid.annotated.proteins.including.multi.annot = data.table::fread(sprintf("%smid.annotated.proteins.including.multi.annot",OUTPUT.DATA.PATH)) %>%
  left_join(families) 

families.and.functions = surely.annotated.proteins %>% 
  inner_join(included.category.sizes) %>%
  rename(fam = family) %>% 
  distinct(fam, annotation.index) 

relaxely.annotated.proteins.including.multi.annot =data.table::fread(sprintf("%srelaxely.annotated.proteins.including.multi.annot",OUTPUT.DATA.PATH)) %>%
  left_join(families)  
mid.annotated.proteins.including.multi.annot =data.table::fread(sprintf("%smid.annotated.proteins.including.multi.annot",OUTPUT.DATA.PATH)) %>%
  left_join(families)  

surely.annotated.proteins.from.included.categories = surely.annotated.proteins %>%
  inner_join(included.category.sizes) %>%  select(qname, family, annotation.index)

surely.annotated.ecod.domains.high.confidence = ecod.domains.hits %>%
  select(qname, qstart, qend, qlength, sstart, send, slength, x_id, h_id, t_id, f_id) %>%
  mutate(h_id = as.character(h_id)) %>%
  # note that some families may be unclassified i.e. d_id = NA
  inner_join(surely.annotated.proteins.from.included.categories)  


name.table.with.metadata = name.table %>%
    mutate(Accession =sub("\\..*", "", ncbi.id)) %>%
  left_join(metadata, by = "Accession")
name.table.with.metadata$lifestyle[is.na(name.table.with.metadata$lifestyle) | name.table.with.metadata$lifestyle %in% unknown_words] = "unknown"
name.table.with.metadata$Family[is.na(name.table.with.metadata$Family) | name.table.with.metadata$Family %in% unknown_words] = "unknown"
name.table.with.metadata$Genus[is.na(name.table.with.metadata$Genus) | name.table.with.metadata$Genus %in% unknown_words] = "unknown"

annotated.families = surely.annotated.proteins %>% 
  inner_join(included.category.sizes) %>%
    left_join(families) %>%
    distinct(family) %>%
    rename(fam = family)

all.proteins.in.genomes = name.table %>%
            mutate(Accession =sub("\\..*", "", ncbi.id)) %>% 
            left_join(metadata, by = "Accession") %>%
           left_join(all.proteins %>% select(repr.name = qname, has.unknown.function, mosaic, mosaic.domain))

genome.sizes.per.family = all.proteins.in.genomes %>%
  group_by(Accession, lifestyle, Genus, Family) %>%
  summarise(genome.size = n_distinct(ncbi.id)) %>%
  group_by(Family) %>%
  summarise(av.genome.size = ceiling(median(genome.size)))
```




```{r}
# number of rHMMs with defined genus/ lifestyle
# we somehow lost 1413 genomes i Wanangwa's table
reprseq.metadata.table %>%
  group_by(lifestyle) %>%
  summarise(n = n()) %>%
  arrange(desc(n))

reprseq.metadata.table %>%
  group_by(host) %>%
  summarise(n = n()) %>%
  arrange(desc(n))
# the data is hugely biased abouk 130k rHMMs rHmms are from Group 1, kingdom = Heunggongvirae, phylum = Uroviricota, class = Caudoviricetes, 1411 are NA (missing data), 1139 have multiple taxa ssigned and only < 300 belong to other taxa.
# When we look at genus  22748 are unclassified, 18199 come from multiple genuses, and 1554 are missing, the other genuses have < 1000 rHMMs
# In terms of order almost 130k are unclassified
# interestingly only the family seems reasonable: bouk 70k are unclassified but the rest is distributed evenly, and rHMMs correspond to one usually
reprseq.metadata.table %>%
  group_by(baltimore.group) %>%
  summarise(n = n()) %>%
  arrange(desc(n))


reprseq.metadata.table %>%
  group_by(kingdom) %>%
  summarise(n = n()) %>%
  arrange(desc(n))

reprseq.metadata.table %>%
  group_by(phylum) %>%
  summarise(n = n()) %>%
  arrange(desc(n))

reprseq.metadata.table %>%
  group_by(class) %>%
  summarise(n = n()) %>%
  arrange(desc(n))

reprseq.metadata.table %>%
  group_by(rder) %>%
  summarise(n = n()) %>%
  arrange(desc(n))

reprseq.metadata.table %>%
  group_by(family) %>%
  summarise(n = n()) %>%
  arrange(desc(n))

reprseq.metadata.table %>%
  group_by(genus) %>%
  summarise(n = n()) %>%
  arrange(desc(n))
  

# there are 62 missing genomes, some have been removed from NCBI, some may be too short
missing.genomes = anti_join( name.table %>%
                              mutate(Accession =sub("\\..*", "", ncbi.id)) %>%
                              distinct(Accession),
                            metadata, by = "Accession") 


```
RNA vs DNA genomes, we know that many RNA genomes are missing in our table
```{r}
ggplot(all.proteins.in.genomes %>% group_by(GeneticMaterial) %>% summarise(num.genomes = n_distinct(Accession))) +
  geom_col(aes(x=GeneticMaterial, y = num.genomes)) +
#+ scale_y_log10() + 
   theme_bw()

ggplot(reprseq.metadata.table %>% group_by(molecule) %>% summarise(num.rHMMs = n_distinct(repr.name))) +
  geom_col(aes(x=molecule, y = num.rHMMs)) +
   scale_y_log10() + 
   theme_bw()
```


```{r}


families.metadata.table = reprseq.metadata.table %>%
  left_join(families %>% select(repr.name = qname, fam = family), by = "repr.name") %>%
  arrange(fam) %>%
  group_by(fam) %>%
  summarise(homogeneous.family = length(unique(family[family != "Unclassified"])) <= 1,
            family = unique.unclassified.rm(family),
            homogeneous.lifestyle = length(unique(lifestyle))==1,
            lifestyle = unique.unclassified.rm(lifestyle),
            num.reprseq.in.fam = n_distinct(repr.name),
            genus = unique.unclassified.rm(genus),
            host = unique.unclassified.rm(host)) %>%
  ungroup() 

annotated.families.metadata.table= families.metadata.table %>%
    inner_join(annotated.families) 


annotated.families.metadata.table %>%
  group_by(lifestyle) %>%
  summarise(n = n()) %>%
  arrange(desc(n))


annotated.families.metadata.table %>%
  group_by(family) %>%
  summarise(n = n()) %>%
  arrange(desc(n))


families.metadata.table %>%
  group_by(host) %>%
  summarise(n = n()) %>%
  arrange(desc(n))


families.metadata.table %>%
  group_by(lifestyle) %>%
  summarise(n = n()) %>%
  arrange(desc(n))


families.metadata.table %>%
  group_by(family) %>%
  summarise(n = n()) %>%
  arrange(desc(n))

all.families = all.proteins %>% 
  group_by(family) %>%
  summarise(mosaic = any(mosaic),
            mosaic.domain = any(mosaic.domain))


# noye that those from uninlded annotations will be called unannotated here
r1=families %>%
  filter(!(qname %in% unaccepted.genes)) %>%
  left_join(surely.annotated.proteins) %>%
  left_join(included.category.sizes) %>%
  mutate(include = if_else(is.na(annotation.index) | is.na(num.proteins), FALSE, include)) %>%
  left_join(all.families) %>%
  group_by(family, mosaic) %>%
  summarise(annotated = any(include)) %>%
  group_by(annotated, mosaic) %>%
  summarise(n = n_distinct(family))


r2=families %>%
  filter(!(qname %in% unaccepted.genes)) %>%
  distinct(qname) %>%
  left_join(surely.annotated.proteins) %>%
  left_join(included.category.sizes) %>%
  mutate(annotated = if_else(is.na(annotation.index) | is.na(num.proteins), FALSE, include)) %>%
  distinct(qname, annotated) %>%
  left_join(all.proteins) %>%
  group_by(annotated, mosaic) %>%
  summarise(n = n_distinct(qname))

g=ggplot(r1) +
  geom_col(aes(alpha=annotated, x = mosaic, y = n), fill = "darkblue") + theme_bw() + ylab("Num. tHMM families") +
  scale_alpha_manual(values = c("TRUE" = 1, "FALSE" = 0.5))
ggsave(g, filename = sprintf("%sannotated_vs_mosaicism.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 6, height = 4)  
g
```

```{r}

num.genomes.per.tax.fam = all.proteins.in.genomes %>%  
           left_join(name.table) %>%
            mutate(Accession =sub("\\..*", "", ncbi.id)) %>% 
            mutate(Family = if_else(is.na(Family), "unknown", Family)) %>%
            group_by(Family) %>% 
            summarise(n.genomes = n_distinct(Accession)) %>%
           ungroup() %>%
          mutate(big.fam = n.genomes >= 50)
num.genomes.per.tax.fam2 = num.genomes.per.tax.fam %>%
  mutate(Family = if_else(big.fam, "other", Family)) %>%
            group_by(Family) %>% 
            summarise(n.genomes = sum(n.genomes)) %>%
           ungroup() %>%
           arrange(desc(n.genomes))
num.genomes.per.tax.fam2$Family = factor(num.genomes.per.tax.fam2$Family, levels = unique(num.genomes.per.tax.fam2$Family))
g=ggplot(num.genomes.per.tax.fam2) +
    #geom_jitter(aes(x=lifestyle, y=n.genes.in.genome,  col = lifestyle), size = 0.5) + 
    #geom_boxplot(aes(x=lifestyle, y=n.genes.in.genome)) + 
geom_col(aes(x = Family, y = n.genomes), fill = "white", col = "darkblue") +
  xlab("") +
  #ylab("Num genes in genome") +
  theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) 
ggsave(g, filename = sprintf("%snum_genomes_per_family.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 12, height = 10)  
g
```



# fams size per lifestyle
```{r}
# genome size per lifestyle


g=ggplot(reprseq.metadata.table %>%  
           left_join(name.table) %>%
            mutate(Accession =sub("\\..*", "", ncbi.id)) %>% 
           left_join(metadata %>% select(Accession, Jumbophage)) %>%
            group_by(lifestyle, Accession, Jumbophage) %>% 
            summarise(n.genes.in.genome = n_distinct(repr.name)) %>%
           filter(lifestyle %in% c("temperate", "virulent"))) + 
    #geom_jitter(aes(x=lifestyle, y=n.genes.in.genome,  col = lifestyle), size = 0.5) + 
    #geom_boxplot(aes(x=lifestyle, y=n.genes.in.genome)) + 
geom_density(aes(x = n.genes.in.genome, fill = lifestyle), alpha = 0.3) +
  xlab("Num genes in genome") +
  #ylab("Num genes in genome") +
  scale_fill_manual(values = c("temperate" = "blue", "virulent" = "red", "multi" = "green", "unknown" = "gray70", "missing data"= "gray50")) + 
  scale_color_manual(values = c("temperate" = "blue", "virulent" = "red", "multi" = "green", "unknown" = "gray70", "missing data"= "gray50")) + 
  theme_bw() +
  facet_grid(.~Jumbophage)
ggsave(g, filename = sprintf("%sgenomesize_vs_lifestyles1.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 12, height = 10)  
g

g=ggplot(reprseq.metadata.table %>%  
           left_join(name.table) %>%
            mutate(Accession =sub("\\..*", "", ncbi.id)) %>% 
            group_by(lifestyle, Accession) %>% 
            summarise(n.genes.in.genome = n_distinct(repr.name)) %>%
           filter(lifestyle %in% c("temperate", "virulent"))) + 
    geom_jitter(aes(x=lifestyle, y=n.genes.in.genome,  col = lifestyle), size = 0.5) + 
    geom_boxplot(aes(x=lifestyle, y=n.genes.in.genome)) + 
  ylab("Num genes in genome") +
  scale_fill_manual(values = c("temperate" = "blue", "virulent" = "red", "multi" = "green", "unknown" = "gray70", "missing data"= "gray50")) + 
  scale_color_manual(values = c("temperate" = "blue", "virulent" = "red", "multi" = "green", "unknown" = "gray70", "missing data"= "gray50")) + 
  theme_bw()
ggsave(g, filename = sprintf("%sgenomesize_vs_lifestyles2.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 12, height = 10)  
g




g=ggplot(reprseq.metadata.table %>% group_by(lifestyle) %>% summarise(n = n_distinct(repr.name))) + geom_col(aes(x=lifestyle, y = n, fill = lifestyle)) + ylab("Num rHMMs") +
      scale_fill_manual(values = c("temperate" = "blue", "virulent" = "red", "multi" = "green", "unknown" = "gray70", "missing data"= "gray50")) + 
  theme_bw()
ggsave(g, filename = sprintf("%srHMMs_vs_lifestyles.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 12, height = 10)  
g

g=ggplot(families.metadata.table %>% group_by(lifestyle) %>% summarise(n = n_distinct(fam))) + geom_col(aes(x=lifestyle, y = n, fill = lifestyle)) + ylab("Num rHMM families") +
      scale_fill_manual(values = c("temperate" = "blue", "virulent" = "red", "multi" = "green", "unknown" = "gray70", "missing data"= "gray50")) + 
  theme_bw()
ggsave(g, filename = sprintf("%srHMMfamilies_vs_lifestyles.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 12, height = 10)  
g

g=ggplot(reprseq.metadata.table %>%
         left_join(families %>% select(fam = family, repr.name = qname)) %>% 
         left_join(name.table) %>%
         mutate(Accession =sub("\\..*", "", ncbi.id)) %>%
         distinct(Accession) %>%
         left_join(metadata) %>%
         group_by(lifestyle) %>% summarise(nyum.genomes = n_distinct(Accession)) %>%
         mutate(lifestyle = if_else(is.na(lifestyle), "missing data", lifestyle))) + 
    geom_col(aes(x=lifestyle, y = nyum.genomes, fill = lifestyle)) + 
    ylab("Num genomes") +
    scale_fill_manual(values = c("temperate" = "blue", "virulent" = "red", "multi" = "green", "unknown" = "gray70", "missing data"= "gray50")) + 
  theme_bw()
ggsave(g, filename = sprintf("%sgenomes_vs_lifestyles.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 12, height = 10)  
g



```


# Number of similar sequences per lifestyle
Clusters of similar sequences in temperate and virulent phages are rather similar.
Families are mostly singletons in both cases


```{r}
# see how many temperate and virulent rHMMs are annotated
reprseq.metadata.table %>% 
  distinct(repr.name, lifestyle) %>%
  left_join(surely.annotated.proteins %>% select(repr.name = qname, annotation.index)) %>%
  left_join(included.category.sizes %>% distinct(annotation.index) %>% mutate(include = TRUE)) %>%
  mutate(include = if_else(!is.na(annotation.index) & !is.na(include), include, FALSE)) %>%
  mutate(known.annnot = include) %>%
  group_by(lifestyle) %>%
  summarise(nun.annotated = sum(known.annnot),
         num.unannotated = sum(!known.annnot))

ggplot(families.metadata.table %>% filter(lifestyle %in% c("temperate", "virulent")) %>% left_join(families.and.functions) %>% mutate(annotated = if_else(!is.na(annotation.index), "annotated", "un-annotated"))) +
  geom_histogram( aes(x=num.reprseq.in.fam, fill = lifestyle), alpha = 0.5) +
  #geom_density(       aes(x=num.reprseq.in.fam, fill = lifestyle), alpha = 0.5) +
  #geom_boxplot(aes(y=num.reprseq.in.fam, x = lifestyle, fill = lifestyle)) +
  theme_bw()  +
  #scale_y_log10()
  xlim(c(0,50)) + 
  #scale_y_log10() 
  scale_y_log10() +
  facet_grid(lifestyle ~ annotated)


dd=rbind(seq.mosaic.pairs.all.pident %>% select(qname, sname, share.any.fragment),
      seq.mosaic.pairs.all.pident %>% select(qname=sname, sname=qname, share.any.fragment)) %>%
  group_by(qname) %>%
  summarise(num.seq.sharing.any.fragment = n_distinct(sname[share.any.fragment])) %>%
  right_join(families) %>%
  filter(!(qname  %in% unaccepted.genes)) %>%
  left_join(surely.annotated.proteins) %>%
  left_join(included.category.sizes) %>%
  left_join(reprseq.metadata.table %>% select(qname = repr.name, lifestyle = lifestyle)) %>%
  mutate(annotated = if_else(is.na(num.proteins), "un-annotated", "annotated"),
         num.seq.sharing.any.fragment = if_else(is.na(num.seq.sharing.any.fragment), 0, num.seq.sharing.any.fragment))


ggplot(dd %>% filter(lifestyle %in% c("temperate", "virulent"))) +
  #geom_histogram() +
  #geom_density(alpha = 0.5,aes(x=num.seq.sharing.any.fragment, fill = qlifestyle)) +
    geom_boxplot(aes(y=num.seq.sharing.any.fragment, x = lifestyle, fill = lifestyle)) +
    facet_grid(. ~ annotated) +
  theme_bw() + 
  #xlim(c(0,300)) + scale_y_sqrt() +
    scale_y_log10()
  #facet_grid(qlifestyle ~ .)
  
  
  
  dd %>% filter(lifestyle %in% c("temperate", "virulent")) %>%
                  mutate(singleton = num.seq.sharing.any.fragment < 1) %>%
    group_by(annotated, lifestyle, singleton) %>%
    summarise(n = n_distinct(qname)) %>%
    ggplot() + geom_col(aes(x = annotated, y = n, fill = singleton)) +
    scale_alpha_manual(values = c("annotated" = 1, "un-annotated" = 0.5)) + theme_bw() + facet_grid(lifestyle ~.)
```


# No let's look at families with at least one annotation
# there is much less consensus wihin these!
```{r}


g=ggplot(families.metadata.table %>% group_by(lifestyle) %>% summarise(n = n_distinct(fam))) + geom_col(aes(x=lifestyle, y = n, fill = lifestyle)) + ylab("Num rHMM families") +
      scale_fill_manual(values = c("temperate" = "blue", "virulent" = "red", "multi" = "green", "unknown" = "gray70", "missing data"= "gray50")) + 
  theme_bw()
ggsave(g, filename = sprintf("%srHMMfamilies_vs_lifestyles.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 12, height = 10)  
g

d1 = surely.annotated.proteins %>% 
#  inner_join(included.category.sizes) %>%
    inner_join(included.category.sizes %>% distinct(annotation.index)) %>%
    right_join(families) %>%
    rename(fam = family) %>%
  group_by(fam) %>%
  summarise(annotated = any(!is.na(annotation.index))) %>%
  ungroup() %>%
  left_join(families.metadata.table) %>%
  group_by(lifestyle) %>% 
  summarise(n = n_distinct(fam),
            n.annot = n_distinct(fam[annotated]),
            n.unannot = n_distinct(fam[!annotated])) %>%
  mutate(prop.annotated = n.annot/n)
g=ggplot(d1 %>%
         rename(annotated = n.annot, unannotated = n.unannot) %>%
         tidyr::gather(key = "annotated", value = "n", annotated, unannotated)) + 
  geom_col(aes(x=lifestyle, y = n, fill = lifestyle, alpha = annotated)) + ylab("Num rHMM families") +
      scale_fill_manual(values = c("temperate" = "blue", "virulent" = "red", "multi" = "green", "unknown" = "gray70", "missing data"= "gray50")) + 
  theme_bw() +
  scale_alpha_manual(values = c("annotated" = 1, "unannotated" = 0.5))
ggsave(g, filename = sprintf("%srHMMfamilies_vs_lifestyles_annot.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 12, height = 10)  
g





```


# Let's look ata family*annotation
```{r, fig.width = 12, fig.height = 8}
annotated.proteins.with.metadata = surely.annotated.proteins %>% 
  inner_join(included.category.sizes) %>%
    left_join(families) %>%
  rename(repr.name = qname, fam = family) %>%
  left_join(reprseq.metadata.table) %>%
  left_join(annotation.indices)


families.per.annotation.index = annotated.proteins.with.metadata %>%
  distinct(fam, annotation.index, num.families)

families.per.annotation.metadata = annotated.families.metadata.table %>%
  left_join(families.per.annotation.index, by = "fam") %>%
  group_by(annotation.index, lifestyle, num.families) %>%
  summarise(no.families = n()) %>%
  left_join(annotation.indices) %>%
  inner_join(top.diversity.classes) %>%
  mutate(prop.families = no.families/num.families) %>%
  tidyr::gather(key = 'data.type', value = 'value', prop.families, no.families)

families.per.annotation.metadata$lifestyle = factor(families.per.annotation.metadata$lifestyle,
                                                    levels =  c("unknown", "multi", "temperate","virulent"))
g=ggplot(families.per.annotation.metadata) +
  geom_col(aes(x=annotation, fill = lifestyle,  y = value), stack = "dodge") +
  coord_flip() +
  theme_bw() +
  scale_fill_manual(values = c("temperate" = "blue", "virulent" = "red", "multi" = "green", "unknown" = "gray50")) +
    facet_grid(category~data.type, scales = "free", space = "free_y")
g
ggsave(g, filename = sprintf("%sannotations_vs_lifestyles.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 12, height = 10)  
```


# now let's see if we can do the same for taxonomy
```{r}
NUM.FAM.THRESHOLD = 3

families.per.annotation.metadata.fam = annotated.families.metadata.table %>%
  left_join(families.per.annotation.index, by = "fam") %>%
  group_by(annotation.index, family, num.families) %>%
  summarise(no.families = n()) %>%
  ungroup() %>%
  mutate(family = if_else(no.families <= NUM.FAM.THRESHOLD, "other", family)) %>%
  group_by(annotation.index, family, num.families) %>%
  summarise(no.families = sum(no.families)) %>%
  ungroup() %>%
  left_join(annotation.indices) %>%
  inner_join(top.diversity.classes) %>%
  mutate(prop.families = no.families/num.families) %>%
  tidyr::gather(key = 'data.type', value = 'value', prop.families, no.families)


colormap.families = unique(families.per.annotation.metadata.fam$family)
colormap.colors =  colorRampPalette(brewer.pal(12, "Set1"))(length(colormap.families))
colormap = data.frame(value = colormap.families, color = colormap.colors)
colormap$color[colormap$value == "multi"] = "green"
colormap$color[colormap$value == "Unclassified"] = "gray90"
colormap$color[colormap$value == "other"] = "gray50"
colm = colormap$color
names(colm) = colormap$value


families.per.annotation.metadata.fam$family = factor(families.per.annotation.metadata.fam$family,
                                                    levels =  c("Unclassified","multi","other", setdiff(unique(families.per.annotation.metadata.fam$family), c("multi", "Unclassified", "other"))))
g=ggplot(families.per.annotation.metadata.fam) +
  geom_col(aes(x=annotation, fill = family,  y = value), stack = "dodge") +
  coord_flip() +
  theme_bw() +
   facet_grid(category~data.type, scales = "free", space = "free_y") +
  scale_fill_manual(values = colm)
g
ggsave(g, filename = sprintf("%sannotations_vs_taxonomy_family.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 12, height = 10)  
```

#Lt's look at taxonomy an lifestyle of all proteins: mosaic and non-mosaic
- mosaic proteins are slightly more often seen in temperate than virulent rHMMs

```{r}
all.proteins.with.metadata = all.proteins %>%
  rename(fam = family) %>%
  left_join(reprseq.metadata.table %>% rename(qname = repr.name))

num.virulent.prot = all.proteins.with.metadata %>% filter(lifestyle == "virulent") %>% nrow()
num.virulent.domain.mosaic.prot = all.proteins.with.metadata %>% filter(lifestyle == "virulent" & mosaic.domain) %>% nrow()
num.virulent.non.domain.mosaic.prot = all.proteins.with.metadata %>% filter(lifestyle == "virulent" & !mosaic.domain) %>% nrow()
num.virulent.mosaic.prot = all.proteins.with.metadata %>% filter(lifestyle == "virulent" & mosaic) %>% nrow()
num.virulent.non.mosaic.prot = all.proteins.with.metadata %>% filter(lifestyle == "virulent" & !mosaic) %>% nrow()

num.temperate.prot = all.proteins.with.metadata %>% filter(lifestyle == "temperate") %>% nrow()
num.temperate.domain.mosaic.prot = all.proteins.with.metadata %>% filter(lifestyle == "temperate" & mosaic.domain) %>% nrow()
num.temperate.non.domain.mosaic.prot = all.proteins.with.metadata %>% filter(lifestyle == "temperate" & !mosaic.domain) %>% nrow()
num.temperate.mosaic.prot = all.proteins.with.metadata %>% filter(lifestyle == "temperate" & mosaic) %>% nrow()
num.temperate.non.mosaic.prot = all.proteins.with.metadata %>% filter(lifestyle == "temperate" & !mosaic) %>% nrow()

dat = data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                 mosaic = c("mosaic","non-mosaic", "mosaic", "non-mosaic", "All", "All"),
                 num = c(num.virulent.domain.mosaic.prot, num.virulent.non.domain.mosaic.prot, num.temperate.domain.mosaic.prot, num.temperate.non.domain.mosaic.prot, num.virulent.prot, num.temperate.prot))

ggplot(dat) +
  geom_col(aes(x = mosaic, y = num, fill = lifestyle), position = "dodge") + scale_y_log10()


# 2 % of virulent rHMMs are mosaic and 3% of temperate rHMMs are mosaic
num.virulent.domain.mosaic.prot/num.virulent.prot
num.temperate.domain.mosaic.prot/num.temperate.prot



dat2 = data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                 mosaic = c("mosaic","non-mosaic", "mosaic", "non-mosaic", "All", "All"),
                 num = c(num.virulent.mosaic.prot, num.virulent.non.mosaic.prot, num.temperate.mosaic.prot, num.temperate.non.mosaic.prot, num.virulent.prot, num.temperate.prot))

ggplot(dat2) +
  geom_col(aes(x = mosaic, y = num, fill = lifestyle), position = "dodge") + scale_y_log10()


# 11 % of virulent rHMMs are mosaic and 14% of temperate rHMMs are mosaic
num.virulent.mosaic.prot/num.virulent.prot
num.temperate.mosaic.prot/num.temperate.prot

```

# Now do the same for families

when we look at mosaicism defined by seq or domain
we see that 20% of virulent families are mosaic, 
and 30% of temperate families are mosaic.
The difference is statistically significant (p < 10^-15)



```{r}

seq.mosaic.qnames =  c(seq.mosaic.pairs.all.pident %>% filter(mosaic) %>% pull(qname), seq.mosaic.pairs.all.pident %>% filter(mosaic) %>% pull(sname)) %>% unique()
all.families.with.metadata = all.proteins %>% 
  mutate(mosaic.seq = qname %in% seq.mosaic.qnames) %>%
  group_by(family) %>%
  summarise(mosaic = any(mosaic),
            mosaic.domain = any(mosaic.domain),
            mosaic.seq = any(mosaic.seq)) %>%
  rename(fam = family) %>%
  left_join(families.metadata.table, by = "fam")


num.virulent.fam = all.families.with.metadata %>% filter(lifestyle == "virulent") %>% nrow()
num.virulent.domain.mosaic.fam = all.families.with.metadata %>% filter(lifestyle == "virulent" & mosaic.domain) %>% nrow()
num.virulent.non.domain.mosaic.fam = all.families.with.metadata %>% filter(lifestyle == "virulent" & !mosaic.domain) %>% nrow()
num.virulent.mosaic.fam = all.families.with.metadata %>% filter(lifestyle == "virulent" & mosaic) %>% nrow()
num.virulent.non.mosaic.fam = all.families.with.metadata %>% filter(lifestyle == "virulent" & !mosaic) %>% nrow()
num.virulent.seq.mosaic.fam = all.families.with.metadata %>% filter(lifestyle == "virulent" & mosaic.seq) %>% nrow()
num.virulent.non.seq.mosaic.fam = all.families.with.metadata %>% filter(lifestyle == "virulent" & !mosaic.seq) %>% nrow()

num.temperate.fam = all.families.with.metadata %>% filter(lifestyle == "temperate") %>% nrow()
num.temperate.domain.mosaic.fam = all.families.with.metadata %>% filter(lifestyle == "temperate" & mosaic.domain) %>% nrow()
num.temperate.non.domain.mosaic.fam = all.families.with.metadata %>% filter(lifestyle == "temperate" & !mosaic.domain) %>% nrow()
num.temperate.mosaic.fam = all.families.with.metadata %>% filter(lifestyle == "temperate" & mosaic) %>% nrow()
num.temperate.non.mosaic.fam = all.families.with.metadata %>% filter(lifestyle == "temperate" & !mosaic) %>% nrow()
num.temperate.seq.mosaic.fam = all.families.with.metadata %>% filter(lifestyle == "temperate" & mosaic.seq) %>% nrow()
num.temperate.non.seq.mosaic.fam = all.families.with.metadata %>% filter(lifestyle == "temperate" & !mosaic.seq) %>% nrow()

dat = data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                 mosaic = c("mosaic","non-mosaic", "mosaic", "non-mosaic", "All", "All"),
                 num = c(num.virulent.domain.mosaic.fam, num.virulent.non.domain.mosaic.fam, num.temperate.domain.mosaic.fam, num.temperate.non.domain.mosaic.fam, num.virulent.fam, num.temperate.fam))

ggplot(dat) +
  geom_col(aes(x = mosaic, y = num, fill = lifestyle), position = "dodge") + scale_y_log10()


# 0.6 % of virulent rHMMs are mosaic and 0.8% of temperate rHMMs are mosaic
num.virulent.domain.mosaic.fam/num.virulent.fam
num.temperate.domain.mosaic.fam/num.temperate.fam



dat2 = data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                 mosaic = c("mosaic","non-mosaic", "mosaic", "non-mosaic", "All", "All"),
                 num = c(num.virulent.mosaic.fam, num.virulent.non.mosaic.fam, num.temperate.mosaic.fam, num.temperate.non.mosaic.fam, num.virulent.fam, num.temperate.fam))

ggplot(dat2) +
  geom_col(aes(x = mosaic, y = num, fill = lifestyle), position = "dodge") 


# 20 % of virulent rHMMs are mosaic and 30% of temperate rHMMs are mosaic
num.virulent.mosaic.fam/num.virulent.fam
num.temperate.mosaic.fam/num.temperate.fam



dat3 = data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                 mosaic = c("mosaic","non-mosaic", "mosaic", "non-mosaic", "All", "All"),
                 num = c(num.virulent.seq.mosaic.fam, num.virulent.non.seq.mosaic.fam, num.temperate.seq.mosaic.fam, num.temperate.non.seq.mosaic.fam, num.virulent.fam, num.temperate.fam))

ggplot(dat3) +
  geom_col(aes(x = mosaic, y = num, fill = lifestyle), position = "dodge") 
num.virulent.seq.mosaic.fam/num.virulent.fam
num.temperate.seq.mosaic.fam/num.temperate.fam

    M1 = matrix(0, nrow=2, ncol = 2)
    M1[1,1] = num.virulent.mosaic.fam
    M1[1,2] = num.virulent.non.mosaic.fam
    M1[2,1] = num.temperate.mosaic.fam
    M1[2,2] = num.temperate.non.mosaic.fam
   chisq.test(M1)

  # significant difference: p < 10^(-15)
```

# Now do the same for annotated proteins only
# alsmost exactly same proportion of annotated virulent/temperate families are mosaic (39%)
```{r}
# remember there may be mulrtiplr rows per each family here
families.and.functions.sel = families.and.functions 
num.virulent.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent") %>% distinct(fam) %>% nrow()
num.virulent.domain.mosaic.fam.annotated  = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & mosaic.domain) %>% distinct(fam) %>% nrow()
num.virulent.non.domain.mosaic.fam.annotated  = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & !mosaic.domain) %>% distinct(fam) %>% nrow()
num.virulent.mosaic.fam.annotated  = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & mosaic) %>% distinct(fam) %>% nrow()
num.virulent.non.mosaic.fam.annotated  = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & !mosaic) %>% distinct(fam) %>% nrow()
num.virulent.seq.mosaic.fam.annotated  = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & mosaic.seq) %>% distinct(fam) %>% nrow()
num.virulent.non.seq.mosaic.fam.annotated  = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & !mosaic.seq) %>% distinct(fam) %>% nrow()

num.temperate.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate") %>% distinct(fam) %>% nrow()
num.temperate.domain.mosaic.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & mosaic.domain) %>% distinct(fam) %>% nrow()
num.temperate.non.domain.mosaic.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & !mosaic.domain) %>% distinct(fam) %>% nrow()
num.temperate.mosaic.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & mosaic) %>% distinct(fam) %>% nrow()
num.temperate.non.mosaic.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & !mosaic) %>% distinct(fam) %>% nrow()
num.temperate.seq.mosaic.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & mosaic.seq) %>% distinct(fam) %>% nrow()
num.temperate.non.seq.mosaic.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & !mosaic.seq) %>% distinct(fam) %>% nrow()

dat = data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                 mosaic = c("mosaic","non-mosaic", "mosaic", "non-mosaic", "All", "All"),
                 num = c(num.virulent.domain.mosaic.fam.annotated, num.virulent.non.domain.mosaic.fam.annotated, num.temperate.domain.mosaic.fam.annotated, num.temperate.non.domain.mosaic.fam.annotated, num.virulent.fam.annotated, num.temperate.fam.annotated))

ggplot(dat) +
  geom_col(aes(x = mosaic, y = num, fill = lifestyle), position = "dodge") + scale_y_log10()


# 0.6 % of virulent rHMMs are mosaic and 0.8% of temperate rHMMs are mosaic
num.virulent.domain.mosaic.fam.annotated/num.virulent.fam.annotated
num.temperate.domain.mosaic.fam.annotated/num.temperate.fam.annotated



dat2 = data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                 mosaic = c("mosaic","non-mosaic", "mosaic", "non-mosaic", "All", "All"),
                 num = c(num.virulent.mosaic.fam.annotated, num.virulent.non.mosaic.fam.annotated, num.temperate.mosaic.fam.annotated, num.temperate.non.mosaic.fam.annotated, num.virulent.fam.annotated, num.temperate.fam.annotated))

ggplot(dat2) +
  geom_col(aes(x = mosaic, y = num, fill = lifestyle), position = "dodge") 


# 20 % of virulent rHMMs are mosaic and 30% of temperate rHMMs are mosaic
num.virulent.mosaic.fam.annotated/num.virulent.fam.annotated
num.temperate.mosaic.fam.annotated/num.temperate.fam.annotated



dat3 = data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                 mosaic = c("mosaic","non-mosaic", "mosaic", "non-mosaic", "All", "All"),
                 num = c(num.virulent.seq.mosaic.fam.annotated, num.virulent.non.seq.mosaic.fam.annotated, num.temperate.seq.mosaic.fam.annotated, num.temperate.non.seq.mosaic.fam.annotated, num.virulent.fam.annotated, num.temperate.fam.annotated))

ggplot(dat3) +
  geom_col(aes(x = mosaic, y = num, fill = lifestyle), position = "dodge") 
num.virulent.seq.mosaic.fam.annotated/num.virulent.fam.annotated
num.temperate.seq.mosaic.fam.annotated/num.temperate.fam.annotated

    M1 = matrix(0, nrow=2, ncol = 2)
    M1[1,1] = num.virulent.mosaic.fam.annotated
    M1[1,2] = num.virulent.non.mosaic.fam.annotated
    M1[2,1] = num.temperate.mosaic.fam.annotated
    M1[2,2] = num.temperate.non.mosaic.fam.annotated
    chisq.test(M1)

  # significant difference: p < 10^(-15)
```


# Now do the same for UNannotated proteins only
6.7 % of virulent unannotated rHMM families are mosaic (seq or domain)
8.2 % of temperate unannotated rHMM families are mosaic (seq or domain)

```{r}
families.and.functions.sel = families.and.functions 
num.virulent.fam.unannotated = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "virulent") %>% distinct(fam) %>% nrow()
num.virulent.domain.mosaic.fam.unannotated  = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & mosaic.domain) %>% distinct(fam) %>% nrow()
num.virulent.non.domain.mosaic.fam.unannotated  = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & !mosaic.domain) %>% distinct(fam) %>% nrow()
num.virulent.mosaic.fam.unannotated  = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & mosaic) %>% distinct(fam) %>% nrow()
num.virulent.non.mosaic.fam.unannotated  = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & !mosaic) %>% distinct(fam) %>% nrow()
num.virulent.seq.mosaic.fam.unannotated  = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & mosaic.seq) %>% distinct(fam) %>% nrow()
num.virulent.non.seq.mosaic.fam.unannotated  = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & !mosaic.seq) %>% distinct(fam) %>% nrow()

num.temperate.fam.unannotated = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "temperate") %>% distinct(fam) %>% nrow()
num.temperate.domain.mosaic.fam.unannotated = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & mosaic.domain) %>% distinct(fam) %>% nrow()
num.temperate.non.domain.mosaic.fam.unannotated = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & !mosaic.domain) %>% distinct(fam) %>% nrow()
num.temperate.mosaic.fam.unannotated = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & mosaic) %>% distinct(fam) %>% nrow()
num.temperate.non.mosaic.fam.unannotated = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & !mosaic) %>% distinct(fam) %>% nrow()
num.temperate.seq.mosaic.fam.unannotated = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & mosaic.seq) %>% distinct(fam) %>% nrow()
num.temperate.non.seq.mosaic.fam.unannotated = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & !mosaic.seq) %>% distinct(fam) %>% nrow()

dat = data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                 mosaic = c("mosaic","non-mosaic", "mosaic", "non-mosaic", "All", "All"),
                 num = c(num.virulent.domain.mosaic.fam.unannotated, num.virulent.non.domain.mosaic.fam.unannotated, num.temperate.domain.mosaic.fam.unannotated, num.temperate.non.domain.mosaic.fam.unannotated, num.virulent.fam.unannotated, num.temperate.fam.unannotated))

ggplot(dat) +
  geom_col(aes(x = mosaic, y = num, fill = lifestyle), position = "dodge") + scale_y_log10()


# 0.6 % of virulent rHMMs are mosaic and 0.8% of temperate rHMMs are mosaic
num.virulent.domain.mosaic.fam.unannotated/num.virulent.fam.unannotated
num.temperate.domain.mosaic.fam.unannotated/num.temperate.fam.unannotated



dat2 = data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                 mosaic = c("mosaic","non-mosaic", "mosaic", "non-mosaic", "All", "All"),
                 num = c(num.virulent.mosaic.fam.unannotated, num.virulent.non.mosaic.fam.unannotated, num.temperate.mosaic.fam.unannotated, num.temperate.non.mosaic.fam.unannotated, num.virulent.fam.unannotated, num.temperate.fam.unannotated))

ggplot(dat2) +
  geom_col(aes(x = mosaic, y = num, fill = lifestyle), position = "dodge") 


# 20 % of virulent rHMMs are mosaic and 30% of temperate rHMMs are mosaic
num.virulent.mosaic.fam.unannotated/num.virulent.fam.unannotated
num.temperate.mosaic.fam.unannotated/num.temperate.fam.unannotated



dat3 = data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                 mosaic = c("mosaic","non-mosaic", "mosaic", "non-mosaic", "All", "All"),
                 num = c(num.virulent.seq.mosaic.fam.unannotated, num.virulent.non.seq.mosaic.fam.unannotated, num.temperate.seq.mosaic.fam.unannotated, num.temperate.non.seq.mosaic.fam.unannotated, num.virulent.fam.unannotated, num.temperate.fam.unannotated))

ggplot(dat3) +
  geom_col(aes(x = mosaic, y = num, fill = lifestyle), position = "dodge") 
num.virulent.seq.mosaic.fam.unannotated/num.virulent.fam.unannotated
num.temperate.seq.mosaic.fam.unannotated/num.temperate.fam.unannotated

    M1 = matrix(0, nrow=2, ncol = 2)
    M1[1,1] = num.virulent.mosaic.fam.unannotated
    M1[1,2] = num.virulent.non.mosaic.fam.unannotated
    M1[2,1] = num.temperate.mosaic.fam.unannotated
    M1[2,2] = num.temperate.non.mosaic.fam.unannotated
    chisq.test(M1)

  # significant difference: p < 10^(-15)
    
    
    
# What are those unannotated families?
    hypot.annots = all.families.with.metadata %>% anti_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & mosaic) %>%
      left_join(mid.annotated.proteins.including.multi.annot %>% rename(fam = family)) %>%
      group_by(annotation, category) %>%
      summarise(num.fam = n_distinct(fam)) %>%
      arrange(desc(num.fam))
      
  hypot.annots
      
```


```{r}
dat = data.frame(lifestyle = character(0), mosaicism.type = character(0), annotation.index = integer(0), All = integer(0), mosaic = integer(0), nonmosaic = integer(0))
for (selected.index in unique(top.diversity.classes$annotation.index)) {
  families.and.functions.sel = families.and.functions %>% 
    filter(annotation.index == selected.index)
  num.virulent.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent") %>% nrow()
  num.virulent.domain.mosaic.fam.annotated  = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & mosaic.domain) %>% nrow()
  num.virulent.non.domain.mosaic.fam.annotated  = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & !mosaic.domain) %>% nrow()
  num.virulent.mosaic.fam.annotated  = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & mosaic) %>% nrow()
  num.virulent.non.mosaic.fam.annotated  = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & !mosaic) %>% nrow()
  num.virulent.seq.mosaic.fam.annotated  = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & mosaic.seq) %>% nrow()
  num.virulent.non.seq.mosaic.fam.annotated  = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "virulent" & !mosaic.seq) %>% nrow()
  
  num.temperate.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate") %>% nrow()
  num.temperate.domain.mosaic.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & mosaic.domain) %>% nrow()
  num.temperate.non.domain.mosaic.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & !mosaic.domain) %>% nrow()
  num.temperate.mosaic.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & mosaic) %>% nrow()
  num.temperate.non.mosaic.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & !mosaic) %>% nrow()
  num.temperate.seq.mosaic.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & mosaic.seq) %>% nrow()
  num.temperate.non.seq.mosaic.fam.annotated = all.families.with.metadata %>% inner_join(families.and.functions.sel) %>% filter(lifestyle == "temperate" & !mosaic.seq) %>% nrow()
  
  dat1= data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                   mosaic = c("mosaic","nonmosaic", "mosaic", "nonmosaic", "All", "All"),
                   num = c(num.virulent.domain.mosaic.fam.annotated, num.virulent.non.domain.mosaic.fam.annotated, num.temperate.domain.mosaic.fam.annotated, num.temperate.non.domain.mosaic.fam.annotated, num.virulent.fam.annotated, num.temperate.fam.annotated)) %>%
    mutate(mosaicism.type = "domain")
  
  dat2 = data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                   mosaic = c("mosaic","nonmosaic", "mosaic", "nonmosaic", "All", "All"),
                   num = c(num.virulent.mosaic.fam.annotated, num.virulent.non.mosaic.fam.annotated, num.temperate.mosaic.fam.annotated, num.temperate.non.mosaic.fam.annotated, num.virulent.fam.annotated, num.temperate.fam.annotated))%>%
    mutate(mosaicism.type = "domain.or.seq")
  
  dat3 = data.frame(lifestyle = c("virulent", "virulent", "temperate", "temperate", "virulent", "temperate"),
                   mosaic = c("mosaic","nonmosaic", "mosaic", "nonmosaic", "All", "All"),
                   num = c(num.virulent.seq.mosaic.fam.annotated, num.virulent.non.seq.mosaic.fam.annotated, num.temperate.seq.mosaic.fam.annotated, num.temperate.non.seq.mosaic.fam.annotated, num.virulent.fam.annotated, num.temperate.fam.annotated))%>%
    mutate(mosaicism.type = "seq")
  
  this.dat = rbind(dat1, dat2, dat3) %>%
    mutate(annotation.index = selected.index) %>%
    tidyr::spread(mosaic, num)
  dat = rbind(dat, this.dat)
  
}

all.dat = dat  %>%
    mutate(prop.mosaic = mosaic/All) %>%
  left_join(annotation.indices) #%>%
  #filter(All >= 10)



g=ggplot(all.dat %>% rowwise() %>% mutate(label = paste0("(out of ", All, " fam.)"),
                                          any.data = if_else(All > 0,1, 0)), 
         aes(x= annotation, y = prop.mosaic, fill = lifestyle, col = lifestyle)) +
  geom_col(aes(col = lifestyle, y = any.data), position = 'dodge', fill = "white", width = 0.5) +
  geom_col( width = 0.5, position = 'dodge') +
  geom_text(aes(label = label), size = 2, nudge_y = 0.05) +
  coord_flip() +
  theme_bw() +
  facet_grid(category ~ mosaicism.type, scales = "free", space = "free_y") +
  ylab("Prop. families mosaic")

g
ggsave(g, filename = sprintf("%sannotation_and_lifestyle_vs_mosaic.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 12, height = 20)  
```

# let's look at mosaic pairs, are they from the same family / same lifestyle

When we look at the number of mosaic and nonmosaic pairs the results are noy conclsuive.
Seems that actually virulent-virulent pairs are overrepresented wihin mosaic pair which is in contrast with the fact that 
families with any mosaic signal are overrrepresented in temperate families.
It may be due to the topology of network of mosaicism (virylent are more interconnected).

```{r}
num.virulent.pairs = num.virulent.fam*(num.virulent.fam - 1)/2
num.temperate.pairs = num.virulent.fam*(num.temperate.fam - 1)/2
num.temperate.virulentpairs = num.virulent.fam*num.temperate.fam
num.potential.pairs = data.frame(pair.lifestyle = c("virulent&virulent", "temperate&temperate", "temperate&virulent"),
                                 num.potential.pairs = c(num.virulent.pairs, num.temperate.pairs, num.temperate.virulentpairs))


sequence.mosaic.family.pairs = seq.mosaic.pairs.all.pident %>% filter(mosaic) %>% distinct(family.pair)
domain.mosaic.family.pairs = qname.pairs.sharing.t.and.disshare.x %>% distinct(family.pair)
mosaic.family.pairs = rbind(sequence.mosaic.family.pairs, domain.mosaic.family.pairs) %>% distinct()


sequence.mosaic.family.pairs.wide =  rbind(sequence.mosaic.family.pairs %>% tidyr::separate(col = "family.pair", into = c("fam.x", "fam.y"), sep = "&", remove = FALSE) %>% select(family.pair, fam.x, fam.y),
                                           sequence.mosaic.family.pairs %>% tidyr::separate(col = "family.pair", into = c("fam.y", "fam.x"), sep = "&", remove = FALSE) %>% select(family.pair, fam.x, fam.y))

mosaic.family.pairs.wide = rbind(mosaic.family.pairs %>% tidyr::separate(col = "family.pair", into = c("fam.x", "fam.y"), sep = "&", remove = FALSE) %>% select(family.pair, fam.x, fam.y),
                                 mosaic.family.pairs %>% tidyr::separate(col = "family.pair", into = c("fam.y", "fam.x"), sep = "&", remove = FALSE) %>% select(family.pair, fam.x, fam.y))


domain.mosaic.family.pairs.wide = rbind(domain.mosaic.family.pairs %>% tidyr::separate(col = "family.pair", into = c("fam.x", "fam.y"), sep = "&", remove = FALSE) %>% select(family.pair, fam.x, fam.y),
                                        domain.mosaic.family.pairs %>% tidyr::separate(col = "family.pair", into = c("fam.y", "fam.x"), sep = "&", remove = FALSE) %>% select(family.pair, fam.x, fam.y))

mosaic.pairs.lifestyle = mosaic.family.pairs.wide %>%
  left_join(families.metadata.table %>% select(fam.x = fam, genus.x = genus, lifestyle.x = lifestyle, family.x = family)) %>%
  left_join(families.metadata.table %>% select(fam.y = fam, genus.y = genus, lifestyle.y = lifestyle, family.y = family)) 





seq.mosaic.pairs.lifestyle = sequence.mosaic.family.pairs.wide %>%
  left_join(families.metadata.table %>% select(fam.x = fam, lifestyle.x = lifestyle, family.x = family)) %>%
  left_join(families.metadata.table %>% select(fam.y = fam, lifestyle.y = lifestyle, family.y = family)) 

domain.mosaic.pairs.lifestyle = domain.mosaic.family.pairs.wide %>%
  left_join(families.metadata.table %>% select(fam.x = fam, lifestyle.x = lifestyle, family.x = family)) %>%
  left_join(families.metadata.table %>% select(fam.y = fam, lifestyle.y = lifestyle, family.y = family)) 


mosaic.pairs.lifestyle.stats = mosaic.pairs.lifestyle %>%
  rowwise() %>%
  mutate(pair.lifestyle =CreatePair(lifestyle.x, lifestyle.y, collapse = "&")) %>%
  filter(pair.lifestyle %in% c("virulent&virulent", "temperate&temperate", "temperate&virulent")) %>%
  group_by(pair.lifestyle) %>%
  summarise(n = n_distinct(family.pair)) %>%
  left_join(num.potential.pairs) %>%
  mutate(num.nonmosaic.pairs = num.potential.pairs - n,
         prop.pairs.mosaic = n/num.potential.pairs)

domain.mosaic.pairs.lifestyle.stats = domain.mosaic.pairs.lifestyle %>%
  rowwise() %>%
  mutate(pair.lifestyle = CreatePair(lifestyle.x, lifestyle.y, collapse = "&")) %>%
  filter(pair.lifestyle %in% c("virulent&virulent", "temperate&temperate", "temperate&virulent")) %>%
  group_by(pair.lifestyle) %>%
  summarise(n = n_distinct(family.pair)) %>%
  left_join(num.potential.pairs) %>%
  mutate(num.nonmosaic.pairs = num.potential.pairs - n,
         prop.pairs.mosaic = n/num.potential.pairs)

seq.mosaic.pairs.lifestyle.stats = seq.mosaic.pairs.lifestyle %>%
  rowwise() %>%
  mutate(pair.lifestyle = CreatePair(lifestyle.x, lifestyle.y, collapse = "&")) %>%
  filter(pair.lifestyle %in% c("virulent&virulent", "temperate&temperate", "temperate&virulent")) %>%
  group_by(pair.lifestyle) %>%
  summarise(n = n_distinct(family.pair)) %>%
  left_join(num.potential.pairs) %>%
  mutate(num.nonmosaic.pairs = num.potential.pairs - n,
         prop.pairs.mosaic = n/num.potential.pairs)

 
```

# Let's look at pairs of specific dunction

When we look at sequence mosaicism and tail fibers.
The proportion of temperate-temperate families that are mosaic is the same as the proportion of virulent-virulent families that are mosaic

The proportion of temperate-vitulent is lower
In endonucleases there are more seq. mosaic pairs within temperate-temperate
In tail spike there are no temperate-temperate mosaic pairs at all

In general when we look into possible pairs of the same function there is a higher proprtion of all possible virulent-virulent pairs thatn temperate-temperate pairs that are mosaic

```{r}
# one family may appear in multiple functions
#selected.annotation.index = 517
#selected.annotation.index = 525 # endocnuclease


num.virulent.pairs.this.function = all.families.with.metadata %>% left_join(families.and.functions)  %>%  filter(lifestyle == "virulent") %>% group_by(annotation.index) %>% summarise(n.fam = n_distinct(fam)) %>% ungroup() %>%
 # filter(annotation.index == selected.annotation.index) %>% 
  mutate(num.potential.pairs = n.fam*(n.fam-1)/2) 

num.temperate.pairs.this.function = all.families.with.metadata %>% left_join(families.and.functions)  %>%  filter(lifestyle == "temperate") %>% group_by(annotation.index) %>% summarise(n.fam = n_distinct(fam)) %>% ungroup() %>%
 # filter(annotation.index == selected.annotation.index) %>% 
  mutate(num.potential.pairs = n.fam*(n.fam-1)/2) 

num.temperate.virulent.pairs.this.function = num.temperate.pairs.this.function %>% select(annotation.index, n.t = n.fam) %>%
  inner_join(num.virulent.pairs.this.function) %>%
  mutate(num.potential.pairs = n.t*n.fam)


num.potential.pairs.this.function = data.frame(pair.lifestyle = c("virulent&virulent", "temperate&temperate", "temperate&virulent"),
                                 num.potential.pairs = c(num.virulent.pairs.this.function %>%
                                                          pull(num.potential.pairs) %>% 
                                                            sum(),
                                                          num.temperate.virulent.pairs.this.function %>%
                                                          pull(num.potential.pairs) %>% 
                                                            sum(), 
                                                          num.temperate.pairs.this.function %>%
                                                          pull(num.potential.pairs) %>% 
                                                            sum()))



sequence.mosaic.family.pairs.this.function = seq.mosaic.pairs.all.pident %>% filter(mosaic) %>% 
  inner_join(surely.annotated.proteins %>% select(qname, qannotation.index = annotation.index)) %>%
  inner_join(surely.annotated.proteins %>% select(sname =qname, sannotation.index = annotation.index)) %>%
  filter(qannotation.index == sannotation.index) %>%
 # filter(qannotation.index == selected.annotation.index) %>%
  distinct(family.pair) %>%
  left_join(mosaic.pairs.lifestyle)


sequence.mosaic.family.pairs.this.function %>%
  rowwise() %>%
  mutate(pair.lifestyle =CreatePair(lifestyle.x, lifestyle.y, collapse = "&")) %>%
  filter(pair.lifestyle %in% c("virulent&virulent", "temperate&temperate", "temperate&virulent")) %>%
  group_by(pair.lifestyle) %>%
  summarise(n = n_distinct(family.pair)) %>%
  left_join(num.potential.pairs.this.function) %>%
  mutate(num.nonmosaic.pairs = num.potential.pairs - n,
         prop.pairs.mosaic = n/num.potential.pairs)





# the numbers are very low for domain mosaicism 
domain.mosaic.family.pairs.this.function = domain.mosaic.pairs.lifestyle %>% 
  inner_join(families.and.functions %>% select(fam.x = fam, annotation.index.x = annotation.index)) %>%
  inner_join(families.and.functions %>% select(fam.y= fam, annotation.index.y = annotation.index)) %>%
  filter(annotation.index.x ==  annotation.index.y) %>%
 # filter(annotation.index.x == selected.annotation.index) %>%
  distinct(family.pair) %>%
  left_join(mosaic.pairs.lifestyle)


domain.mosaic.family.pairs.this.function %>%
  rowwise() %>%
  mutate(pair.lifestyle =CreatePair(lifestyle.x, lifestyle.y, collapse = "&")) %>%
  filter(pair.lifestyle %in% c("virulent&virulent", "temperate&temperate", "temperate&virulent")) %>%
  group_by(pair.lifestyle) %>%
  summarise(n = n_distinct(family.pair)) %>%
  left_join(num.potential.pairs.this.function) %>%
  mutate(num.nonmosaic.pairs = num.potential.pairs - n,
         prop.pairs.mosaic = n/num.potential.pairs)

```

# Look at the taxonomy
```{r}
uncharacterised.family.names = c("multi", "Unclassified", "")
all.taxonomic.families = all.families.with.metadata %>% filter(!(family %in% uncharacterised.family.names)) %>% group_by(family) %>% summarise(num.fam = n_distinct(fam)) %>% arrange(desc(num.fam)) %>% mutate(n.pairs = num.fam*(num.fam-1)/2) 
num.pairs.same.family = all.taxonomic.families %>% pull(n.pairs) %>% sum()
all.family.combinations =  t(combn(all.taxonomic.families$family,2)) %>% as.data.frame() # all pairs of different families
names(all.family.combinations) = c("family.x", "family.y")
all.family.combinations = all.family.combinations %>% 
  left_join(all.taxonomic.families %>% select(family.x = family, n.fam.x = num.fam)) %>% 
  left_join(all.taxonomic.families %>% select(family.y = family, n.fam.y = num.fam)) %>%
  mutate(num.pairs = n.fam.x*n.fam.y) 
num.pairs.different.family = all.family.combinations %>% pull(num.pairs) %>% sum()
num.potential.pairs.per.taxonomy = data.frame(family.pair.type = c("same family", "different family"),
                                 num.potential.pairs = c(num.pairs.same.family, num.pairs.different.family))

mosaic.pairs.taxonomy.stats = mosaic.pairs.lifestyle %>%
  rowwise() %>%
  mutate(
    known.family = !(family.x  %in% uncharacterised.family.names)  & !(family.y %in% uncharacterised.family.names),
    same.family = family.x == family.y & known.family,
    different.family = family.x != family.y & known.family) %>%
  mutate(family.pair.type = if_else(!known.family, "uknown", if_else(same.family, "same family", "different family"))) %>%
  distinct(family.pair, family.pair.type) %>%
  group_by(family.pair.type) %>%
  summarise(n = n_distinct(family.pair)) %>%
  left_join(num.potential.pairs.per.taxonomy) %>%
  mutate(num.nonmosaic.pairs = num.potential.pairs - n,
         prop.pairs.mosaic = n/num.potential.pairs)

# More pairs from the same family
# 0.00733% % of pairs from the same family are mosaic
# only 
#0.00102 % of pairs from different families are mosaic


mosaic.pairs.taxonomy.stats

    M1 = matrix(0, nrow=2, ncol = 2)
    M1[1,1] = mosaic.pairs.taxonomy.stats %>% filter(family.pair.type == "same family") %>% pull(n)
    M1[1,2] = mosaic.pairs.taxonomy.stats %>% filter(family.pair.type == "same family") %>% pull(num.nonmosaic.pairs)
    M1[2,1] = mosaic.pairs.taxonomy.stats %>% filter(family.pair.type == "different family") %>% pull(n)
    M1[2,2] = mosaic.pairs.taxonomy.stats %>% filter(family.pair.type == "different family") %>% pull(num.nonmosaic.pairs)
    fisher_test_res = chisq.test(M1)
    fisher_test_res
```

# looking at the rect mosaicism
```{r}
recentHGT.with.metadata = recent_HGT_pairs_raw.70 %>%
  left_join(reprseq.metadata.table %>% select(qname = repr.name, qfamily = family, qlifestyle = lifestyle, qhost = host, qnum.genomes.with.this.rHMM = num.genomes.with.this.rHMM)) %>%
  left_join(reprseq.metadata.table %>% select(sname = repr.name, sfamily = family, slifestyle = lifestyle, shost = host, snum.genomes.with.this.rHMM = num.genomes.with.this.rHMM)) %>%
    left_join(families %>% select(qname = qname, qfam = family)) %>%
    left_join(families %>% select(sname = qname, sfam = family)) %>%
  rowwise() %>%
  mutate(pair.fam = CreatePair(qfam, sfam),
         pair.lifestyle = CreatePair(qlifestyle, slifestyle, collapse = " & "),
         pair.family = CreatePair(qfamily, sfamily, collapse = " & "),
         reprseq.pair = CreatePair(qname, sname, collapse = "&"),
         fam.pair = CreatePair(qfam, sfam, collapse = "&"),
         host.pair = CreatePair(qhost, shost, collapse = "&")) %>%
  mutate(known.lifestyle = (pair.lifestyle %in% c("virulent & virulent", "temperate & temperate", "temperate & virulent")),
         multi.lifestyle = qlifestyle == "multi" | slifestyle == "multi",
         known.family = qfamily != "unknown"  & sfamily != "unknown",
         known.host  = qhost != "unknown" & shost != "unknown") %>%
  mutate(pair.lifestyle = if_else(multi.lifestyle, "multi", if_else(known.lifestyle, pair.lifestyle, "unknown")),
         pair.family = if_else(known.family, pair.family, "unknown"),
         host.pair = if_else(known.host, host.pair, "unknown"))

ss = recentHGT.with.metadata %>%
  #mutate(qindex=if_else(is.na(qindex), 10000, qindex),
  #       sindex=if_else(is.na(sindex), 10000, sindex)) %>%
  filter(qindex == sindex) %>%
  group_by(qindex, pair.lifestyle) %>%
  summarise(n.prot.pairs = n_distinct(reprseq.pair),
            n.fam.pairs = n_distinct(fam.pair)) %>%
  arrange(desc(n.fam.pairs)) %>%
  rename(annotation.index = qindex) %>%
  left_join(annotation.indices)


g=ggplot(ss) +
  geom_col(aes(x=annotation, y = n.prot.pairs, fill = pair.lifestyle)) +
  coord_flip() +
  theme_bw() +
  facet_grid(category~., scales = "free", space = "free_y") +
  xlab("") +
  ylab("Num rHMM pairs with recent mosaicism (>= 70% pident)")#+
  #ylab("Num family pairs with recent mosaicism (>= 70% pident)")#+
  #scale_fill_manual(values = colm)
g
ggsave(g, filename = sprintf("%sannotations_vs_lifestyle.pairs_prot.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 12, height = 10)  


ss2 = recentHGT.with.metadata %>%
  filter(qindex == sindex) %>%
  group_by(qindex, pair.family) %>%
  summarise(n.prot.pairs = n_distinct(reprseq.pair),
            n.fam.pairs = n_distinct(fam.pair)) %>%
  arrange(desc(n.fam.pairs)) %>%
  rename(annotation.index = qindex) %>%
  left_join(annotation.indices)


g=ggplot(ss2) +
  geom_col(aes(x=annotation, y = n.prot.pairs, fill = pair.family)) +
  coord_flip() +
  theme_bw() +
  facet_grid(category~., scales = "free", space = "free_y") +
  xlab("") +
  ylab("Num family pairs with recent mosaicism (>= 70% pident)")#+
  #scale_fill_manual(values = colm)
g
ggsave(g, filename = sprintf("%sannotations_vs_taxonomy.pairs_prot.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 12, height = 10)  



recentHGT.with.metadata %>% filter(pair.lifestyle == "temperate & virulent" & !is.na(qindex) & !is.na(sindex)) %>%
  left_join(name.table %>% select(qname = repr.name, qncbi.id = ncbi.id)) %>%
  left_join(name.table %>% select(sname = repr.name, sncbi.id = ncbi.id)) %>%
  distinct(qindex, qncbi.id, sncbi.id)

# we have 1 pair of spike rHMM mosaic between only virulent and only temerate

# and 3 pairs of fibers rHMMs mosaic between only virulent and temperate

# additionally there are pairs that include rHMM that is found in both virulent and temperate


# Tail spike: NC_031924.1 (33317-36247(+)) - found as tail spike according to NCBI, this is a Salmonella phage, found by bacphlip as virulent
# MOSAIC WITH NC_054641.1:29238-31259(-: lfestyle of " NC_054641.1" is unknown accroding to bachlip but according to NCBI it is lytic!!
# and  NC_042346.1:14178-16163(+) : found as tail spike according to NCBI lifestyle = temperate, 
# and 	NC_054647.1:8557-10578(-) which has unknown lifestyle

# tail fibers:
# NC_014636.1:76160-80101(+): virulent T4 like Aeromonas phage, tail fiber protein; virulent according to bacphlip and literature (https://pubmed.ncbi.nlm.nih.gov/22226819/)
# AND Escherichia phage D108,	NC_013594.1:31507-33009(+) tail fiber protein, temperate according to bachlip and literature (Escherichia phage D108), tail fiber protein
# AND Enterobacteria phage Mu NC_000929.1:31458-32972(+) tail fiber protein, temperate according to bacphlip and literature
```



# RECENT MOSAICISM N ENTIRE GENOME METADATA
```{r}


recent.pairs.with.metadata = recent_HGT_pairs_raw.70 %>%
  left_join(name.table.with.metadata %>% distinct(qgenome = Accession, qname = repr.name, qlifestyle = lifestyle, qGenus = Genus, qfamily = Family)) %>%
  left_join(name.table.with.metadata %>% distinct(sgenome = Accession, sname = repr.name, slifestyle = lifestyle, sGenus = Genus, sfamily = Family)) %>%
  filter(qindex == sindex) %>%  
  rename(annotation.index = qindex) %>%
  inner_join(top.diversity.classes %>% distinct(annotation.index)) %>%
  rowwise() %>%
  mutate(genome.pair = CreatePair(qgenome, sgenome)) 


  recent.pairs.with.metadata.raw = recent.pairs.with.metadata %>%
    mutate(known.lifestyle = qlifestyle !="unknown" & slifestyle != "unknown") %>%
    rowwise() %>%
    mutate(l.pair = CreatePair(qlifestyle, slifestyle, collapse = " - "),
           same.lifestyle = if_else(!known.lifestyle, "unknwon", if_else(qlifestyle == slifestyle, "same", "different")))
  
  
recent.pairs.with.metadata.l1 = recent.pairs.with.metadata.raw %>%
  group_by(annotation.index, l.pair) %>%
  summarise(n.l = n_distinct(genome.pair)) %>%
  right_join(top.diversity.classes) %>%
  mutate(n.l = if_else(is.na(n.l), 0, n.l)) %>%
  left_join(annotation.indices) 

g=ggplot(recent.pairs.with.metadata.l1) +
  geom_col(aes(x = annotation, y = n.l, fill = l.pair)) +
  facet_grid(category ~., scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) +
  coord_flip() +
  theme_bw() +
  ylab("Num. pairs genomes") +
  xlab("") +
  #scale_fill_manual()
  scale_fill_manual(values = c("temperate - temperate" = "blue", "virulent - virulent" = "red", "temperate - virulent" = "green", "unknown" = "gray70", "missing data"= "gray50"))
g
ggsave(g, filename = sprintf("%sreceent_pairs_liefestyle2.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 6, height = 5)  

recent.pairs.with.metadata %>%
  filter((qlifestyle =="temperate" & slifestyle == "virulent") | (slifestyle =="temperate" & qlifestyle == "virulent") ) %>%
  filter(annotation.index == 184) 
  
# endolysin
#	NC_047906.1:17387-18241(+) i.e, one of two endolysin proteins in the virulent phage Streptomyces phage Rowa (LITERATURE SAYS IT IS TEMPERATE: https://phagesdb.org/phages/Rowa/)
# osaic to endolysin in temeperate pahage Streptomyces phage Spernnie, () NC_054659.1:19549-20445(+)): it is confirmed in literature this phage is temperate: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8407751/
# and to NC_054658.1:19811-20713(+)
# Note that the 	NC_047906.1:17387-18241(+) is found in both virulent and temperate genomes ()according to bacphlip)



```

```{r}
Get.Taxa.And.Lifestyle.Genome.Pairs = function(recent.pairs.with.metadata, top.diversity.classes) {
   recent.pairs.with.metadata.tb = recent.pairs.with.metadata %>% as.data.table()
   recent.pairs.with.metadata.tb[, known.lifestyle := qlifestyle !="unknown" & slifestyle != "unknown"]
   recent.pairs.with.metadata.tb[, known.genus := qGenus !="unknown" & sGenus != "unknown" & qGenus != "Unclassified" & sGenus != "Unclassified"]
   recent.pairs.with.metadata.tb[, known.fam := qfamily !="unknown" & sfamily != "unknown" & qfamily != "Unclassified" & sfamily != "Unclassified"]
   recent.pairs.with.metadata.tb[, same.l := if_else(!known.lifestyle, "unknwon", if_else(qlifestyle == slifestyle, "same", "different"))]
      
recent.pairs.with.metadata.l = recent.pairs.with.metadata.tb %>%
  rename(same = same.l) %>%
    group_by(annotation.index, same) %>%
    summarise(n = n_distinct(genome.pair)) %>%
    right_join(top.diversity.classes) 
 

  
  recent.pairs.with.metadata.tg = recent.pairs.with.metadata.tb %>%
    #rowwise() %>%
    #mutate(tg.pair = CreatePair(qGenus, sGenus, collapse = " - ")) %>%
    mutate(same = if_else(!known.genus, "unknown", if_else(qGenus == sGenus, "same", "different"))) %>%
    group_by(annotation.index, same) %>%
    summarise(n = n_distinct(genome.pair))
  
  
  
  recent.pairs.with.metadata.fam = recent.pairs.with.metadata.tb %>%
    #rowwise() %>%
    #mutate(fam.pair = CreatePair(qfamily, sfamily, collapse = " - ")) %>%
    mutate(same = if_else(!known.fam, "unknown", if_else(qfamily == sfamily, "same", "different"))) %>%
    group_by(annotation.index, same) %>%
    summarise(n = n_distinct(genome.pair)) %>%
    right_join(top.diversity.classes) 

  
  
  recent.pairs.meta.all = recent.pairs.with.metadata.l %>% mutate(type = "lifestyle") %>% select(annotation.index, same, n, type) %>%
    rbind(recent.pairs.with.metadata.fam %>% mutate(type = "family") %>% select(annotation.index, same, n, type)) %>%
    rbind(recent.pairs.with.metadata.tg %>% mutate(type = "genus") %>% select(annotation.index, same, n, type)) %>%
    left_join(annotation.indices)  %>%
    right_join(top.diversity.classes) 
  recent.pairs.meta.all$n[is.na(recent.pairs.meta.all$n)] = 0
  recent.pairs.meta.all$same[is.na(recent.pairs.meta.all$same)] = "unknwon"
  
  return(recent.pairs.meta.all)
}


recent.pairs.meta.all = Get.Taxa.And.Lifestyle.Genome.Pairs(recent.pairs.with.metadata, top.diversity.classes) 
text.size = 8
g=ggplot(recent.pairs.meta.all) +
  geom_col(aes(x = annotation, y = n, fill = same)) +
  facet_grid(category ~type, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) +
  coord_flip() +
  theme_bw() +
  theme( axis.title.y.right = element_text(size = text.size),
         axis.text.y.right = element_text(size = text.size),
          axis.title.y = element_text( size = text.size),
         axis.text.y = element_text(size = text.size),
  ) +
  ylab("Num. pairs genomes") +
  xlab("") +
 #scale_y_log10() +
  #scale_fill_manual()
  scale_fill_manual(values = c("same" = "darkgreen", "different" = "orange", "unknown" = "gray"))
g
ggsave(g, filename = sprintf("%sreceent_pairs_meta.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 15, height = 20)  

```

# all pairs not only recent

```{r}

sequence.mosaic.reprseq.pairs = seq.mosaic.pairs.all.pident %>% filter(mosaic) %>% distinct(reprseq.pair)
domain.mosaic.reprseq.pairs = qname.pairs.sharing.t.and.disshare.x %>% distinct(reprseq.pair)
mosaic.reprseq.pairs = rbind(sequence.mosaic.reprseq.pairs, domain.mosaic.reprseq.pairs) %>% distinct(reprseq.pair)



mosaic.reprseq.pairs.wide = rbind(mosaic.reprseq.pairs %>% tidyr::separate(col = "reprseq.pair", into = c("qname", "sname"), sep = "&", remove = FALSE) %>% select(reprseq.pair, qname, sname),
                                        mosaic.reprseq.pairs %>% tidyr::separate(col = "reprseq.pair", into = c("sname", "qname"), sep = "&", remove = FALSE) %>% select(reprseq.pair, qname, sname))
```

```{r, eval = FALSE}
surely.annotated.proteins.from.top.classes =  surely.annotated.proteins %>%
    inner_join(top.diversity.classes %>% distinct(annotation.index)) 
  
mosaic.reprseq.pairs.lifestyle = mosaic.reprseq.pairs.wide %>%
  #left_join(families.metadata.table %>% select(fam.x = fam, genus.x = genus, lifestyle.x = lifestyle, family.x = family)) %>%
  #left_join(families.metadata.table %>% select(fam.y = fam, genus.y = genus, lifestyle.y = lifestyle, family.y = family)) %>%
  inner_join(surely.annotated.proteins.from.top.classes %>% select(qname = qname, qindex = annotation.index)) %>%
  inner_join(surely.annotated.proteins.from.top.classes %>% select(sname = qname, sindex = annotation.index)) %>%
  filter(qindex == sindex) %>%  
  left_join(name.table.with.metadata %>% distinct(qgenome = Accession, qname = repr.name, qlifestyle = lifestyle, qGenus = Genus, qfamily = Family, qhost = Host)) %>%
  left_join(name.table.with.metadata %>% distinct(sgenome = Accession, sname = repr.name, slifestyle = lifestyle, sGenus = Genus, sfamily = Family, shost = Host)) %>%
  rename(annotation.index = qindex) %>%
  rowwise() %>%
  mutate(genome.pair = CreatePair(qgenome, sgenome)) 

data.table::fwrite(mosaic.reprseq.pairs.lifestyle, file = sprintf("%smosaic.reprseq.pairs.lifestyle.txt", OUTPUT.DATA.PATH))
```

```{r}
mosaic.reprseq.pairs.lifestyle = data.table::fread(file = sprintf("%smosaic.reprseq.pairs.lifestyle.txt", OUTPUT.DATA.PATH))
all.pairs.meta.all = Get.Taxa.And.Lifestyle.Genome.Pairs(mosaic.reprseq.pairs.lifestyle, top.diversity.classes) 
#data.table::fwrite(all.pairs.meta.all, file = sprintf("%sall.pairs.meta.all.txt", OUTPUT.DATA.PATH))
```

```{r}
text.size = 8
g=ggplot(all.pairs.meta.all) +
  geom_col(aes(x = annotation, y = n, fill = same)) +
  facet_grid(category ~type, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) +
  coord_flip() +
  theme_bw() +
  theme( axis.title.y.right = element_text(size = text.size),
         axis.text.y.right = element_text(size = text.size),
          axis.title.y = element_text( size = text.size),
         axis.text.y = element_text(size = text.size),
  ) +
  ylab("Num. pairs genomes") +
  xlab("") +
 #scale_y_log10() +
  #scale_fill_manual()
  scale_fill_manual(values = c("same" = "darkgreen", "different" = "orange", "unknown" = "gray"))
g
ggsave(g, filename = sprintf("%sallpairs_meta.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 15, height = 20)  
```



# Let's do the same of pairs of rHMM families
# it is a bit more tricky because a pair of mosaic tail fibers may actually mean that a tail spike within some family where a tail fiber exist was mosaic 
# Also we will nly look at rHMM families that are found ONLY in one lifestyle/genus



```{r}



Get.Taxa.And.Lifestyle.Fam.Pairs = function(mosaic.pairs.lifestyle2) {
    mosaic.pairs.lifestyle2$genus.x[is.na(mosaic.pairs.lifestyle2$genus.x)] = unclear.name
    mosaic.pairs.lifestyle2$genus.x[mosaic.pairs.lifestyle2$genus.x == "multi"] = unclear.name
    mosaic.pairs.lifestyle2$genus.x[mosaic.pairs.lifestyle2$genus.x == "Unclassified"] = unclear.name
    mosaic.pairs.lifestyle2$family.x[is.na(mosaic.pairs.lifestyle2$family.x)] = unclear.name
    mosaic.pairs.lifestyle2$family.x[mosaic.pairs.lifestyle2$family.x == "multi"] = unclear.name
    mosaic.pairs.lifestyle2$family.x[mosaic.pairs.lifestyle2$family.x == "Unclassified"] = unclear.name
    mosaic.pairs.lifestyle2$genus.y[is.na(mosaic.pairs.lifestyle2$genus.y)] = unclear.name
    mosaic.pairs.lifestyle2$genus.y[mosaic.pairs.lifestyle2$genus.y == "multi"] = unclear.name
    mosaic.pairs.lifestyle2$genus.y[mosaic.pairs.lifestyle2$genus.y == "Unclassified"] = unclear.name
    mosaic.pairs.lifestyle2$family.y[is.na(mosaic.pairs.lifestyle2$family.y)] = unclear.name
    mosaic.pairs.lifestyle2$family.y[mosaic.pairs.lifestyle2$family.y == "multi"] = unclear.name
    mosaic.pairs.lifestyle2$family.y[mosaic.pairs.lifestyle2$family.y == "Unclassified"] = unclear.name
    mosaic.pairs.lifestyle2$lifestyle.x[mosaic.pairs.lifestyle2$lifestyle.x == ""] = unclear.name
    mosaic.pairs.lifestyle2$lifestyle.x[mosaic.pairs.lifestyle2$lifestyle.x == "multi"] = unclear.name
    mosaic.pairs.lifestyle2$lifestyle.x[is.na(mosaic.pairs.lifestyle2$lifestyle.x)] = unclear.name
    mosaic.pairs.lifestyle2$lifestyle.y[mosaic.pairs.lifestyle2$lifestyle.y == ""] = unclear.name
    mosaic.pairs.lifestyle2$lifestyle.y[mosaic.pairs.lifestyle2$lifestyle.y == "multi"] = unclear.name
    mosaic.pairs.lifestyle2$lifestyle.y[is.na(mosaic.pairs.lifestyle2$lifestyle.y)] = unclear.name

   mosaic.pairs.lifestyle2.tb = mosaic.pairs.lifestyle2 %>% as.data.table()
   mosaic.pairs.lifestyle2.tb[, known.lifestyle := lifestyle.x !="unknown" & lifestyle.y != "unknown"]
   mosaic.pairs.lifestyle2.tb[, known.genus := genus.x !="unknown" & genus.y != "unknown"]
   mosaic.pairs.lifestyle2.tb[, known.fam := family.x !="unknown" & family.y != "unknown"]
   mosaic.pairs.lifestyle2.tb[, same.l := if_else(!known.lifestyle, "unknwon", if_else(lifestyle.x == lifestyle.y, "same", "different"))]
      
  all.fam.pairs.with.metadata.l = mosaic.pairs.lifestyle2.tb %>%
    rename(same = same.l) %>%
      group_by(annotation.index, same) %>%
      summarise(n = n_distinct(family.pair)) %>%
      right_join(top.diversity.classes) 
 
  all.fam.pairs.with.metadata.tg = mosaic.pairs.lifestyle2.tb %>%
    #rowwise() %>%
    #mutate(tg.pair = CreatePair(qGenus, sGenus, collapse = " - ")) %>%
    mutate(same = if_else(!known.genus, "unknown", if_else(genus.x == genus.y, "same", "different"))) %>%
    group_by(annotation.index, same) %>%
    summarise(n = n_distinct(family.pair))

  
  all.fam.pairs.with.metadata.fam = mosaic.pairs.lifestyle2.tb %>%
    #rowwise() %>%
    #mutate(fam.pair = CreatePair(qfamily, sfamily, collapse = " - ")) %>%
    mutate(same = if_else(!known.fam, "unknown", if_else(family.x == family.y, "same", "different"))) %>%
    group_by(annotation.index, same) %>%
    summarise(n = n_distinct(family.pair)) %>%
    right_join(top.diversity.classes) 

  
  all.fam.pairs.meta.all = all.fam.pairs.with.metadata.l %>% mutate(type = "lifestyle") %>% select(annotation.index, same, n, type) %>%
    rbind(all.fam.pairs.with.metadata.fam %>% mutate(type = "family") %>% select(annotation.index, same, n, type)) %>%
    rbind(all.fam.pairs.with.metadata.tg %>% mutate(type = "genus") %>% select(annotation.index, same, n, type)) %>%
    left_join(annotation.indices)  %>%
    right_join(top.diversity.classes) 
  all.fam.pairs.meta.all$n[is.na(all.fam.pairs.meta.all$n)] = 0
  all.fam.pairs.meta.all$same[is.na(all.fam.pairs.meta.all$same)] = "unknwon"
  return(all.fam.pairs.meta.all)
  }
  
```

```{r}
families.and.functions.top.diversity = families.and.functions %>% 
  inner_join(top.diversity.classes)
unclear.name = "unknown"
mosaic.pairs.lifestyle2 = mosaic.pairs.lifestyle %>%
  inner_join(families.and.functions.top.diversity %>% select(qindex = annotation.index, fam.x = fam)) %>%
  inner_join(families.and.functions.top.diversity %>% select(sindex = annotation.index, fam.y = fam)) %>%
  filter(qindex == sindex) %>%
  mutate(annotation.index = qindex)
all.fam.pairs.meta.all = Get.Taxa.And.Lifestyle.Fam.Pairs(mosaic.pairs.lifestyle2)
 
text.size = 8
g=ggplot(all.fam.pairs.meta.all) +
  geom_col(aes(x = annotation, y = n, fill = same)) +
  facet_grid(category ~type, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) +
  coord_flip() +
  theme_bw() +
  theme( axis.title.y.right = element_text(size = text.size),
         axis.text.y.right = element_text(size = text.size),
          axis.title.y = element_text( size = text.size),
         axis.text.y = element_text(size = text.size),
  ) +
  ylab("Num. pairs rHMM families") +
  xlab("") +
 #scale_y_log10() +
  #scale_fill_manual()
  scale_fill_manual(values = c("same" = "darkgreen", "different" = "orange", "unknown" = "gray"))
g
ggsave(g, filename = sprintf("%sall_fam_pairs_meta.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 15, height = 20)  




recent.family.pairs = recent.pairs.with.metadata %>%
  distinct(qname,sname) %>%
  left_join(families %>% select(qname, qfam = family)) %>%
  left_join(families %>% select(sname = qname, sfam = family)) %>%
  rowwise() %>%
  mutate(family.pair = CreatePair(qfam, sfam)) %>%
  distinct(family.pair)

recent.fam.pairs.meta.all = Get.Taxa.And.Lifestyle.Fam.Pairs(mosaic.pairs.lifestyle2 = mosaic.pairs.lifestyle2 %>% inner_join(recent.family.pairs))
 
text.size = 8
g=ggplot(recent.fam.pairs.meta.all) +
  geom_col(aes(x = annotation, y = n, fill = same)) +
  facet_grid(category ~type, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) +
  coord_flip() +
  theme_bw() +
  theme( axis.title.y.right = element_text(size = text.size),
         axis.text.y.right = element_text(size = text.size),
          axis.title.y = element_text( size = text.size),
         axis.text.y = element_text(size = text.size),
  ) +
  ylab("Num. pairs rHMM families") +
  xlab("") +
 #scale_y_log10() +
  #scale_fill_manual()
  scale_fill_manual(values = c("same" = "darkgreen", "different" = "orange", "unknown" = "gray"))
g
ggsave(g, filename = sprintf("%srecent_fam_pairs_meta.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 15, height = 20)  



```


# Same of represeqs
```{r}

sure.annotations.from.top.diversity = surely.annotated.proteins %>% inner_join(top.diversity.classes)
all.mosaic.reprseq.pairs = mosaic.reprseq.pairs.wide %>%
  #left_join(families.metadata.table %>% select(fam.x = fam, genus.x = genus, lifestyle.x = lifestyle, family.x = family)) %>%
  #left_join(families.metadata.table %>% select(fam.y = fam, genus.y = genus, lifestyle.y = lifestyle, family.y = family)) %>%
  inner_join(sure.annotations.from.top.diversity %>% select(qname = qname, qindex = annotation.index)) %>%
  inner_join(sure.annotations.from.top.diversity %>% select(sname = qname, sindex = annotation.index)) %>%
  filter(qindex == sindex) %>%
  left_join(reprseq.metadata.table %>% select(qname = repr.name,  lifestyle.x = lifestyle, genus.x = genus, family.x = family, qnum.genomes.with.this.rHMM = num.genomes.with.this.rHMM)) %>%
  left_join(reprseq.metadata.table %>% select(sname = repr.name,  lifestyle.y = lifestyle, genus.y = genus, family.y = family, snum.genomes.with.this.rHMM = num.genomes.with.this.rHMM)) %>%
  mutate(family.pair = reprseq.pair) %>%
  inner_join(sure.annotations.from.top.diversity %>% select(qname, qindex = annotation.index)) %>%
  inner_join(sure.annotations.from.top.diversity %>% select(sname = qname, sindex = annotation.index)) %>%
  filter(qindex == sindex) %>%
  mutate(annotation.index = qindex)

all.reprseq.pairs.meta.all = Get.Taxa.And.Lifestyle.Fam.Pairs(all.mosaic.reprseq.pairs)
 
text.size = 8
g=ggplot(all.reprseq.pairs.meta.all) +
  geom_col(aes(x = annotation, y = n, fill = same)) +
  facet_grid(category ~type, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) +
  coord_flip() +
  theme_bw() +
  theme( axis.title.y.right = element_text(size = text.size),
         axis.text.y.right = element_text(size = text.size),
          axis.title.y = element_text( size = text.size),
         axis.text.y = element_text(size = text.size),
  ) +
  ylab("Num. pairs rHMMs") +
  xlab("") +
 #scale_y_log10() +
  #scale_fill_manual()
  scale_fill_manual(values = c("same" = "darkgreen", "different" = "orange", "unknown" = "gray"))
g
ggsave(g, filename = sprintf("%sall_reprseq_pairs_meta.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 15, height = 20)  






recent_mosaic_reprseq_pairs.with.metadata = recent_HGT_pairs_raw.70 %>%
  # take reprseq lifestyle and taxa
  left_join(reprseq.metadata.table %>% select(qname = repr.name,  lifestyle.x = lifestyle, genus.x = genus, family.x = family, qnum.genomes.with.this.rHMM = num.genomes.with.this.rHMM)) %>%
  left_join(reprseq.metadata.table %>% select(sname = repr.name,  lifestyle.y = lifestyle, genus.y = genus, family.y = family, snum.genomes.with.this.rHMM = num.genomes.with.this.rHMM)) %>%
  # take reprseq rather than family pairs
  mutate(family.pair = pair) %>%
  inner_join(sure.annotations.from.top.diversity %>% select(qname, qindex = annotation.index)) %>%
  inner_join(sure.annotations.from.top.diversity %>% select(sname = qname, sindex = annotation.index)) %>%
  filter(qindex == sindex) %>%
  mutate(annotation.index = qindex)

recent.reprseq.pairs.meta.all = Get.Taxa.And.Lifestyle.Fam.Pairs(recent_mosaic_reprseq_pairs.with.metadata)
 
text.size = 8
g=ggplot(recent.reprseq.pairs.meta.all) +
  geom_col(aes(x = annotation, y = n, fill = same)) +
  facet_grid(category ~type, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) +
  coord_flip() +
  theme_bw() +
  theme( axis.title.y.right = element_text(size = text.size),
         axis.text.y.right = element_text(size = text.size),
          axis.title.y = element_text( size = text.size),
         axis.text.y = element_text(size = text.size),
  ) +
  ylab("Num. pairs rHMMs") +
  xlab("") +
 #scale_y_log10() +
  #scale_fill_manual()
  scale_fill_manual(values = c("same" = "darkgreen", "different" = "orange", "unknown" = "gray"))
g
ggsave(g, filename = sprintf("%srecent_reprseq_pairs_meta.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 15, height = 20)  



```

Let us look at the number of families with mosaic signal that crosses the taxa /lifestyle boarder
```{r, fig.width = 10, fig.height = 10}



all.mosaic.pairs.data = mosaic.reprseq.pairs.wide %>%
  # we will look at wnames from top classes
  inner_join(sure.annotations.from.top.diversity %>% select(qname = qname, qindex = annotation.index)) %>%
  left_join(name.table.with.metadata %>% distinct(qgenome = Accession, qname = repr.name, qlifestyle = lifestyle, qGenus = Genus, qfamily = Family)) %>%
  left_join(name.table.with.metadata %>% distinct(sgenome = Accession, sname = repr.name, slifestyle = lifestyle, sGenus = Genus, sfamily = Family)) %>%
  left_join(families %>% select(fam.q = family, qname = qname)) %>%
  left_join(families %>% select(fam.s = family, sname = qname)) 

all.mosaic.pairs.data[,qlifestyle := if_else(is.na(all.mosaic.pairs.data$qlifestyle), "unknown", all.mosaic.pairs.data$qlifestyle)]
all.mosaic.pairs.data[,slifestyle := if_else(is.na(all.mosaic.pairs.data$slifestyle), "unknown", all.mosaic.pairs.data$slifestyle)]
all.mosaic.pairs.data[,qfamily := if_else(is.na(all.mosaic.pairs.data$qfamily), "unknown", all.mosaic.pairs.data$qfamily)]
all.mosaic.pairs.data[,sfamily := if_else(is.na(all.mosaic.pairs.data$sfamily), "unknown", all.mosaic.pairs.data$sfamily)]
all.mosaic.pairs.data[,qGenus := if_else(is.na(all.mosaic.pairs.data$qGenus), "unknown", all.mosaic.pairs.data$qGenus)]
all.mosaic.pairs.data[,sGenus := if_else(is.na(all.mosaic.pairs.data$sGenus), "unknown", all.mosaic.pairs.data$sGenus)]
all.mosaic.pairs.data[,qlifestyle := if_else(all.mosaic.pairs.data$qlifestyle == "", "unknown", all.mosaic.pairs.data$qlifestyle)]
all.mosaic.pairs.data[,slifestyle := if_else(all.mosaic.pairs.data$slifestyle == "", "unknown", all.mosaic.pairs.data$slifestyle)]
all.mosaic.pairs.data[,qfamily := if_else(all.mosaic.pairs.data$qfamily == "", "unknown", all.mosaic.pairs.data$qfamily)]
all.mosaic.pairs.data[,sfamily := if_else(all.mosaic.pairs.data$sfamily == "", "unknown", all.mosaic.pairs.data$sfamily)]
all.mosaic.pairs.data[,qGenus := if_else(all.mosaic.pairs.data$qGenus == "", "unknown", all.mosaic.pairs.data$qGenus)]
all.mosaic.pairs.data[,sGenus := if_else(all.mosaic.pairs.data$sGenus == "", "unknown", all.mosaic.pairs.data$sGenus)]

   all.mosaic.pairs.data.tb = all.mosaic.pairs.data %>% as.data.table()
   all.mosaic.pairs.data.tb[, known.lifestyle := slifestyle !="unknown" & qlifestyle != "unknown"]
   all.mosaic.pairs.data.tb[, known.genus := qGenus !="unknown" & sGenus != "unknown"]
   all.mosaic.pairs.data.tb[, known.fam := qfamily !="unknown" & sfamily != "unknown"]
   all.mosaic.pairs.data.tb[, same.l := if_else(!known.lifestyle, "unknwon", if_else(qlifestyle == slifestyle, "same", "different"))]
   all.mosaic.pairs.data.tb[, same.g := if_else(!known.genus, "unknwon", if_else(qGenus == sGenus, "same", "different"))]
   all.mosaic.pairs.data.tb[, same.f := if_else(!known.fam, "unknwon", if_else(qfamily == sfamily, "same", "different"))]

  # this is symmetric, every pair is there twice
aa=   all.mosaic.pairs.data.tb %>%
  group_by(fam.q, qindex) %>%
  # if there is any pair of genomes with different lifestyle/taxa we classify ths family*annotation as having mosaicism between lifestyle.taxa
  # otherwise we classify as having mosaic pairs only within the same lifestyle/taxa
  # unless all pairs contain at least one unclassfied genome
  summarise(lifestyle = if_else(any(same.l == "different"), "different", if_else(any(same.l == "same"), "same", "unknown")),
            genus = if_else(any(same.g == "different"), "different", if_else(any(same.g == "same"), "same", "unknown")),
            family = if_else(any(same.f == "different"), "different", if_else(any(same.f == "same"), "same", "unknown")))

top.diversity.sizes = rbind(included.category.sizes %>% inner_join(top.diversity.classes) %>% mutate(crosses = "all families")  %>% mutate(cross.type = "family"),
                            included.category.sizes %>% inner_join(top.diversity.classes) %>% mutate(crosses = "all families")  %>% mutate(cross.type = "genus"),
                            included.category.sizes %>% inner_join(top.diversity.classes) %>% mutate(crosses = "all families")  %>% mutate(cross.type = "lifestyle")) %>%
  select(annotation.index, cross.type, crosses, n = num.families)
  
bb = aa %>% 
  mutate(annotation.index = qindex) %>%
  tidyr::gather(key = "cross.type", value = "crosses", lifestyle, genus, family) %>%
  group_by(annotation.index, cross.type, crosses) %>%
  summarise(n = n()) %>%
  rbind(top.diversity.sizes) %>%
  right_join(top.diversity.classes) %>%
  left_join(annotation.indices, by = "annotation.index") %>%
  left_join(included.category.sizes %>% select(annotation.index, num.families)) %>%
  rowwise() %>%
  mutate(prop = n/num.families)


g= ggplot(bb  %>% filter(crosses != "all families")) +
  geom_col(aes(x= annotation, y = prop, alpha = crosses, fill = category, col = category)) +
  facet_grid(category ~cross.type, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) +
  coord_flip() +
  theme_bw() +
  theme( axis.title.y.right = element_text(size = text.size),
         axis.text.y.right = element_text(size = text.size),
          axis.title.y = element_text( size = text.size),
         axis.text.y = element_text(size = text.size),
  ) +
  scale_fill_manual(values = PHROG.COLOR.MAP) +
    scale_color_manual(values = PHROG.COLOR.MAP) +
  scale_alpha_manual(values = c("all families" = 0, "unknown" = 0.2, "same" = 0.5, "different" = 1)) +
  #ylab("Num. mosaic rHMM families") +
  ylab("Prop. rHMM families") +
  xlab("")
g
ggsave(g, filename = sprintf("%sfamilies_mosaic_signal_cross2.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 15, height = 20)  


###################### RECENT ONLY ###################################

  # this is symmetric, every pair is there twice
aa.recent  =  all.mosaic.pairs.data.tb %>%
  inner_join(recent_HGT_pairs_raw.70 %>% select(reprseq.pair = pair) %>% rowwise() %>% mutate(reprseq.pair = gsub(" AND ", "&", reprseq.pair))) %>%
  group_by(fam.q, qindex) %>%
  # if there is any pair of genomes with different lifestyle/taxa we classify ths family*annotation as having mosaicism between lifestyle.taxa
  # otherwise we classify as having mosaic pairs only within the same lifestyle/taxa
  # unless all pairs contain at least one unclassfied genome
  summarise(lifestyle = if_else(any(same.l == "different"), "some genome pairs from different", if_else(any(same.l =="same"), "all known genome pairs from same", "only genome pairs from unknown")),
            genus = if_else(any(same.g == "different"), "some genome pairs from different", if_else(any(same.g =="same"), "all known genome pairs from same", "only genome pairs from unknown")),
            family = if_else(any(same.f == "different"), "some genome pairs from different", if_else(any(same.f =="same"), "all known genome pairs from same", "only genome pairs from unknown")))

top.diversity.sizes = rbind(included.category.sizes %>% inner_join(top.diversity.classes) %>% mutate(crosses = "all families")  %>% mutate(cross.type = "family"),
                            included.category.sizes %>% inner_join(top.diversity.classes) %>% mutate(crosses = "all families")  %>% mutate(cross.type = "genus"),
                            included.category.sizes %>% inner_join(top.diversity.classes) %>% mutate(crosses = "all families")  %>% mutate(cross.type = "lifestyle")) %>%
  select(annotation.index, cross.type, crosses, n = num.families)
  
bb.recent = aa.recent %>% 
  mutate(annotation.index = qindex) %>%
  tidyr::gather(key = "cross.type", value = "crosses", lifestyle, genus, family) %>%
  group_by(annotation.index, cross.type, crosses) %>%
  summarise(n = n()) %>%
  rbind(top.diversity.sizes) %>%
  right_join(top.diversity.classes) %>%
  left_join(annotation.indices, by = "annotation.index") %>%
  left_join(included.category.sizes %>% select(annotation.index, num.families)) %>%
  rowwise() %>%
  mutate(prop = n/num.families)


g= ggplot(bb.recent  %>% filter(crosses != "all families")) +
  geom_col(aes(x= annotation, y = prop, alpha = crosses, fill = category, col = category)) +
  facet_grid(category ~cross.type, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) +
  coord_flip() +
  theme_bw() +
  theme( axis.title.y.right = element_text(size = text.size),
         axis.text.y.right = element_text(size = text.size),
          axis.title.y = element_text( size = text.size),
         axis.text.y = element_text(size = text.size),
  ) +
  scale_fill_manual(values = PHROG.COLOR.MAP) +
    scale_color_manual(values = PHROG.COLOR.MAP) +
  guides(fill = "none", col = "none") +
  scale_alpha_manual(values = c( "only genome pairs from unknown" = 0, "all known genome pairs from same" = 0.5,  "some genome pairs from different" = 1), name = "Type") +
  #ylab("Num. mosaic rHMM families") +
  ylab("Prop. rHMM families") +
  xlab("")
g
ggsave(g, filename = sprintf("%sfamilies_recent_mosaic_signal_cross2.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 15, height = 20)  

```




















```{r}
all.proteins.in.genomes.dat.by.molecule = all.proteins.in.genomes %>%
  group_by(Accession, Family) %>%
  summarise(num.genes.in.genome = n_distinct(repr.name),
         num.mosaic.genes.in.genome = n_distinct(repr.name[mosaic]),
         num.domain.mosaic.genes.in.genome = n_distinct(repr.name[mosaic.domain])
         ) %>%
  mutate(prop.mosaic.genes.in.genome = num.mosaic.genes.in.genome/num.genes.in.genome,
         prop.domain.mosaic.genes.in.genome = num.domain.mosaic.genes.in.genome/num.genes.in.genome
         ) %>%
  left_join(num.genomes.per.tax.fam) %>%
  rowwise() %>%
  filter(Family != "unknown") %>% 
  left_join(genome.sizes.per.family) %>%
  mutate(Family = paste0(Family, "\n", n.genomes, " genomes\n", "av. genome size:",av.genome.size)) %>%
  arrange(av.genome.size)



kwresult = kruskal.test(prop.mosaic.genes.in.genome ~ Family, data = all.proteins.in.genomes.dat.by.molecule)


library(ggpubr)
all.proteins.in.genomes.dat.by.molecule$Family = factor(all.proteins.in.genomes.dat.by.molecule$Family, levels = unique(all.proteins.in.genomes.dat.by.molecule$Family))
g=ggplot(all.proteins.in.genomes.dat.by.molecule %>% filter(big.fam),
         aes(x=Family, y = prop.mosaic.genes.in.genome)) + 
  geom_jitter(aes(col = Family), size = 0.5, alpha = 0.3) +
  stat_compare_means() +
  #geom_histogram(aes(x=prop.mosaic.genes.in.genome, fill = lifestyle), alpha = 0.5) + facet_grid(lifestyle ~.)
  geom_boxplot(alpha = 0.3) + 
  #geom_label(x="Straboviridae", y = 0.3, label = paste0("Kruskal-Wallis: ", as.numeric(kwresult$statistic),
  #                                                         "\np=", as.numeric(kwresult$p.value))) +
  theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  guides(col ="none") +
  xlab("")
g
ggsave(g, filename = sprintf("%sprop_mosaic_genes_per_genome_and_TaxFam.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 8, height = 6)  


```



```{r}
all.proteins.in.genomes.dat.by.molecule = all.proteins.in.genomes %>%
  group_by(Accession, GeneticMaterial) %>%
  summarise(num.genes.in.genome = n_distinct(repr.name),
         num.mosaic.genes.in.genome = n_distinct(repr.name[mosaic]),
         num.domain.mosaic.genes.in.genome = n_distinct(repr.name[mosaic.domain])
         ) %>%
  mutate(prop.mosaic.genes.in.genome = num.mosaic.genes.in.genome/num.genes.in.genome,
         prop.domain.mosaic.genes.in.genome = num.domain.mosaic.genes.in.genome/num.genes.in.genome
         )


g=ggplot(all.proteins.in.genomes.dat.by.molecule %>% filter(GeneticMaterial %in% c("DNA", "RNA"))) + 
  geom_jitter(aes(x=GeneticMaterial, y = prop.mosaic.genes.in.genome, col = GeneticMaterial), size = 0.5, alpha = 0.3)+
  #geom_histogram(aes(x=prop.mosaic.genes.in.genome, fill = lifestyle), alpha = 0.5) + facet_grid(lifestyle ~.)
  geom_boxplot(aes(x=GeneticMaterial, y = prop.mosaic.genes.in.genome), alpha = 0.3) + 
  theme_bw()
g
ggsave(g, filename = sprintf("%sprop_mosaic_genes_per_genome_and_GeneticMaterial.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 6, height = 4)  

```



# TODO
Look at the entire genomes: temperate vs virulent and the proportion of their genes that are mosaic

```{r}



all.proteins.in.genomes.dat = all.proteins.in.genomes %>%
  group_by(Accession, lifestyle) %>%
  summarise(num.genes.in.genome = n_distinct(repr.name),
         num.mosaic.genes.in.genome = n_distinct(repr.name[mosaic]),
         num.domain.mosaic.genes.in.genome = n_distinct(repr.name[mosaic.domain])
         ) %>%
  mutate(prop.mosaic.genes.in.genome = num.mosaic.genes.in.genome/num.genes.in.genome,
         prop.domain.mosaic.genes.in.genome = num.domain.mosaic.genes.in.genome/num.genes.in.genome
         )

all.proteins.in.genomes.dat %>%
  group_by(lifestyle) %>%
  summarise(num.genomes = n_distinct(Accession),
            av.num.mosaic.genes.in.genome = mean(num.mosaic.genes.in.genome),
            av.prop.mosaic.genes.in.genome = mean(prop.mosaic.genes.in.genome),
            num.genomes.with.any.mosaic.gene = n_distinct(Accession[num.mosaic.genes.in.genome > 0]), 
            prop.genomes.with.any.mosaic.gene = n_distinct(Accession[num.mosaic.genes.in.genome > 0])/num.genomes)


g=ggplot(all.proteins.in.genomes.dat %>% filter(lifestyle %in% c("temperate", "virulent"))) + 
  geom_jitter(aes(x=lifestyle, y = prop.mosaic.genes.in.genome, col = lifestyle), size = 0.5, alpha = 0.3)+
  #geom_histogram(aes(x=prop.mosaic.genes.in.genome, fill = lifestyle), alpha = 0.5) + facet_grid(lifestyle ~.)
  geom_boxplot(aes(x=lifestyle, y = prop.mosaic.genes.in.genome), alpha = 0.3) + 
  theme_bw()
g
ggsave(g, filename = sprintf("%sprop_mosaic_genes_per_genome_and_lifestyle.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 6, height = 4)  



wilcox.test(x = all.proteins.in.genomes.dat %>% filter(lifestyle %in% c("virulent")) %>% pull(prop.mosaic.genes.in.genome),
            y = all.proteins.in.genomes.dat %>% filter(lifestyle %in% c("temperate")) %>% pull(prop.mosaic.genes.in.genome),
            alternative = "less")$p.value

wilcox.test(x = all.proteins.in.genomes.dat %>% filter(lifestyle %in% c("virulent")) %>% pull(prop.domain.mosaic.genes.in.genome),
            y = all.proteins.in.genomes.dat %>% filter(lifestyle %in% c("temperate")) %>% pull(prop.domain.mosaic.genes.in.genome),
            alternative = "greater")$p.value


g=ggplot(all.proteins.in.genomes.dat %>% filter(lifestyle %in% c("temperate", "virulent"))) + 
  geom_jitter(aes(x=lifestyle, y = num.mosaic.genes.in.genome, col = lifestyle), size = 0.5, alpha = 0.3)+
  #geom_histogram(aes(x=prop.mosaic.genes.in.genome, fill = lifestyle), alpha = 0.5) + facet_grid(lifestyle ~.)
  geom_boxplot(aes(x=lifestyle, y = num.mosaic.genes.in.genome), alpha = 0.3) + 
  theme_bw()
g
ggsave(g, filename = sprintf("%snum_mosaic_genes_per_genome_and_lifestyle.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 6, height =4)  


g=ggplot(all.proteins.in.genomes.dat %>% filter(lifestyle %in% c("temperate", "virulent"))) + 
  geom_jitter(aes(x=lifestyle, y = prop.domain.mosaic.genes.in.genome, col = lifestyle), size = 0.5, alpha = 0.3)+
  #geom_histogram(aes(x=prop.mosaic.genes.in.genome, fill = lifestyle), alpha = 0.5) + facet_grid(lifestyle ~.)
  geom_boxplot(aes(x=lifestyle, y = prop.domain.mosaic.genes.in.genome), alpha = 0.3) + 
  theme_bw()
g
ggsave(g, filename = sprintf("%sprop_mosaic_genes_per_genome_and_lifestyle_domain.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 6, height = 4)  

g=ggplot(all.proteins.in.genomes.dat %>% filter(lifestyle %in% c("temperate", "virulent"))) + 
  geom_jitter(aes(x=lifestyle, y = num.domain.mosaic.genes.in.genome, col = lifestyle), size = 0.5, alpha = 0.3)+
  #geom_histogram(aes(x=prop.mosaic.genes.in.genome, fill = lifestyle), alpha = 0.5) + facet_grid(lifestyle ~.)
  geom_boxplot(aes(x=lifestyle, y = num.domain.mosaic.genes.in.genome), alpha = 0.3) + 
  theme_bw()
g
ggsave(g, filename = sprintf("%snum_mosaic_genes_per_genome_and_lifestyle_domain.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 6, height =4)  


```


# take into account annotated and unannotated genes
```{r}
all.proteins.in.genomes.dat2 = all.proteins.in.genomes %>%
  group_by(Accession, lifestyle) %>%
  mutate(num.genes.in.genome = n_distinct(repr.name)) %>%
  filter(!(repr.name %in% unaccepted.genes$qname)) %>%
  group_by(has.unknown.function, Accession, lifestyle, num.genes.in.genome) %>%
  summarise(num.mosaic.genes.in.genome = n_distinct(repr.name[mosaic]),
            num.domain.mosaic.genes.in.genome = n_distinct(repr.name[mosaic.domain])) 



g=ggplot(all.proteins.in.genomes.dat2 %>% filter(lifestyle %in% c("temperate", "virulent"))) + 
  geom_jitter(aes(x=lifestyle, y = num.mosaic.genes.in.genome, col = lifestyle), size = 0.5, alpha = 0.3)+
  #geom_histogram(aes(x=prop.mosaic.genes.in.genome, fill = lifestyle), alpha = 0.5) + facet_grid(lifestyle ~.)
  geom_boxplot(aes(x=lifestyle, y = num.mosaic.genes.in.genome), alpha = 0.3) + 
  theme_bw()+
  facet_grid(.~has.unknown.function)
g
ggsave(g, filename = sprintf("%snum_mosaic_genes_per_genome_and_lifestyle2.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 6, height =4)  


g=ggplot(all.proteins.in.genomes.dat2 %>% filter(lifestyle %in% c("temperate", "virulent"))) + 
  geom_jitter(aes(x=lifestyle, y = num.domain.mosaic.genes.in.genome, col = lifestyle), size = 0.5, alpha = 0.3)+
  #geom_histogram(aes(x=prop.mosaic.genes.in.genome, fill = lifestyle), alpha = 0.5) + facet_grid(lifestyle ~.)
  geom_boxplot(aes(x=lifestyle, y = num.domain.mosaic.genes.in.genome), alpha = 0.3) + 
  theme_bw()+
  facet_grid(.~has.unknown.function)
g
ggsave(g, filename = sprintf("%snum_mosaic_domain_genes_per_genome_and_lifestyle2.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 6, height =4)  


```

```{r}
genomes.per.family = 
all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  left_join(surely.annotated.proteins) %>%
  group_by(Accession, lifestyle, Family) %>%
  summarise(genome.size = n_distinct(ncbi.id)) %>%
  group_by(Family) %>%
  summarise(av.genome.size = median(genome.size),
            num.virulent.genomes = n_distinct(Accession[lifestyle == "virulent"]),
            num.temperate.genomes = n_distinct(Accession[lifestyle == "temperate"]),
            num.genomes = n_distinct(Accession)) %>%
  mutate(prop.virulent.genomes = num.virulent.genomes/num.genomes,
         prop.temperate.genomes = num.temperate.genomes/num.genomes) %>%
  arrange(desc(num.genomes)) %>%
  mutate(Lifestyle = if_else(prop.temperate.genomes > 0.5, "mostly temperate", 
                             if_else(prop.virulent.genomes > 0.5, "mostly virulent", "temperate/virulent")))


functions.per.family = all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  left_join(surely.annotated.proteins) %>%
  group_by(annotation.index, Family) %>%
  summarise(num.genomes.this.function = n_distinct(Accession)) %>%
  left_join(genomes.per.family, by = "Family") %>%
  mutate(prop.genomes.with.this.function = num.genomes.this.function/num.genomes) %>%
  left_join(annotation.indices) %>%
  filter(num.genomes >= 50 & !is.na(Family) & Family != "Unclassified") %>%
  inner_join(top.diversity.classes %>% distinct(annotation.index)) %>%
  mutate(family = paste0(Family, "\n", Lifestyle, "\n", num.genomes, " genomes", "\nav. genome size: ", av.genome.size))
functions.per.family$family = factor(functions.per.family$family, levels = functions.per.family %>% arrange(desc(num.genomes)) %>% pull(family) %>% unique())

g=ggplot(functions.per.family) +
  geom_tile(aes(x = family, y =annotation , fill = prop.genomes.with.this.function)) + #, hjust = 0, vjust = 0
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  facet_grid(category ~., scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) + 
  xlab("") + ylab("") +
  scale_fill_gradient(name = "% genomes", low = "#ADD8E6", high = "#000046")
g  
  ggsave(g, filename = sprintf("%sfunctions_vs_taxonomy_family.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 18, height =15)  
  
  
  
  genomes.per.genera = 
all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  left_join(surely.annotated.proteins) %>%
  group_by(Accession, lifestyle, Genus) %>%
  summarise(genome.size = n_distinct(ncbi.id)) %>%
  group_by(Genus) %>%
  summarise(av.genome.size = median(genome.size),
            num.virulent.genomes = n_distinct(Accession[lifestyle == "virulent"]),
            num.temperate.genomes = n_distinct(Accession[lifestyle == "temperate"]),
            num.genomes = n_distinct(Accession)) %>%
  mutate(prop.virulent.genomes = num.virulent.genomes/num.genomes,
         prop.temperate.genomes = num.temperate.genomes/num.genomes) %>%
  arrange(desc(num.genomes)) %>%
  mutate(Lifestyle = if_else(prop.temperate.genomes > 0.5, "mostly temperate", 
                             if_else(prop.virulent.genomes > 0.5, "mostly virulent", "temperate/virulent")))


functions.per.genera= all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  left_join(surely.annotated.proteins) %>%
  group_by(annotation.index, Genus) %>%
  summarise(num.genomes.this.function = n_distinct(Accession)) %>%
  left_join(genomes.per.genera, by = "Genus") %>%
  mutate(prop.genomes.with.this.function = num.genomes.this.function/num.genomes) %>%
  left_join(annotation.indices) %>%
  filter(num.genomes >= 30 & !is.na(Genus) & Genus != "Unclassified") %>%
  inner_join(top.diversity.classes %>% distinct(annotation.index)) %>%
  mutate(family = paste0(Genus, "\n", Lifestyle, "\n", num.genomes, " genomes", "\nav. genome size: ", av.genome.size))
functions.per.genera$family = factor(functions.per.genera$family, levels = functions.per.genera %>% arrange(desc(num.genomes)) %>% pull(family) %>% unique())

g=ggplot(functions.per.genera) +
  geom_tile(aes(x = family, y =annotation , fill = prop.genomes.with.this.function)) + #, hjust = 0, vjust = 0
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  facet_grid(category ~., scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) + 
  xlab("") + ylab("") +
  scale_fill_gradient(name = "% genomes", low = "#ADD8E6", high = "#000046")
g  
  ggsave(g, filename = sprintf("%sfunctions_vs_taxonomy_genera.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width = 18, height =15)  


```
# Same faceting by lifestyle
```{r}
genomes.per.family = 
all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  group_by(Family) %>%
  mutate(num.genomes.in.family = n_distinct(Accession)) %>%
  group_by(Accession, lifestyle, Family, num.genomes.in.family) %>%
  summarise(genome.size = n_distinct(ncbi.id)) %>%
  group_by(Family, lifestyle, num.genomes.in.family) %>%
  summarise(av.genome.size = median(genome.size),
            num.genomes = n_distinct(Accession)) 


functions.per.family = all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  left_join(surely.annotated.proteins) %>%
  group_by(annotation.index, Family, lifestyle) %>%
  summarise(num.genomes.this.function = n_distinct(Accession)) %>%
  left_join(genomes.per.family, by = c("Family", "lifestyle")) %>%
  mutate(prop.genomes.with.this.function = num.genomes.this.function/num.genomes) %>%
  left_join(annotation.indices) %>%
  filter(num.genomes >= 10 & !is.na(Family) & Family != "Unclassified" & lifestyle != "unknown") %>%
  inner_join(top.diversity.classes %>% distinct(annotation.index)) %>%
  mutate(family = paste0(Family, "\n", num.genomes, " genomes", "\nav. genome size: ", av.genome.size))
functions.per.family$family = factor(functions.per.family$family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family), desc(num.genomes)) %>% pull(family) %>% unique())


g=ggplot(functions.per.family) +
  geom_tile(aes(x = family, y =annotation , fill = prop.genomes.with.this.function)) + #, hjust = 0, vjust = 0
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  facet_grid(category ~ lifestyle, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) + 
  xlab("") + ylab("") +
  scale_fill_gradient(name = "% genomes", low = "#ADD8E6", high = "#000046")
g  
  ggsave(g, filename = sprintf("%sfunctions_vs_taxonomy_and_lifestyle.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width =25, height =15)  

  
  
genomes.per.family = 
all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  group_by(Family) %>%
  mutate(num.genomes.in.family = n_distinct(Accession)) %>%
  group_by(Accession, Genus, Family, num.genomes.in.family) %>%
  summarise(genome.size = n_distinct(ncbi.id)) %>%
  group_by(Genus, Family, num.genomes.in.family) %>%
  summarise(av.genome.size = median(genome.size),
            num.genomes = n_distinct(Accession)) 


functions.per.family = all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  left_join(surely.annotated.proteins) %>%
  group_by(annotation.index, Genus, Family) %>%
  summarise(num.genomes.this.function = n_distinct(Accession),
            num.genomes.with.this.function.mosaic = n_distinct(Accession[mosaic]),
            num.genomes.with.this.function.mosaic.domain = n_distinct(Accession[mosaic.domain])) %>%
  left_join(genomes.per.family, by = c("Genus","Family")) %>%
  mutate(prop.genomes.with.this.function = num.genomes.this.function/num.genomes,
         prop.mosaic = num.genomes.with.this.function.mosaic/num.genomes.this.function,
         prop.mosaic.domain = num.genomes.with.this.function.mosaic.domain/num.genomes.this.function) %>%
  left_join(annotation.indices) %>%
  filter(num.genomes >= 20 & !is.na(Family)) %>%
  inner_join(top.diversity.classes %>% distinct(annotation.index)) %>%
  mutate(family = paste0(Genus, "\n", Family, "\n", num.genomes, " genomes", "\nav. genome size: ", av.genome.size))
functions.per.family$family = factor(functions.per.family$family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family), desc(num.genomes)) %>% pull(family) %>% unique())
functions.per.family$Family = factor(functions.per.family$Family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family)) %>% pull(Family) %>% unique())

  
functions.per.family$Family = factor(functions.per.family$Family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family)) %>% pull(Family) %>% unique())
  g=ggplot(functions.per.family) +
  geom_tile(aes(x = family, y =annotation , fill = prop.genomes.with.this.function)) + #, hjust = 0, vjust = 0
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  facet_grid(category ~ Family, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) + 
  xlab("") + ylab("") +
  scale_fill_gradient(name = "% genomes", low = "#ADD8E6", high = "#000046")
g  
  ggsave(g, filename = sprintf("%sfunctions_vs_taxonomy_family_an_genus.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width =30, height =15)  
  
  
  functions.per.family$Family = factor(functions.per.family$Family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family)) %>% pull(Family) %>% unique())
  g=ggplot(functions.per.family  %>% 
             ungroup() %>%
             filter(Genus != "unknown") %>% 
             select(Family, family, annotation, category, prop.genomes.with.this.function, prop.mosaic) %>%
    tidyr::complete(tidyr::nesting(Family, family), tidyr::nesting(annotation, category), fill  = list(prop.genomes.with.this.function = 0, prop.mosaic = 0)) )+
  geom_tile(aes(x = family, y =annotation , alpha = prop.genomes.with.this.function, fill = prop.mosaic)) + #, hjust = 0, vjust = 0
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  facet_grid(category ~ Family, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) + 
  xlab("") + ylab("") +
  scale_fill_gradient(name = "% mosaic", low = "beige", high = "darkred")
g  
  ggsave(g, filename = sprintf("%sfunctions_vs_taxonomy_family_an_genus3.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width =30, height =15)  
  
  
  
    functions.per.family$Family = factor(functions.per.family$Family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family)) %>% pull(Family) %>% unique())
  g=ggplot(functions.per.family %>% filter(Genus != "unknown")) +
  geom_tile(aes(x = family, y =annotation , alpha = prop.genomes.with.this.function, fill = prop.mosaic.domain)) + #, hjust = 0, vjust = 0
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  facet_grid(category ~ Family, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) + 
  xlab("") + ylab("") +
  scale_fill_gradient(name = "% mosaic", low = "yellow", high = "darkred")
g  
  ggsave(g, filename = sprintf("%sfunctions_vs_taxonomy_family_an_genus3_domain_mosaicism.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width =30, height =15)  
```


# same for temperate vs virulent phages

```{r}
genomes.per.lifestyle = 
all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  group_by(lifestyle) %>%
  mutate(num.genomes.in.lifestyle = n_distinct(Accession)) %>%
  group_by(Accession, lifestyle, num.genomes.in.lifestyle) %>%
  summarise(genome.size = n_distinct(ncbi.id)) %>%
  group_by(lifestyle, num.genomes.in.lifestyle) %>%
  summarise(av.genome.size = median(genome.size),
            num.genomes = n_distinct(Accession)) 


functions.per.lifestyle = all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  left_join(surely.annotated.proteins) %>%
  group_by(annotation.index, lifestyle) %>%
  summarise(num.genomes.this.function = n_distinct(Accession),
            num.genomes.with.this.function.mosaic = n_distinct(Accession[mosaic]),
            num.genomes.with.this.function.mosaic.domain = n_distinct(Accession[mosaic.domain])) %>%
  left_join(genomes.per.lifestyle, by = c("lifestyle")) %>%
  mutate(prop.genomes.with.this.function = num.genomes.this.function/num.genomes,
         prop.mosaic = num.genomes.with.this.function.mosaic/num.genomes.this.function,
         prop.mosaic.domain = num.genomes.with.this.function.mosaic.domain/num.genomes.this.function) %>%
  left_join(annotation.indices) %>%
  inner_join(top.diversity.classes %>% distinct(annotation.index))  %>% 
  filter(lifestyle != "unknown" & !is.na(lifestyle)) %>%
  mutate(lifestyle = paste0(lifestyle, "\n", num.genomes, " genomes", "\nav. genome size: ", av.genome.size))

  
  g=ggplot(functions.per.lifestyle) +
  geom_tile(aes(x = lifestyle, y =annotation , alpha = prop.genomes.with.this.function, fill = prop.mosaic)) + #, hjust = 0, vjust = 0
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  facet_grid(category ~ lifestyle, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) + 
  xlab("") + ylab("") +
  scale_fill_gradient(name = "% mosaic", low = "yellow", high = "darkred")
g  
  ggsave(g, filename = sprintf("%sfunctions_vs_lifestyle_mosaicism1.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width =10, height =15)  
  
    g=ggplot(functions.per.lifestyle) +
  geom_col(aes(y =annotation, x = prop.genomes.with.this.function, fill = prop.mosaic)) + #, hjust = 0, vjust = 0
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  facet_grid(category ~ lifestyle, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) + 
  ylab("") + xlab("% Genomes with this function") +
  scale_fill_gradient(name = "% mosaic", low = "beige", high = "darkred")
g  
  ggsave(g, filename = sprintf("%sfunctions_vs_lifestyle_mosaicism.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width =10, height =15)  

```

############################################## NEX TAXONOMY ##########################################


# same plot using new taxonomy and ecods

```{r, eval = FALSE}

##  need to save this part from script 3b, this will not work at the moment
h.domain.occurence.data = data.table::fread(sprintf("%sFigure2/h.domain.occurence.data.txt", OUTPUT.FIGURES.METADATA.PATH)) %>% 
  mutate(h_id = as.character(h_id)) %>%
  select(h_id, category = most_frequent_category) %>%
  distinct() %>%
  filter(!is.na(category))




genomes.per.family = all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  left_join(surely.annotated.ecod.domains.high.confidence %>% distinct(qname, annotation.index, family, h_id)) %>%
  group_by(Family) %>%
  mutate(num.genomes.in.family = n_distinct(Accession)) %>%
  group_by(Accession, Genus, Family, num.genomes.in.family) %>%
  summarise(genome.size = n_distinct(ncbi.id)) %>%
  group_by(Genus, Family, num.genomes.in.family) %>%
  summarise(av.genome.size = median(genome.size),
            num.genomes = n_distinct(Accession)) 


functions.per.family = all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  left_join(surely.annotated.ecod.domains.high.confidence %>% distinct(qname, annotation.index, family, h_id)) %>%
  group_by(h_id, Genus, Family) %>%
  summarise(num.genomes.this.function = n_distinct(Accession),
            num.genomes.with.this.function.mosaic = n_distinct(Accession[mosaic]),
            num.genomes.with.this.function.mosaic.domain = n_distinct(Accession[mosaic.domain])) %>%
  left_join(genomes.per.family, by = c("Genus","Family")) %>%
  mutate(prop.genomes.with.this.function = num.genomes.this.function/num.genomes,
         prop.mosaic = num.genomes.with.this.function.mosaic/num.genomes.this.function,
         prop.mosaic.domain = num.genomes.with.this.function.mosaic.domain/num.genomes.this.function) %>%
  filter(num.genomes >= 20 & !is.na(Family) & Genus != "") %>%
  mutate(family = paste0(Genus, "\n", Family, "\n", num.genomes, " genomes", "\nav. genome size: ", av.genome.size)) %>%
  left_join(ecod_id_to_names_map %>% filter(level == "H") %>% select(h_id = id, h_name = name)) %>%
  filter(!is.na(h_id)) %>%
  # only look at H_ids known in annotated proteins
  inner_join(h.domain.occurence.data)

functions.per.family$family = factor(functions.per.family$family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family), desc(num.genomes)) %>% pull(family) %>% unique())
functions.per.family$Family = factor(functions.per.family$Family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family)) %>% pull(Family) %>% unique())

  
functions.per.family$Family = factor(functions.per.family$Family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family)) %>% pull(Family) %>% unique())
  g=ggplot(functions.per.family) +
  geom_tile(aes(x = family, y =h_name , fill = prop.genomes.with.this.function)) + #, hjust = 0, vjust = 0
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  facet_grid(category ~ Family, scales = "free", space = "free"
             #labeller = labeller(category = phrog.class.labels)
             ) + 
  xlab("") + ylab("") +
  scale_fill_gradient(name = "% genomes", low = "#ADD8E6", high = "#000046")
g  
  ggsave(g, filename = sprintf("%sfunctions_ECOD_vs_new_taxonomy_family_an_genus.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width =30, height =15)  
```

# same with mid relaxed annotations
```{r}
genomes.per.family = 
all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  group_by(Family) %>%
  mutate(num.genomes.in.family = n_distinct(Accession)) %>%
  group_by(Accession, Genus, Family, num.genomes.in.family) %>%
  summarise(genome.size = n_distinct(ncbi.id)) %>%
  group_by(Genus, Family, num.genomes.in.family) %>%
  summarise(av.genome.size = median(genome.size),
            num.genomes = n_distinct(Accession)) 


functions.per.family = all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  left_join(mid.annotated.proteins.including.multi.annot) %>%
  inner_join(top.diversity.classes %>% distinct(annotation.index)) %>%
  group_by(annotation.index, Genus, Family) %>%
  summarise(num.genomes.this.function = n_distinct(Accession),
            num.genomes.with.this.function.mosaic = n_distinct(Accession[mosaic]),
            num.genomes.with.this.function.mosaic.domain = n_distinct(Accession[mosaic.domain])) %>%
  left_join(genomes.per.family, by = c("Genus","Family")) %>%
  mutate(prop.genomes.with.this.function = num.genomes.this.function/num.genomes,
         prop.mosaic = num.genomes.with.this.function.mosaic/num.genomes.this.function,
         prop.mosaic.domain = num.genomes.with.this.function.mosaic.domain/num.genomes.this.function) %>%
  left_join(annotation.indices) %>%
  filter(num.genomes >= 20 & !is.na(Family) & Genus != "") %>%
  mutate(Family = if_else(Family == "", "Unknwon", Family)) %>%
  mutate(family = paste0(Genus, "\n", Family, "\n", num.genomes, " genomes", "\nav. genome size: ", av.genome.size))
functions.per.family$family = factor(functions.per.family$family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family), desc(num.genomes)) %>% pull(family) %>% unique())
functions.per.family$Family = factor(functions.per.family$Family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family)) %>% pull(Family) %>% unique())

  
functions.per.family$Family = factor(functions.per.family$Family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family)) %>% pull(Family) %>% unique())
  g=ggplot(functions.per.family) +
  geom_tile(aes(x = family, y =annotation , fill = prop.genomes.with.this.function)) + #, hjust = 0, vjust = 0
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  facet_grid(category ~ Family, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) + 
  xlab("") + ylab("") +
  scale_fill_gradient(name = "% genomes", low = "#ADD8E6", high = "#000046")
g  
  ggsave(g, filename = sprintf("%sfunctions_mid_relaxed_annot_vs__new_taxonomy_family_an_genus.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width =30, height =15)  

```



# same with sure annotations
```{r}
genomes.per.family = 
all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  group_by(Family) %>%
  mutate(num.genomes.in.family = n_distinct(Accession)) %>%
  group_by(Accession, Genus, Family, num.genomes.in.family) %>%
  summarise(genome.size = n_distinct(ncbi.id)) %>%
  group_by(Genus, Family, num.genomes.in.family) %>%
  summarise(av.genome.size = median(genome.size),
            num.genomes = n_distinct(Accession)) 


functions.per.family = all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  left_join(surely.annotated.proteins) %>%
  inner_join(top.diversity.classes %>% distinct(annotation.index)) %>%
  group_by(annotation.index, Genus, Family) %>%
  summarise(num.genomes.this.function = n_distinct(Accession),
            num.genomes.with.this.function.mosaic = n_distinct(Accession[mosaic]),
            num.genomes.with.this.function.mosaic.domain = n_distinct(Accession[mosaic.domain])) %>%
  left_join(genomes.per.family, by = c("Genus","Family")) %>%
  mutate(prop.genomes.with.this.function = num.genomes.this.function/num.genomes,
         prop.mosaic = num.genomes.with.this.function.mosaic/num.genomes.this.function,
         prop.mosaic.domain = num.genomes.with.this.function.mosaic.domain/num.genomes.this.function) %>%
  left_join(annotation.indices) %>%
  filter(num.genomes >= 20 & !is.na(Family) & Genus != "") %>%
  mutate(Family = if_else(Family == "", "Unknwon", Family)) %>%
  mutate(family = paste0(Genus, "\n", Family, "\n", num.genomes, " genomes", "\nav. genome size: ", av.genome.size))
functions.per.family$family = factor(functions.per.family$family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family), desc(num.genomes)) %>% pull(family) %>% unique())
functions.per.family$Family = factor(functions.per.family$Family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family)) %>% pull(Family) %>% unique())

  
functions.per.family$Family = factor(functions.per.family$Family, levels = functions.per.family %>% arrange(desc(num.genomes.in.family)) %>% pull(Family) %>% unique())
  g=ggplot(functions.per.family) +
  geom_tile(aes(x = family, y =annotation , fill = prop.genomes.with.this.function)) + #, hjust = 0, vjust = 0
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  facet_grid(category ~ Family, scales = "free", space = "free",
             labeller = labeller(category = phrog.class.labels)) + 
  xlab("") + ylab("") +
  scale_fill_gradient(name = "% genomes", low = "#ADD8E6", high = "#000046")
g  
  ggsave(g, filename = sprintf("%sfunctions_sure_annot_vs_new_taxonomy_family_an_genus.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width =30, height =15)  

```



# Taxonomy vs mosaicism
```{r}
genome.sizes.per.family = all.proteins.in.genomes %>%
  group_by(Accession, lifestyle, Genus, Family) %>%
  summarise(genome.size = n_distinct(ncbi.id))

  genomes.mosaicism..per.genera = 
all.proteins.in.genomes %>%
  rename(qname = repr.name) %>%
  left_join(surely.annotated.proteins) %>%
  group_by(Accession, lifestyle, Genus, Family) %>%
  summarise(genome.size = n_distinct(ncbi.id),
            num.mosaic.genes = n_distinct(ncbi.id[mosaic]),
            prop.mosaic.genes = num.mosaic.genes/genome.size) %>%
  group_by(Genus, Family, lifestyle) %>%
  summarise(av.genome.size = median(genome.size),
            num.virulent.genomes = n_distinct(Accession[lifestyle == "virulent"]),
            num.temperate.genomes = n_distinct(Accession[lifestyle == "temperate"]),
            num.genomes = n_distinct(Accession),
            num.mosaic.genomes = n_distinct(Accession[num.mosaic.genes > 0]),
            av.num.mosaic.genes = median(num.mosaic.genes),
            av.prop.mosaic.genes = median(prop.mosaic.genes)
            ) %>%
  mutate(prop.virulent.genomes = num.virulent.genomes/num.genomes,
         prop.temperate.genomes = num.temperate.genomes/num.genomes,
         prop.mosaic.genomes = num.mosaic.genomes/num.genomes) %>%
  arrange(desc(num.genomes)) %>%
  mutate(Lifestyle = if_else(prop.temperate.genomes > 0.5, "mostly temperate", 
                             if_else(prop.virulent.genomes > 0.5, "mostly virulent", "temperate/virulent"))) %>%
  filter(num.genomes > 10 & Genus != "Unclassified" & !is.na(Genus))



g=ggplot(genomes.mosaicism..per.genera, aes(y = av.prop.mosaic.genes, x = av.genome.size)) +
  geom_point(aes(col = lifestyle)) +
  geom_text(aes(label = Genus, col = lifestyle), nudge_y = 0.005, size = 2) + 
  #geom_smooth(se = FALSE, method = "lm") +
  theme_bw() #+ 
  #acet_grid(.~lifestyle)
g
  ggsave(g, filename = sprintf("%sgenome_size_vs_moasaicism.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width =25, height =15)  

```

# Mosaicism between same/different lifestyle and same/diffeerent genus and same/diffeerent genus in recent and all mosaicism
```{r, eval = FALSE}
library(aphid)
HMM.DIR =  "/Users/bognasmug/Downloads/checkv-db-v1.5 2/hmm_db/checkv_hmms/"
all.hmm.files = list.files(path =HMM.DIR)
dd = data.frame(file = all.hmm.files) %>%
  mutate( name = NA,accession = NA)

for (this.file in all.hmm.files) {
  aa = readPHMM(file = sprintf("%s%s", HMM.DIR, this.file))
  #print(aa$name)
  if (is.null(aa$name)) {
      dd$name[dd$file == this.file] = NA
  } else {
    dd$name[dd$file == this.file] =  aa$name
  }

  if (is.null(aa$accession)) {
      dd$accession[dd$file == this.file] = NA
  } else {
    dd$accession[dd$file == this.file] =  aa$accession
  }
}
cc = read.csv("/Users/bognasmug/Downloads/checkv-db-v1.5 2/hmm_db/checkv_hmms.tsv", sep = "\t")

cc1 = cc %>% filter(!(hmm %in% dd$name) & !((hmm %in% dd$accession)))
cc2 = cc %>% filter((hmm %in% dd$name) | (hmm %in% dd$accession))
```


```{r}
unknown.chars = c("", "unknown")
fam.pairs.with.trait = mosaic.pairs.lifestyle  %>%  inner_join(recent.family.pairs) 


# mosaic.pairs.lifestyle2 is for pairs within annotation only within the annotations from top diversity
multi.fam.stats.lifestyle = fam.pairs.with.trait %>%
  mutate(trait.x = lifestyle.x, 
         trait.y = lifestyle.y) %>%
  #inner_join(domain.mosaic.pairs.lifestyle %>% distinct(family.pair)) %>%
  #rowwise() %>% 
  #mutate(lifestylepair = CreatePair(lifestyle.x, lifestyle.y, collapse = " & ")) %>% 
  mutate(lifestylepair = if_else(trait.x == "multi"  | trait.y == "multi", 
                                 "at least \none rHMM family\nin the pair \n is present in\ngenomes of different traits", 
                                 if_else(trait.x %in% unknown.chars | trait.y %in% unknown.chars, 
                                         "at least \none rHMM family\nin the pair\nhas unknown trait", 
                                         if_else(trait.x == trait.y, "same trait", "different trait")))) %>%
  group_by(lifestylepair) %>% 
  summarise(num.pairs = n_distinct(family.pair)) %>%
  arrange(desc(num.pairs)) %>%
   mutate(mosaicism = "mosaicism defined by ECOD or seq.",
          trait = "lifestyle")

multi.fam.stats.fam = fam.pairs.with.trait %>%
  mutate(trait.x = family.x, trait.y = family.y) %>%
  #inner_join(domain.mosaic.pairs.lifestyle %>% distinct(family.pair)) %>%
  #rowwise() %>% 
  #mutate(lifestylepair = CreatePair(lifestyle.x, lifestyle.y, collapse = " & ")) %>% 
  mutate(lifestylepair = if_else(trait.x == "multi"  | trait.y == "multi", 
                                 "at least \none rHMM family\nin the pair \n is present in\ngenomes of different traits", 
                                 if_else(trait.x %in% unknown.chars | trait.y %in% unknown.chars, 
                                         "at least \none rHMM family\nin the pair\nhas unknown trait", 
                                         if_else(trait.x == trait.y, "same trait", "different trait")))) %>%
  group_by(lifestylepair) %>% 
  summarise(num.pairs = n_distinct(family.pair)) %>%
  arrange(desc(num.pairs)) %>%
   mutate(mosaicism = "mosaicism defined by ECOD or seq.",
          trait = "family")


multi.fam.stats.genus = fam.pairs.with.trait %>%
  mutate(trait.x = genus.x, trait.y = genus.y) %>%
  #inner_join(domain.mosaic.pairs.lifestyle %>% distinct(family.pair)) %>%
  #rowwise() %>% 
  #mutate(lifestylepair = CreatePair(lifestyle.x, lifestyle.y, collapse = " & ")) %>% 
  mutate(lifestylepair = if_else(trait.x == "multi"  | trait.y == "multi", 
                                 "at least \none rHMM family\nin the pair \n is present in\ngenomes of different traits", 
                                 if_else(trait.x %in% unknown.chars | trait.y %in% unknown.chars, 
                                         "at least \none rHMM family\nin the pair\nhas unknown trait", 
                                         if_else(trait.x == trait.y, "same trait", "different trait")))) %>%
  group_by(lifestylepair) %>% 
  summarise(num.pairs = n_distinct(family.pair)) %>%
  arrange(desc(num.pairs)) %>%
   mutate(mosaicism = "mosaicism defined by ECOD or seq.",
          trait = "genus")




g=ggplot(rbind(rbind(multi.fam.stats.lifestyle, multi.fam.stats.fam ), multi.fam.stats.genus)) +
  #rbind(multi.fam.stats, multi.fam.stats.recent)) +
  geom_col(aes(x = lifestylepair, y = num.pairs, fill = lifestylepair)) +
  #xlab("Lifestyles of mosaic rHMM family pairs") + 
  ylab("Num of mosaic rHMM family pairs") +
  xlab("") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  scale_fill_manual(values = c("at least \none rHMM family\nin the pair \n is present in\ngenomes of different traits" = "darkred",
                               "at least \none rHMM family\nin the pair\nhas unknown trait" = "gray",
                               "same trait" = "khaki",
                               "different trait" = "orange"),
                    name = "Lifestyles of mosaic rHMM family pairs") +
  #facet_grid(mosaicism ~., scales = "free")
  facet_grid(trait ~., scales = "free")
  g
  ggsave(g, filename = sprintf("%sdomain_mosaicism_fam_pair_lifestyle2_recent.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width =14, height =8)  
  
   
  # same for recent mosaicism

```


# same on the level of reprseqs

```{r}
unknown.chars = c("", "unknown")
reprseq.pairs.with.trait = mosaic.reprseq.pairs.wide  %>%  
  #  inner_join(recent_HGT_pairs_raw.70 %>% select(reprseq.pair = pair) %>% rowwise() %>% mutate(reprseq.pair = gsub(" AND ", "&", reprseq.pair))) %>%
  left_join(reprseq.metadata.table %>% select(qname = repr.name, lifestyle.x = lifestyle, family.x = family, genus.x = genus, host.x = host)) %>%
  left_join(reprseq.metadata.table %>% select(sname = repr.name, lifestyle.y = lifestyle, family.y = family, genus.y = genus, host.y = host))

# mosaic.pairs.lifestyle2 is for pairs within annotation only within the annotations from top diversity
multi.reprseq.stats.lifestyle = reprseq.pairs.with.trait %>%
  mutate(trait.x = lifestyle.x, 
         trait.y = lifestyle.y) %>%
  #inner_join(domain.mosaic.pairs.lifestyle %>% distinct(family.pair)) %>%
  #rowwise() %>% 
  #mutate(lifestylepair = CreatePair(lifestyle.x, lifestyle.y, collapse = " & ")) %>% 
  mutate(lifestylepair = if_else(trait.x == "multi"  | trait.y == "multi", 
                                 "at least \none rHMM \nin the pair \n is present in\ngenomes of different traits", 
                                 if_else(trait.x %in% unknown.chars | trait.y %in% unknown.chars, 
                                         "at least \none rHMM \nin the pair\nhas unknown trait", 
                                         if_else(trait.x == trait.y, "same trait", "different trait")))) %>%
  group_by(lifestylepair) %>% 
  summarise(num.pairs = n_distinct(reprseq.pair)) %>%
  arrange(desc(num.pairs)) %>%
   mutate(mosaicism = "mosaicism defined by ECOD or seq.",
          trait = "lifestyle")

multi.reprseq.stats.fam = reprseq.pairs.with.trait %>%
  mutate(trait.x = family.x, trait.y = family.y) %>%
  #inner_join(domain.mosaic.pairs.lifestyle %>% distinct(family.pair)) %>%
  #rowwise() %>% 
  #mutate(lifestylepair = CreatePair(lifestyle.x, lifestyle.y, collapse = " & ")) %>% 
  mutate(lifestylepair = if_else(trait.x == "multi"  | trait.y == "multi", 
                                 "at least \none rHMM \nin the pair \n is present in\ngenomes of different traits", 
                                 if_else(trait.x %in% unknown.chars | trait.y %in% unknown.chars, 
                                         "at least \none rHMM \nin the pair\nhas unknown trait", 
                                         if_else(trait.x == trait.y, "same trait", "different trait")))) %>%
  group_by(lifestylepair) %>% 
  summarise(num.pairs = n_distinct(reprseq.pair)) %>%
  arrange(desc(num.pairs)) %>%
   mutate(mosaicism = "mosaicism defined by ECOD or seq.",
          trait = "family")


multi.reprseq.stats.genus = reprseq.pairs.with.trait %>%
  mutate(trait.x = genus.x, trait.y = genus.y) %>%
  #inner_join(domain.mosaic.pairs.lifestyle %>% distinct(family.pair)) %>%
  #rowwise() %>% 
  #mutate(lifestylepair = CreatePair(lifestyle.x, lifestyle.y, collapse = " & ")) %>% 
  mutate(lifestylepair = if_else(trait.x == "multi"  | trait.y == "multi", 
                                 "at least \none rHMM \nin the pair \n is present in\ngenomes of different traits", 
                                 if_else(trait.x %in% unknown.chars | trait.y %in% unknown.chars, 
                                         "at least \none rHMM \nin the pair\nhas unknown trait", 
                                         if_else(trait.x == trait.y, "same trait", "different trait")))) %>%
  group_by(lifestylepair) %>% 
  summarise(num.pairs = n_distinct(reprseq.pair)) %>%
  arrange(desc(num.pairs)) %>%
   mutate(mosaicism = "mosaicism defined by ECOD or seq.",
          trait = "genus")




g=ggplot(rbind(rbind(multi.reprseq.stats.lifestyle, multi.reprseq.stats.fam ), multi.reprseq.stats.genus)) +
  #rbind(multi.fam.stats, multi.fam.stats.recent)) +
  geom_col(aes(x = lifestylepair, y = num.pairs, fill = lifestylepair)) +
  #xlab("Lifestyles of mosaic rHMM family pairs") + 
  ylab("Num of mosaic rHMM pairs") +
  xlab("") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust =1,vjust=1, size = 8),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = 8),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  scale_fill_manual(values = c("at least \none rHMM \nin the pair \n is present in\ngenomes of different traits" = "darkred",
                               "at least \none rHMM \nin the pair\nhas unknown trait" = "gray",
                               "same trait" = "khaki",
                               "different trait" = "orange"),
                    name = "Lifestyles of mosaic rHMM pairs") +
  #facet_grid(mosaicism ~., scales = "free")
  facet_grid(trait ~., scales = "free")
  g
  ggsave(g, filename = sprintf("%sdomain_mosaicism_reprseq_pair_lifestyle2.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width =14, height =8)  
  
   
  # same for recent mosaicism

```

# Let's look at proportion of genomes that show mosaicism to genomes from other taxa / lifestyles
```{r, eval = FALSE}
mosaicism.crosses.boarder.type = mosaic.reprseq.pairs.wide %>%
  left_join(name.table.with.metadata %>% distinct(qgenome = Accession, qname = repr.name, qlifestyle = lifestyle, qGenus = Genus, qfamily = Family)) %>%
  left_join(name.table.with.metadata %>% distinct(sgenome = Accession, sname = repr.name, slifestyle = lifestyle, sGenus = Genus, sfamily = Family)) %>%
  group_by(qgenome, qname, qlifestyle, qGenus, qfamily) %>%
  summarise(different.lifestyle.pair = any(qlifestyle != slifestyle  & slifestyle != "unknwon") & qlifestyle != "unknown",
            different.genus.pair = any(qGenus != sGenus & qGenus != "unknwon") & qGenus != "unknown" ,
            different.family.pair = any(qfamily != sfamily & sfamily != "unknwon")  & qfamily != "unknown") %>% 
  ungroup() %>% 
   select(Accession = qgenome, repr.name = qname, different.lifestyle.pair, different.genus.pair, different.family.pair) %>%
  right_join(all.proteins.in.genomes, by = c("Accession", "repr.name"))

data.table::fwrite(mosaicism.crosses.boarder.type, file = sprintf("%smosaicism.crosses.boarder.typee.txt", OUTPUT.DATA.PATH))
```

```{r}

mosaicism.crosses.boarder.type = data.table::fread(file = sprintf("%smosaicism.crosses.boarder.typee.txt", OUTPUT.DATA.PATH))
dd.stats = mosaicism.crosses.boarder.type %>%
  group_by(lifestyle, Accession) %>%
  summarise(n = n_distinct(repr.name),
            n.mosaic = n_distinct(repr.name[mosaic]),
            n.mosaic.distinct.lifestyle = n_distinct(repr.name[mosaic & different.lifestyle.pair]),
            n.mosaic.distinct.genus = n_distinct(repr.name[mosaic & different.genus.pair]),
            n.mosaic.distinct.family = n_distinct(repr.name[mosaic & different.family.pair])) %>%
  ungroup() %>%
  mutate(prop.mosaic.distinct.lifestyle = n.mosaic.distinct.lifestyle/n,
         prop.mosaic.distinct.genus = n.mosaic.distinct.genus/n,
         prop.mosaic.distinct.family = n.mosaic.distinct.family/n)

g=ggplot(dd.stats %>% tidyr::gather(key = "trait", value = "prop", prop.mosaic.distinct.lifestyle, prop.mosaic.distinct.genus, prop.mosaic.distinct.family) %>% filter(!is.na(lifestyle) & lifestyle != "unknown" & trait == "prop.mosaic.distinct.family"), 
       aes(x=lifestyle, y = prop, col = lifestyle)) +
  geom_boxplot() +
  geom_jitter(size = 0.1, alpha = 0.5) +
  #facet_grid(trait ~ .) + 
  guides(color = "none") +
  theme_bw() +
  ylab("Prop. genes in genome that have mosaic signal to genomes from different family") + xlab("")
    g
  ggsave(g, filename = sprintf("%slifestyle_vs_family_cross.png", OUTPUT.FIGURES.METADATA.PATH), bg = "white", width =5, height =5)  
  
```

```{r}
ANNOT.SEL = 517
aa2 = recentHGT.with.metadata #%>% 
  #rowwise() %>% mutate(same.host = known.host & qhost == shost,
  #same.family = known.family & qfamily == sfamily)
aa2 = recentHGT.with.metadata %>% 
  filter(qindex == ANNOT.SEL | sindex ==ANNOT.SEL) 
  #filter(known.host & known.family) %>%
  #@rowwise() %>% 
  #mutate(same.host = known.host & qhost == shost,
  #       same.family = known.family & qfamily == sfamily)


aa2 = mosaic.reprseq.pairs.lifestyle %>% 
  filter(annotation.index == ANNOT.SEL) %>%
  inner_join(seq.mosaic.pairs.all.pident %>% filter(mosaic) %>% distinct(reprseq.pair), by = "reprseq.pair") 

bb2=aa2 %>%  
  filter(qhost != "unknown" & shost != "unknown" & qfamily != "unknown" & sfamily != "unknown") %>%
  #mutate(known.host  = qhost != "unknown" & shost != "unknown",
  #       known.family = qfamily != "unknown" & sfamily != "unknown") %>%
  #filter(known.host & known.family) %>%
  mutate(same.host = qhost == shost,
         same.family = qfamily == sfamily) %>%
  group_by(same.family, same.host) %>% 
  summarise(n = n_distinct(reprseq.pair))

M =M = t(matrix(bb2$n, nrow=2,ncol=2 ))
M[1,1] = bb2 %>% filter(same.family & same.host) %>% pull(n)
M[2,2] = bb2 %>% filter(!same.family & !same.host) %>% pull(n)
M[1,2] = bb2 %>% filter(!same.family & same.host) %>% pull(n)
M[2,1] = bb2 %>% filter(same.family & !same.host) %>% pull(n)
chisq.test(M)
cor.test(rf$same.family %>% as.numeric(), rf$same.host %>% as.numeric())
# look at this at the genus level and only at tail fibers?
```



# TODO
yellow-beige picture where instead of the phage taxa we have host genus to see where the signal comes from
also boxplots of % modular genomes for the most frequent families and hosts + ANOVA



# TODO
# maybe the problem was we were looking for functions that are unique and from included group, what if we look at those that don;t have any funtion?
# field: has.unknown.function in all.proteins