---
title: "Look into examples of mosaicism"
author: "Bogna Smug"
date: '2023-04-24'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

unaccepted.genes = data.table::fread(file = sprintf("%sunaccepted.genes.txt",OUTPUT.DATA.PATH))
families = data.table::fread(file = sprintf("%sfamilies.txt", OUTPUT.DATA.PATH)) %>%
  anti_join(unaccepted.genes %>% select(qname)) 
recent_HGT_pairs_raw.70 = data.table::fread(sprintf("%sFigure_Supplementary/FigS11_recent_HGT/recent_HGT_pars_70.txt", OUTPUT.FIGURES.PATH))
recent_HGT_pairs_raw.50 = data.table::fread(sprintf("%sFigure_Supplementary/FigS11_recent_HGT/recent_HGT_pars_50.txt", OUTPUT.FIGURES.PATH))
recent_HGT_pairs_raw.90 = data.table::fread(sprintf("%sFigure_Supplementary/FigS11_recent_HGT/recent_HGT_pars_90.txt", OUTPUT.FIGURES.PATH))
surely.annotated.proteins.including.multi.annot = data.table::fread( sprintf("%ssurely.annotated.proteins.including.multi.annot",OUTPUT.DATA.PATH))  %>%
  left_join(families) %>%
  anti_join(unaccepted.genes %>% select(qname))
ecod_id_to_names_map = data.table::fread(sprintf("%secod_id_to_names_map.txt",OUTPUT.DATA.PATH)) %>% as.data.frame()
relaxely.annotated.proteins.including.multi.annot =data.table::fread(sprintf("%srelaxely.annotated.proteins.including.multi.annot",OUTPUT.DATA.PATH)) %>%
  left_join(families)  %>%
  anti_join(unaccepted.genes %>% select(qname))
qname.pairs.sharing.t.and.disshare.x.distinct = data.table::fread(file = sprintf("%sFigure_Supplementary/FigS7_folds_shared_by_mosaic/qname.pairs.sharing.t.and.disshare.x.distinct.txt", OUTPUT.FIGURES.PATH))


hhr.table = read.csv("/Users/bognasmug/MGG Dropbox/Projects/divRBP/phage-pp-workdir-refseq-hhblits/output/prot-families/all-by-all/hhblits/table-hhr.txt", 
                     sep = ",",
                     header = TRUE) %>%
  anti_join(unaccepted.genes %>% select(qname)) %>%
  anti_join(unaccepted.genes %>% select(sname = qname))
ecod.domains.hits = data.table::fread(sprintf("%secod.domains.hits.txt",OUTPUT.DATA.PATH)) %>%
  anti_join(unaccepted.genes %>% select(qname)) 
included.category.sizes = data.table::fread(file = sprintf("%sincluded.category.sizes",OUTPUT.DATA.PATH))
surely.annotated.proteins = GetUniquelyAnnotatedProteins(surely.annotated.proteins.including.multi.annot, GENERAL.ANTIDEFENSE.ANNOTATION.INDEX)
surely.annotated.proteins.from.included.categories = surely.annotated.proteins %>%
  inner_join(included.category.sizes) %>%
  select(qname, family, annotation.index)

name.table = read.table(NAME_TABLE_PATH, sep = ",", header = TRUE)
name.table.full = name.table %>% tidyr::separate(col = ncbi.id, sep = ":", into = c("genome.id", "gene.id"), remove = FALSE) %>%
  tidyr::separate(col = gene.id, sep = "-", into = c("gene.start", "end"), remove = FALSE) %>%
  mutate(gene.end = str_remove_all(end, "[(+)]")) %>%
  select(-end) %>%
  rename(qname = repr.name) %>%
  left_join(surely.annotated.proteins)
```









# look into annotated pairs with recent HGT
```{r}
recent_HGT_pairs_raw.50.annotated = recent_HGT_pairs_raw.50 %>% 
  #left_join(relaxely.annotated.proteins.including.multi.annot %>% select(qname = qname, qannotation = annotation)) %>%
  #left_join(relaxely.annotated.proteins.including.multi.annot %>% select(sname = qname, sannotation = annotation)) %>%
  left_join(surely.annotated.proteins.including.multi.annot %>% select(qname = qname, qannotation = annotation)) %>%
  left_join(surely.annotated.proteins.including.multi.annot %>% select(sname = qname, sannotation = annotation)) %>%
  filter(!is.na(qannotation) | !is.na(sannotation)) 

recent_HGT_pairs_raw.70.annotated = recent_HGT_pairs_raw.70 %>% 
  #left_join(relaxely.annotated.proteins.including.multi.annot %>% select(qname = qname, qannotation = annotation)) %>%
  #left_join(relaxely.annotated.proteins.including.multi.annot %>% select(sname = qname, sannotation = annotation)) %>%
  left_join(surely.annotated.proteins.including.multi.annot %>% select(qname = qname, qannotation = annotation)) %>%
  left_join(surely.annotated.proteins.including.multi.annot %>% select(sname = qname, sannotation = annotation)) %>%
  filter(!is.na(qannotation) | !is.na(sannotation)) 
  

recent_HGT_pairs_raw.50.annotated.not.found.by.ecod.mosaicism = recent_HGT_pairs_raw.50.annotated %>%
  anti_join(qname.pairs.sharing.t.and.disshare.x.distinct %>% select(qname = qname.x, sname = qname.y) %>% distinct()) %>%
  anti_join(qname.pairs.sharing.t.and.disshare.x.distinct %>% select(qname = qname.y, sname = qname.x) %>% distinct())


recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism = recent_HGT_pairs_raw.70.annotated %>%
  anti_join(qname.pairs.sharing.t.and.disshare.x.distinct %>% select(qname = qname.x, sname = qname.y) %>% distinct()) %>%
  anti_join(qname.pairs.sharing.t.and.disshare.x.distinct %>% select(qname = qname.y, sname = qname.x) %>% distinct())

```




```{r, eval = FALSE}

recent_HGT_pairs_pident70_with_ecod = Get.Domain.coverage.In.Recent.HGT(
      recent_HGT_pairs_raw = recent_HGT_pairs_raw.70.annotated, 
    # note that technically pident and prob in the raw hhr.tabl may not be the same as within our summarised table we used      before
    # make the hhr table symmetric so that every reprseq is seen on qname position
    hhr.table = hhr.table %>% filter(prob >= MINIMUM.PROB.FOR.PAIRWISE.HIT & pident >= 70),
    surely.annotated.proteins.from.included.categories, 
    all.ecod.hits %>% filter(prob > 70 & scov > 0.5) %>% mutate(t_id = annot)) %>%
  filter(prop.fragment.covered.by.domain >= 0.2)
  
recent_HGT_pairs_pident70_with_pfam = Get.Domain.coverage.In.Recent.HGT(
      recent_HGT_pairs_raw = recent_HGT_pairs_raw.70.annotated, 
      # note that technically pident and prob in the raw hhr.tabl may not be the same as within our summarised table we used       before
      # make the hhr table symmetric so that every reprseq is seen on qname position
      hhr.table = hhr.table %>% filter(prob >= MINIMUM.PROB.FOR.PAIRWISE.HIT & pident >= 70),
      surely.annotated.proteins.from.included.categories, 
      ecod.domains.hits = all.pfam.hits %>% filter(prob > 70 & scov > 0.5) %>% mutate(t_id = annot)) %>%
    filter(prop.fragment.covered.by.domain >= 0.2)
 


recent_HGT_pairs_raw.70.annotated.domin.annot.col = recent_HGT_pairs_raw.70.annotated %>%
  left_join(recent_HGT_pairs_pident70_with_ecod %>% mutate(ecod.mosaicism.detected = TRUE), by = c("qname", "sname")) %>%
  left_join(recent_HGT_pairs_pident70_with_pfam %>% mutate(pfam.mosaicism.detected = TRUE), by = c("qname", "sname")) %>%
  mutate(ecod.mosaicism.detected = if_else(is.na(ecod.mosaicism.detected), FALSE,ecod.mosaicism.detected)) %>%
  mutate(pfam.mosaicism.detected = if_else(is.na(pfam.mosaicism.detected), FALSE,pfam.mosaicism.detected)) %>%
  distinct(qname, sname, qannotation, sannotation, ecod.mosaicism.detected, pfam.mosaicism.detected) 

fwrite(mid.annotated.proteins.including.multi.annot %>% 
  filter(qname %in% c(recent_HGT_pairs_raw.70.annotated.domin.annot.col$qname, recent_HGT_pairs_raw.70.annotated.domin.annot.col$sname)) %>% select(qname, annotation, category),
  file = sprintf("%stables/mid.confident.annotatetions.for.Rafal.csv", OUTPUT.FIGURES.PATH))

fwrite(recent_HGT_pairs_raw.70.annotated.domin.annot.col, 
       file = sprintf("%stables/recent_HGT_pairs_raw.70.annotated.for.Rafal.csv", OUTPUT.FIGURES.PATH))

recent_HGT_pairs_raw.70.annotated.not.covered.by.domains = recent_HGT_pairs_raw.70.annotated.domin.annot.col  %>%
  filter(!ecod.mosaicism.detected & !pfam.mosaicism.detected) %>%
  distinct(qname, sname, qannotation, sannotation) 

recent_HGT_pairs_from_categories.not.found.in.ecod = recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  anti_join(diversity.data1 %>% filter(top.quantile) %>% distinct(annotation) %>% select(qannotation = annotation)) %>%
  anti_join(diversity.data1 %>% filter(top.quantile) %>% distinct(annotation) %>% select(sannotation = annotation))  #%>%
  #filter(qannotation %in% unique(included.category.sizes$annotation) | sannotation %in% unique(included.category.sizes$annotation))
  #filter(qname == "reprseq025732")

```





```{r}
all.ecod.hits = data.table::fread(file = ECOD.DOMAIN.HITS.PATH) %>%
  Parse.Hhr.Raw.Result()
all.pfam.hits = data.table::fread(file = PFAM.DOMAIN.HITS.PATH) %>%
  Parse.Hhr.Raw.Result()
hhr.annotated = data.table::fread(file = sprintf("%shhr.annotated",OUTPUT.DATA.PATH)) %>%
  anti_join(unaccepted.genes %>% select(qname)) 
recent.mosaicism.hhr = data.table::fread(file = sprintf("%srecent.mosaicism.hhr.txt",OUTPUT.DATA.PATH)) %>%
  anti_join(unaccepted.genes %>% select(qname)) %>%
  anti_join(unaccepted.genes %>% select(sname = qname))
all.fastas = seqinr::read.fasta("/Users/bognasmug/MGG Dropbox/Projects/divRBP/phage-pp-workdir-refseq-hhblits/input/coding-seqs/cds-aa.fa", as.string = TRUE)
prot.names = read.table(PROTEIN_NAMES_MAPPING_PATH, sep = ",", header = TRUE) 
all.cds = seqinr::read.fasta("/Users/bognasmug/MGG Dropbox/Projects/divRBP/phage-pp-workdir-refseq-hhblits/input/coding-seqs/cds-aa.fa", as.string= TRUE)
name.table = read.table(NAME_TABLE_PATH, sep = ",", header = TRUE)
```


```{r}

selected.annot = "antidefense"#"tail length tape measure"##
q=#recent_HGT_pairs_raw.70.annotated.not.covered.by.domains %>% 
  recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  filter(qannotation == selected.annot) %>%
  pull(qname)
s=recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  #recent_HGT_pairs_raw.70.annotated.not.covered.by.domains %>% 
  filter(sannotation == selected.annot) %>%
  pull(sname)

qn = unique(c(q,s))
anots.from.hypothetical = relaxely.annotated.proteins.including.multi.annot %>% 
  filter(qname %in% qn) %>%
  group_by(qname) %>%
  summarise(annot = paste0(sort(unique(annotation)), collapse = " & ")) %>%
  left_join(hhr.annotated, by = "qname") %>%
  group_by(qname) %>%
  filter(prob > 95) %>%
  summarise(max.cov = max(qcov*scov),
            max.cov.annot = annotation[qcov*scov == max.cov])

#surely.annotated.proteins %>% filter(qname == "reprseq054994")
# "reprseq025371") has very confident hits to two classes: DNA methyltransferase and antidefense
#hhr.annotated %>% filter(qname == "reprseq025371") %>% mutate(qs = qcov*scov) %>% arrange(desc(qs))
```

```{r}
recent_HGT_ttm_raw.50.annotated =
  recent_HGT_pairs_raw.50.annotated.not.found.by.ecod.mosaicism %>%
  #recent_HGT_pairs_raw.70.annotated.not.covered.by.domains %>% 
  filter(qannotation == "tail length tape measure" | sannotation == "tail length tape measure") %>%
 inner_join(hhr.table) %>%
  rowwise() %>%
  mutate(qcov = (qend - qstart + 1)/qlength,
         scov = (send - sstart + 1)/slength) %>%
  ungroup() %>%
  arrange(desc(prob), desc(pident))  %>%
  slice(1)


#recent.mosaicism.hhr.raw = data.table::fread(file = HHALIGN_RECENT_MOSAICISM_PATH) 
recent_HGT_tail_spike.pairs_raw.70.annotated = 
  #recent_HGT_pairs_from_categories.not.found.in.ecod %>%
  recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  #recent_HGT_pairs_raw.70.annotated.not.covered.by.domains %>% 
  filter(qannotation == "tail spike" | sannotation == "tail spike") %>%
 inner_join(hhr.table) %>%
  rowwise() %>%
  mutate(qcov = (qend - qstart + 1)/qlength,
         scov = (send - sstart + 1)/slength) %>%
  ungroup() %>%
  arrange(desc(prob), desc(pident)) %>%
  slice(1)


recent_HGT_antidefense.pairs_raw.70.annotated = recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  #recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  #recent_HGT_pairs_raw.70.annotated.not.covered.by.domains %>% 
  filter(qannotation == "antidefense" | sannotation == "antidefense") %>%
 inner_join(hhr.table) %>%
  rowwise() %>%
  mutate(qcov = (qend - qstart + 1)/qlength,
         scov = (send - sstart + 1)/slength) %>%
  ungroup() %>%
  arrange(desc(prob), desc(pident)) %>%
  slice(3)

# nice example:  reprseq084783  and reprseq122084 
recent_HGT_ttmp.pairs_raw.70.annotated = recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  #recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  #recent_HGT_pairs_raw.70.annotated.not.covered.by.domains %>% 
  filter(qannotation == "tail length tape measure" | sannotation == "tail length tape measure") %>%
 inner_join(hhr.table) %>%
      rowwise() %>%
  mutate(qcov = (qend - qstart + 1)/qlength,
         scov = (send - sstart + 1)/slength) %>%
  ungroup() %>%
  arrange(desc(pident)) %>%
  slice(1)

# nice example reprseq014946 reprseq089720   
recent_HGT_tf.pairs_raw.70.annotated = recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  #recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  #recent_HGT_pairs_raw.70.annotated.not.covered.by.domains %>% 
  filter(qannotation == "tail fiber" & sannotation == "tail fiber") %>%
   inner_join(hhr.table) %>%
    rowwise() %>%
  mutate(qcov = (qend - qstart + 1)/qlength,
         scov = (send - sstart + 1)/slength) %>%
  ungroup() %>%
  arrange(desc(prob), qcov, scov, desc(pident)) %>%
  slice(1)

recent_HGT_endolysins.pairs_raw.70.annotated = recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>% 
  #recent_HGT_pairs_raw.70.annotated.not.covered.by.domains %>% 
  filter(qannotation == "endolysin" & sannotation == "endolysin") %>%
   inner_join(hhr.table) %>%
    rowwise() %>%
  mutate(qcov = (qend - qstart + 1)/qlength,
         scov = (send - sstart + 1)/slength) %>%
  ungroup() %>%
  arrange(desc(prob), desc(pident), qcov, scov) %>%
  slice(1)

# nice example: reprseq006081    reprseq077270
recent_HGT_dnapolimerase.pairs_raw.70.annotated = recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  #recent_HGT_pairs_raw.70.annotated.not.covered.by.domains %>% 
  filter(qannotation == "DNA polymerase" & sannotation == "DNA polymerase") %>%
   inner_join(hhr.table) %>%
    rowwise() %>%
  mutate(qcov = (qend - qstart + 1)/qlength,
         scov = (send - sstart + 1)/slength) %>%
  ungroup() %>%
  arrange(desc(pident)) %>%
  slice(1)

recent_HGT_neck.passage.pairs_raw.70.annotated = recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  #recent_HGT_pairs_raw.70.annotated.not.covered.by.domains %>% 
  filter(qannotation == "neck passage" & sannotation == "tail completion") %>%
   inner_join(hhr.table) %>%
    rowwise() %>%
  mutate(qcov = (qend - qstart + 1)/qlength,
         scov = (send - sstart + 1)/slength) %>%
  ungroup() %>%
  arrange(desc(pident)) 


recent_HGT_neck.passage2.pairs_raw.70.annotated = recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  #recent_HGT_pairs_raw.70.annotated.not.covered.by.domains %>% 
  filter(qannotation == "neck passage" & sannotation == "minor capsid") %>%
  filter(qname == "reprseq028936" & sname == "reprseq067546") %>%
   inner_join(hhr.table) %>%
    rowwise() %>%
  mutate(qcov = (qend - qstart + 1)/qlength,
         scov = (send - sstart + 1)/slength) %>%
  ungroup() %>%
  arrange(desc(pident)) 


recent_HGT_tf_unknwin.pairs_raw.70.annotated = recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>% 
  filter((qannotation == "tail fiber" & is.na(sannotation)) | (sannotation == "tail fiber" & is.na(qannotation))) %>%
  inner_join(hhr.table) %>%
    rowwise() %>%
  mutate(qcov = (qend - qstart + 1)/qlength,
         scov = (send - sstart + 1)/slength) %>%
  ungroup() %>%
  arrange(desc(prob), desc(pident), qcov, scov) %>%
  slice(1)
  
tfpair = recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  filter(qname == "reprseq003460" & sname == "reprseq125591") %>%
    inner_join(hhr.table)  %>%
    rowwise() %>%
    mutate(qcov = (qend - qstart + 1)/qlength,
          scov = (send - sstart + 1)/slength) %>%
    ungroup()

recent_HGT_to_look_at = recent_HGT_ttmp.pairs_raw.70.annotated %>%
  rbind(recent_HGT_tf.pairs_raw.70.annotated) %>%
  rbind(recent_HGT_endolysins.pairs_raw.70.annotated) %>%
  rbind(recent_HGT_dnapolimerase.pairs_raw.70.annotated) %>%
  rbind(recent_HGT_antidefense.pairs_raw.70.annotated) %>%
  rbind(recent_HGT_neck.passage.pairs_raw.70.annotated) %>%
  rbind(recent_HGT_tf_unknwin.pairs_raw.70.annotated) %>%
  rbind(recent_HGT_ttm_raw.50.annotated) %>%
  rbind(tfpair)

recent_HGT_to_look_at.n = hhr.table  %>% #recent.mosaicism.hhr
  inner_join(recent_HGT_to_look_at %>% distinct(qname, sname), by = c("qname", "sname")) %>%
   left_join(surely.annotated.proteins.including.multi.annot %>% select(qname = qname, qannotation = annotation, qcategory = category)) %>%
  left_join(surely.annotated.proteins.including.multi.annot %>% select(sname = qname, sannotation = annotation, scategory = category)) %>%
  group_by(qname, sname, qstart, qend, qlength, sstart, send, slength, pident, bitscore, eval, prob,pval) %>%
  summarise(qannotation = paste(unique(sort(qannotation)), collapse = "_OR_"),
            sannotation = paste(unique(sort(sannotation)), collapse = "_OR_"))
  
 


seq.to.look.at = data.frame(repr.name = unique(c(recent_HGT_to_look_at.n$qname, recent_HGT_to_look_at.n$sname))) %>%
  inner_join(prot.names)

all.fastas.to.look.at = all.fastas[which(names(all.fastas) %in% seq.to.look.at$prot.name)]
df = data.frame(prot.name = names(all.fastas.to.look.at)) %>% left_join(seq.to.look.at)
names(all.fastas.to.look.at) = df$repr.name


all.ecod.hits %>% filter(qname %in% df$repr.name) %>%
  arrange(qname)

seqinr::write.fasta(all.fastas.to.look.at, 
                    names = names(all.fastas.to.look.at),
                    file.out = sprintf("%stables/recent_HGT_examples.fa", OUTPUT.FIGURES.PATH))

for (pair.index in 1:nrow(recent_HGT_to_look_at.n)) {
  names.to.include = c(recent_HGT_to_look_at.n$qname[pair.index], recent_HGT_to_look_at.n$sname[pair.index])
  annotation = paste(c(recent_HGT_to_look_at.n$qannotation[pair.index], recent_HGT_to_look_at.n$sannotation[pair.index]), collapse = " & ")
  pair = paste(c(recent_HGT_to_look_at.n$qname[pair.index], recent_HGT_to_look_at.n$sname[pair.index]), collapse = " & ")
  fastas.to.incldude = all.fastas.to.look.at[which(names(all.fastas.to.look.at) %in% names.to.include)]
  seqinr::write.fasta(fastas.to.incldude, 
                    names = names(fastas.to.incldude),
                    file.out = sprintf("%stables/recent_HGT_pair_%s_%s.fa", OUTPUT.FIGURES.PATH, pair, annotation))
  
}
write.csv(recent_HGT_to_look_at.n,
          file = sprintf("%stables/recent_HGT_no_ecod_mosaicim_examples.csv", OUTPUT.FIGURES.PATH),
          row.names = FALSE)





recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>% filter(qannotation == "minor capsid") %>%
  inner_join(hhr.table) %>%
    rowwise() %>%
  mutate(qcov = (qend - qstart + 1)/qlength,
         scov = (send - sstart + 1)/slength) %>%
  ungroup() %>%
  arrange(desc(prob), desc(pident), qcov, scov)
```

```{r}
# see in which genomes they are

```

```{r}
# tail fiber pair
# reprseq003460 reprseq125591
#  NC_047780.1:33401-37132(+) and # NC_028977.1:33253-35964(+)
# Two Klebsiella phage, very similar genome ()88% similarity, seems that the tail fiber is one of  only two protein that differs between the genomes! (NC_028977.1:33253-35964(+) and NC_028977.1:35977-37512(+)) i.e. reprseq125591 and reprseq000001. These are two tail fiber proteins next to each other
# within them the first gene 150 aa are similar, the rest is different
# the similar fragment is onot covered by ECOD but it actually is covered by pfam    pf03906.17 ; phage_t7_tail ; phage t7 tail fibre protein 


# TTM pair
#reprseq084783 reprseq122084
# NC_002321.1:8926-11010(+) NC_008689.1:28527-31859(+)
# in their genomes there are 2 adjacent TMP.
# They all align very well but they are split in a differnt place hence the observed mosaicism



# Tfiber pair: reprseq018733 and reprseq100089
# i.e. NC_007291.1:39796-42012(+) NC_023743.1:12476-15469(+)
# genomes with 57%  coverage
# the parts that are not aligned are: NC_007291.1:38794-39717(+),  NC_007291.1:39796-42012(+) , NC_007291.1:42041-42508(+)
# i.e. DNA primase and two tail fiber proteins
# another part that differs is around positions 13405-18048 which seem to annotate structural module (capsid and capsid assembly)
# the shared N terminal is found by HHpred as T7 tail fiber domain.
# inetstingly they sahre even an antidefense protein
# that shared fragment is also shared as a fragment with ~ 60 other proteins
# 19 are on pident > 50
# only one is on pident > 70



# antidefense vs unknoqn
# reprseq053259 reprseq088441
# NC_023552.1:37789-39126(+) NC_051650.1:44260-46806(+): Mycobacterium phages
# they belong two two genomes that are not similar at all
# but align on one minor protein NC_023552.1:25474-25833(+) and a fragment with two DNA methyltransferase proteins, some alignments with gaps in the region: 37960-40220



# neck passage and tail completion proteins:
# reprseq028936	reprseq037076
# both are seen in multiple genomes
# let's look at an example (cluster representatives)
# NC_049374.1:7400-9001(+) and NC_021860.1:6853-9039(+)
# both genomes are Lactoccocus phage jm2
# there is ~ 70% coverage between the genomes
#. But they differ in some fragments
# break in NC_021860.1 between 7376 - 9079: this is about 75% of the tail fiber gene. The N terminal stays the same
# also break in NC_021860.1 between 20024 and 20643: two unknown proteins
# the shared part (1-177 positions in reprseq028936) includes: N-terminal Ig-like domain in baseplate protein ORF48 (positions 8-106) while the shared fragment is shared up to ~ 177
# tail completion protein has two ecod domains: 
# tail completion protein contain: 
# a) N-terminal Ig-like domain in baseplate protein ORF48 (positions 8-106)
# b) Pectin lyase-like (~ 476-685)
# neck passag (reprseq028936) protein conatin: 
# a) N-terminal Ig-like domain in baseplate protein ORF48 (positions 7-117)
# b) gp11/gp12 receptor-binding domain  F: GP11_1st | Protein: BASEPLATE STRUCTURAL PROTEIN GP11: 465-533 (prob < 95%) (positions 465-533)
# This actually should be detected with domain approach if we lowered the prob threshold

# unknown protein and a tail fiber
# reprseq025732 and reprseq053508
# reprseq053508 is only found in one genome
# reprseq025732 found in multiple genomes
# let us look at the genomes of cluster representative
#  NC_049839.1:29929-32682(-) and NC_047825.1:20349-22691(-)
# botrh are from Klebsiella phage KOX1, almost 80% of genomes are aligned
# alignment in NC_049839.1 not conserved in the beggining (positions: 1-2592) and in the middle (positions: 28717 to 32209)
# the middle: three genes breaking at tail fiber we have NC_049839.1:25437-28883 (unknown), NC_049839.1:28916-29236, NC_049839.1:29242-29901(+), (unknoqn), NC_049839.1:29929-32682: tail fiber: the beginning is aligned to the other genome
# tail fiber domains reprseq053508 Pectin lyase-like (219-699), no ecod nor pfam at  N terminal
# unknown domains: reprseq025732 tail spike domain (748-784), Protein: L-SHAPED TAIL FIBER PROTEIN: (786  917)
# the domain that is shared is unknown (16-174) by ecod and pfam
# But HHpred finds hits to: collar spike or depolymerase!


# minor capsid and a structural protein
#reprseq067546	reprseq073249 (neck passage)


#reprseq028936 (neck passage) same as described above!! and  reprseq067546 (minor capsid)
# reprseq028936 is seen in multiple genomes, reprseq067546 only in one
# let's look at an example (cluster representatives)
# NC_049374.1:7400-9001(+) and NC_021855.1:6850-8775(+)
# both genomes are Lactoccocus phage, coverage > 70%
# they differ in multiple fragments:
# eg: NC_049374.1: 7927 - 89565
# we see that the first ~ 550 positions (i.e 186 aa positions) are shared between the two proteins, the rest differs
# this fragment is also shared with the tail comletion protein described above (reprseq037076)
# it includes N-terminal Ig-like domain in baseplate protein ORF48 (positions 8-106) 
# neck passage (reprseq028936) protein conatins: 
# a) N-terminal Ig-like domain in baseplate protein ORF48 (positions 7-117)
# b) gp11/gp12 receptor-binding domain  F: GP11_1st | Protein: BASEPLATE STRUCTURAL PROTEIN GP11: 465-533 (prob < 95%) (positions 465-533)
# minor capsid reprseq067546
# a) Glycoside hydrolase/deacetylase in positions ~ 370-600, 
# b) N-terminal Ig-like domain in baseplate protein ORF48 (positions 1-110)
# in fact these re two different X groups so almost fulfil the ecod mosaicism

# endolysin & endolysin
# reprseq094548 [seem in 7 genomes, representative = NC_018836.1:21373-22425(+)] & reprseq118348 [NC_001978.3:17087-18193(+)] (shared fragment, position within reprseq094548: 223-362 )
# reprseq094548 has
# a) amidase on positions: ~ 3-155
# b) Putative peptidoglycan binding domain 218-345 (prob < 85%$, pfam, < 65% ecod)
# reprseq118348 has 
# a) Cysteine proteinases 1-153
# b) putative peptidoglycan binding domai (306-363, pfam, 78%)
# so the shared fragment is nt that well described, but the unshared ones are very well described by domain dbs
# genomes: are completely dissimilar, the only simialrity is;
# NC_018836.1 positions 21902 to 22427 (C terminal of endolysin)
# NC_018836.1 positions 26245 to 26462 (middle part of dCMP deaminase, NC_018836.1:26164-26676(+))


# reprseq047669 AND reprseq116831 (tail and   tail spike/tail fiber/haemolysin-type calcium binding/structural protein): shared fragment 1-491
# reprseq047669 (tail) has the following domains:
# a) Putative tailspike protein Orf210 N-terminal domain ( 341 - 491 ) <so it is a tail spike!>
# b) pectin lyase-like (Glyco_hydro_28): (578 - 876)
# also (below scov threshold)
# c) protein: baseplate structural protein gp10/gp9 n-terminal domain-related (91 - 177)
# reprseq116831 has the following domains:
# a) Putative tailspike protein Orf210 N-terminal domain (340-491)
# b)  Pectin lyase-like PhageP22-tail_N (485-  881)
# looks like we don;t see strong similarity in  Pectin lyase-like PhageP22-tail_N (485-  881) and pectin lyase-like (Glyco_hydro_28): (578 - 876) but they are from the same X group (Single-stranded right-handed beta-helix)
# also after aa 900 there are no domains detected. But this looks like it could be due to different rates of divergence



# unknown (shows similarity to endolysin) and TTM
# reprseq099845 (NC_023722.1:20629-25836(+), only in one genome) reprseq107503 (NC_048070.1:11025-16616(+), only in one genome)
#Rhodoccocus phage vs Cyanobacterium phage, share only 1% of the genome
# they share positions;  21072 - 21638 of NC_023722.1 i.e. a part of the NC_023722.1:20629-25836(+) gene which in our piepeline shows similarity to endolysin but in NCBI is annotated as tail length tape measure protein
# shared fragment 152-369 (in reprseq099845) and 66-283 (reprseq107503), pident = 61%
# none of the prots described by any dpmain in ECOD
# PFAM descriebes positions 550 -602 and 88-222 in reprseq107503 as Tape measure protein (so the shared fragment includes PF20155.2)
#PFAM also descriebes positions 251-307 in reprseq099845 as Tape measure protein (so the shared fragment includes PF20155.2), and other domains with small prob (< 45%) as eg bacteriocin



# TTM and TTM
#reprseq024035 and reprseq054815
```

```{r}
name.table %>% filter()
```


```{r}
recent_HGT_to_look_at.n1 = recent_HGT_pairs_raw.70.annotated.not.found.by.ecod.mosaicism %>%
  left_join(name.table %>% distinct(qname = repr.name, qncbi.id = ncbi.id)) %>%
  left_join(name.table %>% distinct(sname = repr.name, sncbi.id = ncbi.id)) %>%
  #filter(qannotation == "antidefense")
  filter(qname == "reprseq047669" & sname == "reprseq116831")

#  NC_047780.1:33401-37132(+) and # NC_028977.1:33253-35964(+)
# Two Klebsiella phage, very similar genome ()88% similarity, seems that the tail fiber is one of  only two protein that differs between the genomes! (NC_028977.1:33253-35964(+) and NC_028977.1:35977-37512(+)) i.e. reprseq125591 and reprseq000001. These are two tail fiber proteins next to each other
# within them the first gene 150 aa are similar, the rest is different
n  = names(all.cds) 
idx = which(grepl("NC_018836.1 ", n))
all.cds[idx] 

# the domain that stays same is covered by pfam: pf03906.17 ; phage_t7_tail ; phage t7 tail fibre protein (prob 77%)
# enterobacteria phage and Escherichia phage

# 

```

```{r, eval = FALSE}
recent_HGT_to_look_at.n1 = recent_HGT_to_look_at.n %>%
  left_join(name.table %>% distinct(qname = repr.name, qncbi.id = ncbi.id)) %>%
  left_join(name.table %>% distinct(sname = repr.name, sncbi.id = ncbi.id)) %>%
  filter(qannotation == "tail length tape measure")

#TMP genomes: both are from Staphylococus phage arrying Panton-Valentine leukocidin genes
#https://www.ncbi.nlm.nih.gov/nuccore/NC_002321.1/
#https://www.ncbi.nlm.nih.gov/nuccore/NC_008689.1/

tmp1_genome1 = all.cds[names(all.cds) == "NC_002321.1:8926-11010(+)"] %>% as.character()

# FROM ANNOTATE WE DONT GET THIS CDS BUT IT EXIST ON NCB
#tmp2_genome1 = all.cds[names(all.cds) == "NC_002321.1:11241-12860(+)"] %>% as.character()
# WE GET THIS INSTEAD
tmp2_genome1 = all.cds[names(all.cds) == "NC_002321.1:11112-12860(+)"] %>% as.character()

#002321 protein, 8926-11010 is a 694 aminoacid long tmp;
#aa = "MPNPIGNMVIKVDLDGSGFNRGVTGLNRQMKMVSRELSANLSQFSRYDNSLEKSKIKVEGLSKKQKVQAQITKELKDSYDKLSKETGENSAKTQAAAAKYNEAYAKLNQYERELNQATQELKDMQREQKALNTAMGKLGTNFNNFGPKLQEIGNSMKNVGRNMTMYVTAPVVAGFAVAAKKGIEFDDSMRKVKATSGATGEEFEALKKKAREMGATTKFSASDSAEALNYMALAGWDSKQMMEGLSGVMDLAAASGEELGAVSDIVTDGLTAFGLKAKDSGHLADVLAQTSSKANTDVRGLGEAFKYVAPVAGALGYTIEDTSIAIGLMSNAGIKGEKAGTALRTMFTNLSSPTRAMGNEMERLGISITDSNGKMIPMRKLLDQLREKFKHLSKDQQASSAATIFGKEAMSGALAIINASDEDYQKLTKSIDSSTGASKRMADTMESGLGGKLRTLRSQLEELALTIYDRIEPALKIIVSAFSKVVTWVTKLPTSIQLAVVGFGLFVA VLGPLVFMFGLFISVMGNAMTVLGPLLINVNKASGLFAFLRTKIASLVKLFPILGVSISSLTLPITLIVGALVGIGIAFYQAYKRSETFRNIVNQAISGVANAFKAAKLALQGFFDLFKGDSKGAVTLEKIFPPETVAGIQNVVNTIRTTFFKVVDAIVGFAKEIGAQLASFWKENGSEITQALQI"
#aa = gsub(" ", "", aa)

a1=substr(tmp1_genome1, start = 1, stop = 442)
a2=substr(tmp1_genome1, start = 443, stop = 694) # 694 is the end of this protein
a3=substr(tmp2_genome1, start = 1, stop = 650)
#"DTMESGLGGKLRTLRSQLEELALTIYDRIEPALKIIVSAFSKVVTWVTKLPTSIQLAVVGFGLFVAVLGPLVFMFGLFISVMGNAMTVLGPLLINVNKASGLFAFLRTKIASLVKLFPILGVSISSLTLPITLIVGALVGIGIAFYQAYKRSETFRNIVNQAISGVANAFKAAKLALQGFFDLFK"



#NC_008689.1
tmp1_genome2 =  all.cds[names(all.cds) == "NC_008689.1:27206-28522(+)"] %>% as.character()
tmp2_genome2 =  all.cds[names(all.cds) == "NC_008689.1:28527-31859(+)"] %>% as.character()
#bb = "MADTMESGLGGKLRTLRSQLEELALTIYDRIEPALKIIVSAFSK VVTWVTKLPTSIQLAVVGFGLFVAVLGPLVFMFGLFISVMGNAMTVLGPLLINVNKAS GLFAFLRTKIASLVKLFPILGVSISSLTLPITLIVGALVGIGIAFYQAYKRSETFRNI VNQAISGVANAFKAAKLALQGFFDLFKGDSKGAVTLEKIFPPETVAGIQNVVNTIRTT FFKVVDAIVGFAKEIGAQLASFWKENGSEITQALQNIAGFIKATFEFIFNFIIKPIMF AIWQVMQFIWPAVKALIVSTWENIKGVIQGAINIILGIIKVFSSLFTGNWRGVWDGIV MILKGTVQLIWNLIQLWFVGKILGVVRYFGGLLKGLISGIWGVIKGIFTKSLSAIWNA TKSIFGFLYNSVKSIFTNMKNWLSSTWNNIKSNTVGKAHSLFTGVRSKFTSLWNATKD IFTKLRNWMSNIWNSIKDNTVGIAGRLWDKVRNIFGNMRDGLKSIIGKIKDHIGGMVD AIKKGLNKLIEGLNWVGGKLGMDEIPRLHTGTEHTHTTTRLVKNGKIARDTFATVGDK GRGNGPNGFRNEMIEFPNGKRVITPNTDTTAYLPKGSKVYNGAQTYSMLNGTLPRFHF GTTMWKDIKSSASSAFNWTKDQIGKGTKWLGDKVGDVMDFIDNPGKLLNYVLQAFGVD FSSLTKGMGIAGDITKAAWSKIKKSAIKWLEDAFAESGDGGVLDMSKLRYLYGHTAAY TRETGRPFHEGLDFDYIYEPVPSTINGRAQVMPFHNGGYGKWVKIVKGALEVIYAHLS KYKVKTGQQVRVGQTVGISGNTGFSTGPHLHYEMRWNGRHRDPLPWLRKNNGGGKSTP GGNGAANARRAIKAAQNILGGRYKASWITNEMMRVASRESNYTANAVNNWDSNARAGI PSRGMFQMIDPSFRAYAKSGYNNPLNPTHQAISAMRYIVGKWVPRTGSWRAAFKRAGD YAYASGGKVYNGLYHLGEEGYPEWIIPTDPSRANEAHKLLALAANDIDNRSKNKRPNN LPNPSISNSDTNYIHTLENKLDAVINCLVSLVESNQVIADKDYEPVINKYVFEDEVNN SIDKRERHESTRVRFRRGGTII"
#bb = gsub(" ", "", bb)
b1=substr(tmp1_genome2, start = 1, stop = 438)
b2=substr(tmp2_genome2, start = 3, stop = 254)
b3=substr(tmp2_genome2, start = 292, stop = 950)
#"DTMESGLGGKLRTLRSQLEELALTIYDRIEPALKIIVSAFSKVVTWVTKLPTSIQLAVVGFGLFVAVLGPLVFMFGLFISVMGNAMTVLGPLLINVNKASGLFAFLRTKIASLVKLFPILGVSISSLTLPITLIVGALVGIGIAFYQAYKRSETFRNIVNQAISGVANAFKAAKLALQGFFDLFK"

print(a2)
print(b2)


print(a1)
print(b1)


print(a3)
print(b3)
```

```{r}

qname1 = "reprseq125483"#"reprseq028936"#"reprseq125483"#"reprseq053259"
qname2 = "reprseq111516"#"reprseq037076"#"reprseq088441"
# find the genome
g1=name.table.full %>% filter(qname == qname1 & cluster.name == ncbi.id)
g2=name.table.full %>% filter(qname == qname2 & cluster.name == ncbi.id)

# find other genes in the genome
name.table.full %>% filter(genome.id == unique(g1$genome.id)) %>% mutate(gene.start = as.integer(gene.start)) %>% arrange(gene.start)
name.table.full %>% filter(genome.id == unique(g2$genome.id)) %>% mutate(gene.start = as.integer(gene.start)) %>% arrange(gene.start)
hhr.this = hhr.table %>% filter((qname == qname1 & sname == qname2 )) #| (qname == qname2 & sname == qname1 ))
hhr.this2 = hhr.table %>% filter((qname == qname2 & sname == qname1 )) #| (qname == qname2 & sname == qname1 ))


g1 %>% mutate(hit.start = as.integer(gene.start) + as.integer(hhr.this$qstart)*3, 
              hit.end = as.integer(gene.start) + as.integer(hhr.this$qend)*3)
g2 %>% mutate(hit.start = as.integer(gene.start) + as.integer(hhr.this$sstart)*3, 
              hit.end = as.integer(gene.start) + as.integer(hhr.this$send)*3)

g1 %>% mutate(hit.start = as.integer(gene.start) + as.integer(hhr.this2$sstart)*3, 
              hit.end = as.integer(gene.start) + as.integer(hhr.this2$send)*3)
g2 %>% mutate(hit.start = as.integer(gene.start) + as.integer(hhr.this2$qstart)*3, 
              hit.end = as.integer(gene.start) + as.integer(hhr.this2$qend)*3)


ecod.domains.hits %>% filter(qname == qname1)
all.ecod.hits %>% filter(qname == qname1)  %>% filter(qstart < 40)
all.pfam.hits %>% filter(qname == qname1) %>% arrange(desc(qcov))# %>% filter(qstart < 450 & qend > 110)

ecod.domains.hits %>% filter(qname == qname2)
all.ecod.hits %>% filter(qname == qname2) %>% arrange(desc(prob)) %>% filter(sname == "ECOD_000009376_e1eg2A1")# %>% filter(qstart < 700 & qend > 145) -> a1 #%>% filter(qstart < 450 & qend > 110)
all.pfam.hits %>% filter(qname == qname2) %>% arrange(desc(prob)) %>% filter(qstart < 700 & qend > 160)
a1
all.cds[which(names(all.cds) == g1$ncbi.id)]
all.cds[which(names(all.cds) == g2$ncbi.id)]
```
