---
title: "generate_figures"
author: "Bogna Smug"
date: '2022-09-28'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(cowplot)
library(RColorBrewer)
library(gridExtra)
ecod_id_to_names_map = data.table::fread(sprintf("%secod_id_to_names_map.txt",OUTPUT.DATA.PATH)) %>% as.data.frame()
```

# 
```{r}
# PHROG CLASS LEGEND
dat = stack(PHROG.COLOR.MAP)
names(dat) = c("values", "Category")
ggp <- ggplot(dat, aes(x=values, y=values, fill = Category)) +    # Create ggplot2 plot
  geom_col(size = 5) +
  scale_fill_manual(values = PHROG.COLOR.MAP, name = "Category")

jpeg(sprintf("%sCategory_Legend.jpg", OUTPUT.FIGURES.PATH),  width = 10, height = 10, units = "cm", res = 300)
ggp_legend <- cowplot::get_legend(ggp) 
grid::grid.newpage()                                     # Draw empty plot window
grid::grid.draw(ggp_legend)  
dev.off()
```


# Fig 1
```{r, eval = FALSE}
col1.uniqueness <- "darkgreen"
col2.coverage <- "darkred"
point.size = 3
line.size = 1
text.size = 12
annotation.uncertainty.data.vs.cov = data.table::fread(sprintf("%sFigure_Supplementary/FigS1_tradeoff/Tradeoff_cov.txt", OUTPUT.FIGURES.PATH)) %>% as.data.frame()
p.a=ggplot(data = annotation.uncertainty.data.vs.cov %>%
          mutate(`probability\nthreshold [%]` = as.factor(minimum.probability),
                  annotation.coverage = 100*annotation.coverage)) +
  geom_point(aes(x = annotation.coverage, y = perc.unique.annotation, linetype = `probability\nthreshold [%]`), col = col1.uniqueness, size = point.size) + 
  geom_line(aes(x = annotation.coverage, y = perc.unique.annotation, linetype = `probability\nthreshold [%]`), col = col1.uniqueness, size = line.size) + 
  geom_point(aes(x = annotation.coverage, y = perc.annotated, linetype = `probability\nthreshold [%]`), col = col2.coverage, size = point.size) + 
  geom_line(aes(x = annotation.coverage, y = perc.annotated, linetype = `probability\nthreshold [%]`), col = col2.coverage, size = line.size) + 
  Get.Theme(legend.position = "right") +
  xlab("sequence coverage threshold [%]") +
  theme( axis.title.y.right = element_text(colour = col2.coverage, size = text.size),
         axis.text.y.right = element_text(colour = col2.coverage, size = text.size),
          axis.title.y = element_text(colour = col1.uniqueness, size = text.size),
         axis.text.y = element_text(colour = col1.uniqueness, size = text.size),
           plot.title = element_text(size=14, face="bold", hjust = 0.5)) +
  scale_y_continuous(name = "functional uniqueness \n(% of annotated proteins with hits to a single function)",
                     limits = c(0,101),
                     breaks = seq(0,100,10),
                     sec.axis=sec_axis(~.*1,name="functional coverage \n(% proteins with a hit to any known function)"))  + 
  labs(title = "Impact of sequence coverage on functional annotation")


annotation.uncertainty.data.vs.prob = data.table::fread(sprintf("%sFigure_Supplementary/FigS1_tradeoff/Tradeoff_prob.txt", OUTPUT.FIGURES.PATH))
p.b=ggplot(data = annotation.uncertainty.data.vs.prob %>%
             filter(annotation.coverage %in% c(0.1, 0.8)) %>%
          mutate(`sequence coverage\nthreshold [%]`  = as.factor(annotation.coverage*100))) +
  geom_point(aes(x = minimum.probability, y = perc.unique.annotation, linetype = `sequence coverage\nthreshold [%]` ), col = col1.uniqueness, size = point.size) + 
  geom_line(aes(x = minimum.probability, y = perc.unique.annotation, linetype = `sequence coverage\nthreshold [%]` ), col = col1.uniqueness, size = line.size) + 
  geom_point(aes(x = minimum.probability, y = perc.annotated, linetype = `sequence coverage\nthreshold [%]` ), col = col2.coverage, size = point.size) + 
  geom_line(aes(x = minimum.probability, y = perc.annotated, linetype = `sequence coverage\nthreshold [%]` ), col = col2.coverage, size = line.size) + 
  Get.Theme(legend.position = "right") +
  xlab("probability [%]") +
  theme( axis.title.y.right = element_text(colour = col2.coverage, size = text.size),
         axis.text.y.right = element_text(colour = col2.coverage, size = text.size),
          axis.title.y = element_text(colour = col1.uniqueness, size = text.size),
         axis.text.y = element_text(colour = col1.uniqueness, size = text.size),
           plot.title = element_text(size=14, face="bold", hjust = 0.5)) +
  scale_y_continuous(name = "functional uniqueness \n(% of annotated proteins with hits to a single function)",
                     limits = c(0,101),
                     breaks = seq(0,100,10),
                     sec.axis=sec_axis(~.*1,name="functional coverage \n(% proteins with a hit to any known function)")) + 
  labs(title = "Impact of hit probability on functional annotation")
```


```{r, eval = TRUE}
p.figure <- plot_grid(p.a, p.b, labels = c("A", "B"), label_size = 25, ncol = 1)
ggsave(p.figure, filename = sprintf("%sFigure_Supplementary/FigS1_tradeoff/FigureS1.png", OUTPUT.FIGURES.PATH), width = 12, height = 10)
```


# Domain  network
```{r}
domain.network.metadata =   data.table::fread(sprintf("%sFigure5/domain.network.metadata.txt", OUTPUT.FIGURES.PATH)) 
domain.network.selected.categories = data.table::fread(sprintf("%sFigure5/domain.network.selected.categories.txt", OUTPUT.FIGURES.PATH))

for (selected_category in unique(domain.network.metadata$category)) {
  annotated.domain.network = Annotate_Graph(domain.network.selected.categories %>% filter(category == selected_category) %>% select(from, to), 
                    edge.colors = NULL,
                    node.color.map = domain.network.metadata %>%
                      filter(category == selected_category) %>%
                                        select(node, color),
                    # node these sizes are defined at high threshold
                    node.size.map = domain.network.metadata %>%
                      filter(category == selected_category) %>%
                      mutate(size = sqrt(n.prot)) %>%
                      select(node, size),
                    node.labels = domain.network.metadata %>% 
                      filter(category == selected_category) %>%
                      select(node, label) %>%
                      rowwise() %>%
                      mutate(label = gsub(" ", "\n", label))%>%
                      mutate(label = if_else(label == "Intramolecular\nchaperone\ndomain\nin\nvirus\ntail\nspike\nprotein", "Intramolecular\nchaperone\ndomain in\nvirus tail\nspike protein", label)) ,
                    line.color.variable =  "from")
  
  annotated.domain.network.figure = VisualiseGGnetwork(annotated.domain.network, 
                     cex = 0.1, label_size = 1.5, layout = igraph::nicely(), main = "", 
                     nudge_y = 0, alpha = 0.3)
  annotated.domain.network.figure
  ggsave(annotated.domain.network.figure, filename = sprintf("%sFigure5/annotated.domain.network.figure_%s.png", OUTPUT.FIGURES.PATH, selected_category), width = 8, height = 8)
}


```

```{r}
all.domain.network.metadata =   data.table::fread(sprintf("%sFigure5/all.domain.network.metadata.txt", OUTPUT.FIGURES.PATH)) 
all.domain.network = data.table::fread(sprintf("%sFigure5/all.domain.network.txt", OUTPUT.FIGURES.PATH))

  annotated.all.domain.network = Annotate_Graph(all.domain.network, 
                    edge.colors = NULL,
                    node.color.map = all.domain.network.metadata %>%
                                        select(node, color),
                    # node these sizes are defined at high threshold
                    node.size.map = all.domain.network.metadata %>%
                                        select(node, size),
                    node.labels = all.domain.network.metadata %>% 
                      select(node, label) %>%
                      rowwise() %>%
                      mutate(label = gsub(" ", "\n", label))%>%
                      mutate(label = if_else(label == "Intramolecular\nchaperone\ndomain\nin\nvirus\ntail\nspike\nprotein", "Intramolecular\nchaperone\ndomain in\nvirus tail\nspike protein", label)) ,
                    line.color.variable =  "from")
  
  annotated.domain.network.figure = VisualiseGGnetwork(annotated.all.domain.network, 
                     cex = 0.1, label_size = 1.5, layout = igraph::nicely(), main = "", 
                     nudge_y = 0, alpha = 1)
  annotated.domain.network.figure
  ggsave(annotated.domain.network.figure, filename = sprintf("%sFigure5/annotated.ALL.domain.network.figure.png", OUTPUT.FIGURES.PATH, selected_category), width = 28, height = 28)


```


```{r}

annotation.netwok.relaxed.threshold = data.table::fread( 
                   sprintf("%sFigure_Supplementary/FigS2_annot_network/annotation.netwok.relaxed.threshold.txt", OUTPUT.FIGURES.PATH))
annotation.netwok.default.threshold = data.table::fread( 
                   sprintf("%sFigure_Supplementary/FigS2_annot_network/annotation.netwok.default.threshold.txt", OUTPUT.FIGURES.PATH))
annotation.network.metadata = data.table::fread( 
                   sprintf("%sFigure_Supplementary/FigS2_annot_network/annotation.network.metadata.txt", OUTPUT.FIGURES.PATH)) %>%
  rowwise() %>%
  mutate(label = if_else(label == "DNA\npolymerase/primase", "DNA\npolymerase/\nprimase",  gsub("\n", " ", label)))

  
annotated.annotation.network.relaxed.threshold = Annotate_Graph(annotation.netwok.relaxed.threshold, 
                  edge.colors = NULL,
                  node.color.map = annotation.network.metadata%>%
                                      select(node, color),
                  # node these sizes are defined at high threshold
                  node.size.map = annotation.network.metadata %>%
                                      select(node, size),
                  node.labels = annotation.network.metadata %>% 
                    select(node, label),
                  line.color.variable =  "from")
annotated.annotation.network.default.threshold = Annotate_Graph(annotation.netwok.default.threshold, 
                  edge.colors = NULL,
                  node.color.map = annotation.network.metadata%>%
                                      select(node, color),
                  # node these sizes are defined at high threshold
                  node.size.map = annotation.network.metadata %>%
                                      select(node, size),
                  node.labels = annotation.network.metadata %>% 
                    select(node, label),
                  line.color.variable =  "from")



network.relaxed.figure = VisualiseGGnetwork(annotated.annotation.network.relaxed.threshold, 
                   cex = 0.1, label_size = 2.35, layout = igraph::nicely(), main = "", nudge_y = 0.005, alpha = 0.3) #nudge_y = 0.015,
network.relaxed.figure
#ggsave(network.relaxed.figure, filename = sprintf("%sFigure_Supplementary/FigS2_annot_network/Relaxed_network.png", OUTPUT.FIGURES.PATH), width = 9, height = 10)

network.default.figure = VisualiseGGnetwork(annotated.annotation.network.default.threshold, 
                   cex = 0.1, label_size = 2.35, layout = igraph::nicely(), main = "", nudge_y = 0.03, alpha = 0.5)
network.default.figure
#ggsave(network.default.figure, filename = sprintf("%sFigure_Supplementary/FigS2_annot_network/Default_network.png", OUTPUT.FIGURES.PATH), width = 5, height = 5)

p.figure <- plot_grid(network.default.figure, network.relaxed.figure, labels = c("A", "B"), label_size = 25, ncol = 1, rel_heights = c(0.27, 0.73))
ggsave(p.figure, filename = sprintf("%sFigure_Supplementary/FigS2_annot_network/Supplementary_Figure2.pdf", OUTPUT.FIGURES.PATH), width =8, height = 12)

```

```{r}
recent_HGT_pairs = data.table::fread(sprintf("%sFigure_Supplementary/FigS11_recent_HGT/recent_HGT_network.txt", OUTPUT.FIGURES.PATH)) %>%
  select(from, to, num.prot.pairs)



annotation.network.metadata = data.table::fread( 
                   sprintf("%sFigure_Supplementary/FigS2_annot_network/annotation.network.metadata.txt", OUTPUT.FIGURES.PATH)) %>%
  rbind(data.frame(node = 0, color = "#808080", size = 4, label = "unknown"))

  
annotated.recent_HGT_pairs = Annotate_Graph(recent_HGT_pairs, 
                  edge.colors = NULL,
                  node.color.map = annotation.network.metadata%>%
                                      select(node, color),
                  # node these sizes are defined at high threshold
                  node.size.map = annotation.network.metadata %>%
                                      select(node, size),
                  node.labels = annotation.network.metadata %>% 
                    select(node, label),
                  line.color.variable =  "from")



jpeg(filename = sprintf("%sFigure_Supplementary/FigS11_recent_HGT//network_recent_HGT_igraph.jpg", OUTPUT.FIGURES.PATH), width = 15, height = 15, units = "cm", res = 300)
VisualiseIgraphnetwork(data.for.network = recent_HGT_pairs %>% mutate(weight = log10(num.prot.pairs)/3),
                      edge.colors = NULL,
                      node.color.map = annotation.network.metadata%>%
                                      select(node, color),
                      # node these sizes are defined at high threshold
                      node.size.map = annotation.network.metadata %>%
                                      select(node, size),
                      node.labels = annotation.network.metadata %>% 
                    select(node, label),
                      line.color.variable = "from",
                      line.color.legend.name = NULL,
                      font.size = 0.75,
                      vertex.size = 1,
                      main = "",
                      vertex.label.dist = 1,
                      include.labels = TRUE,
                      alpha = 0.3)
dev.off()

```

# Fig 3 network
```{r}
mosaic.network = data.table::fread(sprintf("%sFigure3/3a_mosaic.network.txt", OUTPUT.FIGURES.PATH))
#mosaic.metadata = data.table::fread(sprintf("%sFigure3/3a_mosaic.metadata.txt", OUTPUT.FIGURES.PATH), sep = "\t")



jpeg(filename = sprintf("%sFigure3/network.jpg", OUTPUT.FIGURES.PATH), width = 10, height = 10, units = "cm", res = 300)
VisualiseIgraphnetwork(data.for.network = mosaic.network %>% mutate(weight = log10(num.reprseq.any.mossaic)/20) %>% 
                         select(from = from.index, to = to.index, weight, dataset),
                      edge.colors = data.frame(value = c("ecod", "seq", "both"), color = c("gray", "gray", "darkred")),
                      edge.linetypes = data.frame(value = c("ecod", "seq", "both"), lty = c("solid", "dotted", "solid")),
                      node.color.map = annotation.network.metadata%>%
                                      select(node, color),
                      # node these sizes are defined at high threshold
                      node.size.map = annotation.network.metadata %>%
                                      select(node, size) %>% mutate(size = 2*size),
                      node.labels = annotation.network.metadata %>% 
                    select(node, label),
                      line.color.variable = "dataset",
                      line.color.legend.name = "Mosaicism via",
                      font.size = 0.25,
                      main = "",
                      vertex.label.dist = 1,
                      include.labels = TRUE,
                      alpha = 1,
                    label.font = 2,
                    legend.title = "Mosaicism via:",
                    node.label.color = "black")
dev.off()
```





# Fig 2: heatmap of ECODS found in multiple categories
```{r}

text_size.x = 9
text_size.y = 9

h.domain.occurence.data = data.table::fread(sprintf("%sFigure2/H_domains.txt", OUTPUT.FIGURES.PATH)) %>% 
  as.data.frame() 
h.domain.occurence.data$category <- factor(h.domain.occurence.data$category, levels = phrogs.class.order)

phrog.class.labels2 = phrog.class.labels
phrog.class.labels2[which(names(phrog.class.labels2) == "head and packaging")] = "head\nand\npack."
phrog.class.labels2[which(names(phrog.class.labels2) == "DNA, RNA and nucleotide metabolism")] = "DNA,\nRNA\nand\nnucleot.\nmetab."
ggplot(h.domain.occurence.data) +
  geom_tile(aes(x=h_name, y = annotation, fill = category)) + #, hjust = 0, vjust = 0
  theme(axis.text.x = element_text(angle=45, hjust =1,vjust=1, size = text_size.x),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = text_size.y),
        strip.text.y = element_text(face = "bold", size = 6, angle = 360),
        strip.text.x = element_text(face = "bold", size = 6, angle = 360)) +
  facet_grid(category ~all.categories, scales = "free", space = "free", 
             labeller = labeller(category = phrog.class.labels,
                                 all.categories = c(phrog.class.labels2, "multiple categories" = "multiple\ncategories"))) + 
  xlab("") + ylab("") +
  guides(fill = "none") +
  scale_fill_manual(values = PHROG.COLOR.MAP)

ggsave(sprintf("%sFigure2/H_domains.jpg", OUTPUT.FIGURES.PATH), width = 7, height = 10)


num.f.within.h = data.table::fread(sprintf("%sFigure_Supplementary/FigS3_num_F_per_H/num.f.within.h.txt", OUTPUT.FIGURES.PATH)) 

num.f.within.h.bars = num.f.within.h %>%
  filter(in.ecod.heatmap) %>%
  rename(family = num_f, topology = num_t) %>%
  tidyr::gather(key = "ECOD level", value = "num.groups", family, topology)

text_size.x = 8
num.f.fig = ggplot(num.f.within.h.bars) +
  geom_col(aes(x = h_name, y = num.groups, fill = all.categories, alpha = `ECOD level`), position = position_dodge2(preserve = "single")) +
    theme(axis.text.x = element_text(angle=45, hjust =1,vjust=1, size = text_size.x),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = text_size.y),
        strip.text.y = element_text(face = "bold", size = 6, angle = 360),
        strip.text.x = element_text(face = "bold", size = 6, angle = 360),
        legend.position = "top") +
  facet_grid(. ~all.categories, scales = "free", space = "free", 
             labeller = labeller(all.categories = c(phrog.class.labels2, "multiple categories" = "multiple\ncategories"))) + 
  xlab("") + ylab("Num. of domain families in the group.") +
  guides(fill = "none") +
  scale_fill_manual(values = c(PHROG.COLOR.MAP, "multiple categories" = "#000000")) +
  scale_y_sqrt(breaks = c(1,10,25,50,100,150)) +
  scale_alpha_manual(values = c("family" = 0.5, "topology" =1))


ct=cor.test(num.f.within.h$num_f, num.f.within.h$num_prot, method = "spearman")
cor.fig = ggplot(num.f.within.h) +
  geom_point(aes(x = num_prot, y = num_f)) +
  scale_x_log10() +
  scale_y_log10() +
  geom_text(aes(x = 2, y = 100, label = paste0("Rho = ", as.numeric(round(ct$estimate,2)), ",\np.val = ",  as.numeric(round(ct$p.value,5))))) +
  theme_bw() + 
  theme(axis.text.x = element_text(size = text_size.x),
        axis.text.y = element_text(size = text_size.y)
        #plot.margin = unit(c(0, 0, 6, 0), "cm")
        ) +
  xlab("Number of representative proteins.") +
  ylab("Number of ECOD family groups.")

fig_diversity_ecod_f = plot_grid(num.f.fig, cor.fig, labels = c("A", "B"), label_size = 25, ncol = 1, rel_heights =  c(0.6, 0.4))#ncol = 2,rel_widths = c(0.6, 0.4))
ggsave(sprintf("%sFigure_Supplementary/FigS3_num_F_per_H/num.f.within.h.jpg", OUTPUT.FIGURES.PATH),width = 9, height = 12) #)width = 15, height = 7)


t.domain.occurence.data.full = data.table::fread(sprintf("%sFigure_Supplementary/FigS4_all_T_domains/T_domains.txt", OUTPUT.FIGURES.PATH)) %>%
  as.data.frame() 
t.domain.occurence.data.full$category <- factor(t.domain.occurence.data.full$category, levels = phrogs.class.order)
ggplot(t.domain.occurence.data.full) +
  geom_tile(aes(y=t_name, x = annotation, fill = category)) + #, hjust = 0, vjust = 0
  theme(axis.text.x = element_text(angle=45, hjust =1,vjust=1, size = text_size.x),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = text_size.y),
        strip.text.y = element_text(face = "bold", size = 6, angle = 360),
        strip.text.x = element_text(face = "bold", size = 6, angle = 360)) +
  facet_grid(all.categories ~category, scales = "free", space = "free", 
             labeller = labeller(category = phrog.class.labels,
                                 all.categories = c(phrog.class.labels2, "multiple categories" = "multiple\ncategories"))) + 
  xlab("") + ylab("") +
  guides(fill = "none") +
  scale_fill_manual(values = PHROG.COLOR.MAP)
ggsave(sprintf("%sFigure_Supplementary/FigS4_all_T_domains/T_domains.jpg", OUTPUT.FIGURES.PATH), width = 20, height = 20)


```





```{r}

most.mosaic.classes = c("tail fiber","tail spike", "endolysin", "DNA polymerase")
  tile.data = data.table::fread(sprintf("%sFigure5/domain.positions.txt", OUTPUT.FIGURES.PATH)) %>%
    filter(domain.level == "T" & annotation %in% most.mosaic.classes)

domain.position.plots =Plot.Domain.Combinations(tile.data, text.size.axis = 9, text.size.panel = 12)$domain.position.plots
pdf(sprintf("%sFigure5/domain.positions_most_mosaic.pdf", OUTPUT.FIGURES.PATH), 
         width = 17,
         height = 20)
do.call("grid.arrange", c(domain.position.plots, ncol=1))
dev.off()


tile.data.dnap.final = data.table::fread(sprintf("%sFigure5/domain.positions.dnap.uniclust.txt", OUTPUT.FIGURES.PATH))
domain.position.plots =Plot.Domain.Combinations(tile.data.dnap.final, text.size.axis = 9, text.size.panel = 12, interactive.plot = TRUE)$domain.position.plots
pdf(sprintf("%sFigure5/domain.positions_hypothetical_dnap.uniclust.pdf", OUTPUT.FIGURES.PATH), 
         width = 17,
         height = 12)
#gg_rect
#ggiraph:girafe(ggobj=gg_rect)
do.call("grid.arrange", c(domain.position.plots, ncol=1))
dev.off()

```




# Fig5: SHINY PLOTS
```{r, eval = FALSE}
text.size.axis = 6
text.size.panel = 12

  domain.position.plots = list()
  tile.data = data.table::fread(sprintf("%sFigure5/domain.pbottomositions.txt", OUTPUT.FIGURES.PATH)) %>% filter(domain.level == "T")

  colormap = tile.data %>% distinct(domain) %>% filter(!(domain %in% c("undetected", "multiple domains")))
  colormap$color = randomcoloR::distinctColorPalette(nrow(colormap))
  colormap = rbind(colormap, data.frame(domain = c("undetected", "multiple domains"), 
                                                  color = c( "#FFFFFF",  "#000000"), stringsAsFactors = FALSE))
  colm = colormap$color
  names(colm) = colormap$domain
        
      for (this.annotation in unique(tile.data$annotation)) {
        this.tile.data = tile.data %>% 
          filter(annotation == this.annotation)  %>% 
          mutate(annotation = this.annotation) %>%
          arrange(row)
        
        gg_rect = Show.Domain.Position.Within.Data(
          this.tile.data,
          colm, 
          theme.no.verical) +
          #scale_y_discrete(breaks =this.tile.data$qname, labels = this.tile.data$family) +
          
            theme(strip.text.x = element_text(size = text.size.panel), 
            legend.position = "none") +
              theme(
                axis.title.x = element_text(size = text.size.axis),
                axis.title.y = element_text(size = text.size.axis),
                #axis.text.x = element_blank(),
                axis.text.y = element_blank(),
                #axis.text.y = element_text(size = text.size.axis),
                #axis.ticks.x = element_blank(),
                axis.ticks.y = element_blank())
                
        domain.position.plots[[this.annotation]] =  gg_rect
  }
  
  
  #Create a subfigure with legend only, take only the most frequent
  colormap.with.names = colormap %>% inner_join(tile.data %>% 
                                                 group_by(domain_name) %>%
                                                 mutate(num_qnames = n_distinct(qname)) %>%
                                                 filter(num_qnames >= 3) %>%
                                                 distinct(domain, domain_name)) %>% distinct(color, domain_name)
  colm.with.names = colormap.with.names$color
  names(colm.with.names) = colormap.with.names$domain_name
  
  g.dummy = ggplot(tile.data, aes(x=qname, y = domain, fill = domain_name)) + geom_col() + 
    scale_fill_manual(values = colm.with.names, name = 'Domain topology') +
    theme(legend.margin =margin(r=0,l=0,t=0,b=0), text = element_text(size = 7)) +
    guides(fill=guide_legend(ncol=6))
  
  leg <- ggpubr::get_legend(g.dummy, position = "right")
  ggleg = ggpubr::as_ggplot(leg)
#grid::widthDetails(leg)
#grid::heightDetails(leg)


  
    ncol = ceiling(length(domain.position.plots)/4)
    jpeg(sprintf("%sFigure5/domain.positions.jpg", OUTPUT.FIGURES.PATH), 
         width = 50, 
         height = 30, units = "cm", res = 300)
    
    ggpubr::ggarrange(
      egg::ggarrange(plots = domain.position.plots,
                     ncol = ncol, nrow = 4),
      ggleg,
      nrow = 2,
      heights = c(0.77,0.25))
      #legend.prob = ggleg,
      #nrow = 1,
      #legend = "bottom")
  dev.off()

```

# Fig. 4: diversity
```{r}
diversity = data.table::fread(sprintf("%sFigure_Supplementary/FigS9_ecod_hist/Diversity_vs_mosaicism.txt", OUTPUT.FIGURES.PATH))

Fig4a = ggplot(diversity, aes(x=prop.proteins.with.ecod.hit)) + 
  geom_histogram(fill = "blue", position="dodge", bins = 20, color = "black") + 
  theme_minimal() + 
  labs(x = "Proportion of proteins with ECOD hit",
       y = "Number of functional classes")
ggsave(sprintf("%sFigure_Supplementary/FigS9_ecod_hist/ecod_histogram.pdf", OUTPUT.FIGURES.PATH), Fig4a, width = 12, height = 10)


# label the ones that have many families or fold types

diversity.ok.ecod = diversity %>% filter(!poorly.ecod.annotated)  %>%
  mutate(label = if_else(diverse.for.label,annotation , ""))


Fig4b = ggplot(diversity.ok.ecod) + 
  #geom_point(aes(x = num.proteins, y = num.topology.combinations, colour = category)) + 
  #geom_vline(aes(xintercept = median.num.prot), col = "red", linetype = "dashed") +
  geom_point(aes(x = num.families, y = num.topology.combinations, colour = category)) + 
  geom_vline(aes(xintercept = median.num.fam), col = "red", linetype = "dashed") +
  geom_hline(aes(yintercept = median.num.topology.combinations), col = "red", linetype = "dashed") +
  scale_x_log10() + 
  #ggrepel::geom_label_repel(aes(x = num.proteins, y = num.topology.combinations, colour = category, label = label), size = 3) + 
  ggrepel::geom_label_repel(aes(x = num.families, y = num.topology.combinations, colour = category, label = label), size = 3) + 
  scale_colour_manual(values = PHROG.COLOR.MAP) + 
  labs(x = "Number of families", y = "Number of fold combinations") + 
  guides(colour = "none") + theme_minimal()

# Fig 4c
# plot the relation between genetic and structural diversity



Get.Labeller = function(diversity.top) {
  levels = c("mean.num.t", 
             "prop.mosaic.pairs.sequence.thr0p1",
             "prop.mosaic.pairs.sequence.thr0p3",
             "prop.mosaic.pairs.sequence.thr0p5", 
             "num.theoretical.fold.type.pairs.with.multi.x",
             "num.mosaic.pairs")
   diversity.top = diversity.top %>%
    filter(proportion %in% levels)
  diversity.top = diversity.top %>% mutate(proportion = factor(proportion, levels = levels))
  diversity.levels <- levels(diversity.top$proportion)
  diversity.labeller <- character(length(diversity.levels))
  diversity.labeller[diversity.levels == "mean.num.t"] <- "Mean number of folds\nper fold architecture"
  diversity.labeller[diversity.levels == "prop.mosaic.pairs.sequence.thr0p1"] <- "Proportion of mosaic pairs\nby sequence (10%)"
  diversity.labeller[diversity.levels == "prop.mosaic.pairs.sequence.thr0p3"] <- "Proportion of mosaic pairs\nby sequence (30%)"
  diversity.labeller[diversity.levels == "prop.mosaic.pairs.sequence.thr0p5"] <- "Proportion of mosaic pairs\nby sequence (50%)"
  diversity.labeller[diversity.levels == "num.theoretical.fold.type.pairs.with.multi.x"] <- "Number of possible\npairs of fold types\nwith multiple X groups"
  diversity.labeller[diversity.levels == "num.mosaic.pairs"] <- "Number of pairs\nmosaic by domain"
  #diversity.labeller[diversity.levels == "num.reprseq.pairs.domain"] <- "Number of protein pairs\nwith domain mosaicism"
  #diversity.labeller[diversity.levels == "prop.mosaic"] <- "Prop. of proteins\nhaving any mosaic signal."
  names(diversity.labeller) <- diversity.levels
  return(diversity.labeller)
}

  
Plot.Mosaicism.Table = function(diversity.top, PHROG.COLOR.MAP) {
  p <- ggplot(diversity.top, aes(y = annotation, x = value, fill = category)) +
    geom_bar(stat = "identity", position=position_dodge()) +
    scale_fill_manual(values = PHROG.COLOR.MAP) + theme_minimal() + 
    theme(strip.text.y = element_text(face = "bold", size = 7, angle = 360)) + 
    labs(x = "", y = "") +
    scale_x_continuous(labels=function(x) sprintf("%.2f", x))
  
  p.most.diverse <- p + guides(fill = "none")
  return(p.most.diverse)
}

diversity.top = data.table::fread(sprintf("%sFigure4/mosaicism.top.diverse.classes.txt", OUTPUT.FIGURES.PATH)) 
diversity.labeller = Get.Labeller(diversity.top)

p.most.diverse = Plot.Mosaicism.Table(diversity.top %>% filter(proportion %in% names(diversity.labeller)), 
                                      PHROG.COLOR.MAP = PHROG.COLOR.MAP) +
    facet_grid(category ~ proportion, scales = "free", space = "free_y", 
                labeller = labeller(category = phrog.class.labels, proportion = diversity.labeller))

diversity.data.per.category = data.table::fread(sprintf("%sFigure4/mosaicism.per.category.txt", OUTPUT.FIGURES.PATH))
diversity.data.per.category$category[diversity.data.per.category$category == "DNA, RNA and nucleotide metabolism"] = "DNA, RNA\nand nucleotide metabolism"
diversity.data.per.category$category[diversity.data.per.category$category == "moron, auxiliary metabolic gene and host takeover"] =
"moron, auxiliary metabolic\ngene and host takeover"
PHROG.COLOR.MAP2 = PHROG.COLOR.MAP
idx = which(names(PHROG.COLOR.MAP2) == "DNA, RNA and nucleotide metabolism")
names(PHROG.COLOR.MAP2)[idx] =  "DNA, RNA\nand nucleotide metabolism"
idx = which(names(PHROG.COLOR.MAP2) == "moron, auxiliary metabolic gene and host takeover")
names(PHROG.COLOR.MAP2)[idx] =  "moron, auxiliary metabolic\ngene and host takeover"
fug4c = Plot.Mosaicism.Table(diversity.data.per.category %>% mutate(annotation = category) %>% filter(proportion %in% names(diversity.labeller)), 
                             PHROG.COLOR.MAP = PHROG.COLOR.MAP2)  +
    facet_grid(. ~ proportion, scales = "free", space = "free_y", 
                labeller = labeller(proportion = diversity.labeller)) +
  theme(plot.margin = unit(c(0, 6, 0, 0), "cm"))


p.fig4 <- plot_grid(Fig4b, p.most.diverse,fug4c, ncol = 1, labels = c("A", "B", "C"), label_size = 18, rel_heights = c(0.3, 0.5, 0.2))
p.fig4.filename <- sprintf("%sFigure4/Figure4.png", OUTPUT.FIGURES.PATH)
ggsave(p.fig4.filename, p.fig4, width = 12, height = 15)
ggsave(sprintf("%sFigure4/Figure4.pdf", OUTPUT.FIGURES.PATH), p.fig4, width = 15, height = 15)
## supplementary version
# Fig 4c
# plot the relation between genetic and structural diversity
```

```{r, eval = FALSE}
levels.supp = c("mean.num.t.per.prot", 
           "prob.mosaicism.if.sharing.fold.and.multi.x.family.measure")
diversity.top.supp = data.table::fread(sprintf("%sFigure4/mosaicism.top.diverse.classes.txt", OUTPUT.FIGURES.PATH)) %>%
  filter(proportion %in% levels.supp)
diversity.top.supp = diversity.top.supp %>% mutate(proportion = factor(proportion, levels = levels.supp))
diversity.levels.supp <- levels(diversity.top.supp$proportion)
diversity.labeller <- character(length(diversity.levels.supp))
diversity.labeller[diversity.levels.supp == "mean.num.t.per.prot"] <- "Mean number of folds\nper protein family"
diversity.labeller[diversity.levels.supp == "prob.mosaicism.if.sharing.fold.and.multi.x.family.measure"] <- "Probability of domain mosaicism\nbetween protein families \nif sharing a fold \nand having multiple X domains"

names(diversity.labeller) <- diversity.levels.supp

p <- ggplot(diversity.top.supp, aes(y = annotation, x = value, fill = category)) +
  geom_bar(stat = "identity", position=position_dodge()) +
  facet_grid(category ~ proportion, scales = "free", space = "free_y", 
                    labeller = labeller(category = phrog.class.labels, proportion = diversity.labeller)) +
  scale_fill_manual(values = PHROG.COLOR.MAP) + theme_minimal() + 
  theme(strip.text.y = element_text(face = "bold", size = 7, angle = 360)) + 
  labs(x = "", y = "")
p.most.diverse.supp <- p + guides(fill = "none")
p.most.diverse.supp

p.fig4.filename <- sprintf("%sFigure4/Figure4c_supplementary_version.png", OUTPUT.FIGURES.PATH)
ggsave(p.fig4.filename, p.most.diverse.supp, width = 15, height = 8)


```

Poorly annotated ECOD analysis
```{r}
classes.with.few.ecod.annotated.prots.new = data.table::fread(sprintf("%sFigure_Supplementary/FigS10_poor_ecod/classes.with.few.ecod.annotated.txt", OUTPUT.FIGURES.PATH))
classes.with.few.ecod.annotated.prots.new$annotation = factor(classes.with.few.ecod.annotated.prots.new$annotation , levels = unique(classes.with.few.ecod.annotated.prots.new$annotation))
classes.with.few.ecod.annotated.prots.long = classes.with.few.ecod.annotated.prots.new %>% 
  select(annotation, category, `prot. with any ECOD hit`,  `proteome covered by ECOD hits`) %>%
  tidyr::gather(key = "stat", value = "prop",`prot. with any ECOD hit`,  `proteome covered by ECOD hits`)

p=ggplot(classes.with.few.ecod.annotated.prots.long, aes(x = annotation, y = prop, fill = category, alpha = stat)) +
   geom_bar(stat = "identity", position=position_dodge()) +
  facet_grid(. ~ category, scales = "free_x", space = "free_x") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1 )) +
  scale_y_continuous(limits = c(0,1))+ 
  guides(fill = "none")+ 
  labs(x = " function", y = "Proportion") +
  scale_fill_manual(values = PHROG.COLOR.MAP) +
  scale_alpha_discrete(range = c(1, .4))
p


ggsave(sprintf("%sFigure_Supplementary/FigS10_poor_ecod/poor_ecod_annotation.png", OUTPUT.FIGURES.PATH), p,width = 20, height = 10)

```




```{r, fig.height =5, fig.width =10}
annot.stats.per.category.and.mosaicism = data.table::fread(file = sprintf("%sFigure3/unknown_functions_mosaic_vs_nonmosaic.txt", OUTPUT.FIGURES.PATH)) %>%
  arrange(category, annotation)


p.annot.low.cov = ggplot(annot.stats.per.category.and.mosaicism %>%
                           filter(!is.na(category) & category != "") %>%
  mutate(mosaic = if_else(mosaic, "Mosaic", "Not mosaic")) %>% as.data.frame()) +
  geom_col(aes(x=annotation, y = prop.proteins.has.this.annot, fill = category, alpha = mosaic), col = "black", position = position_dodge2(preserve = "single") ) +
  facet_grid(.~category, scales = "free", space = "free_x", labeller = labeller(category = phrog.class.labels)) +
  scale_alpha_manual(values = c("Mosaic" = 1, "Not mosaic" = 0.3)) +
  scale_fill_manual(values = PHROG.COLOR.MAP) +
  theme(axis.text.x = element_text(angle=45, size = 8, vjust = 1, hjust = 1),
        axis.text.y = element_text(size = 9)) +
  guides(fill = "none") + 
  theme(legend.text=element_text(size=4)) + 
  labs(x = "", y = "Proportion of proteins with this annotation") 
ggsave(filename = sprintf("%sFigure_Supplementary/FigS8_annot_low_cov/annotations.at.low.cov.jpg", 
                          OUTPUT.FIGURES.PATH) ,p.annot.low.cov , width = 15, height = 12)





annots.within.proteins.with.significant.higher.mosaic = annot.stats.per.category.and.mosaicism %>%
   filter(corrected.pval < 0.05 & odds.ratio > 1) %>%
  select(annotation, category, mosaic, prop.proteins.has.this.annot, odds.ratio, corrected.pval) %>%
   arrange(desc(odds.ratio))
write.xlsx(x = annots.within.proteins.with.significant.higher.mosaic, 
           file = sprintf("%stables/Supplementary_table_mosaic_vs_ninmosaic.xlsx", OUTPUT.FIGURES.PATH), 
           sheetName = "Annotations in mosaic vs non-mosaic proteins", 
  col.names = TRUE, row.names = TRUE, append = TRUE)


ss=annot.stats.per.category.and.mosaicism %>%
  arrange(desc(odds.ratio)) %>%
  filter(corrected.pval < 0.05) %>%
  head(60) %>%
  mutate(mosaic = if_else(mosaic, "Mosaic", "Not mosaic"))
  
ss$mosaic = factor(ss$mosaic, levels = c("Not mosaic", "Mosaic"))
ss$annotation = factor(ss$annotation, levels = unique(ss %>% arrange((odds.ratio)) %>% pull(annotation)))
ss$category[ss$category == "DNA, RNA and nucleotide metabolism"] = "DNA, RNA\nand nucleotide metabolism"
ss$category[ss$category =="moron, auxiliary metabolic gene and host takeover" ] = "moron,\nauxiliary metabolic gene\nand host takeover" 
PHROG.COLOR.MAP2 = PHROG.COLOR.MAP
idx=which(names(PHROG.COLOR.MAP2) == "DNA, RNA and nucleotide metabolism")
names(PHROG.COLOR.MAP2)[idx] = "DNA, RNA\nand nucleotide metabolism"
idx=which(names(PHROG.COLOR.MAP2) ==  "moron, auxiliary metabolic gene and host takeover")
names(PHROG.COLOR.MAP2)[idx] = "moron,\nauxiliary metabolic gene\nand host takeover" 

p.odds.low.cov = ggplot(ss  %>% distinct(annotation, category, odds.ratio), 
                        aes(x=annotation, y = odds.ratio, fill = category)) +
  geom_bar(stat="identity", width = 0.8) + 
  scale_fill_manual(values = PHROG.COLOR.MAP2) +
  theme_minimal() + 
  #guides(fill = "none") + 
  theme(legend.text=element_text(size=8)) + 
  labs(x = "", y = "Odds ratio", fill = "category") + 
  coord_flip() 
p.odds.low.cov
ggsave(filename = sprintf("%sFigure3/low.cov.odds.ratio.jpg", OUTPUT.FIGURES.PATH) ,p.odds.low.cov , width = 10, height = 7)


```


# new Fif 3c shared domains by mosaic pairs (known and unknown)
```{r}
domains.shared.in.mosaic.pairs = data.table::fread(file = sprintf("%sFigure_Supplementary/FigS7_folds_shared_by_mosaic/domains-shared-by-mosaic-pairs.txt", OUTPUT.FIGURES.PATH))

included.ecods = domains.shared.in.mosaic.pairs %>% 
  filter(prop.protein.pairs >= 0.005) %>%
  pull(t_id)

ecod.table <- domains.shared.in.mosaic.pairs %>% 
  filter(t_id %in% included.ecods)%>%
  arrange(prop.protein.pairs)

ecod.table$t_name = factor(ecod.table$t_name, 
                           levels = ecod.table %>% pull(t_name) %>% unique())

display.h.name <- ecod.table$h_name
display.h.name[ecod.table$t_name == ecod.table$h_name] <- "same"
display.t.name <- sprintf("%s\n(H: %s)", ecod.table$t_name, display.h.name)

getPalette <- colorRampPalette(RColorBrewer::brewer.pal(9, "Set1"))
no.hom <- length(unique(ecod.table$h_name))
p.shared.ecods <- ggplot(data=ecod.table, aes(x=t_name, y=prop.protein.pairs, fill=h_name)) + 
  geom_bar(stat="identity", position = position_dodge2(width = 0.9, preserve = "single"),
           alpha = 1, color="black") +#, fill = "gray90") +
  scale_x_discrete(breaks = ecod.table$t_name, labels = display.t.name) +
  theme_minimal() + theme(legend.text=element_text(size=6)) +
  labs(x = "", y = "Proportion of representative proteins pairs") + 
  guides(fill = "none", color = "none") + 
  coord_flip()

p.shared.ecods
#p.panelc <- plot_grid(p.seq, p.ecod, labels = c("C", "D"), label_size = 25, ncol = 2)
ggsave(filename = sprintf("%sFigure_Supplementary/FigS7_folds_shared_by_mosaic/folds_shared_by_mosaic_proteins.jpg", OUTPUT.FIGURES.PATH), p.shared.ecods, width = 12, height = 6)

```


# Fig 3c shared ecods in knwona and unknown prots
```{r}
domains.within.proteins = data.table::fread(file = sprintf("%sFigure3/domains-in-known-and-unknown.txt", OUTPUT.FIGURES.PATH))


domains.within.proteins.with.significant.higher.mosaic = domains.within.proteins %>%
   filter(corrected.pval < 0.05 & odds.ratio > 1) %>%
  select(t_id, t_name, mosaic, prop.proteins.with.this.ecod.hit = freq.q, odds.ratio, corrected.pval) %>%
   arrange(desc(odds.ratio))
write.xlsx(x = domains.within.proteins.with.significant.higher.mosaic, 
           file = sprintf("%stables/Supplementary_table_mosaic_vs_ninmosaic.xlsx", OUTPUT.FIGURES.PATH), 
           sheetName = "Domains in mosaic vs non-mosaic proteins", 
  col.names = TRUE, row.names = TRUE, append = TRUE)


included.ecods = domains.within.proteins %>% 
  filter(corrected.pval < 0.05) %>%
  arrange(desc(odds.ratio)) %>%
  head(30) %>%
  #filter(freq.q >= 0.025) %>%
  pull(t_id) %>% unique()

ecod.table <- domains.within.proteins %>% 
  filter(t_id %in% included.ecods)%>%
  arrange(mosaic, odds.ratio)

ecod.table$t_name = factor(ecod.table$t_name, 
                           levels = ecod.table %>% filter(mosaic == "Mosaic") %>% arrange(odds.ratio) %>% pull(t_name) %>% unique())
ecod.table$mosaic = factor(ecod.table$mosaic, levels = c("Not mosaic", "Mosaic"))

display.h.name <- ecod.table$h_name
display.h.name[ecod.table$t_name == ecod.table$h_name] <- "same"
display.t.name <- sprintf("%s\n(H: %s)", ecod.table$t_name, display.h.name)

getPalette <- colorRampPalette(RColorBrewer::brewer.pal(9, "Set1"))
no.hom <- length(unique(ecod.table$h_name))
p <- ggplot(data=ecod.table, aes(x=t_name, y=freq.q, alpha = mosaic)) + #fill=h_name, 
  geom_bar(stat="identity", position = position_dodge2(width = 0.9, preserve = "single"), col = "black")
#p <- p + scale_fill_manual(values = getPalette(no.hom)) +
#  scale_color_manual(values = getPalette(no.hom))
p <- p + scale_x_discrete(breaks = ecod.table$t_name, labels = display.t.name)
p <- p + theme_minimal() + theme(legend.text=element_text(size=4))
p <- p + labs(x = "", y = "Proportion of representative proteins")
p <- p + guides(fill = "none", color = "none", alpha = "none") + 
  scale_alpha_manual(values = c("Mosaic" = 1, "Not mosaic" = 0.3))
p.domains <- p + coord_flip()
p.domains

#p.panelc <- plot_grid(p.seq, p.ecod, labels = c("C", "D"), label_size = 25, ncol = 2)
ggsave(filename = sprintf("%sFigure3/folds_within_mosaic_proteins.jpg", OUTPUT.FIGURES.PATH), p.domains, width = 12, height = 12)


p.odds = ggplot(ecod.table %>% distinct(t_name, odds.ratio), aes(x=t_name, y = odds.ratio)) +
  geom_bar(stat="identity", width = 0.8, fill = "gray90", col = "black") + 
  scale_x_discrete(breaks = ecod.table$t_name, labels = display.t.name) +
  theme_minimal() + 
  guides(fill = "none") + 
  theme(legend.text=element_text(size=4)) + 
  labs(x = "", y = "Odds ratio", fill = "category") + 
  coord_flip() 
p.odds
ggsave(filename = sprintf("%sFigure3/folds_within_mosaic_proteins.odds.ratio.jpg", OUTPUT.FIGURES.PATH) ,p.odds , width = 15, height = 12)
```





# Density plot: Figure 3a
```{r} 
annotated.hhr.table = data.table::fread(sprintf("%sFigure_Supplementary/FigS5_p_vs_cov/cov.vs.pident.txt", OUTPUT.FIGURES.PATH))
spectral.scale = rev(RColorBrewer::brewer.pal(11, "Spectral"))
fig3a = ggplot() + 
  stat_binhex(data = annotated.hhr.table, bins = 50, aes(x=pident, y=max.cov)) + 
  geom_rect(aes(xmin = MINIMUM.PIDENT.FOR.PAIRWISE.HIT/100, xmax = 1, 
            ymin = 0, ymax = MINIMUM.COV.FOR.PROTEIN.SIMILARITY, 
            color = "red"), fill = NA, size = 2) + guides(color = "none") +
  geom_text(aes(x = mean(c(MINIMUM.PIDENT.FOR.PAIRWISE.HIT/100, 1)), 
                y = mean(c(0, MINIMUM.COV.FOR.PROTEIN.SIMILARITY)), 
                label = "Mosaic\n pairs", col = "red")) +
  scale_fill_gradientn(colours = spectral.scale) + 
  scale_x_continuous(limits = c(0,1)) + 
  scale_y_continuous(limits = c(0,1)) + 
  theme_bw() + 
  labs(x = "Percentage identity", y = "Pairwise sequence coverage", fill = "no. pairs")
fig3a
ggsave(sprintf("%sFigure_Supplementary/FigS5_p_vs_cov/panelA_pairwise-seq.png", OUTPUT.FIGURES.PATH), fig3a,  width = 5, height = 4)
  
```

#join fig 3 panels
```{r}
fig3bc = plot_grid(p.odds, p.odds.low.cov, labels = c("B", "C"), label_size = 25, ncol = 2,rel_widths = c(0.45, 0.55))
ggsave(sprintf("%s/Figure3/panelBC.pdf", OUTPUT.FIGURES.PATH), fig3bc,  width = 15, height = 6)
```



