---
title: "generate_figures"
author: "Bogna Smug"
date: '2022-09-28'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(cowplot)
library(RColorBrewer)
library(gridExtra)
library(ggh4x)
ecod_id_to_names_map = data.table::fread(sprintf("%secod_id_to_names_map.txt",OUTPUT.DATA.PATH)) %>% as.data.frame()
included.category.sizes = data.table::fread(file = sprintf("%sincluded.category.sizes",OUTPUT.DATA.PATH))
unspecific.annotation.indices = data.table::fread(file = sprintf("%sunspecific.annotation.indices",OUTPUT.DATA.PATH))
included.categories.metadata = data.table::fread( 
                   sprintf("%stables/included.categories.metadata.txt", OUTPUT.FIGURES.PATH)) %>%
    rowwise() %>%
    mutate(label = gsub(" ", "\n", label))
  
diversity = data.table::fread(sprintf("%sFigure4/diversity.top.diverse.classes.txt", OUTPUT.FIGURES.PATH))
top.diversity.classes = diversity %>% filter(data.type == "top.diversity") %>% distinct(annotation.index)

PHROG.COLOR.MAP2 = PHROG.COLOR.MAP
idx = which(names(PHROG.COLOR.MAP2) == "DNA, RNA and nucleotide metabolism")
names(PHROG.COLOR.MAP2)[idx] =  "DNA, RNA\nand nucleotide metabolism"
idx = which(names(PHROG.COLOR.MAP2) == "moron, auxiliary metabolic gene and host takeover")
names(PHROG.COLOR.MAP2)[idx] =  "moron, auxiliary metabolic\ngene and host takeover"
MIN.NUM.FAMILIES.CATEGORY.COVERAGE.PLOT = 5
text_size.x = 9
text_size.y = 9
phrog.class.labels2 = phrog.class.labels
phrog.class.labels2[which(names(phrog.class.labels2) == "head and packaging")] = "head\nand\npack."
phrog.class.labels2[which(names(phrog.class.labels2) == "DNA, RNA and nucleotide metabolism")] = "DNA,\nRNA\nand\nnucleot.\nmetab."
```

# Main category legend 
```{r}
# PHROG CLASS LEGEND
dat = stack(PHROG.COLOR.MAP)
names(dat) = c("values", "Category")
ggp <- ggplot(dat, aes(x=values, y=values, fill = Category)) +    # Create ggplot2 plot
  geom_col(size = 5) +
  scale_fill_manual(values = PHROG.COLOR.MAP, name = "Category")

jpeg(sprintf("%sCategory_Legend.jpg", OUTPUT.FIGURES.PATH),  width = 10, height = 10, units = "cm", res = 300)
ggp_legend <- cowplot::get_legend(ggp) 
grid::grid.newpage()                                     # Draw empty plot window
grid::grid.draw(ggp_legend)  
dev.off()
```


# Annotation trade-off
```{r, eval = TRUE}
col1.uniqueness <- "darkblue"
col2.coverage <- "sandybrown"
point.size = 3
line.size = 1
text.size = 12
annotation.uncertainty.data.vs.cov = data.table::fread(sprintf("%sFigure_Supplementary/tradeoff/Tradeoff_cov.txt", OUTPUT.FIGURES.PATH)) %>% as.data.frame()
p.a=ggplot(data = annotation.uncertainty.data.vs.cov %>%
          mutate(`probability\nthreshold [%]` = as.factor(minimum.probability),
                  annotation.coverage = 100*annotation.coverage)) +
  geom_point(aes(x = annotation.coverage, y = perc.unique.annotation, linetype = `probability\nthreshold [%]`), col = col1.uniqueness, size = point.size) + 
  geom_line(aes(x = annotation.coverage, y = perc.unique.annotation, linetype = `probability\nthreshold [%]`), col = col1.uniqueness, size = line.size) + 
  geom_point(aes(x = annotation.coverage, y = perc.annotated, linetype = `probability\nthreshold [%]`), col = col2.coverage, size = point.size) + 
  geom_line(aes(x = annotation.coverage, y = perc.annotated, linetype = `probability\nthreshold [%]`), col = col2.coverage, size = line.size) + 
  Get.Theme(legend.position = "right") +
  xlab("sequence coverage threshold [%]") +
  theme( axis.title.y.right = element_text(colour = col2.coverage, size = text.size),
         axis.text.y.right = element_text(colour = col2.coverage, size = text.size),
          axis.title.y = element_text(colour = col1.uniqueness, size = text.size),
         axis.text.y = element_text(colour = col1.uniqueness, size = text.size),
           plot.title = element_text(size=14, face="bold", hjust = 0.5)) +
  scale_y_continuous(name = "functional uniqueness \n(% of annotated proteins with hits to a single function)",
                     limits = c(0,101),
                     breaks = seq(0,100,10),
                     sec.axis=sec_axis(~.*1,name="functional coverage \n(% proteins with a hit to any known function)"))  + 
  labs(title = "Impact of sequence coverage on functional annotation")


annotation.uncertainty.data.vs.prob = data.table::fread(sprintf("%sFigure_Supplementary/tradeoff/Tradeoff_prob.txt", OUTPUT.FIGURES.PATH))
p.b=ggplot(data = annotation.uncertainty.data.vs.prob %>%
             filter(annotation.coverage %in% c(0.1, 0.8)) %>%
          mutate(`sequence coverage\nthreshold [%]`  = as.factor(annotation.coverage*100))) +
  geom_point(aes(x = minimum.probability, y = perc.unique.annotation, linetype = `sequence coverage\nthreshold [%]` ), col = col1.uniqueness, size = point.size) + 
  geom_line(aes(x = minimum.probability, y = perc.unique.annotation, linetype = `sequence coverage\nthreshold [%]` ), col = col1.uniqueness, size = line.size) + 
  geom_point(aes(x = minimum.probability, y = perc.annotated, linetype = `sequence coverage\nthreshold [%]` ), col = col2.coverage, size = point.size) + 
  geom_line(aes(x = minimum.probability, y = perc.annotated, linetype = `sequence coverage\nthreshold [%]` ), col = col2.coverage, size = line.size) + 
  Get.Theme(legend.position = "right") +
  xlab("probability [%]") +
  theme( axis.title.y.right = element_text(colour = col2.coverage, size = text.size),
         axis.text.y.right = element_text(colour = col2.coverage, size = text.size),
          axis.title.y = element_text(colour = col1.uniqueness, size = text.size),
         axis.text.y = element_text(colour = col1.uniqueness, size = text.size),
           plot.title = element_text(size=14, face="bold", hjust = 0.5)) +
  scale_y_continuous(name = "functional uniqueness \n(% of annotated proteins with hits to a single function)",
                     limits = c(0,101),
                     breaks = seq(0,100,10),
                     sec.axis=sec_axis(~.*1,name="functional coverage \n(% proteins with a hit to any known function)")) + 
  labs(title = "Impact of hit probability on functional annotation")

p.figure <- plot_grid(p.a, p.b, labels = c("A", "B"), label_size = 25, ncol = 1)
ggsave(p.figure, filename = sprintf("%sFigure_Supplementary/tradeoff/FigureS1.png", OUTPUT.FIGURES.PATH), bg = "white", width = 12, height = 10)
```





# Domain  network
```{r}
domain.network.metadata =   data.table::fread(sprintf("%sFigure_Supplementary/domain_network/domain.network.metadata.txt", OUTPUT.FIGURES.PATH)) 
domain.network.selected.categories = data.table::fread(sprintf("%sFigure_Supplementary/domain_network/domain.network.selected.categories.txt", OUTPUT.FIGURES.PATH))

for (selected_category in unique(domain.network.metadata$category)) {
  annotated.domain.network = Annotate_Graph(domain.network.selected.categories %>% filter(category == selected_category) %>% select(from, to), 
                    edge.colors = NULL,
                    node.color.map = domain.network.metadata %>%
                      filter(category == selected_category) %>%
                                        select(node, color),
                    # node these sizes are defined at high threshold
                    node.size.map = domain.network.metadata %>%
                      filter(category == selected_category) %>%
                      mutate(size = sqrt(n.prot)) %>%
                      select(node, size),
                    node.labels = domain.network.metadata %>% 
                      filter(category == selected_category) %>%
                      select(node, label) %>%
                      rowwise() %>%
                      mutate(label = gsub(" ", "\n", label))%>%
                      mutate(label = if_else(label == "Intramolecular\nchaperone\ndomain\nin\nvirus\ntail\nspike\nprotein", "Intramolecular\nchaperone\ndomain in\nvirus tail\nspike protein", label)) ,
                    line.color.variable =  "from")
  
  annotated.domain.network.figure = VisualiseGGnetwork(annotated.domain.network, 
                     cex = 0.1, label_size = 1.5, layout = igraph::nicely(), main = "", 
                     nudge_y = 0, alpha = 0.3)
  annotated.domain.network.figure
  ggsave(annotated.domain.network.figure, filename = sprintf("%sFigure_Supplementary/domain_network/annotated.domain.network.figure_%s.png", OUTPUT.FIGURES.PATH, selected_category), bg = "white", width = 8, height = 8)
}


```



# Annotation tradeoff: annotation network
```{r, eval = TRUE}

annotation.netwok.relaxed.threshold = data.table::fread( 
                   sprintf("%sFigure_Supplementary/annot_networks/annotation.netwok.relaxed.threshold.txt", OUTPUT.FIGURES.PATH))
annotation.netwok.default.threshold = data.table::fread( 
                   sprintf("%sFigure_Supplementary/annot_networks/annotation.netwok.default.threshold.txt", OUTPUT.FIGURES.PATH))
annotation.network.metadata = included.categories.metadata %>%
  rowwise() %>%
  mutate(label = if_else(label == "DNA\npolymerase/primase", "DNA\npolymerase/\nprimase",  gsub("\n", " ", label)))

  
annotated.annotation.network.relaxed.threshold = Annotate_Graph(annotation.netwok.relaxed.threshold, 
                  edge.colors = NULL,
                  node.color.map = included.categories.metadata%>%
                                      select(node, color),
                  # node these sizes are defined at high threshold
                  node.size.map = included.categories.metadata %>%
                                      select(node, size),
                  node.labels = included.categories.metadata %>% 
                    select(node, label),
                  line.color.variable =  "from")
annotated.annotation.network.default.threshold = Annotate_Graph(annotation.netwok.default.threshold, 
                  edge.colors = NULL,
                  node.color.map = included.categories.metadata%>%
                                      select(node, color),
                  # node these sizes are defined at high threshold
                  node.size.map = included.categories.metadata %>%
                                      select(node, size),
                  node.labels = included.categories.metadata %>% 
                    select(node, label),
                  line.color.variable =  "from")



network.relaxed.figure = VisualiseGGnetwork(annotated.annotation.network.relaxed.threshold, 
                   cex = 0.1, label_size = 2.35, layout = igraph::nicely(), main = "", nudge_y = 0.005, alpha = 0.3) #nudge_y = 0.015,
network.relaxed.figure
#ggsave(network.relaxed.figure, filename = sprintf("%sFigure_Supplementary/FigS2_annot_network/Relaxed_network.png", OUTPUT.FIGURES.PATH), width = 9, height = 10)

network.default.figure = VisualiseGGnetwork(annotated.annotation.network.default.threshold, 
                   cex = 0.1, label_size = 2.35, layout = igraph::nicely(), main = "", nudge_y = 0.03, alpha = 0.5)
network.default.figure
#ggsave(network.default.figure, filename = sprintf("%sFigure_Supplementary/FigS2_annot_network/Default_network.png", OUTPUT.FIGURES.PATH), width = 5, height = 5)

p.figure <- plot_grid(network.default.figure, network.relaxed.figure, labels = c("A", "B"), label_size = 25, ncol = 1, rel_heights = c(0.27, 0.73))
ggsave(p.figure, filename = sprintf("%sFigure_Supplementary/annot_networks/annot_networkss.pdf", OUTPUT.FIGURES.PATH), width =8, height = 12)

```





# Main mosaicism network
```{r}



mosaic.network = data.table::fread(sprintf("%sFigure2/2b_mosaic.network.txt", OUTPUT.FIGURES.PATH))
pdf(file = sprintf("%sFigure2/Figure2b_main_network_by_domain.pdf", OUTPUT.FIGURES.PATH), width = 5, height = 5)

VisualiseIgraphnetwork(data.for.network = mosaic.network %>% 
                         filter(num.mosaic.fold.combination.pairs > 3) %>%
                         #filter(num.reprseq.pairs.domain > 5)  %>% 
                         #mutate(weight = log10(num.reprseq.pairs.domain)/20) %>%
                         rowwise() %>%
                         mutate(weight = log10(1+num.mosaic.fold.combination.pairs/10),
                                dataset = if_else(num.family.pairs.recent.mosaic.seq.pident50 > 0, ">=50% pident",  "ancient")) %>%
                         ungroup() %>%
                         select(from = from.index, to = to.index, weight, dataset),
                      edge.colors = data.frame(value = c(">=50% pident","ancient"), color = c("darkred", "gray50")),
                        #data.frame(value = c("recent", "ancient"), color = c("darkred", "gray")),
                      #edge.linetypes = data.frame(value = c("ecod", "seq", "both"), lty = c("solid", "dotted", "solid")),
                      node.color.map = included.categories.metadata%>%
                                      select(node, color),
                      # node these sizes are defined at high threshold
                      node.size.map = included.categories.metadata %>%
                                      select(node, size) %>% mutate(size = 2*size),
                      node.labels = included.categories.metadata %>% 
                    select(node, label),
                      line.color.variable = "dataset",
                      line.color.legend.name = NULL,
                      font.size = 0.25,
                      main = "",
                      vertex.label.dist = 2,
                      include.labels = TRUE,
                      alpha = 1,
                    label.font = 2,
                    legend.title = "Mosaicism:",
                    node.label.color = "black",
                    add.legend = FALSE)
dev.off()

```





# Heatmap of ECODS found in multiple categories
```{r}



h.domain.occurence.data = data.table::fread(sprintf("%sFigure1/H_domains.txt", OUTPUT.FIGURES.PATH)) %>% 
  as.data.frame() 
h.domain.occurence.data$category <- factor(h.domain.occurence.data$category, levels = phrogs.class.order)
#h.domain.occurence.data = h.domain.occurence.data %>%
#  tidyr::unite(col = "h_name2", h_name, h_id, sep = "\n")

g0=ggplot(h.domain.occurence.data) +
  geom_tile(aes(x=h_name, y = annotation, fill = num.prot.with.this.h.domain)) + #, hjust = 0, vjust = 0
  theme_bw() +
  theme(axis.text.x = element_text(angle=45, hjust =1,vjust=1, size = text_size.x),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = text_size.y),
        strip.text.y = element_text(face = "bold", size = 8, angle = 360),
        strip.text.x = element_text(face = "bold", size = 8, angle = 360)) +
  facet_grid(category ~all.categories, scales = "free", space = "free", 
             labeller = labeller(category = phrog.class.labels,
                                 all.categories = c(phrog.class.labels2, "multiple categories" = "multiple\ncategories"))) + 
  xlab("") + ylab("") +
  #scale_fill_gradient(name = "number\nof rHHMs", low = "#9999D1", high = "#000046", trans = "log10")
  scale_fill_gradient(name = "number\nof rHHMs", low = "seashell2", high = "seashell4", trans = "log10")

  #scale_alpha_continuous(name = "number\nof rHHMs", trans = "log10") +
  #  scale_fill_manual(values = PHROG.COLOR.MAP) +
  #guides(fill = "none")

g0

# now change the strip colors
g <- ggplot_gtable(ggplot_build(g0))
stripr <- which(grepl('strip-r', g$layout$name))
fills <-PHROG.COLOR.MAP[phrogs.class.order] %>% as.character()
k <- 1
for (i in stripr) {
  j <- which(grepl('rect', g$grobs[[i]]$grobs[[1]]$childrenOrder))
  g$grobs[[i]]$grobs[[1]]$children[[j]]$gp$fill <- fills[k]
  k <- k+1
}


pdf(file = sprintf("%sFigure1/Figure1_H_domains.pdf", OUTPUT.FIGURES.PATH), width = 11, height = 12)
grid::grid.draw(g)
dev.off()

tiff(sprintf("%sFigure1/Figure1_H_domains.tiff", OUTPUT.FIGURES.PATH), width = 11, height = 12)
grid::grid.draw(g)
dev.off()
```

```{r}
num.f.within.h = data.table::fread(sprintf("%sFigure_Supplementary/domain-div-freq-h/num.f.within.h.txt", OUTPUT.FIGURES.PATH))

num.f.within.h.bars = num.f.within.h %>%
  filter(in.ecod.heatmap) %>%
  rename(family = num_f, topology = num_t) %>%
  tidyr::gather(key = "ECOD level", value = "num.groups", family, topology)

text_size.x = 8
num.f.fig = ggplot(num.f.within.h.bars) +
  geom_col(aes(x = h_name, y = num.groups, fill = all.categories, alpha = `ECOD level`), position = position_dodge2(preserve = "single")) +
    theme(axis.text.x = element_text(angle=45, hjust =1,vjust=1, size = text_size.x),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = text_size.y),
        strip.text.y = element_text(face = "bold", size = 6, angle = 360),
        strip.text.x = element_text(face = "bold", size = 6, angle = 360),
        legend.position = "top") +
  facet_grid(. ~all.categories, scales = "free", space = "free", 
             labeller = labeller(all.categories = c(phrog.class.labels2, "multiple categories" = "multiple\ncategories"))) + 
  xlab("") + ylab("Num. of domain families in the group.") +
  guides(fill = "none") +
  scale_fill_manual(values = c(PHROG.COLOR.MAP, "multiple categories" = "#000000")) +
  scale_y_sqrt(breaks = c(1,10,25,50,100,150)) +
  scale_alpha_manual(values = c("family" = 0.5, "topology" =1))


ct=cor.test(num.f.within.h$num_f, num.f.within.h$num_prot, method = "spearman")
cor.fig = ggplot(num.f.within.h, aes(x = num_prot, y = num_f)) +
  geom_point() +
  scale_x_log10() +
  scale_y_log10() +
    stat_cor(method = "spearman", alternative = "two.sided", label.x = 0.2, label.y = 2, label.sep = "\n", cor.coef.name = "rho") +
  #geom_text(aes(x = 2, y = 100, label = paste0("Rho = ", as.numeric(round(ct$estimate,2)), ",\np.val = ",  as.numeric(round(ct$p.value,5))))) +
  theme_bw() + 
  theme(axis.text.x = element_text(size = text_size.x),
        axis.text.y = element_text(size = text_size.y)
        #plot.margin = unit(c(0, 0, 6, 0), "cm")
        ) +
  xlab("Number of representative proteins.") +
  ylab("Number of ECOD family groups.")

fig_diversity_ecod_f = plot_grid(num.f.fig, cor.fig, labels = c("A", "B"), label_size = 25, ncol = 1, rel_heights =  c(0.6, 0.4))#ncol = 2,rel_widths = c(0.6, 0.4))
ggsave(sprintf("%sFigure_Supplementary/domain-div-freq-h/domain-div-freq-h.jpg", OUTPUT.FIGURES.PATH),width = 9, height = 12) #)width = 15, height = 7)

```

```{r}
t.domain.occurence.data.full = data.table::fread(sprintf("%sFigure_Supplementary/T_domains/T_domains.txt", OUTPUT.FIGURES.PATH)) %>%
  as.data.frame() 
t.domain.occurence.data.full$category <- factor(t.domain.occurence.data.full$category, levels = phrogs.class.order)
ggplot(t.domain.occurence.data.full) +
  geom_tile(aes(y=t_name, x = annotation, fill = category)) + #, hjust = 0, vjust = 0
  theme(axis.text.x = element_text(angle=45, hjust =1,vjust=1, size = text_size.x),
        axis.text.y = element_text(hjust = 1,vjust=0.5, size = text_size.y),
        strip.text.y = element_text(face = "bold", size = 6, angle = 360),
        strip.text.x = element_text(face = "bold", size = 6, angle = 360)) +
  facet_grid(all.categories ~category, scales = "free", space = "free", 
             labeller = labeller(category = phrog.class.labels,
                                 all.categories = c(phrog.class.labels2, "multiple categories" = "multiple\ncategories"))) + 
  xlab("") + ylab("") +
  guides(fill = "none") +
  scale_fill_manual(values = PHROG.COLOR.MAP)
ggsave(sprintf("%sFigure_Supplementary/T_domains/T_domains.jpg", OUTPUT.FIGURES.PATH), width = 20, height = 20)


```



# Shiny plots for most mosaic classes
```{r, fig.width = 10}
all.tiles.data = data.table::fread(sprintf("%sFigure3/domain.positions.txt", OUTPUT.FIGURES.PATH))
most.mosaic.classes = c("DNA polymerase", "tail fiber","tail spike", "endolysin")
i=0
domain.position.plots = list()
for (mosaic.class in most.mosaic.classes) {
  i=i+1
  tile.data = all.tiles.data %>%
    filter(domain.level == "T" & annotation %in% mosaic.class)
  domain.position.plots[[i]] =Plot.Domain.Combinations(tile.data, text.size.axis = 8, text.size.panel = 12)$domain.position.plots[[1]]
}
pdf(sprintf("%sFigure3/domain.positions_most_mosaic.pdf", OUTPUT.FIGURES.PATH), 
           width = 17,
           height = 20)
do.call("grid.arrange", c(domain.position.plots, ncol=1))
dev.off()
```



# diversity scatterplot and ecod coverage
```{r}

# label the ones that have many families or fold types
diversity.ok.ecod = diversity %>% 
  #filter(!poorly.ecod.annotated)  %>%
  mutate(label = if_else(diverse.for.label,annotation , ""))


Fig4b = ggplot(diversity.ok.ecod, aes(x = num.families, y = num.topology.combinations)) + 
  geom_point(aes(colour = category)) + 
  stat_cor(method = "spearman", alternative = "two.sided", label.x = 0.2, label.y = 35, label.sep = "\n", cor.coef.name = "rho") +
  #geom_vline(aes(xintercept = threshold.num.fam), col = "red", linetype = "dashed") +
  #geom_hline(aes(yintercept = threshold.num.topology.combinations), col = "red", linetype = "dashed") +
  scale_x_log10() + 
  ggrepel::geom_label_repel(aes(x = num.families, y = num.topology.combinations, colour = category, label = label), size = 3) + 
  scale_colour_manual(values = PHROG.COLOR.MAP) + 
  labs(x = "Number of families", y = "Number of fold combinations") + 
  guides(colour = "none") + theme_minimal()

dir.create(sprintf("%sFigure_Supplementary/diversity_scatterplot/", OUTPUT.FIGURES.PATH))
ggsave(sprintf("%sFigure_Supplementary/diversity_scatterplot/diversity_scatterplot.pdf", OUTPUT.FIGURES.PATH), Fig4b, width = 10, height = 6)
ggsave(sprintf("%sFigure_Supplementary/diversity_scatterplot/diversity_scatterplot.png", OUTPUT.FIGURES.PATH), Fig4b,bg = "white", width = 10, height = 6)
## supplementary version
```

```{r}
  levels = c("mean.length",
             "max.num.h",
            "mean.num.t", 
             "prop.proteins.with.ecod.hit",
             "prop.with.multi.x.domains")
  
Get.Labeller = function(diversity.top, levels) {
   diversity.top = diversity.top %>%
    filter(proportion %in% levels)
  diversity.top = diversity.top %>% mutate(proportion = factor(proportion, levels = levels))
  diversity.levels <- levels(diversity.top$proportion)
  diversity.labeller <- character(length(diversity.levels))
  diversity.labeller[diversity.levels == "mean.num.t"] <- "Mean number of folds\nper fold architecture"
  #diversity.labeller[diversity.levels == "prop.mosaic.pairs.sequence.thr0p1"] <- "Proportion of mosaic pairs\nby sequence (10%)"
  #diversity.labeller[diversity.levels == "prop.mosaic.pairs.sequence.thr0p3"] <- "Proportion of mosaic pairs\nby sequence (30%)"
  #diversity.labeller[diversity.levels == "prop.mosaic.pairs.sequence.thr0p5"] <- "Proportion of mosaic pairs\nby sequence (50%)"
  #diversity.labeller[diversity.levels == "num.theoretical.fold.type.pairs.with.multi.x"] <- "Number of possible\npairs of fold types\nwith multiple X groups"
 # diversity.labeller[diversity.levels == "num.mosaic.pairs"] <- "Number of pairs\nmosaic by domain"
  diversity.labeller[diversity.levels == "prop.with.multi.x.domains"] <- "Prop. of prot. with >= 2 X groups"
  #diversity.labeller[diversity.levels == "num.reprseq.pairs.domain"] <- "Number of protein pairs\nwith domain mosaicism"
  #diversity.labeller[diversity.levels == "prop.mosaic"] <- "Prop. of proteins\nhaving any mosaic signal."
  #diversity.labeller[diversity.levels == "proteome covered by ECOD hits"] <- "proteome covered by ECOD hits"
  diversity.labeller[diversity.levels == "prop.proteins.with.ecod.hit"] <-  "Prop. of prot. with >= 1 X group"
  diversity.labeller[diversity.levels == "mean.length"] = "Mean protein length (num AA)"
  diversity.labeller[diversity.levels == "max.num.h"] = "Max number of H domains"
  names(diversity.labeller) <- diversity.levels
  return(diversity.labeller)
}


diversity.top = diversity %>%
  filter(!(annotation.index %in% unspecific.annotation.indices)) %>%
  filter(num.families > MIN.NUM.FAMILIES.CATEGORY.COVERAGE.PLOT) %>%
  #inner_join(top.diversity.classes) %>%
    select(mean.num.t, annotation, category, 
           prop.with.multi.x.domains,
           prop.proteins.with.ecod.hit,
           mean.length,
           max.num.h) %>%
  tidyr::gather(key = "proportion", value = "value", -c("annotation", "category")) %>%
  filter(proportion %in% levels)
diversity.top$proportion = factor(diversity.top$proportion, levels = levels)
diversity.labeller = Get.Labeller(diversity.top, levels)
p.most.diverse = Plot.Mosaicism.Table(diversity.top %>% 
                                        filter(proportion %in% names(diversity.labeller)), 
                                      PHROG.COLOR.MAP = PHROG.COLOR.MAP) +
    facet_grid(category ~ proportion, scales = "free", space = "free_y", 
                labeller = labeller(category = phrog.class.labels, proportion = diversity.labeller)) 
dir.create(sprintf("%sFigure_Supplementary/ecod-coverage", OUTPUT.FIGURES.PATH))
ggsave(sprintf("%sFigure_Supplementary/ecod-coverage/ecod-coverage.pdf", OUTPUT.FIGURES.PATH), 
      p.most.diverse, width = 15, height = 11)
ggsave(sprintf("%sFigure_Supplementary/ecod-coverage/ecod-coverage.png", OUTPUT.FIGURES.PATH), 
     p.most.diverse, bg = "white", width = 15, height = 11)

```



# Mosaicism by sequence vs ecod
```{r}
# mosaic family in a given annotatioh i.e. at least one protein in this family having this annotation is mosaic
mosaicism.underestimated.data  = data.table::fread(sprintf("%sFigure4/mosaic.domain.classes.txt", OUTPUT.FIGURES.PATH)) %>%
   select(annotation.index, annotation, category, num.families, num.families.mosaic.domain, num.families.mosaic.seq.or.domain, pval.seq.or.domain.corr, pval.domain.corr)  %>%
  inner_join(included.category.sizes %>% select(annotation.index)) %>%
  inner_join(top.diversity.classes)

mosaicism.underestimated.data = mosaicism.underestimated.data %>%
  select(annotation.index, annotation, category, 
         num.families,
         num.families.mosaic.domain,
         num.families.mosaic.seq.or.domain,
         pval.seq.or.domain.corr,
         pval.domain.corr,
         num.families.mosaic.domain) %>%
  filter(!(annotation.index %in% unspecific.annotation.indices)) %>%
  rename(`All` = num.families,
         `Mosaic by ECOD` =  num.families.mosaic.domain,#num.prot.mosaic.domain.within.and.to.other.specific.class, 
         `Mosaic by ECOD or seq` =   num.families.mosaic.seq.or.domain) %>%
  tidyr::gather(key = "type", value = "num.families", `All`, `Mosaic by ECOD`, `Mosaic by ECOD or seq`)  %>%
  mutate(label_domain = if_else(pval.domain.corr <= 0.05 & type == "Mosaic by ECOD", "*", ""),
         label_seq_or_domain = if_else(pval.seq.or.domain.corr <= 0.05 & type == "Mosaic by ECOD or seq","*", ""))
mosaicism.underestimated.data$type = factor(mosaicism.underestimated.data$type,
                                            levels = c(
                                              'All',
                                              'Mosaic by ECOD or seq',
                                              'Mosaic by ECOD'))



g=ggplot(mosaicism.underestimated.data, aes(y = annotation, x = num.families, fill = category, col = category)) +
  geom_bar(aes(alpha = type), stat = "identity", position  = "identity") +
  geom_text(aes(label = label_domain), nudge_x = 2, size = 5, color = "black", fontface='bold') +
  geom_text(aes(label = label_seq_or_domain), nudge_x = 2, size = 5, color = "gray30", fontface='bold') +
  #geom_text(aes(label = significant), nudge_x = 10) +
  scale_alpha_manual(values = c(All = 0.05,
                                `Mosaic by ECOD` = 1,
                                `Mosaic by ECOD or seq` = 0.3)) +
    scale_fill_manual(values = PHROG.COLOR.MAP) + 
    scale_color_manual(values = PHROG.COLOR.MAP) + 
   theme_minimal() + 
    theme(strip.text.y = element_text(face = "bold", size = 7, angle = 360)) + 
    labs(y = "", x = "Num. rHMM families") +
    scale_x_continuous(labels=function(x) sprintf("%.2f", x)) +
    facet_grid(category ~ ., scales = "free", space = "free") +
  guides(color = "none", fill = "none")
  
g
ggsave(sprintf("%sFigure4/Ecod_vs_signal_mosaicism_families.new.pdf", OUTPUT.FIGURES.PATH), g,
       width = 13, height = 7)
ggsave(sprintf("%sFigure4/Ecod_vs_signal_mosaicism_families.new.png", OUTPUT.FIGURES.PATH), g, bg = "white",
       width = 13, height = 7)

```






# Fig 3c shared ecods in knwonn and unknown prots

```{r}
t.domain.occurence.data = data.table::fread(sprintf("%sFigure2/t.domain.occurence.data.txt", OUTPUT.FIGURES.PATH)) %>%
  as.data.frame() 
domain_fill_colors = t.domain.occurence.data %>% 
  select(t_id, category = most_frequent_category) %>%
  distinct() %>%
  left_join(PHROG.COLOR.MAP %>% stack() %>% rename(category = ind, color = values)) %>%
  mutate(color = if_else(is.na(color), "#FFFFFF", color)) %>%
  as.data.frame()

all.t.in.fold.types = data.table::fread(file = sprintf("%sFigure2/all.t.in.fold.types.txt", OUTPUT.FIGURES.PATH))
p.odds = Plot.Domains.Within.Mosaic.Proteins(all.t.in.fold.types, colors = domain_fill_colors) + theme(text = element_text(size = 20))
#ggsave(filename = sprintf("%sAdditional_Figures/folds_within_mosaic_fold_types.jpg", OUTPUT.FIGURES.PATH) ,p.odds , width = 15, height = 12)
ggsave(filename = sprintf("%sFigure2/folds_within_mosaic_fold_types.pdf", OUTPUT.FIGURES.PATH) ,p.odds , width = 10, height = 9)

all.t.in.fold.types_pident50 = data.table::fread(file = sprintf("%sFigure2/all.t.in.fold.types_pident50.txt", OUTPUT.FIGURES.PATH))
p.odds = Plot.Domains.Within.Mosaic.Proteins(all.t.in.fold.types_pident50, colors = domain_fill_colors) + theme(text = element_text(size = 20))
#ggsave(filename = sprintf("%sAdditional_Figures/folds_within_mosaic_fold_types_pident50.jpg", OUTPUT.FIGURES.PATH) ,p.odds , width = 15, height = 12)
ggsave(filename = sprintf("%sFigure2/folds_within_mosaic_fold_types_pident50.pdf", OUTPUT.FIGURES.PATH) ,p.odds , width = 10, height = 9)
```



# cov vs pident heatmap
```{r} 
spectral.scale = rev(RColorBrewer::brewer.pal(11, "Spectral"))
qname.pairs.sharing.t.and.disshare.x = data.table::fread(file = sprintf("%stables/qname.pairs.sharing.t.and.disshare.x.txt", OUTPUT.FIGURES.PATH)) %>%
  distinct(reprseq.pair)

qname.pairs.sharing.t.and.disshare.x.symmetric = rbind(
  qname.pairs.sharing.t.and.disshare.x %>%
      tidyr::separate(col = reprseq.pair, into = c('qname', 'sname'), sep = "&", remove = FALSE) ,
  qname.pairs.sharing.t.and.disshare.x %>%
      tidyr::separate(col = reprseq.pair, into = c('sname', 'qname'), sep = "&", remove = FALSE)) 

cov.vs.pident = data.table::fread(sprintf("%sFigure_Supplementary/pident-maxcov/cov.vs.pident.txt", OUTPUT.FIGURES.PATH))

cov.vs.pident.plot = ggplot() + 
  stat_binhex(data = cov.vs.pident, bins = 50, aes(x=pident, y=max.cov)) + 
#  geom_rect(aes(xmin = MINIMUM.PIDENT.FOR.PAIRWISE.HIT/100, xmax = 1, 
#            ymin = 0, ymax = MINIMUM.COV.FOR.PROTEIN.SIMILARITY, 
#            color = "red"), fill = NA, size = 2) + guides(color = "none") +
#  geom_text(aes(x = mean(c(MINIMUM.PIDENT.FOR.PAIRWISE.HIT/100, 1)), 
#                y = mean(c(0, MINIMUM.COV.FOR.PROTEIN.SIMILARITY)), 
#                label = "Mosaic\n pairs", col = "red")) +
  scale_fill_gradientn(colours = spectral.scale, trans = 'log10') + 
  scale_x_continuous(limits = c(0,1)) + 
  scale_y_continuous(limits = c(0,1)) + 
  theme_bw() + 
  labs(x = "Percentage identity", y = "Pairwise sequence coverage", fill = "no. pairs")

#ggsave(sprintf("%sFigure_Supplementary/FigS5_p_vs_cov/panelA_pairwise-seq.png", OUTPUT.FIGURES.PATH), fig3a, bg = "white",  width = 5, height = 4)

#bb = qname.pairs.sharing.t.and.disshare.x.symmetric %>%
#  anti_join(cov.vs.pident)

domain.mosaic.pair.cov.vs.pident.data = cov.vs.pident %>%
  inner_join(qname.pairs.sharing.t.and.disshare.x.symmetric)

domain.mosaic.pair.cov.vs.pident.plot = ggplot() + 
  stat_binhex(data = domain.mosaic.pair.cov.vs.pident.data, bins = 50, aes(x=pident, y=max.cov)) + 
  #geom_rect(aes(xmin = MINIMUM.PIDENT.FOR.PAIRWISE.HIT/100, xmax = 1, 
  #          ymin = 0, ymax = MINIMUM.COV.FOR.PROTEIN.SIMILARITY, 
  #          color = "red"), fill = NA, size = 2) + guides(color = "none") +
  #geom_text(aes(x = mean(c(MINIMUM.PIDENT.FOR.PAIRWISE.HIT/100, 1)), 
  #              y = mean(c(0, MINIMUM.COV.FOR.PROTEIN.SIMILARITY)), 
  #              label = "Mosaic\n pairs", col = "red")) +
  scale_fill_gradientn(colours = spectral.scale, trans = 'log10') + 
  scale_x_continuous(limits = c(0,1)) + 
  scale_y_continuous(limits = c(0,1)) + 
  theme_bw() + 
  labs(x = "Percentage identity", y = "Pairwise sequence coverage", fill = "no. pairs")



not.domain.mosaic.pair.cov.vs.pident.data = cov.vs.pident %>%
  anti_join(qname.pairs.sharing.t.and.disshare.x.symmetric) 

not.domain.mosaic.pair.cov.vs.pident.plot = ggplot() + 
  stat_binhex(data = not.domain.mosaic.pair.cov.vs.pident.data, bins = 50, aes(x=pident, y=max.cov)) + 
  #geom_rect(aes(xmin = MINIMUM.PIDENT.FOR.PAIRWISE.HIT/100, xmax = 1, 
  #          ymin = 0, ymax = MINIMUM.COV.FOR.PROTEIN.SIMILARITY, 
  #          color = "red"), fill = NA, size = 2) + guides(color = "none") +
  #geom_text(aes(x = mean(c(MINIMUM.PIDENT.FOR.PAIRWISE.HIT/100, 1)), 
  #              y = mean(c(0, MINIMUM.COV.FOR.PROTEIN.SIMILARITY)), 
  #              label = "Mosaic\n pairs", col = "red")) +
  scale_fill_gradientn(colours = spectral.scale, trans = 'log10') + 
  scale_x_continuous(limits = c(0,1)) + 
  scale_y_continuous(limits = c(0,1)) + 
  theme_bw() + 
  labs(x = "Percentage identity", y = "Pairwise sequence coverage", fill = "no. pairs")

cov.vs.pident.figure <- plot_grid(cov.vs.pident.plot, not.domain.mosaic.pair.cov.vs.pident.plot,domain.mosaic.pair.cov.vs.pident.plot, 
                                  labels = c("A", "B", "C"), 
                                  label_size = 25, ncol = 1)

ggsave(sprintf("%sFigure_Supplementary/pident-maxcov/pident-maxcov.png", OUTPUT.FIGURES.PATH), cov.vs.pident.figure, bg = "white",  width = 5, height = 12)

```

```{r}

recent_HGT_pairs_70 = data.table::fread(sprintf("%sFigure_Supplementary/network_recent_HGT/recent_HGT_network_70.txt", OUTPUT.FIGURES.PATH)) %>%
  select(from, to, num.prot.pairs)
recent_HGT_pairs_90 = data.table::fread(sprintf("%sFigure_Supplementary/network_recent_HGT/recent_HGT_network_90.txt", OUTPUT.FIGURES.PATH)) %>%
  select(from, to, num.prot.pairs)


pdf(sprintf("%sFigure_Supplementary/network_recent_HGT/network_recent_HGT.pdf", OUTPUT.FIGURES.PATH), width = 6, height = 12) #, units = "cm", res = 300)
par(mfrow=c(2,1))
VisualiseIgraphnetwork(data.for.network = recent_HGT_pairs_70 %>% mutate(weight = log10(num.prot.pairs+1)/3),
                      edge.colors = NULL,
                      node.color.map = included.categories.metadata %>%
                                      select(node, color),
                      # node these sizes are defined at high threshold
                      node.size.map = included.categories.metadata %>%
                                      select(node, size),
                      node.labels = included.categories.metadata %>% 
                    select(node, label),
                      line.color.variable = "from",
                      line.color.legend.name = NULL,
                      font.size = 0.5,
                      vertex.size = 1,
                      main = "",
                      vertex.label.dist = 1,
                      include.labels = TRUE,
                      alpha = 0.3)

VisualiseIgraphnetwork(data.for.network = recent_HGT_pairs_90 %>% mutate(weight = log10(num.prot.pairs+1)/3),
                      edge.colors = NULL,
                      node.color.map = included.categories.metadata %>%
                                      select(node, color),
                      # node these sizes are defined at high threshold
                      node.size.map = included.categories.metadata  %>%
                                      select(node, size),
                      node.labels = included.categories.metadata  %>% 
                    select(node, label),
                      line.color.variable = "from",
                      line.color.legend.name = NULL,
                      font.size = 0.5,
                      vertex.size = 1,
                      main = "",
                      vertex.label.dist = 1,
                      include.labels = TRUE,
                      alpha = 0.3)
dev.off()
```

# Supplementary Table
```{r}

mosaic.domain.classes = data.table::fread(sprintf("%sFigure4/mosaic.domain.classes.txt", OUTPUT.FIGURES.PATH))
write.xlsx(x = mosaic.domain.classes %>% left_join(annotation.indices) %>% select(annotation.index, annotation, category, 
                                                                                      odds.ratio.domain, pval.domain, pval.domain.corr,
                                                                                      odds.ratio.seq.or.domain, pval.seq.or.domain, pval.seq.or.domain.corr), 
           file= sprintf("%sFigure_Supplementary/Supplementary_Tables/Statistical_Tests.xls", OUTPUT.FIGURES.PATH), 
           sheetName = "Overrepresentation_of_functional_classes_in_mosaic_rHMMs", 
           col.names = TRUE, row.names = FALSE, append = FALSE)

all.t.in.fold.types.pident50 = data.table::fread(file = sprintf("%sFigure2/all.t.in.fold.types_pident50.txt", OUTPUT.FIGURES.PATH))
all.t.in.fold.types = data.table::fread(file = sprintf("%sFigure2/all.t.in.fold.types.txt", OUTPUT.FIGURES.PATH))
stat.dat = all.t.in.fold.types  %>% select(t_id, t_name, h.index, h_name, odds.ratio, pval) %>%
  full_join(all.t.in.fold.types.pident50 %>% select(t_id, t_name, h.index, h_name, pident50.odds.ratio = odds.ratio, pident50.pval = pval))

write.xlsx(x = stat.dat, 
           file= sprintf("%sFigure_Supplementary/Supplementary_Tables/Statistical_Tests.xls", OUTPUT.FIGURES.PATH), 
           sheetName = "Overrepresentation_of_ECOD_domains_in_mosaic_rHMMs", 
           col.names = TRUE, row.names = FALSE, append = TRUE)

```



```{r, eval = TRUE}
 tmp <- installed.packages()
 save(tmp, file=sprintf("%sFigure_Supplementary/Supplementary_Tables/R_installedpkgs_all.rda", OUTPUT.FIGURES.PATH))
``` 


```{r, eval = FALSE}
library(devtools)  
 load(file=sprintf("%sFigure_Supplementary/Supplementary_Tables/R_installedpkgs_all.rda", OUTPUT.FIGURES.PATH))
 installedpackages = tmp
 tmp[which(names(tmp[,1]) == "plotly"),]
 tmp[which(names(tmp[,1]) == "ggplot2"),]
pcknames = as.vector(installedpackages[is.na(installedpackages[,"Priority"]), 1])
pckversions = as.vector(installedpackages[is.na(installedpackages[,"Priority"]), 3])
for (count in 1:length(installedpackages)) {
    install_version(pcknames[count], version = pckversions[count])
}
```


